{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5338cfab-cf15-4c63-a091-c1cbb51bf8dd",
   "metadata": {},
   "source": [
    "# Process Bigraph Schema\n",
    "This tutorial covers Process Bigraph Schema, starting with single `store` and increasing in detail as you scroll towards the bottom. It uses the libraries [bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) for the schema validation and type registry and [bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) for visualization.\n",
    "\n",
    "## Bigraph-schema and Bigraph-viz\n",
    "\n",
    "[Bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) is a serializable type schema for Process Bigraph simulations. From this library, we will use the `validate_schema` method to check our schemas, and `fill` to fill in missing details to prepare the schema for simulation. We will also use the `type_registry` to access different schemas, and register new ones.\n",
    "\n",
    "[Bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) is an easy-to-use plotting tool for compositional bigraph schema, built on top of Graphviz. It reads raw Compositional Bigraph Schemas and generates a figure that can be displayed in a notebook or saved to file. Bigraph-viz's main plotting function is `plot_bigraph` for the bigraph structure. `plot_flow` and `plot_multitimestep` make plots of process orchestration."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df69e36f-3da8-4bfb-8bc0-0aeb04a6f7d7",
   "metadata": {},
   "source": [
    "### Installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "63fb7d1a-cd49-4aac-8d4f-31bb54496a51",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bigraph-schema==0.0.13\n",
      "bigraph-viz==0.0.30\n"
     ]
    }
   ],
   "source": [
    "# !pip install bigraph-schema\n",
    "# !pip install bigraph-viz\n",
    "!pip freeze | grep bigraph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f700afa2-bfee-49bf-a4b1-e23f0540461d",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "2ab3cb32-e007-45c4-807e-b1b442ac1608",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from bigraph_viz import plot_bigraph, plot_flow, plot_multitimestep, pf\n",
    "from bigraph_schema import TypeSystem\n",
    "from bigraph_viz.dict_utils import replace_regex_recursive\n",
    "\n",
    "plot_settings = {'remove_process_place_edges': True}\n",
    "save_images = False\n",
    "if save_images:\n",
    "    plot_settings.update({'out_dir': 'out','dpi': '250'})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3b85fe8-899d-4c61-8c75-0d572e44e493",
   "metadata": {},
   "source": [
    "### Initialize Type System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b1166de0-3811-4b35-90c1-3fbfa8343259",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "types = TypeSystem()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e716c23-5fbe-4c08-a868-09611ca40ec4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Stores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a6e9cc2-e0df-4f73-b4c3-eddb1055b088",
   "metadata": {},
   "source": [
    "### Simple stores\n",
    "A store can be declared in a JSON dictionary with the store's name mapped to its `value`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d711b952-089a-4938-817d-d1ccad9de56e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"59pt\" height=\"59pt\"\n",
       " viewBox=\"0.00 0.00 60.33 60.33\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 56.33)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-56.33 56.33,-56.33 56.33,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"26.16\" cy=\"-26.16\" rx=\"26.33\" ry=\"26.33\"/>\n",
       "<text text-anchor=\"start\" x=\"11.66\" y=\"-29.56\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "<text text-anchor=\"start\" x=\"18.16\" y=\"-16.56\" font-family=\"Times,serif\" font-size=\"12.00\">1.0</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d40ca0>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simple_store_state = {\n",
    "    'store1': 1.0,\n",
    "}\n",
    "plot_bigraph(simple_store_state, **plot_settings, show_values=True, filename='simple_store')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db1cead7-e691-4291-920e-827d69c00727",
   "metadata": {},
   "source": [
    "To include values *and* types in a node label, you can instead have the store name map to a dict with `_value` and `_type` keys."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f3e5fa71-7ffc-4a8f-a1a0-dc547c886f84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"78pt\" height=\"78pt\"\n",
       " viewBox=\"0.00 0.00 80.12 80.12\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 76.12)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-76.12 76.12,-76.12 76.12,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"36.06\" cy=\"-36.06\" rx=\"36.12\" ry=\"36.12\"/>\n",
       "<text text-anchor=\"start\" x=\"21.56\" y=\"-39.46\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "<text text-anchor=\"start\" x=\"14.56\" y=\"-26.46\" font-family=\"Times,serif\" font-size=\"12.00\">1.0::float</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d40eb0>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "typed_store_spec = {\n",
    "    'store1': {\n",
    "        '_value': 1.0, \n",
    "        '_type': 'float',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(typed_store_spec, **plot_settings, show_values=True, show_types=True, filename='store')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de1011cb-903b-4c15-baed-a62f4afdd6b9",
   "metadata": {},
   "source": [
    "### Hierarchy\n",
    "A hierarchy is a place graph of nested stores. Stores can be placed within stores using JSON dictionary nesting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "0f3f0824-f89d-4c6e-830e-2c5a321f7b33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"246pt\" height=\"355pt\"\n",
       " viewBox=\"0.00 0.00 253.25 364.70\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 360.7)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-360.7 249.25,-360.7 249.25,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"156.86\" cy=\"-333.92\" rx=\"22.56\" ry=\"22.56\"/>\n",
       "<text text-anchor=\"start\" x=\"142.36\" y=\"-330.82\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"119.86\" cy=\"-246.83\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"100.86\" y=\"-243.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M147.9,-312.32C142.81,-300.62 136.39,-285.86 130.98,-273.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"194.86\" cy=\"-246.83\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"175.86\" y=\"-243.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M166.06,-312.32C171.28,-300.62 177.87,-285.86 183.43,-273.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"33.86\" cy=\"-148.65\" rx=\"33.72\" ry=\"33.72\"/>\n",
       "<text text-anchor=\"start\" x=\"10.36\" y=\"-145.55\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M101.2,-224.96C88.06,-210.26 70.39,-190.51 56.45,-174.92\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"119.86\" cy=\"-148.65\" rx=\"33.72\" ry=\"33.72\"/>\n",
       "<text text-anchor=\"start\" x=\"96.36\" y=\"-145.55\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M119.86,-217.75C119.86,-206.88 119.86,-194.36 119.86,-183.03\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"205.86\" cy=\"-148.65\" rx=\"33.72\" ry=\"33.72\"/>\n",
       "<text text-anchor=\"start\" x=\"182.36\" y=\"-145.55\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M138.52,-224.96C151.66,-210.26 169.32,-190.51 183.26,-174.92\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"205.86\" cy=\"-39.4\" rx=\"39.3\" ry=\"39.3\"/>\n",
       "<text text-anchor=\"start\" x=\"177.86\" y=\"-36.3\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M205.86,-114.01C205.86,-103.08 205.86,-90.85 205.86,-79.54\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108ce7bb0>"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hierarchy_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': {\n",
    "            'store1.1.1': 'Any',\n",
    "            'store1.1.2': 'Any',\n",
    "            'store1.1.3': {\n",
    "                'store1.1.3.1': 'Any',\n",
    "            },\n",
    "        },\n",
    "        'store1.2': 'Any',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(hierarchy_spec, **plot_settings, filename='hierarchy')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c825e07c-8cd5-4df1-8958-aa0f36d61045",
   "metadata": {},
   "source": [
    "## Processes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ea1a83-aeac-4048-9593-7b30588b0ad8",
   "metadata": {},
   "source": [
    "### Single process\n",
    "Processes require ports, which are declared with the `_ports` key mapped to a dict that has port names matched to types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7c74237b-0d66-4e8f-a8ca-79174fadad8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"118pt\" height=\"95pt\"\n",
       " viewBox=\"0.00 0.00 121.00 98.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 94)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-94 117,-94 117,4 -4,4\"/>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-63 0,-63 0,-27 54,-27 54,-63\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-41.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.06,-71.7C105.93,-70.05 80.86,-61.99 59.59,-55.15\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"57.22\" cy=\"-54.39\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-69\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.06,-18.3C105.93,-19.95 80.86,-28.01 59.59,-34.85\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"57.22\" cy=\"-35.61\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-33\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d40e20>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_spec = {\n",
    "    '_type': 'python_process',\n",
    "    'process1': {\n",
    "        '_type': 'edge',\n",
    "        '_ports': {\n",
    "            'port1': 'Any',\n",
    "            'port2': 'Any',\n",
    "        },\n",
    "    },\n",
    "}\n",
    "# validate_schema(process_spec)  # validate it follows the schema types TODO -- are disconnected ports allowed?\n",
    "plot_bigraph(process_spec, **plot_settings, rankdir='RL', filename='process')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68a2c480-32fe-43d3-b618-8fc6c647b8ca",
   "metadata": {},
   "source": [
    "### Multiple processes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "8c2457a3-fc47-4850-8b9e-9ee98e81567d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"249pt\" height=\"123pt\"\n",
       " viewBox=\"0.00 0.00 256.50 127.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 123)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-123 252.5,-123 252.5,4 -4,4\"/>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"67.5,-119 13.5,-119 13.5,-83 67.5,-83 67.5,-119\"/>\n",
       "<text text-anchor=\"start\" x=\"19.5\" y=\"-97.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"157.5,-119 103.5,-119 103.5,-83 157.5,-83 157.5,-119\"/>\n",
       "<text text-anchor=\"start\" x=\"109.5\" y=\"-97.9\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;process3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"238.5,-119 184.5,-119 184.5,-83 238.5,-83 238.5,-119\"/>\n",
       "<text text-anchor=\"start\" x=\"190.5\" y=\"-97.9\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1.93,-23.99C4.5,-32.6 10.28,-50.71 17.5,-65 19.71,-69.38 22.36,-73.88 25.05,-78.14\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"26.39\" cy=\"-80.21\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"17.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M44.39,-36.98C43.66,-48.83 42.69,-64.45 41.91,-77.17\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"41.76\" cy=\"-79.58\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"43.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process2&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;process2&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M100.97,-27C102.83,-36.32 106.41,-52.1 111.5,-65 113.21,-69.34 115.35,-73.83 117.56,-78.08\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"118.6\" cy=\"-80.03\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"111.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process2&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;process2&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M143,-27.67C140.95,-39.53 137.2,-61.24 134.34,-77.79\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"133.98\" cy=\"-79.84\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"138.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process3&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;process3&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M192.13,-36.7C191.9,-45.35 192.29,-55.84 194.5,-65 195.52,-69.24 197.12,-73.58 198.92,-77.68\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"199.89\" cy=\"-79.76\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"194.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process3&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;process3&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M237.05,-23.35C233.7,-33.41 225.22,-58.83 218.95,-77.66\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"218.2\" cy=\"-79.92\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"226.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d40250>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_schema = {\n",
    "    '_type': 'python_process',\n",
    "    '_ports': {\n",
    "        'port1': 'Any',\n",
    "        'port2': 'Any'\n",
    "    }\n",
    "}\n",
    "\n",
    "processes_spec = {\n",
    "    'process1': process_schema,\n",
    "    'process2': process_schema,\n",
    "    'process3': process_schema,\n",
    "}\n",
    "# validate_schema(processes_spec)  # validate it follows the schema types TODO -- are disconnected ports allowed?\n",
    "plot_bigraph(processes_spec, **plot_settings, rankdir='BT', filename='multiple_processes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e5f5a6-8dd4-41c3-ba37-33ed9e3e0830",
   "metadata": {},
   "source": [
    "## Wires\n",
    "To connect a process's ports to stores, add wiring with a `_wires` key that maps port names to relative paths in the store hierarchy. The ports used by the wires need to match the ports in the schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a6ecad5d-8948-4636-910c-79284a3e8679",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"180pt\" height=\"134pt\"\n",
       " viewBox=\"0.00 0.00 184.78 137.78\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 133.78)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-133.78 180.78,-133.78 180.78,4 -4,4\"/>\n",
       "<!-- (&#39;node1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;node1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"23.39\" cy=\"-106.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"8.39\" y=\"-103.29\" font-family=\"Times,serif\" font-size=\"12.00\">node1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"115.39,-36 61.39,-36 61.39,0 115.39,0 115.39,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"67.39\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M37.5,-86.65C47.67,-73.13 61.42,-54.85 72.07,-40.7\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"73.48\" cy=\"-38.82\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"62.39\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;node2&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"88.39\" cy=\"-106.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"73.39\" y=\"-103.29\" font-family=\"Times,serif\" font-size=\"12.00\">node2</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M88.39,-82.24C88.39,-69.62 88.39,-54.07 88.39,-41.55\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"88.39\" cy=\"-39.23\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"88.39\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;node3&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"153.39\" cy=\"-106.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"138.39\" y=\"-103.29\" font-family=\"Times,serif\" font-size=\"12.00\">node3</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M139.29,-86.65C129.12,-73.13 115.36,-54.85 104.72,-40.7\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"103.31\" cy=\"-38.82\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"122.39\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port3</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108c55190>"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "connected_process_spec = {\n",
    "    'process1': {\n",
    "        '_type': 'python_process',\n",
    "        '_ports': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'int',\n",
    "            'port3': 'int',\n",
    "        },\n",
    "        'wires': {\n",
    "            'port1': 'node1',\n",
    "            'port2': 'node2',\n",
    "            'port3': 'node3',\n",
    "        }\n",
    "    },\n",
    "    'node1': 'float',\n",
    "    'node2': 'int',\n",
    "    'node3': 'int',\n",
    "}\n",
    "# report = validate_schema(connected_process_spec)  # validate it follows the schema types\n",
    "# print(report)\n",
    "plot_bigraph(connected_process_spec, **plot_settings, filename='wires')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e037285d-116a-404d-96ec-ba47318a2e8b",
   "metadata": {},
   "source": [
    "### Advanced wiring\n",
    "**TODO**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b17201-2489-45c8-a06b-9fdc29a714a7",
   "metadata": {},
   "source": [
    "## Composites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf5d428-db23-4985-80ff-6b93d22973aa",
   "metadata": {},
   "source": [
    "### Flat composite\n",
    "A flat composite connects stores with multiple connected processes, without nesting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "781a84e6-bbe2-4708-b441-47efc611b46f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"172pt\" height=\"143pt\"\n",
       " viewBox=\"0.00 0.00 176.63 146.63\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 142.63)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-142.63 172.63,-142.63 172.63,4 -4,4\"/>\n",
       "<!-- (&#39;store1.1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1.1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"140.32\" cy=\"-110.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"121.32\" y=\"-107.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-123.32 0,-123.32 0,-87.32 54,-87.32 54,-123.32\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-102.22\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.23,-109.06C95.58,-108.35 76.05,-107.48 59.84,-106.75\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"57.43\" cy=\"-106.64\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-110.32\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-57.32 0,-57.32 0,-21.32 54,-21.32 54,-57.32\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-36.22\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M115.71,-95.26C99.1,-84.66 76.7,-70.38 58.7,-58.9\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"56.81\" cy=\"-57.69\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-83.32\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1.2&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"140.32\" cy=\"-28.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"121.32\" y=\"-25.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M112.26,-35.41C99.33,-39.61 84.09,-45.8 72,-54.32 61.03,-62.05 51.05,-73.1 43.38,-82.93\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"41.97\" cy=\"-84.79\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-57.32\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M112.94,-18.53C100.45,-15.19 85.37,-13.03 72,-16.32 67.7,-17.37 63.35,-18.93 59.14,-20.75\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"56.92\" cy=\"-21.77\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-19.32\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d40af0>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flat_composite_spec = {\n",
    "    'store1.1': 'float',\n",
    "    'store1.2': 'int',\n",
    "    'process1': {\n",
    "        '_type': 'python_process:temporal',\n",
    "        'wires': {\n",
    "            'port1': 'store1.1',\n",
    "            'port2': 'store1.2',\n",
    "        }\n",
    "    },\n",
    "    'process2': {\n",
    "        '_type': 'python_process:temporal',\n",
    "        'wires': {\n",
    "            'port1': 'store1.1',\n",
    "            'port2': 'store1.2',\n",
    "        }\n",
    "    },\n",
    "}\n",
    "# validate_schema(flat_composite_spec)  # validate it follows the schema types\n",
    "plot_bigraph(flat_composite_spec, **plot_settings, rankdir='RL', filename='flat_composite')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398509c7-cf5c-4cce-a179-3a39752b0dde",
   "metadata": {},
   "source": [
    "### Nested composite\n",
    "A nested composite has store hierarchies with multiple connected processes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "701e759d-7f9b-41bb-911c-a8208c892462",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"281pt\" height=\"234pt\"\n",
       " viewBox=\"0.00 0.00 289.00 240.19\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 236.19)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-236.19 285,-236.19 285,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143\" cy=\"-209.41\" rx=\"22.56\" ry=\"22.56\"/>\n",
       "<text text-anchor=\"start\" x=\"128.5\" y=\"-206.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"124\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M143,-186.12C143,-172.48 143,-154.91 143,-140.34\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"68\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"49\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M129.24,-190.79C116.9,-174.96 98.71,-151.67 85.37,-134.57\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;process1&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-36 0,-36 0,0 54,0 54,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<!-- (&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;process2&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"189,-36 135,-36 135,0 189,0 189,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"141\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<!-- (&#39;process3&#39;,) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;process3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"281,-129.32 227,-129.32 227,-93.32 281,-93.32 281,-129.32\"/>\n",
       "<text text-anchor=\"start\" x=\"233\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M164.1,-199.34C178.84,-192.35 198.41,-181.67 213,-168.63 224.09,-158.73 233.98,-145.36 241.27,-134.13\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"242.49\" cy=\"-132.2\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"223\" y=\"-160.63\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M120.63,-92.71C101.42,-77.59 73.69,-55.75 53.39,-39.78\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"51.66\" cy=\"-38.41\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"85\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M148.7,-82.92C151.44,-69.75 154.69,-54.15 157.29,-41.66\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"157.77\" cy=\"-39.31\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"155\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M46.64,-92.11C39.32,-84.55 31.97,-75.17 28,-65 25.11,-57.61 24.31,-49.02 24.42,-41.28\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"24.5\" cy=\"-39.09\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"28\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M87.93,-90.95C103.27,-76.06 124.43,-55.5 140.2,-40.17\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"141.89\" cy=\"-38.53\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"124\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108ce70a0>"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nested_composite_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': 'float',\n",
    "        'store1.2': 'int',\n",
    "        'process1': {\n",
    "            '_type': 'python_process:temporal',\n",
    "            'wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'python_process:temporal',\n",
    "            'wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "    },\n",
    "    'process3': {\n",
    "        '_type': 'python_process:temporal',\n",
    "        'wires': {\n",
    "            'port1': 'store1',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "# validate_schema(nested_composite_spec)  # validate it follows the schema types\n",
    "plot_bigraph(nested_composite_spec, **plot_settings, filename='nested_composite')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c8c1704-2331-44eb-bd05-72e36ccb6858",
   "metadata": {},
   "source": [
    "### Composite process\n",
    "Composite processes are processes with internal stores and internal processes, which can run as their own simulation. The composite process has schema, allowing it to connect to a super-simulation. This allows for improved distributed computation, with composite processes able to run on their own computer and syncronize to the super-simulation via message passing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "638c5dc7-4cfd-46e4-b66a-8c42ca98ff1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"201pt\" height=\"313pt\"\n",
       " viewBox=\"0.00 0.00 206.62 321.63\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 317.63)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-317.63 202.62,-317.63 202.62,4 -4,4\"/>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"124.82\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"105.82\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"192.82,-36 138.82,-36 138.82,0 192.82,0 192.82,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"144.82\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M136.24,-84.88C142.41,-71.13 149.96,-54.31 155.89,-41.11\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"156.81\" cy=\"-39.07\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"148.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"57.82,-36 3.82,-36 3.82,0 57.82,0 57.82,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"9.82\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M116.64,-83.46C112.53,-73.31 106.72,-62.24 98.82,-54 89,-43.75 75.53,-35.96 63.17,-30.36\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"60.96\" cy=\"-29.4\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"106.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"49.82\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"30.82\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M55.99,-83.01C59.47,-72.8 64.8,-61.81 72.82,-54 89.38,-37.9 114.05,-29.06 133.82,-24.3\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"135.78\" cy=\"-23.85\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M37.9,-84.92C35.43,-78.55 33.17,-71.64 31.82,-65 30.27,-57.35 29.76,-48.84 29.72,-41.24\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"29.73\" cy=\"-39.14\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"31.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"150.82,-226.63 52.82,-226.63 52.82,-190.63 150.82,-190.63 150.82,-226.63\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"154.82,-230.63 48.82,-230.63 48.82,-186.63 154.82,-186.63 154.82,-230.63\"/>\n",
       "<text text-anchor=\"start\" x=\"56.82\" y=\"-205.53\" font-family=\"Times,serif\" font-size=\"12.00\">composite_process</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M154.92,-185.85C161.25,-181.1 166.87,-175.41 170.82,-168.63 179.85,-153.15 164.71,-137.52 149.44,-126.66\"/>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M108.16,-181.36C111.29,-168.4 115.06,-152.77 118.22,-139.67\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"107.63\" cy=\"-183.57\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"112.82\" y=\"-160.63\" font-family=\"Times,serif\" font-size=\"10.00\">tunnel port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M90.02,-186.01C82.05,-171.39 71.51,-152.08 63.22,-136.88\"/>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M44.17,-190.03C34.89,-184.64 26.49,-177.65 20.82,-168.63 13.39,-156.81 20.47,-142.86 29.49,-131.8\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"46\" cy=\"-191.03\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"20.82\" y=\"-160.63\" font-family=\"Times,serif\" font-size=\"10.00\">tunnel port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port1&#39;)&#45;&gt;(&#39;composite_process&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;port1&#39;)&#45;&gt;(&#39;composite_process&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M78.33,-278.2C77.98,-269.3 78.29,-258.21 80.82,-248.63 81.94,-244.37 83.62,-240.07 85.53,-235.95\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"86.59\" cy=\"-233.81\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"80.82\" y=\"-251.63\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port2&#39;)&#45;&gt;(&#39;composite_process&#39;,) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;port2&#39;)&#45;&gt;(&#39;composite_process&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M123.45,-289.57C120.67,-279.29 114.19,-255.33 109.04,-236.31\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"108.42\" cy=\"-234.01\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"113.82\" y=\"-251.63\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d12250>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "composite_process_spec = {\n",
    "    'composite_process': {\n",
    "        'store1.1': 'Any',\n",
    "        'store1.2': 'Any',\n",
    "        'process1': {\n",
    "            '_type': 'python_process:temporal',\n",
    "            'wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'python_process:temporal',\n",
    "            'wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        '_ports': {\n",
    "            'port1': 'Any', \n",
    "            'port2': 'Any', \n",
    "        },\n",
    "        '_tunnels': {\n",
    "            'port1': 'store1.1',\n",
    "            'port2': 'store1.2',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "# validate_schema(composite_process_spec)  # validate it follows the schema types\n",
    "plot_bigraph(composite_process_spec, **plot_settings, filename='composite_process')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "137bf506-d71c-447b-9bf3-7f310c26a5c5",
   "metadata": {},
   "source": [
    "## Updates\n",
    "\n",
    "- state update\n",
    "- structure update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f28a8137-280e-4cd5-bfc3-d0d512684283",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "schemaa = {\n",
    "    'base': {\n",
    "        'a': {'_type': 'int'},\n",
    "    },\n",
    "\n",
    "    # cannot override existing keys unless it is of a subtype\n",
    "    'second': {\n",
    "        'b': {'_type': 'int'},\n",
    "        '_super': 'base',\n",
    "    },\n",
    "}\n",
    "\n",
    "types.type_registry.register_multiple(schemaa, force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1506f428-649d-4c38-aa47-9d85132faa99",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"119pt\" height=\"55pt\"\n",
       " viewBox=\"0.00 0.00 122.08 56.08\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 52.08)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-52.08 118.08,-52.08 118.08,4 -4,4\"/>\n",
       "<!-- (&#39;a&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;a&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"24.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"21.04\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">a</text>\n",
       "<text text-anchor=\"start\" x=\"17.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">11</text>\n",
       "</g>\n",
       "<!-- (&#39;b&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;b&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"90.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"86.54\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">b</text>\n",
       "<text text-anchor=\"start\" x=\"83.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">13</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d128b0>"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema = {'_type': 'second'}\n",
    "state = {\n",
    "    'a': 11,\n",
    "    'b': 13,\n",
    "}\n",
    "plot_bigraph(state, show_values=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "36f1fa08-5bb2-43c6-b9d8-3cba63646def",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"119pt\" height=\"55pt\"\n",
       " viewBox=\"0.00 0.00 122.08 56.08\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 52.08)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-52.08 118.08,-52.08 118.08,4 -4,4\"/>\n",
       "<!-- (&#39;a&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;a&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"24.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"21.04\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">a</text>\n",
       "<text text-anchor=\"start\" x=\"17.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">11</text>\n",
       "</g>\n",
       "<!-- (&#39;b&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;b&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"90.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"86.54\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">b</text>\n",
       "<text text-anchor=\"start\" x=\"86.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">8</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d12d90>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "update = {\n",
    "    'b': -5\n",
    "}\n",
    "new_state = types.apply(\n",
    "    schema,\n",
    "    state,\n",
    "    update\n",
    ")\n",
    "plot_bigraph(new_state, show_values=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "aa40b177-e9c6-4722-b996-1c1dba40eca4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"119pt\" height=\"55pt\"\n",
       " viewBox=\"0.00 0.00 122.08 56.08\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 52.08)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-52.08 118.08,-52.08 118.08,4 -4,4\"/>\n",
       "<!-- (&#39;a&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;a&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"24.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"21.04\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">a</text>\n",
       "<text text-anchor=\"start\" x=\"17.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">16</text>\n",
       "</g>\n",
       "<!-- (&#39;b&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;b&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"90.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"86.54\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">b</text>\n",
       "<text text-anchor=\"start\" x=\"86.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">8</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d126d0>"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "update = {\n",
    "    'a': 5\n",
    "}\n",
    "new_state = types.apply(\n",
    "    schema,\n",
    "    new_state,\n",
    "    update\n",
    ")\n",
    "plot_bigraph(new_state, show_values=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "be2541ae-d8a7-4926-8a18-6f9127f0744a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from bigraph_schema.registry import apply_registry\n",
    "from bigraph_viz.dict_utils import deep_merge\n",
    "def apply_merge(current, update, bindings=None, types=None):\n",
    "    assert isinstance(state, dict)\n",
    "    assert isinstance(update, dict)\n",
    "    updated = deep_merge(state, update)\n",
    "    return updated\n",
    "types.apply_registry.register('merge', apply_merge, force=True)\n",
    "\n",
    "schemab = {\n",
    "    'merge_branch': {\n",
    "        '_default': {'a1': 'any'},\n",
    "        '_apply': 'merge',\n",
    "        \n",
    "    },\n",
    "\n",
    "    # cannot override existing keys unless it is of a subtype\n",
    "    'second': {\n",
    "        'a1': {'a2': 'any'},\n",
    "        '_super': 'merge_branch',\n",
    "    },\n",
    "}\n",
    "\n",
    "types.type_registry.register_multiple(schemab, force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "ed5c34bf-b357-4800-b126-73a32b6646c2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"55pt\" height=\"125pt\"\n",
       " viewBox=\"0.00 0.00 56.08 128.08\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 124.08)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-124.08 52.08,-124.08 52.08,4 -4,4\"/>\n",
       "<!-- (&#39;a1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;a1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"24.04\" cy=\"-102.08\" rx=\"18\" ry=\"18\"/>\n",
       "<text text-anchor=\"start\" x=\"18.04\" y=\"-98.98\" font-family=\"Times,serif\" font-size=\"12.00\">a1</text>\n",
       "</g>\n",
       "<!-- (&#39;a1&#39;, &#39;a2&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;a1&#39;, &#39;a2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"24.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"18.04\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">a2</text>\n",
       "<text text-anchor=\"start\" x=\"17.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">11</text>\n",
       "</g>\n",
       "<!-- (&#39;a1&#39;,)&#45;&gt;(&#39;a1&#39;, &#39;a2&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;a1&#39;,)&#45;&gt;(&#39;a1&#39;, &#39;a2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M24.04,-83.45C24.04,-73.18 24.04,-60.1 24.04,-48.92\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108c55670>"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema = {'_type': 'merge_branch'}\n",
    "state = {\n",
    "    'a1': {'a2': 11},\n",
    "}\n",
    "plot_bigraph(state, show_values=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "072c2502-6fdd-4c10-af91-cf00e4de1656",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"119pt\" height=\"125pt\"\n",
       " viewBox=\"0.00 0.00 122.08 128.08\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 124.08)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-124.08 118.08,-124.08 118.08,4 -4,4\"/>\n",
       "<!-- (&#39;a1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;a1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"57.04\" cy=\"-102.08\" rx=\"18\" ry=\"18\"/>\n",
       "<text text-anchor=\"start\" x=\"51.04\" y=\"-98.98\" font-family=\"Times,serif\" font-size=\"12.00\">a1</text>\n",
       "</g>\n",
       "<!-- (&#39;a1&#39;, &#39;a2&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;a1&#39;, &#39;a2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"24.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"18.04\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">a2</text>\n",
       "<text text-anchor=\"start\" x=\"17.54\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">11</text>\n",
       "</g>\n",
       "<!-- (&#39;a1&#39;,)&#45;&gt;(&#39;a1&#39;, &#39;a2&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;a1&#39;,)&#45;&gt;(&#39;a1&#39;, &#39;a2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M49.88,-84.58C45.1,-73.58 38.77,-58.98 33.57,-46.99\"/>\n",
       "</g>\n",
       "<!-- (&#39;a1&#39;, &#39;a3&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;a1&#39;, &#39;a3&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"90.04\" cy=\"-24.04\" rx=\"24.08\" ry=\"24.08\"/>\n",
       "<text text-anchor=\"start\" x=\"84.04\" y=\"-27.44\" font-family=\"Times,serif\" font-size=\"12.00\">a3</text>\n",
       "<text text-anchor=\"start\" x=\"85.04\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">&#45;5</text>\n",
       "</g>\n",
       "<!-- (&#39;a1&#39;,)&#45;&gt;(&#39;a1&#39;, &#39;a3&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;a1&#39;,)&#45;&gt;(&#39;a1&#39;, &#39;a3&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M64.2,-84.58C68.98,-73.58 75.31,-58.98 80.51,-46.99\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d4d0d0>"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "update = {\n",
    "    'a1': {'a3': -5},\n",
    "}\n",
    "new_state = types.apply(\n",
    "    schema,\n",
    "    state,\n",
    "    update\n",
    ")\n",
    "# print(pf(new_state))\n",
    "plot_bigraph(new_state, show_values=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "847cef26-65a1-4da6-9dd3-eb5839277970",
   "metadata": {},
   "source": [
    "## Orchestration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "e76c97a8-3037-464a-929e-ba4414c73c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_settings2={}\n",
    "if save_images:\n",
    "    plot_settings2.update({\n",
    "        'out_dir': 'out',\n",
    "        'dpi': '250'\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48975890-06a4-462a-9b1d-6bddd51779fe",
   "metadata": {},
   "source": [
    "### Multi-timestepping\n",
    "\n",
    "Temporal processes can be run by multi-timestepping, with each process updating the state according to its `sync_state` state, which it can update adaptively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f0eff7b0-f0e2-4d56-8b16-ef0fc1ce5aba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: flow Pages: 1 -->\n",
       "<svg width=\"231pt\" height=\"165pt\"\n",
       " viewBox=\"0.00 0.00 238.00 170.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 166)\">\n",
       "<title>flow</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-166 234,-166 234,4 -4,4\"/>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"65.01,-38 0.01,-38 0.01,0 65.01,0 65.01,-38\"/>\n",
       "<text text-anchor=\"start\" x=\"8.01\" y=\"-22.8\" font-family=\"Times,serif\" font-size=\"14.00\">temporal</text>\n",
       "<text text-anchor=\"start\" x=\"8.51\" y=\"-7.8\" font-family=\"Times,serif\" font-size=\"14.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"131.43\" cy=\"-19\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"131.43\" y=\"-16.8\" font-family=\"Times,serif\" font-size=\"9.00\">1.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M66,-19C85.73,-19 109.26,-19 122.04,-19\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"213.87\" cy=\"-19\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"213.87\" y=\"-16.8\" font-family=\"Times,serif\" font-size=\"9.00\">2.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M140.89,-19C156.79,-19 188.7,-19 204.52,-19\"/>\n",
       "</g>\n",
       "<!-- (&#39;state&#39;,) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;state&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"161.54\" cy=\"-90.61\" rx=\"21.4\" ry=\"21.4\"/>\n",
       "<text text-anchor=\"start\" x=\"149.04\" y=\"-86.91\" font-family=\"Times,serif\" font-size=\"14.00\">state</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M137.16,-32.62C140.99,-41.74 146.16,-54.02 150.75,-64.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"138.85,-32.13 135.3,-28.19 135.62,-33.48 138.85,-32.13\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"149.1,-65.54 152.65,-69.47 152.33,-64.19 149.1,-65.54\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) end -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) end</title>\n",
       "<ellipse fill=\"none\" stroke=\"none\" stroke-width=\"2\" cx=\"230.36\" cy=\"-19\" rx=\"0\" ry=\"0\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) end -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) end</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M223.27,-19C223.43,-19 223.59,-19 223.74,-19\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"221.43,-20.75 226.43,-19 221.43,-17.25 221.43,-20.75\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M205.18,-30.89C197.74,-41.07 186.84,-55.99 177.79,-68.38\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"206.48,-32.07 208.02,-27.01 203.66,-30.01 206.48,-32.07\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"176.53,-67.13 175,-72.2 179.36,-69.19 176.53,-67.13\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"65.01,-161.65 0.01,-161.65 0.01,-123.65 65.01,-123.65 65.01,-161.65\"/>\n",
       "<text text-anchor=\"start\" x=\"8.01\" y=\"-146.45\" font-family=\"Times,serif\" font-size=\"14.00\">temporal</text>\n",
       "<text text-anchor=\"start\" x=\"8.51\" y=\"-131.45\" font-family=\"Times,serif\" font-size=\"14.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"90.22\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"90.22\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">0.5</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M65.75,-142.65C71.36,-142.65 76.74,-142.65 81.03,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"131.43\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"131.43\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">1.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M99.36,-142.65C106.1,-142.65 115.23,-142.65 122.03,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M102.33,-133.81C112.43,-126.45 127.09,-115.75 139.29,-106.85\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"101.47,-132.28 98.46,-136.64 103.53,-135.1 101.47,-132.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"140.04,-108.46 143.05,-104.1 137.98,-105.64 140.04,-108.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"172.65\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"172.65\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">1.5</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M140.58,-142.65C147.32,-142.65 156.45,-142.65 163.24,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M138.88,-129.78C141.56,-125.14 144.68,-119.75 147.73,-114.48\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"137.51,-128.66 136.52,-133.87 140.54,-130.41 137.51,-128.66\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"149.13,-115.56 150.12,-110.35 146.1,-113.8 149.13,-115.56\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"213.87\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"213.87\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">2.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M181.8,-142.65C188.53,-142.65 197.67,-142.65 204.46,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M169.55,-128.14C168.86,-124.88 168.1,-121.32 167.33,-117.73\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"167.8,-128.32 170.56,-132.85 171.22,-127.59 167.8,-128.32\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"169.04,-117.37 166.29,-112.85 165.62,-118.1 169.04,-117.37\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) end -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) end</title>\n",
       "<ellipse fill=\"none\" stroke=\"none\" stroke-width=\"2\" cx=\"230.36\" cy=\"-142.65\" rx=\"0\" ry=\"0\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) end -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) end</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M223.27,-142.65C223.43,-142.65 223.59,-142.65 223.74,-142.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"221.43,-144.4 226.43,-142.65 221.43,-140.9 221.43,-144.4\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M203.57,-132.41C197.06,-125.94 188.45,-117.38 180.72,-109.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"202.14,-133.45 206.92,-135.74 204.6,-130.97 202.14,-133.45\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"182.33,-108.82 177.55,-106.53 179.86,-111.3 182.33,-108.82\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d12520>"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multitimestep_spec = {\n",
    "    'temporal process2': {\n",
    "        '_ports': {'port1': 'Any'},\n",
    "        'wires': {'port1': 'state'},\n",
    "        '_sync_step': 1.0,\n",
    "    },\n",
    "    'temporal process1': {\n",
    "        '_ports': {'port1': 'Any'},\n",
    "        'wires': {'port1': 'state'},\n",
    "        '_sync_step': 0.5,\n",
    "    },\n",
    "}\n",
    "multitimestep_spec = replace_regex_recursive(multitimestep_spec)\n",
    "plot_multitimestep(multitimestep_spec, total_time=2.0, **plot_settings2, filename='multitimestep')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a00cf677-ebac-4309-8ac9-a5b2855724af",
   "metadata": {},
   "source": [
    "### Flows\n",
    "\n",
    "A directed acyclic graph (DAG) – a type of workflow that declares the dependencies between processes. A flow run is not temporal, and the order is entirely based on dependencies. It can triggered between time steps, running the steps in the order as determined by their dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "db45c99b-b97b-47bc-a73e-c669882f3969",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: flow Pages: 1 -->\n",
       "<svg width=\"130pt\" height=\"183pt\"\n",
       " viewBox=\"0.00 0.00 134.00 188.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 184)\">\n",
       "<title>flow</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-184 130,-184 130,4 -4,4\"/>\n",
       "<!-- (&#39;step1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;step1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-180 0,-180 0,-144 54,-144 54,-180\"/>\n",
       "<text text-anchor=\"start\" x=\"12.5\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">step1</text>\n",
       "</g>\n",
       "<!-- (&#39;step2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;step2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-108 0,-108 0,-72 54,-72 54,-108\"/>\n",
       "<text text-anchor=\"start\" x=\"12.5\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">step2</text>\n",
       "</g>\n",
       "<!-- (&#39;step1&#39;,)&#45;&gt;(&#39;step2&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;step1&#39;,)&#45;&gt;(&#39;step2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M27,-143.34C27,-136.75 27,-129.08 27,-121.67\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"30.5,-121.93 27,-111.93 23.5,-121.93 30.5,-121.93\"/>\n",
       "</g>\n",
       "<!-- (&#39;step4&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;step4&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"90,-36 36,-36 36,0 90,0 90,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"48.5\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">step4</text>\n",
       "</g>\n",
       "<!-- (&#39;step2&#39;,)&#45;&gt;(&#39;step4&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;step2&#39;,)&#45;&gt;(&#39;step4&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M36.08,-71.34C39.68,-64.34 43.91,-56.12 47.94,-48.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"50.95,-50.09 52.41,-39.59 44.72,-46.89 50.95,-50.09\"/>\n",
       "</g>\n",
       "<!-- (&#39;step3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;step3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"126,-108 72,-108 72,-72 126,-72 126,-108\"/>\n",
       "<text text-anchor=\"start\" x=\"84.5\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">step3</text>\n",
       "</g>\n",
       "<!-- (&#39;step3&#39;,)&#45;&gt;(&#39;step4&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;step3&#39;,)&#45;&gt;(&#39;step4&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M89.92,-71.34C86.32,-64.34 82.09,-56.12 78.06,-48.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"81.28,-46.89 73.59,-39.59 75.05,-50.09 81.28,-46.89\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d407f0>"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "step_schema = {\n",
    "    '_depends_on': 'list[string]',\n",
    "}\n",
    "types.type_registry.register('step_process', step_schema)\n",
    "\n",
    "flow = {\n",
    "    'step1': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': [], \n",
    "    },\n",
    "    'step2': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': 'step1', \n",
    "    },\n",
    "    'step3': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': [], \n",
    "    },\n",
    "    'step4': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': ['step2', 'step3'], \n",
    "    },\n",
    "}\n",
    "\n",
    "plot_flow(flow, **plot_settings2, filename='flow')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5882581a-c5ef-40b5-acbd-efddad269807",
   "metadata": {},
   "source": [
    "## Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff81178f-36e4-4200-982a-ed8155b87934",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Multiscale map of a cell\n",
    "\n",
    "Bigraphs can represent the hierarchical nesting of cellular structures. Below is a multiscale map of a cell from [Qin et al. A multi-scale map of cell structure fusing protein images and interactions. Nature. 2021.](https://pubmed.ncbi.nlm.nih.gov/34819669/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "47bdf8e5-0bdd-4cac-b57e-2449c75fa178",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"582pt\" height=\"700pt\"\n",
       " viewBox=\"0.00 0.00 598.78 720.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.547089 0.547089) rotate(0) translate(4 1275.5)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-1275.5 1060.09,-1275.5 1060.09,4 -4,4\"/>\n",
       "<!-- (&#39;cell&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;cell&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"423.28\" cy=\"-1253.5\" rx=\"18\" ry=\"18\"/>\n",
       "<text text-anchor=\"start\" x=\"414.78\" y=\"-1250.4\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"286.28\" cy=\"-1164.41\" rx=\"35.18\" ry=\"35.18\"/>\n",
       "<text text-anchor=\"start\" x=\"261.78\" y=\"-1161.31\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M407.83,-1242.68C385.64,-1228.58 344.33,-1202.32 316.09,-1184.36\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"654.28\" cy=\"-1164.41\" rx=\"27.9\" ry=\"27.9\"/>\n",
       "<text text-anchor=\"start\" x=\"635.78\" y=\"-1161.31\" font-family=\"Times,serif\" font-size=\"12.00\">nucleus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M440.64,-1245.95C480.72,-1230.84 579.91,-1193.45 627.7,-1175.43\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"115.28\" cy=\"-1040.29\" rx=\"40.11\" ry=\"40.11\"/>\n",
       "<text text-anchor=\"start\" x=\"93.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">secretory</text>\n",
       "<text text-anchor=\"start\" x=\"90.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M255.25,-1146.46C231.73,-1133.11 199,-1113.45 172.28,-1093.32 162.64,-1086.07 152.74,-1077.48 143.94,-1069.4\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"227.28\" cy=\"-1040.29\" rx=\"45.92\" ry=\"45.92\"/>\n",
       "<text text-anchor=\"start\" x=\"198.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"202.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M271.08,-1131.96C263.78,-1116.84 254.92,-1098.52 247.09,-1082.31\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"344.28\" cy=\"-1040.29\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"310.78\" y=\"-1056.69\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"320.28\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"323.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"328.78\" y=\"-1017.69\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M301.36,-1131.64C307.54,-1118.64 314.82,-1103.31 321.61,-1089.01\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"89.28\" cy=\"-890.44\" rx=\"57.05\" ry=\"57.05\"/>\n",
       "<text text-anchor=\"start\" x=\"52.78\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"68.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"70.78\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M108.38,-1000.06C105.55,-983.98 102.22,-965.06 99.14,-947.56\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"57.28\" cy=\"-728.58\" rx=\"57.05\" ry=\"57.05\"/>\n",
       "<text text-anchor=\"start\" x=\"49.28\" y=\"-744.98\" font-family=\"Times,serif\" font-size=\"12.00\">ion</text>\n",
       "<text text-anchor=\"start\" x=\"20.78\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"36.28\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"38.78\" y=\"-705.98\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M78.08,-833.53C74.97,-817.95 71.58,-801.04 68.46,-785.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"227.28\" cy=\"-890.44\" rx=\"45.92\" ry=\"45.92\"/>\n",
       "<text text-anchor=\"start\" x=\"204.78\" y=\"-906.84\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"222.28\" y=\"-893.84\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"198.78\" y=\"-880.84\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"202.78\" y=\"-867.84\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M227.28,-993.72C227.28,-975.74 227.28,-955.11 227.28,-937.11\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"173.28\" cy=\"-728.58\" rx=\"40.11\" ry=\"40.11\"/>\n",
       "<text text-anchor=\"start\" x=\"150.28\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"148.78\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M212.66,-846.18C204.35,-821.59 194.05,-791.07 186.06,-767.43\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"296.28\" cy=\"-728.58\" rx=\"65.11\" ry=\"65.11\"/>\n",
       "<text text-anchor=\"start\" x=\"254.28\" y=\"-738.48\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleoprotein</text>\n",
       "<text text-anchor=\"start\" x=\"275.78\" y=\"-725.48\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"280.78\" y=\"-712.48\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M245.4,-847.44C253.07,-829.68 262.14,-808.67 270.48,-789.34\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"69.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"46.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"64.28\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"46.28\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"44.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M150.19,-694.94C133.44,-671.25 110.81,-639.24 93.74,-615.11\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"177.28\" cy=\"-579.51\" rx=\"48.03\" ry=\"48.03\"/>\n",
       "<text text-anchor=\"start\" x=\"142.28\" y=\"-576.41\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondrion</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M174.36,-687.78C174.86,-669.49 175.45,-647.65 175.97,-628.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"292.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"278.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"266.78\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"271.78\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"288.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M294.51,-662.64C294.14,-649.1 293.76,-635.16 293.42,-622.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"396.28\" cy=\"-579.51\" rx=\"43.68\" ry=\"43.68\"/>\n",
       "<text text-anchor=\"start\" x=\"374.28\" y=\"-589.41\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "<text text-anchor=\"start\" x=\"371.28\" y=\"-576.41\" font-family=\"Times,serif\" font-size=\"12.00\">biogenesis</text>\n",
       "<text text-anchor=\"start\" x=\"369.28\" y=\"-563.41\" font-family=\"Times,serif\" font-size=\"12.00\">community</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M332.94,-673.65C345.88,-654.63 360.09,-633.73 371.77,-616.55\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"292.28\" cy=\"-453.07\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"278.78\" y=\"-469.47\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"273.78\" y=\"-456.47\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"271.78\" y=\"-443.47\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"288.78\" y=\"-430.47\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M292.28,-536.38C292.28,-523.41 292.28,-509.13 292.28,-496.16\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"390.28\" cy=\"-453.07\" rx=\"32.02\" ry=\"32.02\"/>\n",
       "<text text-anchor=\"start\" x=\"368.28\" y=\"-449.97\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M394.18,-534.99C393.4,-518.87 392.53,-500.9 391.82,-486\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"482.28\" cy=\"-453.07\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"465.28\" y=\"-469.47\" font-family=\"Times,serif\" font-size=\"12.00\">nucleic</text>\n",
       "<text text-anchor=\"start\" x=\"472.28\" y=\"-456.47\" font-family=\"Times,serif\" font-size=\"12.00\">acid</text>\n",
       "<text text-anchor=\"start\" x=\"463.78\" y=\"-443.47\" font-family=\"Times,serif\" font-size=\"12.00\">binding</text>\n",
       "<text text-anchor=\"start\" x=\"461.78\" y=\"-430.47\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M421.15,-542.52C432.86,-525.58 446.76,-505.46 458.34,-488.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"390.28\" cy=\"-335.75\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"366.78\" y=\"-339.15\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"372.78\" y=\"-326.15\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M390.28,-420.25C390.28,-406.39 390.28,-390 390.28,-375.44\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"161.28\" cy=\"-200.05\" rx=\"54.39\" ry=\"54.39\"/>\n",
       "<text text-anchor=\"start\" x=\"126.78\" y=\"-216.45\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondirial</text>\n",
       "<text text-anchor=\"start\" x=\"149.28\" y=\"-203.45\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n",
       "<text text-anchor=\"start\" x=\"137.78\" y=\"-190.45\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"143.78\" y=\"-177.45\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M353.14,-322.37C318.22,-309.8 265.32,-288.29 224.28,-260.86 215.5,-254.99 206.81,-247.86 198.82,-240.6\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"272.28\" cy=\"-200.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"249.28\" y=\"-209.95\" font-family=\"Times,serif\" font-size=\"12.00\">mito&#45;cyto</text>\n",
       "<text text-anchor=\"start\" x=\"248.78\" y=\"-196.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"256.78\" y=\"-183.95\" font-family=\"Times,serif\" font-size=\"12.00\">cluster</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M363.13,-307.18C349.93,-293.5 333.98,-276.57 320.28,-260.86 312.14,-251.53 303.58,-241.08 295.97,-231.57\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"390.28\" cy=\"-200.05\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"355.78\" y=\"-229.45\" font-family=\"Times,serif\" font-size=\"12.00\">cotranslational</text>\n",
       "<text text-anchor=\"start\" x=\"373.28\" y=\"-216.45\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n",
       "<text text-anchor=\"start\" x=\"369.28\" y=\"-203.45\" font-family=\"Times,serif\" font-size=\"12.00\">targeting</text>\n",
       "<text text-anchor=\"start\" x=\"385.28\" y=\"-190.45\" font-family=\"Times,serif\" font-size=\"12.00\">to</text>\n",
       "<text text-anchor=\"start\" x=\"365.28\" y=\"-177.45\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "<text text-anchor=\"start\" x=\"373.78\" y=\"-164.45\" font-family=\"Times,serif\" font-size=\"12.00\">system</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M390.28,-296.08C390.28,-285.28 390.28,-273.28 390.28,-261.55\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"508.28\" cy=\"-200.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"484.78\" y=\"-209.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"487.78\" y=\"-196.95\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"504.78\" y=\"-183.95\" font-family=\"Times,serif\" font-size=\"12.00\">5</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M417.42,-307.18C430.63,-293.5 446.57,-276.57 460.28,-260.86 468.42,-251.53 476.98,-241.08 484.58,-231.57\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"604.28\" cy=\"-200.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"580.78\" y=\"-209.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"583.78\" y=\"-196.95\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"600.78\" y=\"-183.95\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M427.95,-323.71C463.5,-312.04 517.05,-291.09 556.28,-260.86 566.19,-253.22 575.29,-243.13 582.85,-233.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"161.28\" cy=\"-51.62\" rx=\"51.74\" ry=\"51.74\"/>\n",
       "<text text-anchor=\"start\" x=\"149.28\" y=\"-74.52\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n",
       "<text text-anchor=\"start\" x=\"137.78\" y=\"-61.52\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"143.78\" y=\"-48.52\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "<text text-anchor=\"start\" x=\"132.28\" y=\"-35.52\" font-family=\"Times,serif\" font-size=\"12.00\">subcomplex</text>\n",
       "<text text-anchor=\"start\" x=\"157.78\" y=\"-22.52\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M161.28,-144.94C161.28,-131.58 161.28,-117.3 161.28,-104.07\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"344.28\" cy=\"-890.44\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"310.78\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"320.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"318.78\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M344.28,-986.33C344.28,-972.58 344.28,-957.77 344.28,-944.03\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"488.28\" cy=\"-1040.29\" rx=\"38.37\" ry=\"38.37\"/>\n",
       "<text text-anchor=\"start\" x=\"474.28\" y=\"-1050.19\" font-family=\"Times,serif\" font-size=\"12.00\">DNA</text>\n",
       "<text text-anchor=\"start\" x=\"465.28\" y=\"-1037.19\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"466.28\" y=\"-1024.19\" font-family=\"Times,serif\" font-size=\"12.00\">assembly</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M628.97,-1151.12C604.1,-1138.43 565.71,-1117.17 536.28,-1093.32 528.06,-1086.67 519.95,-1078.61 512.8,-1070.85\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"575.28\" cy=\"-1040.29\" rx=\"30.31\" ry=\"30.31\"/>\n",
       "<text text-anchor=\"start\" x=\"557.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"560.28\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">lumen</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M639.04,-1139.86C625.41,-1118.79 605.49,-1087.99 591.55,-1066.44\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"654.28\" cy=\"-1040.29\" rx=\"30.31\" ry=\"30.31\"/>\n",
       "<text text-anchor=\"start\" x=\"636.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"641.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">body</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M654.28,-1135.7C654.28,-1116.56 654.28,-1091.11 654.28,-1071.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"753.28\" cy=\"-1040.29\" rx=\"50.82\" ry=\"50.82\"/>\n",
       "<text text-anchor=\"start\" x=\"721.28\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"749.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M671.97,-1141.58C685.66,-1124.7 704.96,-1100.89 721.32,-1080.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"873.28\" cy=\"-1040.29\" rx=\"50.82\" ry=\"50.82\"/>\n",
       "<text text-anchor=\"start\" x=\"841.28\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"869.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M681.1,-1154.39C713.73,-1142.8 769.83,-1120.85 813.28,-1093.32 821.08,-1088.38 828.91,-1082.46 836.21,-1076.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"995.28\" cy=\"-1040.29\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"977.78\" y=\"-1050.19\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"961.78\" y=\"-1037.19\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"977.28\" y=\"-1024.19\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M682.67,-1160.54C734.83,-1154.37 848.32,-1136.71 933.28,-1093.32 941.41,-1089.17 949.41,-1083.77 956.82,-1078.02\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.28\" cy=\"-890.44\" rx=\"33.49\" ry=\"33.49\"/>\n",
       "<text text-anchor=\"start\" x=\"538.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">nucleolus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M572.44,-1009.37C570.12,-984.82 566.83,-950.08 564.42,-924.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"645.28\" cy=\"-890.44\" rx=\"33.47\" ry=\"33.47\"/>\n",
       "<text text-anchor=\"start\" x=\"627.78\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"626.78\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"627.28\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M588.28,-1011.83C600.42,-986.19 618.49,-948.02 631.02,-921.56\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"440.28\" cy=\"-728.58\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"401.28\" y=\"-738.48\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleprotein</text>\n",
       "<text text-anchor=\"start\" x=\"419.78\" y=\"-725.48\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"424.78\" y=\"-712.48\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M541.02,-862.69C523.53,-839.58 497.82,-805.61 476.76,-777.78\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.28\" cy=\"-728.58\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"547.78\" y=\"-744.98\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"535.78\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"540.78\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"545.78\" y=\"-705.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M561.28,-856.32C561.28,-831.81 561.28,-798.23 561.28,-771.8\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"667.28\" cy=\"-728.58\" rx=\"45.01\" ry=\"45.01\"/>\n",
       "<text text-anchor=\"start\" x=\"639.28\" y=\"-738.48\" font-family=\"Times,serif\" font-size=\"12.00\">splicosomal</text>\n",
       "<text text-anchor=\"start\" x=\"646.78\" y=\"-725.48\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"651.78\" y=\"-712.48\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M579.74,-861.6C597.12,-835.39 623.27,-795.94 642.47,-766.99\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"531.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"517.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"505.78\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"510.78\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"527.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M552.84,-686.22C548.73,-666.09 543.81,-641.94 539.7,-621.81\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"634.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"620.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"615.78\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"613.78\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"618.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M580.07,-689.71C590.93,-667.83 604.55,-640.41 615.41,-618.51\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"763.28\" cy=\"-890.44\" rx=\"48.58\" ry=\"48.58\"/>\n",
       "<text text-anchor=\"start\" x=\"732.78\" y=\"-906.84\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n",
       "<text text-anchor=\"start\" x=\"733.78\" y=\"-893.84\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"742.78\" y=\"-880.84\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"747.78\" y=\"-867.84\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M756.69,-988.82C757.76,-972.93 758.95,-955.42 760.01,-939.69\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"778.28\" cy=\"-728.58\" rx=\"47.25\" ry=\"47.25\"/>\n",
       "<text text-anchor=\"start\" x=\"754.28\" y=\"-744.98\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"748.78\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"757.78\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"762.78\" y=\"-705.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M767.78,-841.48C769.71,-820.88 771.96,-796.89 773.88,-776.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"778.28\" cy=\"-579.51\" rx=\"34.8\" ry=\"34.8\"/>\n",
       "<text text-anchor=\"start\" x=\"765.78\" y=\"-589.41\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"757.78\" y=\"-576.41\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"762.78\" y=\"-563.41\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M778.28,-680.64C778.28,-659.29 778.28,-634.48 778.28,-614.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"778.28\" cy=\"-453.07\" rx=\"34.8\" ry=\"34.8\"/>\n",
       "<text text-anchor=\"start\" x=\"762.78\" y=\"-462.97\" font-family=\"Times,serif\" font-size=\"12.00\">NuA4</text>\n",
       "<text text-anchor=\"start\" x=\"765.78\" y=\"-449.97\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"757.78\" y=\"-436.97\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M778.28,-543.85C778.28,-526.59 778.28,-505.78 778.28,-488.55\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"873.28\" cy=\"-890.44\" rx=\"39.7\" ry=\"39.7\"/>\n",
       "<text text-anchor=\"start\" x=\"849.28\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"849.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"852.78\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M873.28,-988.82C873.28,-969.92 873.28,-948.73 873.28,-930.96\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"995.28\" cy=\"-890.44\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"975.28\" y=\"-919.84\" font-family=\"Times,serif\" font-size=\"12.00\">negative</text>\n",
       "<text text-anchor=\"start\" x=\"971.28\" y=\"-906.84\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"990.28\" y=\"-893.84\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"981.78\" y=\"-880.84\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"966.28\" y=\"-867.84\" font-family=\"Times,serif\" font-size=\"12.00\">biosynthesis</text>\n",
       "<text text-anchor=\"start\" x=\"977.28\" y=\"-854.84\" font-family=\"Times,serif\" font-size=\"12.00\">process</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M995.28,-986.33C995.28,-975.15 995.28,-963.26 995.28,-951.83\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d40910>"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "music_map = {\n",
    "    'cell': {\n",
    "        'cytoplasm': {\n",
    "            'secretory organelles': {\n",
    "                'transmembrane transport systems': {\n",
    "                    'ion transmembrane transport systems': {}}},\n",
    "            'cytoplasmic organelles': {\n",
    "                'subgroup of cytoplasmic organelles': {\n",
    "                    'metabolic organelles': {\n",
    "                        'subgroup of metabolic organelles': {},\n",
    "                        'mitochondrion': {}},\n",
    "                    'ribonucleoprotein complex family': {\n",
    "                        'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
    "                        'ribosome biogenesis community': {\n",
    "                            'ribosome': {\n",
    "                                'ribosomal subunit': {\n",
    "                                    'mitochondirial large ribosomal subunit': {\n",
    "                                        'large ribosomal subunit subcomplex 1': {}},\n",
    "                                    'mito-cyto ribosomal cluster': {},\n",
    "                                    'cotranslational protein targeting to membrane system': {},\n",
    "                                    'ribosomal complex 5': {},\n",
    "                                    'ribosomal complex 2': {}}},\n",
    "                            'nucleic acid binding complex': {}}}}},\n",
    "            'transcriptional regulation complex family': {\n",
    "                'transcriptional regulation complexes': {}}},\n",
    "        'nucleus': {\n",
    "            'DNA metabolic assembly': {},\n",
    "            'nuclear lumen': {\n",
    "                'nucleolus': {\n",
    "                    'ribonucleprotein complex family': {},\n",
    "                    'RNA processing complex family': {\n",
    "                        'RNA processing complex 1': {},\n",
    "                        'RNA splicing complex family': {}},\n",
    "                    'splicosomal complex family': {}},\n",
    "                'nuclear splicing speckle': {}},\n",
    "            'nuclear body': {},\n",
    "            'nucleo-plasm 1': {\n",
    "                'chromosome organization complex family': {\n",
    "                    'chromatin organization complex family': {\n",
    "                        'HAT complex family': {\n",
    "                            'NuA4 HAT complex': {}}}}},\n",
    "            'nucleo-plasm 2': {\n",
    "                'chromatin regulation complex': {}},\n",
    "            'nuclear transcriptional speckle': {\n",
    "                'negative regulation of RNA biosynthesis process': {}}}}}\n",
    "music_map1 = replace_regex_recursive(music_map)\n",
    "plot_bigraph(music_map1, **plot_settings, filename='cell_hierarchy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "ab431982-d156-4b7a-9918-384a5ea5132a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cell': {'cytoplasm': {'secretory organelles': {'transmembrane transport systems': {'ion transmembrane transport systems': {}}},\n",
       "   'cytoplasmic organelles': {'subgroup of cytoplasmic organelles': {'metabolic organelles': {'subgroup of metabolic organelles': {},\n",
       "      'mitochondrion': {}},\n",
       "     'ribonucleoprotein complex family': {'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
       "      'ribosome biogenesis community': {'ribosome': {'ribosomal subunit': {'mitochondirial large ribosomal subunit': {'large ribosomal subunit subcomplex 1': {}},\n",
       "         'mito-cyto ribosomal cluster': {},\n",
       "         'cotranslational protein targeting to membrane system': {},\n",
       "         'ribosomal complex 5': {},\n",
       "         'ribosomal complex 2': {}}},\n",
       "       'nucleic acid binding complex': {}}}}},\n",
       "   'transcriptional regulation complex family': {'transcriptional regulation complexes': {}}},\n",
       "  'nucleus': {'DNA metabolic assembly': {},\n",
       "   'nuclear lumen': {'nucleolus': {'ribonucleprotein complex family': {},\n",
       "     'RNA processing complex family': {'RNA processing complex 1': {},\n",
       "      'RNA splicing complex family': {}},\n",
       "     'splicosomal complex family': {}},\n",
       "    'nuclear splicing speckle': {}},\n",
       "   'nuclear body': {},\n",
       "   'nucleo-plasm 1': {'chromosome organization complex family': {'chromatin organization complex family': {'HAT complex family': {'NuA4 HAT complex': {}}}}},\n",
       "   'nucleo-plasm 2': {'chromatin regulation complex': {}},\n",
       "   'nuclear transcriptional speckle': {'negative regulation of RNA biosynthesis process': {}}}}}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "music_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "5a76608a-fff0-4005-9f32-c02afef39d9e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"776pt\" height=\"700pt\"\n",
       " viewBox=\"0.00 0.00 798.32 720.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.426819 0.426819) rotate(0) translate(4 1636.04)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-1636.04 1814.45,-1636.04 1814.45,4 -4,4\"/>\n",
       "<!-- (&#39;cell&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;cell&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"929.45\" cy=\"-1614.04\" rx=\"18\" ry=\"18\"/>\n",
       "<text text-anchor=\"start\" x=\"920.95\" y=\"-1610.94\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"819.45\" cy=\"-1523.95\" rx=\"35.18\" ry=\"35.18\"/>\n",
       "<text text-anchor=\"start\" x=\"794.95\" y=\"-1520.85\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M915.52,-1601.88C898.4,-1588.18 868.97,-1564.61 847.18,-1547.16\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1413.45\" cy=\"-1523.95\" rx=\"27.9\" ry=\"27.9\"/>\n",
       "<text text-anchor=\"start\" x=\"1394.95\" y=\"-1520.85\" font-family=\"Times,serif\" font-size=\"12.00\">nucleus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M947.95,-1610.33C985.71,-1604.76 1074.36,-1591.38 1148.45,-1578.04 1235.26,-1562.41 1337.48,-1541.11 1385.66,-1530.89\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;transcription&#39;) -->\n",
       "<g id=\"node46\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;transcription&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"1076.95,-355.75 1009.95,-355.75 1009.95,-319.75 1076.95,-319.75 1076.95,-355.75\"/>\n",
       "<text text-anchor=\"start\" x=\"1013.95\" y=\"-334.65\" font-family=\"Times,serif\" font-size=\"12.00\">transcription</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"node47\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;translation&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"143.95,-355.75 86.95,-355.75 86.95,-319.75 143.95,-319.75 143.95,-355.75\"/>\n",
       "<text text-anchor=\"start\" x=\"90.95\" y=\"-334.65\" font-family=\"Times,serif\" font-size=\"12.00\">translation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;metabolism&#39;) -->\n",
       "<g id=\"node48\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"539.95,-495.07 476.95,-495.07 476.95,-459.07 539.95,-459.07 539.95,-495.07\"/>\n",
       "<text text-anchor=\"start\" x=\"480.95\" y=\"-473.97\" font-family=\"Times,serif\" font-size=\"12.00\">metabolism</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n",
       "<g id=\"node49\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"1722.45,-643.51 1668.45,-643.51 1668.45,-607.51 1722.45,-607.51 1722.45,-643.51\"/>\n",
       "<text text-anchor=\"start\" x=\"1686.95\" y=\"-628.91\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "<text text-anchor=\"start\" x=\"1682.95\" y=\"-615.91\" font-family=\"Times,serif\" font-size=\"12.00\">cycle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;signalling&#39;) -->\n",
       "<g id=\"node50\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;signalling&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"665.45,-907.13 611.45,-907.13 611.45,-871.13 665.45,-871.13 665.45,-907.13\"/>\n",
       "<text text-anchor=\"start\" x=\"615.45\" y=\"-886.03\" font-family=\"Times,serif\" font-size=\"12.00\">signalling</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"node51\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"943.45,-907.13 889.45,-907.13 889.45,-871.13 943.45,-871.13 943.45,-907.13\"/>\n",
       "<text text-anchor=\"start\" x=\"899.45\" y=\"-892.53\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n",
       "<text text-anchor=\"start\" x=\"895.45\" y=\"-879.53\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;) -->\n",
       "<g id=\"node52\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"1261.95,-643.51 1202.95,-643.51 1202.95,-607.51 1261.95,-607.51 1261.95,-643.51\"/>\n",
       "<text text-anchor=\"start\" x=\"1218.95\" y=\"-628.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"1206.95\" y=\"-615.91\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"861.45\" cy=\"-1398.83\" rx=\"40.11\" ry=\"40.11\"/>\n",
       "<text text-anchor=\"start\" x=\"839.95\" y=\"-1402.23\" font-family=\"Times,serif\" font-size=\"12.00\">secretory</text>\n",
       "<text text-anchor=\"start\" x=\"836.95\" y=\"-1389.23\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M830.81,-1489.63C836.3,-1473.54 842.92,-1454.16 848.59,-1437.52\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"633.45\" cy=\"-1398.83\" rx=\"45.92\" ry=\"45.92\"/>\n",
       "<text text-anchor=\"start\" x=\"604.95\" y=\"-1402.23\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"608.95\" y=\"-1389.23\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M787.1,-1508.33C759.71,-1495.28 719.97,-1474.73 688.45,-1451.86 680.57,-1446.15 672.66,-1439.4 665.38,-1432.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"750.45\" cy=\"-1398.83\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"716.95\" y=\"-1415.23\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"726.45\" y=\"-1402.23\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"729.95\" y=\"-1389.23\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"734.95\" y=\"-1376.23\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M802.21,-1492.2C794.39,-1478.24 784.97,-1461.43 776.32,-1446\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"771.45\" cy=\"-1172.62\" rx=\"57.05\" ry=\"57.05\"/>\n",
       "<text text-anchor=\"start\" x=\"734.95\" y=\"-1182.52\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"750.45\" y=\"-1169.52\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"752.95\" y=\"-1156.52\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M846.48,-1360.55C831.79,-1323.94 809.23,-1267.74 792.68,-1226.52\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"edge66\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M862.43,-1357.92C864.11,-1302.52 868.38,-1199.28 878.45,-1111.81 886.9,-1038.37 903.63,-952.33 911.77,-912.48\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"912.22\" cy=\"-910.3\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"878.45\" y=\"-1170.12\" font-family=\"Times,serif\" font-size=\"10.00\">secretory</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"781.45\" cy=\"-748.8\" rx=\"57.05\" ry=\"57.05\"/>\n",
       "<text text-anchor=\"start\" x=\"773.45\" y=\"-765.2\" font-family=\"Times,serif\" font-size=\"12.00\">ion</text>\n",
       "<text text-anchor=\"start\" x=\"744.95\" y=\"-752.2\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"760.45\" y=\"-739.2\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"762.95\" y=\"-726.2\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M772.8,-1114.73C774.71,-1034.11 778.18,-887.48 780.1,-806.8\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;) -->\n",
       "<g id=\"edge63\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M743.95,-1121.89C739.22,-1112.66 734.52,-1103.03 730.45,-1093.81 707.22,-1041.23 710.84,-1024.24 686.45,-972.19 676.51,-950.97 662.83,-928.05 652.62,-911.84\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"651.44\" cy=\"-909.98\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"730.45\" y=\"-1030.5\" font-family=\"Times,serif\" font-size=\"10.00\">transport</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"edge65\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M786.78,-1116.91C788.78,-1109.18 790.73,-1101.31 792.45,-1093.81 804.73,-1040.18 787.27,-1019.43 815.45,-972.19 832,-944.42 861.61,-922.06 884.5,-907.73\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"886.56\" cy=\"-906.46\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"815.45\" y=\"-1036\" font-family=\"Times,serif\" font-size=\"10.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"827.95\" y=\"-1025\" font-family=\"Times,serif\" font-size=\"10.00\">transport</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;) -->\n",
       "<g id=\"edge59\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M783.94,-691.15C782.62,-654.98 775.16,-609.12 749.45,-577.49 697.9,-514.09 599.18,-490.66 545.48,-482.36\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"543.21\" cy=\"-482.02\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"794.95\" y=\"-628.51\" font-family=\"Times,serif\" font-size=\"10.00\">ion</text>\n",
       "<text text-anchor=\"start\" x=\"783.45\" y=\"-617.51\" font-family=\"Times,serif\" font-size=\"10.00\">transport</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"485.45\" cy=\"-1172.62\" rx=\"45.92\" ry=\"45.92\"/>\n",
       "<text text-anchor=\"start\" x=\"462.95\" y=\"-1189.02\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"480.45\" y=\"-1176.02\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"456.95\" y=\"-1163.02\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"460.95\" y=\"-1150.02\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M608.17,-1359.54C580.93,-1318.27 537.82,-1252.96 510.62,-1211.75\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"511.45\" cy=\"-889.13\" rx=\"40.11\" ry=\"40.11\"/>\n",
       "<text text-anchor=\"start\" x=\"488.45\" y=\"-892.53\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"486.95\" y=\"-879.53\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M489.65,-1126.08C494.68,-1071.62 503.02,-981.36 507.79,-929.69\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"268.45\" cy=\"-889.13\" rx=\"65.11\" ry=\"65.11\"/>\n",
       "<text text-anchor=\"start\" x=\"226.45\" y=\"-899.03\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleoprotein</text>\n",
       "<text text-anchor=\"start\" x=\"247.95\" y=\"-886.03\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"252.95\" y=\"-873.03\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M457.3,-1135.11C419.29,-1085.79 351.33,-997.64 308.14,-941.62\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"624.45\" cy=\"-625.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"601.95\" y=\"-641.91\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"619.45\" y=\"-628.91\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"601.45\" y=\"-615.91\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"599.95\" y=\"-602.91\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M527.28,-851.48C548.54,-802.26 586.14,-715.2 607.8,-665.04\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"437.45\" cy=\"-625.51\" rx=\"48.03\" ry=\"48.03\"/>\n",
       "<text text-anchor=\"start\" x=\"402.45\" y=\"-622.41\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondrion</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M500.53,-849.53C487.18,-802.34 464.62,-722.58 450.41,-672.32\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;) -->\n",
       "<g id=\"edge57\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M511.15,-848.07C510.53,-763.22 509.1,-567.71 508.61,-500.92\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"508.6\" cy=\"-498.47\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"510.45\" y=\"-623.01\" font-family=\"Times,serif\" font-size=\"10.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;) -->\n",
       "<g id=\"edge58\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M436.4,-576.78C437.68,-563.53 440.58,-549.5 446.45,-537.49 453.81,-522.44 466.65,-509.15 478.75,-498.98\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"480.54\" cy=\"-497.51\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"446.45\" y=\"-545.99\" font-family=\"Times,serif\" font-size=\"10.00\">mitochondrion</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"185.45\" cy=\"-625.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"171.95\" y=\"-641.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"159.95\" y=\"-628.91\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"164.95\" y=\"-615.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"181.95\" y=\"-602.91\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M248.78,-826.13C233.15,-776.87 211.75,-709.41 198.26,-666.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"289.45\" cy=\"-625.51\" rx=\"43.68\" ry=\"43.68\"/>\n",
       "<text text-anchor=\"start\" x=\"267.45\" y=\"-635.41\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "<text text-anchor=\"start\" x=\"264.45\" y=\"-622.41\" font-family=\"Times,serif\" font-size=\"12.00\">biogenesis</text>\n",
       "<text text-anchor=\"start\" x=\"262.45\" y=\"-609.41\" font-family=\"Times,serif\" font-size=\"12.00\">community</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M273.64,-823.46C277.48,-775.65 282.6,-711.83 285.96,-669.98\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"185.45\" cy=\"-477.07\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"171.95\" y=\"-493.47\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"166.95\" y=\"-480.47\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"164.95\" y=\"-467.47\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"181.95\" y=\"-454.47\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M185.45,-582.55C185.45,-563.02 185.45,-539.85 185.45,-520.29\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"283.45\" cy=\"-477.07\" rx=\"32.02\" ry=\"32.02\"/>\n",
       "<text text-anchor=\"start\" x=\"261.45\" y=\"-473.97\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M287.66,-580.97C286.73,-558.21 285.61,-530.83 284.76,-510.05\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"389.45\" cy=\"-477.07\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"372.45\" y=\"-493.47\" font-family=\"Times,serif\" font-size=\"12.00\">nucleic</text>\n",
       "<text text-anchor=\"start\" x=\"379.45\" y=\"-480.47\" font-family=\"Times,serif\" font-size=\"12.00\">acid</text>\n",
       "<text text-anchor=\"start\" x=\"370.95\" y=\"-467.47\" font-family=\"Times,serif\" font-size=\"12.00\">binding</text>\n",
       "<text text-anchor=\"start\" x=\"368.95\" y=\"-454.47\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M313.5,-587.95C323.85,-572.34 336.14,-553.96 347.45,-537.49 353.02,-529.37 359.1,-520.68 364.86,-512.52\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"283.45\" cy=\"-337.75\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"259.95\" y=\"-341.15\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"265.95\" y=\"-328.15\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M283.45,-444.22C283.45,-424.33 283.45,-398.6 283.45,-377.53\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge56\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M259.75,-454.75C252.35,-448.22 244.12,-441.06 236.45,-434.64 204.63,-408.04 167.07,-378.57 142.34,-359.41\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"140.46\" cy=\"-357.96\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"214.45\" y=\"-403.14\" font-family=\"Times,serif\" font-size=\"10.00\">ribosomes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"54.45\" cy=\"-201.05\" rx=\"54.39\" ry=\"54.39\"/>\n",
       "<text text-anchor=\"start\" x=\"19.95\" y=\"-217.45\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondirial</text>\n",
       "<text text-anchor=\"start\" x=\"42.45\" y=\"-204.45\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n",
       "<text text-anchor=\"start\" x=\"30.95\" y=\"-191.45\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"36.95\" y=\"-178.45\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M246.32,-324.11C211.41,-311.3 158.52,-289.46 117.45,-261.86 108.68,-255.97 100,-248.83 92.01,-241.57\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"165.45\" cy=\"-201.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"142.45\" y=\"-210.95\" font-family=\"Times,serif\" font-size=\"12.00\">mito&#45;cyto</text>\n",
       "<text text-anchor=\"start\" x=\"141.95\" y=\"-197.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"149.95\" y=\"-184.95\" font-family=\"Times,serif\" font-size=\"12.00\">cluster</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M256.34,-308.76C243.15,-294.88 227.2,-277.73 213.45,-261.86 205.34,-252.5 196.79,-242.05 189.18,-232.53\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"283.45\" cy=\"-201.05\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"248.95\" y=\"-230.45\" font-family=\"Times,serif\" font-size=\"12.00\">cotranslational</text>\n",
       "<text text-anchor=\"start\" x=\"266.45\" y=\"-217.45\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n",
       "<text text-anchor=\"start\" x=\"262.45\" y=\"-204.45\" font-family=\"Times,serif\" font-size=\"12.00\">targeting</text>\n",
       "<text text-anchor=\"start\" x=\"278.45\" y=\"-191.45\" font-family=\"Times,serif\" font-size=\"12.00\">to</text>\n",
       "<text text-anchor=\"start\" x=\"258.45\" y=\"-178.45\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "<text text-anchor=\"start\" x=\"266.95\" y=\"-165.45\" font-family=\"Times,serif\" font-size=\"12.00\">system</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M283.45,-298.15C283.45,-287.03 283.45,-274.63 283.45,-262.53\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"401.45\" cy=\"-201.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"377.95\" y=\"-210.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"380.95\" y=\"-197.95\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"397.95\" y=\"-184.95\" font-family=\"Times,serif\" font-size=\"12.00\">5</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M310.55,-308.76C323.75,-294.88 339.7,-277.73 353.45,-261.86 361.56,-252.5 370.11,-242.05 377.71,-232.53\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"497.45\" cy=\"-201.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"473.95\" y=\"-210.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"476.95\" y=\"-197.95\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"493.95\" y=\"-184.95\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M321.12,-325.45C356.66,-313.54 410.21,-292.26 449.45,-261.86 459.34,-254.19 468.44,-244.09 476,-234.38\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"54.45\" cy=\"-51.62\" rx=\"51.74\" ry=\"51.74\"/>\n",
       "<text text-anchor=\"start\" x=\"42.45\" y=\"-74.52\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n",
       "<text text-anchor=\"start\" x=\"30.95\" y=\"-61.52\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"36.95\" y=\"-48.52\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "<text text-anchor=\"start\" x=\"25.45\" y=\"-35.52\" font-family=\"Times,serif\" font-size=\"12.00\">subcomplex</text>\n",
       "<text text-anchor=\"start\" x=\"50.95\" y=\"-22.52\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M54.45,-145.99C54.45,-132.35 54.45,-117.74 54.45,-104.22\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n",
       "<g id=\"edge54\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M425.64,-453.72C438.45,-446.66 453.22,-439.46 467.45,-434.64 662,-368.74 909.43,-347.09 1004.14,-340.92\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1006.58\" cy=\"-340.77\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"606.45\" y=\"-408.64\" font-family=\"Times,serif\" font-size=\"10.00\">nucleic</text>\n",
       "<text text-anchor=\"start\" x=\"610.95\" y=\"-397.64\" font-family=\"Times,serif\" font-size=\"10.00\">acids</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"643.45\" cy=\"-1172.62\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"609.95\" y=\"-1182.52\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"619.45\" y=\"-1169.52\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"617.95\" y=\"-1156.52\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M727.62,-1350.01C709.52,-1312.07 684.33,-1259.29 666.24,-1221.37\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n",
       "<g id=\"edge53\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M670.33,-1125.97C675.23,-1115.71 679.66,-1104.64 682.45,-1093.81 711.04,-982.6 647.46,-678.3 702.45,-577.49 769.54,-454.48 930.61,-380.86 1004.82,-352.37\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1006.81\" cy=\"-351.61\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"702.45\" y=\"-623.01\" font-family=\"Times,serif\" font-size=\"10.00\">regulation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;) -->\n",
       "<g id=\"edge64\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M640.4,-1118.89C640.02,-1110.5 639.68,-1101.92 639.45,-1093.81 637.6,-1028.18 637.91,-950.61 638.21,-912.95\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"638.23\" cy=\"-910.62\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"643.45\" y=\"-1036\" font-family=\"Times,serif\" font-size=\"10.00\">trancript</text>\n",
       "<text text-anchor=\"start\" x=\"639.45\" y=\"-1025\" font-family=\"Times,serif\" font-size=\"10.00\">regulation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1572.45\" cy=\"-1289.61\" rx=\"38.37\" ry=\"38.37\"/>\n",
       "<text text-anchor=\"start\" x=\"1558.45\" y=\"-1299.51\" font-family=\"Times,serif\" font-size=\"12.00\">DNA</text>\n",
       "<text text-anchor=\"start\" x=\"1549.45\" y=\"-1286.51\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"1550.45\" y=\"-1273.51\" font-family=\"Times,serif\" font-size=\"12.00\">assembly</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1441.25,-1517.14C1473.9,-1508.75 1526.93,-1489.87 1553.45,-1451.86 1578.69,-1415.68 1580.17,-1363.48 1577.51,-1328.36\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1514.45\" cy=\"-1398.83\" rx=\"30.31\" ry=\"30.31\"/>\n",
       "<text text-anchor=\"start\" x=\"1496.95\" y=\"-1402.23\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"1499.45\" y=\"-1389.23\" font-family=\"Times,serif\" font-size=\"12.00\">lumen</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1432.31,-1502.15C1444.82,-1488.16 1461.44,-1469.18 1475.45,-1451.86 1482.55,-1443.09 1490.05,-1433.23 1496.56,-1424.47\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1312.45\" cy=\"-1398.83\" rx=\"30.31\" ry=\"30.31\"/>\n",
       "<text text-anchor=\"start\" x=\"1294.95\" y=\"-1402.23\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"1299.95\" y=\"-1389.23\" font-family=\"Times,serif\" font-size=\"12.00\">body</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1394.58,-1502.15C1382.08,-1488.16 1365.46,-1469.18 1351.45,-1451.86 1344.35,-1443.09 1336.84,-1433.23 1330.33,-1424.47\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1651.45\" cy=\"-1398.83\" rx=\"50.82\" ry=\"50.82\"/>\n",
       "<text text-anchor=\"start\" x=\"1619.45\" y=\"-1402.23\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"1647.95\" y=\"-1389.23\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1441.72,-1519.18C1477.65,-1513.37 1540.5,-1499.74 1586.45,-1470.86 1599.19,-1462.86 1611,-1451.7 1620.97,-1440.6\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1213.45\" cy=\"-1398.83\" rx=\"50.82\" ry=\"50.82\"/>\n",
       "<text text-anchor=\"start\" x=\"1181.45\" y=\"-1402.23\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"1209.95\" y=\"-1389.23\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1387.36,-1512.02C1358.43,-1499.38 1310.88,-1477.04 1273.45,-1451.86 1265.78,-1446.71 1258.02,-1440.69 1250.73,-1434.61\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1413.45\" cy=\"-1398.83\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"1395.95\" y=\"-1408.73\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"1379.95\" y=\"-1395.73\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"1395.45\" y=\"-1382.73\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1413.45,-1495.32C1413.45,-1482.7 1413.45,-1467.31 1413.45,-1452.61\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n",
       "<g id=\"edge60\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1569.16,-1250.62C1563.27,-1170.48 1554.74,-978.79 1591.45,-824.08 1607.94,-754.57 1654.58,-682.76 1679.42,-647.97\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1680.58\" cy=\"-646.35\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"1568.45\" y=\"-1030.5\" font-family=\"Times,serif\" font-size=\"10.00\">DNA</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1502.45\" cy=\"-1172.62\" rx=\"33.49\" ry=\"33.49\"/>\n",
       "<text text-anchor=\"start\" x=\"1479.45\" y=\"-1169.52\" font-family=\"Times,serif\" font-size=\"12.00\">nucleolus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1512.83,-1367.59C1510.59,-1325.76 1506.56,-1250.47 1504.24,-1207.02\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1418.45\" cy=\"-1172.62\" rx=\"33.47\" ry=\"33.47\"/>\n",
       "<text text-anchor=\"start\" x=\"1400.95\" y=\"-1182.52\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"1399.95\" y=\"-1169.52\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"1400.45\" y=\"-1156.52\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1502.68,-1370.34C1484.65,-1328.24 1450.33,-1248.09 1431.54,-1204.21\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1272.45\" cy=\"-889.13\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"1233.45\" y=\"-899.03\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleprotein</text>\n",
       "<text text-anchor=\"start\" x=\"1251.95\" y=\"-886.03\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1256.95\" y=\"-873.03\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1486.23,-1142.45C1479.31,-1131.84 1470.49,-1120.37 1460.45,-1111.81 1446.53,-1099.95 1437.52,-1105.49 1423.45,-1093.81 1372.13,-1051.21 1328.67,-987.7 1301.74,-942.95\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1502.45\" cy=\"-889.13\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"1488.95\" y=\"-905.53\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"1476.95\" y=\"-892.53\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"1481.95\" y=\"-879.53\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1486.95\" y=\"-866.53\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1502.45,-1138.22C1502.45,-1086.82 1502.45,-988.36 1502.45,-932.18\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1396.45\" cy=\"-889.13\" rx=\"45.01\" ry=\"45.01\"/>\n",
       "<text text-anchor=\"start\" x=\"1368.45\" y=\"-899.03\" font-family=\"Times,serif\" font-size=\"12.00\">splicosomal</text>\n",
       "<text text-anchor=\"start\" x=\"1375.95\" y=\"-886.03\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1380.95\" y=\"-873.03\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1490.7,-1140.42C1471.56,-1089.6 1433.65,-988.93 1412.2,-931.97\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1329.45\" cy=\"-625.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"1315.95\" y=\"-641.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"1303.95\" y=\"-628.91\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"1308.95\" y=\"-615.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1325.95\" y=\"-602.91\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1478.58,-853.51C1468.58,-838.91 1456.88,-821.72 1446.45,-806.08 1413.31,-756.39 1375.74,-698.43 1352.31,-662.09\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1432.45\" cy=\"-625.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"1418.95\" y=\"-641.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"1413.95\" y=\"-628.91\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"1411.95\" y=\"-615.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1416.95\" y=\"-602.91\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1491.51,-847.25C1478.36,-798.1 1456.42,-716.12 1443.31,-667.1\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n",
       "<g id=\"edge52\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1511.7,-846.84C1523.5,-783.45 1536.51,-660.76 1483.45,-577.49 1389.5,-430.07 1171.53,-366.92 1082.45,-346.63\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1080.23\" cy=\"-346.13\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"1479.45\" y=\"-551.49\" font-family=\"Times,serif\" font-size=\"10.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"1468.45\" y=\"-540.49\" font-family=\"Times,serif\" font-size=\"10.00\">processing</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;) -->\n",
       "<g id=\"edge68\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1480.66,-851.96C1472.5,-841.34 1462.28,-830.79 1450.45,-824.08 1401.63,-796.39 1371.2,-838.6 1325.45,-806.08 1271.28,-767.57 1246.37,-687.82 1237.07,-648.88\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1236.57\" cy=\"-646.78\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"1325.45\" y=\"-751.8\" font-family=\"Times,serif\" font-size=\"10.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"1329.45\" y=\"-740.8\" font-family=\"Times,serif\" font-size=\"10.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1651.45\" cy=\"-1172.62\" rx=\"48.58\" ry=\"48.58\"/>\n",
       "<text text-anchor=\"start\" x=\"1620.95\" y=\"-1189.02\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n",
       "<text text-anchor=\"start\" x=\"1621.95\" y=\"-1176.02\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"1630.95\" y=\"-1163.02\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1635.95\" y=\"-1150.02\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1651.45,-1347.23C1651.45,-1309.78 1651.45,-1259.05 1651.45,-1222.11\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1647.45\" cy=\"-889.13\" rx=\"47.25\" ry=\"47.25\"/>\n",
       "<text text-anchor=\"start\" x=\"1623.45\" y=\"-905.53\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"1617.95\" y=\"-892.53\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"1626.95\" y=\"-879.53\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1631.95\" y=\"-866.53\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1650.76,-1123.23C1650.02,-1071.04 1648.85,-988.52 1648.11,-936.95\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n",
       "<g id=\"edge61\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1664.72,-1124.94C1676.66,-1081.18 1693.76,-1013.76 1703.45,-954.19 1714.05,-888.96 1713.56,-872.09 1716.45,-806.08 1718.68,-755.21 1724.37,-741.82 1716.45,-691.53 1714.15,-676.92 1709.1,-661.11 1704.54,-648.75\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1703.77\" cy=\"-646.7\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"1715.45\" y=\"-886.63\" font-family=\"Times,serif\" font-size=\"10.00\">chromosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1565.45\" cy=\"-625.51\" rx=\"34.8\" ry=\"34.8\"/>\n",
       "<text text-anchor=\"start\" x=\"1552.95\" y=\"-635.41\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"1544.95\" y=\"-622.41\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"1549.95\" y=\"-609.41\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1633.35,-843.16C1617,-791 1590.44,-706.25 1575.85,-659.71\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n",
       "<g id=\"edge62\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1651.54,-841.24C1655.6,-801.14 1662.91,-742.08 1674.45,-691.53 1677.74,-677.11 1682.81,-661.31 1687.14,-648.9\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1687.86\" cy=\"-646.85\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"1674.45\" y=\"-746.3\" font-family=\"Times,serif\" font-size=\"10.00\">chromatin</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1565.45\" cy=\"-477.07\" rx=\"34.8\" ry=\"34.8\"/>\n",
       "<text text-anchor=\"start\" x=\"1549.95\" y=\"-486.97\" font-family=\"Times,serif\" font-size=\"12.00\">NuA4</text>\n",
       "<text text-anchor=\"start\" x=\"1552.95\" y=\"-473.97\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"1544.95\" y=\"-460.97\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1565.45,-589.81C1565.45,-566.55 1565.45,-536.04 1565.45,-512.78\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1213.45\" cy=\"-1172.62\" rx=\"39.7\" ry=\"39.7\"/>\n",
       "<text text-anchor=\"start\" x=\"1189.45\" y=\"-1182.52\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"1189.45\" y=\"-1169.52\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"1192.95\" y=\"-1156.52\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1213.45,-1347.23C1213.45,-1306.68 1213.45,-1250.57 1213.45,-1213.23\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1221.45\" cy=\"-1033\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"1201.45\" y=\"-1062.4\" font-family=\"Times,serif\" font-size=\"12.00\">negative</text>\n",
       "<text text-anchor=\"start\" x=\"1197.45\" y=\"-1049.4\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"1216.45\" y=\"-1036.4\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"1207.95\" y=\"-1023.4\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"1192.45\" y=\"-1010.4\" font-family=\"Times,serif\" font-size=\"12.00\">biosynthesis</text>\n",
       "<text text-anchor=\"start\" x=\"1203.45\" y=\"-997.4\" font-family=\"Times,serif\" font-size=\"12.00\">process</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1388.74,-1351.01C1353.04,-1283.36 1287.2,-1158.59 1249.64,-1087.43\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n",
       "<g id=\"edge55\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1206.42,-973.49C1169.13,-828.23 1073.55,-456.01 1049.18,-361.06\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"1048.62\" cy=\"-358.91\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"1131.95\" y=\"-628.51\" font-family=\"Times,serif\" font-size=\"10.00\">negative</text>\n",
       "<text text-anchor=\"start\" x=\"1128.45\" y=\"-617.51\" font-family=\"Times,serif\" font-size=\"10.00\">regulation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"node53\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1056.45\" cy=\"-1172.62\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"1021.95\" y=\"-1202.02\" font-family=\"Times,serif\" font-size=\"12.00\">cotranslational</text>\n",
       "<text text-anchor=\"start\" x=\"1039.45\" y=\"-1189.02\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n",
       "<text text-anchor=\"start\" x=\"1035.45\" y=\"-1176.02\" font-family=\"Times,serif\" font-size=\"12.00\">targeting</text>\n",
       "<text text-anchor=\"start\" x=\"1051.45\" y=\"-1163.02\" font-family=\"Times,serif\" font-size=\"12.00\">to</text>\n",
       "<text text-anchor=\"start\" x=\"1031.45\" y=\"-1150.02\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "<text text-anchor=\"start\" x=\"1039.95\" y=\"-1137.02\" font-family=\"Times,serif\" font-size=\"12.00\">system</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"edge67\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1029.45,-1117.34C998.5,-1055.1 949.21,-956.01 927.42,-912.2\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"926.39\" cy=\"-910.12\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"1014.45\" y=\"-1030.5\" font-family=\"Times,serif\" font-size=\"10.00\">cotranslation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;translation&#39;, &#39;membrane&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;translation&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge69\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;translation&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M55.98,-469.82C52.98,-455.72 47.34,-420.64 58.45,-394.64 64.27,-381.02 75.17,-369.02 85.89,-359.68\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"87.61\" cy=\"-358.22\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"58.45\" y=\"-403.14\" font-family=\"Times,serif\" font-size=\"10.00\">membrane</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108d40a00>"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import copy\n",
    "\n",
    "transcription_process = {\n",
    "    '_ports': {\n",
    "        'RNA processing': 'Any',\n",
    "        'regulation': 'Any',\n",
    "        'nucleic acids': 'Any',\n",
    "        'negative regulation': 'Any'\n",
    "    },\n",
    "    'wires': {\n",
    "        'RNA processing': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "        'regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "        'nucleic acids': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', \n",
    "            'ribonucleoprotein complex family', 'ribosome biogenesis community', 'nucleic acid binding complex'],\n",
    "        'negative regulation': [\n",
    "            'nucleus', 'nuclear transcriptional speckle', 'negative regulation of RNA biosynthesis process'],\n",
    "    }\n",
    "}\n",
    "\n",
    "translation_process = {\n",
    "    '_ports': {\n",
    "        'ribosomes': 'Any',\n",
    "        'membrane': 'Any'\n",
    "    },\n",
    "    'wires': {\n",
    "        'ribosomes': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', 'ribosome biogenesis community', 'ribosome'],\n",
    "        # 'membrane': ['],\n",
    "    }\n",
    "}\n",
    "\n",
    "metabolism_process = {\n",
    "    '_ports': {\n",
    "        'organelles': 'Any',\n",
    "        'mitochondrion': 'Any',\n",
    "        'ion transport': 'Any',\n",
    "    },\n",
    "    'wires': {\n",
    "        'organelles': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', ],\n",
    "        'mitochondrion': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', 'mitochondrion'],\n",
    "        'ion transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems', 'ion transmembrane transport systems'],\n",
    "    }\n",
    "}\n",
    "\n",
    "cell_cycle_process = {\n",
    "    '_ports': {\n",
    "        'DNA': 'Any',\n",
    "        'chromosome': 'Any',\n",
    "        'chromatin': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'DNA': ['nucleus', 'DNA metabolic assembly'],\n",
    "        'chromosome': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family',],\n",
    "        'chromatin': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family', 'chromatin organization complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "signalling_process = {\n",
    "    '_ports': {\n",
    "        'transport': 'Any',\n",
    "        'trancript regulation': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'trancript regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "    }\n",
    "}\n",
    "\n",
    "protein_transport_process = {\n",
    "    '_ports': {\n",
    "        'transmembrane transport': 'Any',\n",
    "        'secretory': 'Any',\n",
    "        'cotranslation': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'transmembrane transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'secretory': [\n",
    "            'cytoplasm', 'secretory organelles'],\n",
    "        'cotranslation': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', \n",
    "            'ribosome biogenesis community', 'cotranslational protein targeting to membrane system']\n",
    "    }\n",
    "}\n",
    "\n",
    "rna_processing_process = {\n",
    "    '_ports': {\n",
    "        'processing complex': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'processing complex': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "# compose schema\n",
    "music_map['cell']['transcription'] = transcription_process\n",
    "music_map['cell']['translation'] = translation_process\n",
    "music_map['cell']['metabolism'] = metabolism_process\n",
    "music_map['cell']['cell cycle'] = cell_cycle_process\n",
    "music_map['cell']['signalling'] = signalling_process\n",
    "music_map['cell']['protein transport'] = protein_transport_process\n",
    "music_map['cell']['RNA processing'] = rna_processing_process\n",
    "\n",
    "music_map2 = replace_regex_recursive(music_map)\n",
    "\n",
    "# plot\n",
    "plot_settingsx = copy.deepcopy(plot_settings)\n",
    "plot_settingsx['node_border_colors'] = {\n",
    "    ('cell', 'translation'): 'blue',\n",
    "    ('cell', 'transcription'): 'blue',\n",
    "    ('cell', 'metabolism'): 'blue',\n",
    "    ('cell', 'cell<br/>cycle'): 'blue',\n",
    "    ('cell', 'signalling'): 'blue',\n",
    "    ('cell', 'protein<br/>transport'): 'blue',\n",
    "    ('cell', 'RNA<br/>processing'): 'blue',\n",
    "}\n",
    "# plot_settingsx.update({'out_dir': 'out','dpi': '250'})\n",
    "plot_bigraph(music_map2, **plot_settingsx, filename='cell_hierarchy_functions')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae52f73c-e384-4880-b18f-47b4c25ce203",
   "metadata": {},
   "source": [
    "### Cell structure and function\n",
    "Processes represent their functions, we show a more minimal cell structure with added processes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b39417ce-d9a1-4985-abce-ce22df816897",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"656pt\" height=\"672pt\"\n",
       " viewBox=\"0.00 0.00 674.26 691.30\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 687.3)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-687.3 670.26,-687.3 670.26,4 -4,4\"/>\n",
       "<!-- (&#39;cell&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;cell&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"353.78\" cy=\"-665.3\" rx=\"18\" ry=\"18\"/>\n",
       "<text text-anchor=\"start\" x=\"345.28\" y=\"-662.2\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"141.78\" cy=\"-574.6\" rx=\"35.91\" ry=\"35.91\"/>\n",
       "<text text-anchor=\"start\" x=\"116.78\" y=\"-571.5\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M336.79,-657.19C301.81,-642.56 221.2,-608.83 175.32,-589.63\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"353.78\" cy=\"-574.6\" rx=\"35.18\" ry=\"35.18\"/>\n",
       "<text text-anchor=\"start\" x=\"329.28\" y=\"-571.5\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M353.78,-646.35C353.78,-636.13 353.78,-622.96 353.78,-610.76\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;) -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"525.78\" cy=\"-574.6\" rx=\"30.33\" ry=\"30.33\"/>\n",
       "<text text-anchor=\"start\" x=\"505.28\" y=\"-571.5\" font-family=\"Times,serif\" font-size=\"12.00\">nucleoid</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M369.96,-655.96C399.48,-640.73 462.37,-608.3 498.57,-589.63\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"38.78\" cy=\"-412.02\" rx=\"38.57\" ry=\"38.57\"/>\n",
       "<text text-anchor=\"start\" x=\"11.28\" y=\"-408.92\" font-family=\"Times,serif\" font-size=\"12.00\">transporters</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M118.76,-545.74C108.48,-532.77 96.51,-516.91 86.78,-501.9 75.53,-484.53 64.46,-464.29 55.78,-447.51\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"116.78\" cy=\"-412.02\" rx=\"20.86\" ry=\"20.86\"/>\n",
       "<text text-anchor=\"start\" x=\"103.78\" y=\"-408.92\" font-family=\"Times,serif\" font-size=\"12.00\">lipids</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M136.3,-538.36C131.29,-506.23 124.08,-459.87 119.97,-433.47\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"253.28,-136.09 128.28,-136.09 128.28,-100.09 253.28,-100.09 253.28,-136.09\"/>\n",
       "<text text-anchor=\"start\" x=\"132.28\" y=\"-114.99\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane transport</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M40.47,-372.69C44.49,-321.86 58.08,-231.54 103.78,-171.18 113.88,-157.83 128.48,-147.22 142.84,-139.14\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"144.69\" cy=\"-138.13\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"103.78\" y=\"-235.16\" font-family=\"Times,serif\" font-size=\"10.00\">transporters</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"231.78\" cy=\"-412.02\" rx=\"38.33\" ry=\"38.33\"/>\n",
       "<text text-anchor=\"start\" x=\"204.78\" y=\"-408.92\" font-family=\"Times,serif\" font-size=\"12.00\">metabolites</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M326.35,-551.23C311.14,-537.94 292.58,-520.15 278.78,-501.9 266.26,-485.32 255.22,-464.95 247,-447.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"392.78\" cy=\"-237.66\" rx=\"66.47\" ry=\"66.47\"/>\n",
       "<text text-anchor=\"start\" x=\"342.78\" y=\"-234.56\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal complexes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M344.01,-539.78C331.47,-490.68 313.92,-397.28 336.78,-322.14 340.08,-311.29 345.32,-300.57 351.29,-290.61\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;) -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"435.78\" cy=\"-412.02\" rx=\"89.75\" ry=\"89.75\"/>\n",
       "<text text-anchor=\"start\" x=\"366.78\" y=\"-408.92\" font-family=\"Times,serif\" font-size=\"12.00\">transcript regulation complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M369.79,-542.26C377.06,-528.02 386.02,-510.46 394.98,-492.92\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"502.28,-36 445.28,-36 445.28,0 502.28,0 502.28,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"449.28\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">translation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M226.46,-373.15C217.81,-311.55 200.95,-191.52 193.95,-141.65\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"193.61\" cy=\"-139.24\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"216.78\" y=\"-235.16\" font-family=\"Times,serif\" font-size=\"10.00\">internal</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M396.12,-170.4C399.5,-142.39 406.08,-110.07 418.78,-83 426.28,-67.01 438.91,-51.89 450.18,-40.37\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"451.86\" cy=\"-38.68\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"418.78\" y=\"-115.59\" font-family=\"Times,serif\" font-size=\"10.00\">p1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;) -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"473.78\" cy=\"-118.09\" rx=\"35.18\" ry=\"35.18\"/>\n",
       "<text text-anchor=\"start\" x=\"449.28\" y=\"-114.99\" font-family=\"Times,serif\" font-size=\"12.00\">transcripts</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M456.47,-323.73C457.7,-317.12 458.83,-310.54 459.78,-304.14 467.54,-252.2 471.09,-191.38 472.65,-154.05\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M473.78,-81.95C473.78,-68.58 473.78,-53.71 473.78,-41.77\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"473.78\" cy=\"-39.35\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"473.78\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">p2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"623.78\" cy=\"-412.02\" rx=\"42.45\" ry=\"42.45\"/>\n",
       "<text text-anchor=\"start\" x=\"593.28\" y=\"-408.92\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M545.61,-550.72C552.98,-541.71 561.16,-531.1 567.78,-520.9 582.15,-498.77 595.83,-472.53 606.08,-451.47\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"623.78\" cy=\"-118.09\" rx=\"22.32\" ry=\"22.32\"/>\n",
       "<text text-anchor=\"start\" x=\"609.78\" y=\"-114.99\" font-family=\"Times,serif\" font-size=\"12.00\">genes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M623.78,-368.57C623.78,-305.63 623.78,-189.79 623.78,-141.32\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;external&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;external&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;external&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M287.78,-393.24C284.77,-344.4 274.64,-209.24 252.78,-171.18 245.8,-159.02 234.84,-148.42 224.07,-140.02\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"222.11\" cy=\"-138.54\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"280.78\" y=\"-235.16\" font-family=\"Times,serif\" font-size=\"10.00\">external</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108c554f0>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_struct_func = {\n",
    "    'cell': {\n",
    "        'membrane': {\n",
    "            'transporters': 'concentrations',\n",
    "            'lipids': 'concentrations',\n",
    "            'transmembrane transport': {\n",
    "                '_value': {\n",
    "                    '_process': 'transport URI',\n",
    "                    '_config': {'parameter': 1}\n",
    "                },\n",
    "                'wires': {\n",
    "                    'transporters': 'transporters',\n",
    "                    'internal': ['..', 'cytoplasm', 'metabolites']},\n",
    "                '_ports': {\n",
    "                    'transporters': 'concentrations',\n",
    "                    'internal': 'concentrations',\n",
    "                    'external': 'concentrations'\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "        'cytoplasm': {\n",
    "            'metabolites': {\n",
    "                '_value': 1.1,\n",
    "                '_type': 'concentrations'\n",
    "            },\n",
    "            'ribosomal complexes': {\n",
    "                '_value': 2.2,\n",
    "                '_type': 'concentrations'\n",
    "            },\n",
    "            'transcript regulation complex': {\n",
    "                '_value': 0.01,\n",
    "                'transcripts': {\n",
    "                    '_value': 0.1,\n",
    "                    '_type': 'concentrations'\n",
    "                }\n",
    "            },\n",
    "            'translation': {\n",
    "                'wires': {\n",
    "                    'p1': 'ribosomal complexes',\n",
    "                    'p2': ['transcript regulation complex', 'transcripts']}}},\n",
    "        'nucleoid': {\n",
    "            'chromosome': {\n",
    "                'genes': 'sequences'\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(cell_struct_func, remove_process_place_edges=True, out_dir='out', filename='cell')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "188eb4a6-c526-4b38-a22d-97f7fb8b7244",
   "metadata": {},
   "source": [
    "### Whole-cell *E. coli* model\n",
    "\n",
    "We here plot the processes and wiring of a whole-cell model of *E. coli* from [Macklin et al. Simultaneous cross-evaluation of heterogeneous E. coli datasets via mechanistic simulation. Science. 2020.](https://pubmed.ncbi.nlm.nih.gov/32703847/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "16960d1b-6e35-421e-bb7d-3efe5052e4fc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/ecoli\n"
     ]
    }
   ],
   "source": [
    "ecoli = { \n",
    "    'chromosome-structure': { \n",
    "        'wires': {\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'DnaA_boxes': ('unique', 'DnaA_box')\n",
    "        }\n",
    "    },\n",
    "    'metabolism': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'catalysts': ('bulk',),\n",
    "            'kinetics_enzymes': ('bulk',),\n",
    "            'kinetics_substrates': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'environment': ('environment',),\n",
    "            'amino_acids_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'tf-binding': { \n",
    "        'wires': {\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'active_tfs_total': ('bulk',),\n",
    "            'inactive_tfs_total': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'molecules': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'molecules': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'ntps': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'rna-degradation': { \n",
    "        'wires': {\n",
    "            'charged_trna': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'nmps': ('bulk',),\n",
    "            'fragmentMetabolites': ('bulk',),\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'endoRnases': ('bulk',),\n",
    "            'exoRnases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'RNA': ('unique', 'RNA'),\n",
    "            'subunits': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'molecules': ('bulk',),\n",
    "            'monomers': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'ppgpp_reaction_metabolites': ('bulk',),\n",
    "            'uncharged_trna': ('bulk',),\n",
    "            'charged_trna': ('bulk',),\n",
    "            'charging_molecules': ('bulk',),\n",
    "            'synthetases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules_total': ('bulk',),\n",
    "            'amino_acids_total': ('bulk',),\n",
    "            'charged_trna_total': ('bulk',),\n",
    "            'uncharged_trna_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'complexation': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',), \n",
    "        }\n",
    "    },\n",
    "    'two-component-system': { \n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'equilibrium': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'protein-degradation': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'proteins': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'chromosome-replication': { \n",
    "        'wires': {\n",
    "            'replisome_trimers': ('bulk',),\n",
    "            'replisome_monomers': ('bulk',),\n",
    "            'dntps': ('bulk',),\n",
    "            'ppi': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'environment': ('environment',)\n",
    "        }\n",
    "    },\n",
    "    'unique': {\n",
    "        'chromosome_domain': {},\n",
    "        'full_chromosome': {},\n",
    "        'oriC': {},\n",
    "        'active_replisome': {},\n",
    "        'RNA': {},\n",
    "        'active_ribosome': {},\n",
    "        'DnaA_box': {},\n",
    "        'promoter': {},\n",
    "    },\n",
    "    'bulk': {},\n",
    "    'environment': {},\n",
    "}\n",
    "node_groups = [[ #line up the top nodes for visual effect\n",
    "        ('unique',), ('bulk',), ('environment',), \n",
    "    ],\n",
    "]\n",
    "# plot settings\n",
    "plot_settings_ecoli = {\n",
    "    'remove_process_place_edges': True,\n",
    "    'rankdir': 'RL',\n",
    "    'dpi': '100',\n",
    "    'out_dir': 'out',\n",
    "}\n",
    "fig = plot_bigraph(ecoli, **plot_settings_ecoli, node_groups=node_groups, filename='ecoli')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54e1102d-521c-49cc-8ed2-be26c2b84c3d",
   "metadata": {
    "tags": []
   },
   "source": [
    "[See E. coli wiring diagram](https://github.com/vivarium-collective/bigraph-viz/blob/main/doc/_static/ecoli.png?raw=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e5848a-e92f-45f3-b8a4-30320268640c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
