{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5338cfab-cf15-4c63-a091-c1cbb51bf8dd",
   "metadata": {},
   "source": [
    "# Process Bigraph Schema\n",
    "This tutorial covers Process Bigraph Schema, starting with single `store` and increasing in detail as you scroll towards the bottom. It uses the libraries [bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) for the schema validation and type registry and [bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) for visualization.\n",
    "\n",
    "## Bigraph-schema and Bigraph-viz\n",
    "\n",
    "[Bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) is a serializable type schema for Process Bigraph simulations. From this library, we will use the `validate_schema` method to check our schemas, and `fill` to fill in missing details to prepare the schema for simulation. We will also use the `type_registry` to access different schemas, and register new ones.\n",
    "\n",
    "[Bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) is an easy-to-use plotting tool for compositional bigraph schema, built on top of Graphviz. It reads raw Compositional Bigraph Schemas and generates a figure that can be displayed in a notebook or saved to file. Bigraph-viz's main plotting function is `plot_bigraph` for the bigraph structure. `plot_flow` and `plot_multitimestep` make plots of process orchestration."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df69e36f-3da8-4bfb-8bc0-0aeb04a6f7d7",
   "metadata": {},
   "source": [
    "### Installation"
   ]
  },
  {
   "cell_type": "code",
   "id": "63fb7d1a-cd49-4aac-8d4f-31bb54496a51",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.215613Z",
     "start_time": "2025-02-12T16:39:54.839523Z"
    }
   },
   "source": [
    "# !pip install bigraph-schema\n",
    "# !pip install bigraph-viz\n",
    "!pip freeze | grep bigraph"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-e git+https://github.com/vivarium-collective/bigraph-schema.git@c0f51b28193de6ec66a9c95a5e8f0291f55ab2ca#egg=bigraph_schema\r\n",
      "-e git+https://github.com/vivarium-collective/bigraph-viz.git@a5a42e1bd20b47a04268a6708de7c8fc20508449#egg=bigraph_viz\r\n",
      "-e git+https://github.com/vivarium-collective/process-bigraph.git@c0940409886b0ae97982ea949c3e384b331f500c#egg=process_bigraph\r\n"
     ]
    }
   ],
   "execution_count": 39
  },
  {
   "cell_type": "markdown",
   "id": "f700afa2-bfee-49bf-a4b1-e23f0540461d",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "id": "2ab3cb32-e007-45c4-807e-b1b442ac1608",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.221717Z",
     "start_time": "2025-02-12T16:39:55.219533Z"
    }
   },
   "source": [
    "# from bigraph_viz import plot_bigraph, plot_flow, plot_multitimestep, pf\n",
    "from bigraph_viz import plot_bigraph, replace_regex_recursive, VisualizeTypes\n",
    "\n",
    "plot_settings = {\n",
    "    'remove_process_place_edges': True\n",
    "}\n",
    "save_images = False\n",
    "if save_images:\n",
    "    plot_settings.update({'out_dir': 'out','dpi': '250'})"
   ],
   "outputs": [],
   "execution_count": 40
  },
  {
   "cell_type": "markdown",
   "id": "0e716c23-5fbe-4c08-a868-09611ca40ec4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Stores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a6e9cc2-e0df-4f73-b4c3-eddb1055b088",
   "metadata": {},
   "source": [
    "### Simple stores\n",
    "A store can be declared in a JSON dictionary with the store's name mapped to its `value`. "
   ]
  },
  {
   "cell_type": "code",
   "id": "d711b952-089a-4938-817d-d1ccad9de56e",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.263159Z",
     "start_time": "2025-02-12T16:39:55.227240Z"
    }
   },
   "source": [
    "simple_store_state = {\n",
    "    'store1': 1.0,\n",
    "}\n",
    "plot_bigraph(simple_store_state,\n",
    "             **plot_settings,\n",
    "             show_values=True,\n",
    "             filename='simple_store')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/simple_store\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"74pt\" height=\"74pt\"\n viewBox=\"0.00 0.00 76.40 76.40\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 72.4)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-72.4 72.4,-72.4 72.4,4 -4,4\"/>\n<!-- (&#39;store1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"34.2\" cy=\"-34.2\" rx=\"34.2\" ry=\"34.2\"/>\n<text text-anchor=\"start\" x=\"8.87\" y=\"-30.6\" font-family=\"Times,serif\" font-size=\"12.00\">store1: 1.0</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x10fe361f0>"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 41
  },
  {
   "cell_type": "markdown",
   "id": "db1cead7-e691-4291-920e-827d69c00727",
   "metadata": {},
   "source": [
    "To include values *and* types in a node label, you can instead have the store name map to a dict with `_value` and `_type` keys."
   ]
  },
  {
   "cell_type": "code",
   "id": "f3e5fa71-7ffc-4a8f-a1a0-dc547c886f84",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.306861Z",
     "start_time": "2025-02-12T16:39:55.270824Z"
    }
   },
   "source": [
    "typed_store_spec = {\n",
    "    'store1': {\n",
    "        '_value': 1.0, \n",
    "        '_type': 'float',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(typed_store_spec,\n",
    "             **plot_settings,\n",
    "             show_values=True,\n",
    "             show_types=True,\n",
    "             filename='store')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/store\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"58pt\" height=\"58pt\"\n viewBox=\"0.00 0.00 59.66 59.66\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 55.66)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-55.66 55.66,-55.66 55.66,4 -4,4\"/>\n<!-- (&#39;store1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"25.83\" cy=\"-25.83\" rx=\"25.83\" ry=\"25.83\"/>\n<text text-anchor=\"start\" x=\"11.16\" y=\"-28.23\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n<text text-anchor=\"start\" x=\"11.16\" y=\"-16.23\" font-family=\"Times,serif\" font-size=\"12.00\">[float]</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x11940ca00>"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 42
  },
  {
   "cell_type": "markdown",
   "id": "de1011cb-903b-4c15-baed-a62f4afdd6b9",
   "metadata": {},
   "source": [
    "### Hierarchy\n",
    "A hierarchy is a place graph of nested stores. Stores can be placed within stores using JSON dictionary nesting."
   ]
  },
  {
   "cell_type": "code",
   "id": "0f3f0824-f89d-4c6e-830e-2c5a321f7b33",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.352787Z",
     "start_time": "2025-02-12T16:39:55.315674Z"
    }
   },
   "source": [
    "hierarchy_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': {\n",
    "            'store1.1.1': 'any',\n",
    "            'store1.1.2': 'any',\n",
    "            'store1.1.3': {\n",
    "                'store1.1.3.1': 'any',\n",
    "            },\n",
    "        },\n",
    "        'store1.2': 'any',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(hierarchy_spec,\n",
    "             **plot_settings,\n",
    "             filename='hierarchy')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/hierarchy\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"235pt\" height=\"343pt\"\n viewBox=\"0.00 0.00 241.78 352.56\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 348.56)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-348.56 237.78,-348.56 237.78,4 -4,4\"/>\n<!-- (&#39;store1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"155.55\" cy=\"-322.97\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"140.89\" y=\"-319.37\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"119.55\" cy=\"-238.47\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"100.39\" y=\"-234.87\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n</g>\n<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M147.02,-302.42C141.99,-290.88 135.57,-276.19 130.21,-263.9\"/>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.2&#39;) -->\n<g id=\"node7\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"191.55\" cy=\"-238.47\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"172.39\" y=\"-234.87\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n</g>\n<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M164.08,-302.42C169.11,-290.88 175.52,-276.19 180.89,-263.9\"/>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"37.55\" cy=\"-143.33\" rx=\"32.23\" ry=\"32.23\"/>\n<text text-anchor=\"start\" x=\"13.89\" y=\"-139.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M101.76,-217.26C89.13,-202.92 72.13,-183.61 58.8,-168.46\"/>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"119.55\" cy=\"-143.33\" rx=\"32.23\" ry=\"32.23\"/>\n<text text-anchor=\"start\" x=\"95.89\" y=\"-139.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M119.55,-211.03C119.55,-200.27 119.55,-187.74 119.55,-176.46\"/>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"201.55\" cy=\"-143.33\" rx=\"32.23\" ry=\"32.23\"/>\n<text text-anchor=\"start\" x=\"177.89\" y=\"-139.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.2</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M137.34,-217.26C149.97,-202.92 166.97,-183.61 180.3,-168.46\"/>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"37.55\" cy=\"-37.55\" rx=\"37.55\" ry=\"37.55\"/>\n<text text-anchor=\"start\" x=\"9.39\" y=\"-33.95\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3.1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M37.55,-110.34C37.55,-99.5 37.55,-87.3 37.55,-76.07\"/>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x10fe37a60>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 43
  },
  {
   "cell_type": "markdown",
   "id": "c825e07c-8cd5-4df1-8958-aa0f36d61045",
   "metadata": {},
   "source": [
    "## Processes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ea1a83-aeac-4048-9593-7b30588b0ad8",
   "metadata": {},
   "source": [
    "### Single process\n",
    "Processes require ports, which are declared with the `_ports` key mapped to a dict that has port names matched to types."
   ]
  },
  {
   "cell_type": "code",
   "id": "7c74237b-0d66-4e8f-a8ca-79174fadad8b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.400009Z",
     "start_time": "2025-02-12T16:39:55.363037Z"
    }
   },
   "source": [
    "process_spec = {\n",
    "    'process1': {\n",
    "        '_type': 'edge',\n",
    "        '_inputs': {\n",
    "            'port1': 'any',\n",
    "            'port2': 'any',\n",
    "        },\n",
    "    },\n",
    "}\n",
    "plot_bigraph(process_spec,\n",
    "             **plot_settings,\n",
    "             rankdir='RL',\n",
    "             filename='process')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/process\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"142pt\" height=\"78pt\"\n viewBox=\"0.00 0.00 146.26 80.15\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 76.15)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-76.15 142.26,-76.15 142.26,4 -4,4\"/>\n<!-- (&#39;process1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;process1&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-53.58 0,-53.58 0,-17.58 54,-17.58 54,-53.58\"/>\n<text text-anchor=\"start\" x=\"6.01\" y=\"-31.98\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M110.78,-55.47C99.18,-52.69 81.54,-48.45 65.52,-44.6\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"66.8,-41.3 56.26,-42.37 65.17,-48.11 66.8,-41.3\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-52.58\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_input -->\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_input&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_input&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M110.78,-16.54C99.29,-19.19 81.88,-23.19 65.99,-26.84\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"65.23,-23.43 56.26,-29.08 66.79,-30.25 65.23,-23.43\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-26.58\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x11940c640>"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 44
  },
  {
   "cell_type": "markdown",
   "id": "68a2c480-32fe-43d3-b618-8fc6c647b8ca",
   "metadata": {},
   "source": [
    "### Multiple processes"
   ]
  },
  {
   "cell_type": "code",
   "id": "8c2457a3-fc47-4850-8b9e-9ee98e81567d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.441400Z",
     "start_time": "2025-02-12T16:39:55.403050Z"
    }
   },
   "source": [
    "process_schema = {\n",
    "    '_type': 'process',\n",
    "    '_inputs': {\n",
    "        'port1': 'Any',\n",
    "    },\n",
    "    '_outputs': {\n",
    "        'port2': 'Any'\n",
    "    },\n",
    "}\n",
    "\n",
    "processes_spec = {\n",
    "    'process1': process_schema,\n",
    "    'process2': process_schema,\n",
    "    'process3': process_schema,\n",
    "}\n",
    "plot_bigraph(processes_spec,\n",
    "             **plot_settings,\n",
    "             rankdir='BT',\n",
    "             filename='multiple_processes')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/multiple_processes\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"266pt\" height=\"114pt\"\n viewBox=\"0.00 0.00 273.15 117.15\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 113.15)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-113.15 269.15,-113.15 269.15,4 -4,4\"/>\n<!-- (&#39;process1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;process1&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"80.58,-109.15 26.58,-109.15 26.58,-73.15 80.58,-73.15 80.58,-109.15\"/>\n<text text-anchor=\"start\" x=\"32.58\" y=\"-87.55\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;process2&#39;,) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;process2&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"170.58,-109.15 116.58,-109.15 116.58,-73.15 170.58,-73.15 170.58,-109.15\"/>\n<text text-anchor=\"start\" x=\"122.58\" y=\"-87.55\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;process3&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;process3&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"251.58,-109.15 197.58,-109.15 197.58,-73.15 251.58,-73.15 251.58,-109.15\"/>\n<text text-anchor=\"start\" x=\"203.58\" y=\"-87.55\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n</g>\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M18.16,-27.29C21.26,-35.43 25.66,-46.13 30.47,-55.15 31.81,-57.68 33.29,-60.26 34.83,-62.82\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"31.78,-64.55 40.1,-71.11 37.69,-60.79 31.78,-64.55\"/>\n<text text-anchor=\"start\" x=\"30.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M115.64,-27.61C117.72,-35.69 120.76,-46.2 124.47,-55.15 125.44,-57.49 126.52,-59.88 127.66,-62.25\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"124.49,-63.74 132.2,-71.01 130.71,-60.52 124.49,-63.74\"/>\n<text text-anchor=\"start\" x=\"123.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process3&#39;,) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process3&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M205.92,-27.92C205.84,-35.92 206.3,-46.25 208.47,-55.15 209.02,-57.44 209.73,-59.76 210.54,-62.05\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"207.21,-63.16 214.26,-71.07 213.68,-60.49 207.21,-63.16\"/>\n<text text-anchor=\"start\" x=\"207.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output -->\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M56.94,-39.31C56.23,-50.04 55.41,-62.38 54.76,-72.21\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"60.41,-39.81 57.58,-29.6 53.43,-39.35 60.41,-39.81\"/>\n<text text-anchor=\"start\" x=\"55.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output -->\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M153.05,-39.04C151.04,-49.88 148.72,-62.39 146.88,-72.33\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"156.48,-39.7 154.86,-29.23 149.6,-38.42 156.48,-39.7\"/>\n<text text-anchor=\"start\" x=\"151.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output -->\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process3&#39;,) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process3&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M243.21,-37.99C239.28,-49 234.66,-61.93 231,-72.16\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"246.42,-39.42 246.49,-28.82 239.83,-37.06 246.42,-39.42\"/>\n<text text-anchor=\"start\" x=\"239.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x11943bd00>"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 45
  },
  {
   "cell_type": "markdown",
   "id": "15e5f5a6-8dd4-41c3-ba37-33ed9e3e0830",
   "metadata": {},
   "source": [
    "## Wires\n",
    "To connect a process's ports to stores, add wiring with a `_wires` key that maps port names to relative paths in the store hierarchy. The ports used by the wires need to match the ports in the schema."
   ]
  },
  {
   "cell_type": "code",
   "id": "a6ecad5d-8948-4636-910c-79284a3e8679",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.486556Z",
     "start_time": "2025-02-12T16:39:55.448225Z"
    }
   },
   "source": [
    "connected_process_spec = {\n",
    "    'process1': {\n",
    "        '_type': 'edge',\n",
    "        '_inputs': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'integer',\n",
    "            'port3': 'integer',\n",
    "        },\n",
    "        'inputs': {\n",
    "            'port1': ['node1'],\n",
    "            'port2': ['node2'],\n",
    "            'port3': ['node3'],\n",
    "        }\n",
    "    },\n",
    "    'node1': 'float',  # TODO -- shouldn't these fill?\n",
    "    'node2': 'integer',\n",
    "    'node3': 'integer',\n",
    "}\n",
    "plot_bigraph(connected_process_spec,\n",
    "             **plot_settings,\n",
    "             filename='wires')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/wires\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"168pt\" height=\"129pt\"\n viewBox=\"0.00 0.00 173.18 133.18\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 129.18)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-129.18 169.18,-129.18 169.18,4 -4,4\"/>\n<!-- (&#39;node1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;node1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"21.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"6.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node1</text>\n</g>\n<!-- (&#39;process1&#39;,) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;process1&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"109.59,-36 55.59,-36 55.59,0 109.59,0 109.59,-36\"/>\n<text text-anchor=\"start\" x=\"61.6\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M34.23,-85.27C42.54,-73.88 53.59,-58.75 63.02,-45.81\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"65.59,-48.24 68.65,-38.1 59.93,-44.11 65.59,-48.24\"/>\n<text text-anchor=\"start\" x=\"55.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;node2&#39;,) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;node2&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"82.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"67.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node2</text>\n</g>\n<!-- (&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M82.59,-81.06C82.59,-71.08 82.59,-59.07 82.59,-48.28\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"86.09,-48.3 82.59,-38.3 79.09,-48.3 86.09,-48.3\"/>\n<text text-anchor=\"start\" x=\"82.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;node3&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;node3&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"128.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node3</text>\n</g>\n<!-- (&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M130.95,-85.27C122.64,-73.88 111.59,-58.75 102.16,-45.81\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"105.25,-44.11 96.53,-38.1 99.59,-48.24 105.25,-44.11\"/>\n<text text-anchor=\"start\" x=\"114.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port3</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1194907c0>"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 46
  },
  {
   "cell_type": "markdown",
   "id": "e037285d-116a-404d-96ec-ba47318a2e8b",
   "metadata": {},
   "source": [
    "### Advanced wiring\n",
    "**TODO**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b17201-2489-45c8-a06b-9fdc29a714a7",
   "metadata": {},
   "source": [
    "## Composites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf5d428-db23-4985-80ff-6b93d22973aa",
   "metadata": {},
   "source": [
    "### Flat composite\n",
    "A flat composite connects stores with multiple connected processes, without nesting."
   ]
  },
  {
   "cell_type": "code",
   "id": "781a84e6-bbe2-4708-b441-47efc611b46f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.532550Z",
     "start_time": "2025-02-12T16:39:55.492685Z"
    }
   },
   "source": [
    "flat_composite_spec = {\n",
    "    'store1.1': 'float',\n",
    "    'store1.2': 'integer',\n",
    "    'process1': {\n",
    "        '_type': 'process',\n",
    "        '_outputs': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'integer',\n",
    "        },\n",
    "        'outputs': {\n",
    "            'port1': ['store1.1'],\n",
    "            'port2': ['store1.2'],\n",
    "        }\n",
    "    },\n",
    "    'process2': {\n",
    "        '_type': 'process',\n",
    "        '_inputs': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'integer',\n",
    "        },\n",
    "        'inputs': {\n",
    "            'port1': ['store1.1'],\n",
    "            'port2': ['store1.2'],\n",
    "        }\n",
    "    },\n",
    "}\n",
    "plot_bigraph(flat_composite_spec,\n",
    "             **plot_settings,\n",
    "             rankdir='RL',\n",
    "             filename='flat_composite')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/flat_composite\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"168pt\" height=\"137pt\"\n viewBox=\"0.00 0.00 172.93 140.82\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 136.82)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-136.82 168.93,-136.82 168.93,4 -4,4\"/>\n<!-- (&#39;store1.1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store1.1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"138.02\" cy=\"-105.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"118.86\" y=\"-102.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n</g>\n<!-- (&#39;process1&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;process1&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-118.91 0,-118.91 0,-82.91 54,-82.91 54,-118.91\"/>\n<text text-anchor=\"start\" x=\"6.01\" y=\"-97.31\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M99.1,-104.17C84.49,-103.5 68.12,-102.75 54.77,-102.14\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"98.69,-107.66 108.84,-104.62 99.01,-100.66 98.69,-107.66\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-104.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process2&#39;,) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;process2&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-53.91 0,-53.91 0,-17.91 54,-17.91 54,-53.91\"/>\n<text text-anchor=\"start\" x=\"6.01\" y=\"-32.31\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M114.43,-91.41C99.94,-82.1 80.79,-69.81 64.21,-59.16\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"66.37,-56.39 56.07,-53.93 62.59,-62.28 66.37,-56.39\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-78.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;store1.2&#39;,) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;store1.2&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"138.02\" cy=\"-26.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"118.86\" y=\"-23.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n</g>\n<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M100.28,-36.89C90.61,-40.45 80.51,-45.07 72,-50.91 59.8,-59.28 48.8,-71.78 40.85,-82.19\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"101.12,-40.3 109.48,-33.79 98.88,-33.67 101.12,-40.3\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-52.91\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.91,-17.52C99.8,-14.2 85.1,-11.94 72,-14.91 69.83,-15.4 67.65,-16.01 65.47,-16.71\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"64.26,-13.43 56.19,-20.29 66.78,-19.96 64.26,-13.43\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-16.91\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x119490a60>"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 47
  },
  {
   "cell_type": "markdown",
   "id": "398509c7-cf5c-4cce-a179-3a39752b0dde",
   "metadata": {},
   "source": [
    "### Nested composite\n",
    "A nested composite has store hierarchies with multiple connected processes."
   ]
  },
  {
   "cell_type": "code",
   "id": "701e759d-7f9b-41bb-911c-a8208c892462",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.578742Z",
     "start_time": "2025-02-12T16:39:55.538756Z"
    }
   },
   "source": [
    "nested_composite_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': 'float',\n",
    "        'store1.2': 'integer',\n",
    "        'process1': {\n",
    "            '_type': 'process',\n",
    "            '_inputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'inputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'process',\n",
    "            '_outputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'outputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "    },\n",
    "    'process3': {\n",
    "        '_type': 'process',\n",
    "        '_inputs': {\n",
    "            'port1': 'tree',\n",
    "        },\n",
    "        'inputs': {\n",
    "            'port1': ['store1'],\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(nested_composite_spec, \n",
    "             **plot_settings,\n",
    "             filename='nested_composite')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/nested_composite\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"203pt\" height=\"227pt\"\n viewBox=\"0.00 0.00 209.18 233.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 229)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-229 205.18,-229 205.18,4 -4,4\"/>\n<!-- (&#39;store1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"102.18\" cy=\"-203.41\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"87.52\" y=\"-199.81\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"102.18\" cy=\"-108.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"83.02\" y=\"-105.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n</g>\n<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M102.18,-180.96C102.18,-167.72 102.18,-150.64 102.18,-136.54\"/>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.2&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"30.18\" cy=\"-108.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"11.02\" y=\"-105.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n</g>\n<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M88.98,-185.44C77.04,-170.11 59.43,-147.48 46.59,-130.99\"/>\n</g>\n<!-- (&#39;process3&#39;,) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;process3&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"201.18,-126.91 147.18,-126.91 147.18,-90.91 201.18,-90.91 201.18,-126.91\"/>\n<text text-anchor=\"start\" x=\"153.19\" y=\"-105.31\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n</g>\n<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;process3&#39;,) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;store1&#39;,)&#45;&gt;(&#39;process3&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M115.39,-185.44C126.04,-171.75 141.23,-152.25 153.49,-136.49\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"156.01,-138.95 159.39,-128.91 150.49,-134.65 156.01,-138.95\"/>\n<text text-anchor=\"start\" x=\"138.18\" y=\"-155.82\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;process1&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;process1&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"57.18,-36 3.18,-36 3.18,0 57.18,0 57.18,-36\"/>\n<text text-anchor=\"start\" x=\"9.19\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M88.31,-84.86C82,-75.01 74.17,-63.6 66.18,-54 63.65,-50.96 60.89,-47.9 58.05,-44.92\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"60.6,-42.52 51.07,-37.89 55.64,-47.45 60.6,-42.52\"/>\n<text text-anchor=\"start\" x=\"74.18\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;process2&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;store1&#39;, &#39;process2&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"144.18,-36 90.18,-36 90.18,0 144.18,0 144.18,-36\"/>\n<text text-anchor=\"start\" x=\"96.19\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M109.11,-70.22C109.48,-68.12 109.84,-66.03 110.18,-64 111.68,-55.11 113.23,-45.22 114.49,-36.97\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"105.68,-69.51 107.38,-79.97 112.58,-70.74 105.68,-69.51\"/>\n<text text-anchor=\"start\" x=\"112.18\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M12.94,-87.3C8.14,-80.34 3.55,-72.23 1.07,-64 -0.77,-57.88 0.47,-51.79 3.28,-46.18\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"6.17,-48.16 8.85,-37.91 0.36,-44.25 6.17,-48.16\"/>\n<text text-anchor=\"start\" x=\"1.18\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n<g id=\"edge7\" class=\"edge\">\n<title>(&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M33.12,-70C34.85,-64.23 37.39,-58.65 41.07,-54 47.53,-45.85 70.52,-35.88 89.56,-28.67\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"29.72,-69.15 31.02,-79.67 36.57,-70.64 29.72,-69.15\"/>\n<text text-anchor=\"start\" x=\"41.18\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1194796a0>"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 48
  },
  {
   "cell_type": "markdown",
   "id": "0c8c1704-2331-44eb-bd05-72e36ccb6858",
   "metadata": {},
   "source": [
    "### Composite process\n",
    "Composite processes are processes with internal stores and internal processes, which can run as their own simulation. The composite process has schema, allowing it to connect to a super-simulation. This allows for improved distributed computation, with composite processes able to run on their own computer and syncronize to the super-simulation via message passing."
   ]
  },
  {
   "cell_type": "code",
   "id": "638c5dc7-4cfd-46e4-b66a-8c42ca98ff1d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.624028Z",
     "start_time": "2025-02-12T16:39:55.584995Z"
    }
   },
   "source": [
    "composite_process_spec = {\n",
    "    'composite_process': {\n",
    "        '_type': 'composite',\n",
    "        'store1.1': 'any',\n",
    "        'store1.2': 'any',\n",
    "        'process1': {\n",
    "            '_type': 'process',\n",
    "            '_outputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'outputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'process',\n",
    "            '_outputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'outputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "        # '_inputs': {\n",
    "        #     'port1': 'Any', \n",
    "        #     'port2': 'Any', \n",
    "        # },\n",
    "        # '_tunnels': {\n",
    "        #     'port1': 'store1.1',\n",
    "        #     'port2': 'store1.2',\n",
    "        # }\n",
    "    }\n",
    "}\n",
    "\n",
    "plot_bigraph(\n",
    "    composite_process_spec, \n",
    "    **plot_settings, filename='composite_process')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/composite_process\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"148pt\" height=\"211pt\"\n viewBox=\"0.00 0.00 152.58 216.82\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 212.82)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-212.82 148.58,-212.82 148.58,4 -4,4\"/>\n<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"30.58\" cy=\"-108.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"11.42\" y=\"-105.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;process1&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"57.58,-36 3.58,-36 3.58,0 57.58,0 57.58,-36\"/>\n<text text-anchor=\"start\" x=\"9.59\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M7.32,-77.59C4.93,-73.22 2.87,-68.63 1.48,-64 -1.48,-54.2 3.48,-44.51 10.1,-36.65\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"4.3,-79.36 12.55,-86.02 10.25,-75.67 4.3,-79.36\"/>\n<text text-anchor=\"start\" x=\"1.58\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;process2&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"144.58,-36 90.58,-36 90.58,0 144.58,0 144.58,-36\"/>\n<text text-anchor=\"start\" x=\"96.59\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M33.53,-70C35.26,-64.23 37.79,-58.65 41.48,-54 47.93,-45.85 70.92,-35.88 89.96,-28.67\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"30.13,-69.15 31.43,-79.67 36.97,-70.64 30.13,-69.15\"/>\n<text text-anchor=\"start\" x=\"41.58\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"102.58\" cy=\"-108.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"83.42\" y=\"-105.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M82.4,-75.27C77.5,-68.06 72.08,-60.6 66.58,-54 61.68,-48.1 55.89,-42.13 50.41,-36.82\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"79.45,-77.17 87.88,-83.59 85.3,-73.32 79.45,-77.17\"/>\n<text text-anchor=\"start\" x=\"73.58\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M109.52,-70.22C109.88,-68.12 110.24,-66.03 110.58,-64 112.09,-55.11 113.64,-45.22 114.9,-36.97\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"106.09,-69.51 107.78,-79.97 112.98,-70.74 106.09,-69.51\"/>\n<text text-anchor=\"start\" x=\"111.58\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;composite_process&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;composite_process&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"114.84,-208.82 16.33,-208.82 16.33,-172.82 114.84,-172.82 114.84,-208.82\"/>\n<text text-anchor=\"start\" x=\"19.93\" y=\"-187.22\" font-family=\"Times,serif\" font-size=\"12.00\">composite_process</text>\n</g>\n<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M57.82,-172.08C52.98,-161.05 46.7,-146.7 41.38,-134.56\"/>\n</g>\n<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M73.8,-172.08C78.95,-160.96 85.66,-146.46 91.31,-134.26\"/>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x119479d90>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 49
  },
  {
   "cell_type": "markdown",
   "id": "847cef26-65a1-4da6-9dd3-eb5839277970",
   "metadata": {},
   "source": [
    "## Orchestration"
   ]
  },
  {
   "cell_type": "code",
   "id": "e76c97a8-3037-464a-929e-ba4414c73c34",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.631727Z",
     "start_time": "2025-02-12T16:39:55.629792Z"
    }
   },
   "source": [
    "plot_settings2={}\n",
    "if save_images:\n",
    "    plot_settings2.update({\n",
    "        'out_dir': 'out',\n",
    "        'dpi': '250'\n",
    "    })"
   ],
   "outputs": [],
   "execution_count": 50
  },
  {
   "cell_type": "markdown",
   "id": "48975890-06a4-462a-9b1d-6bddd51779fe",
   "metadata": {},
   "source": [
    "### Multi-timestepping\n",
    "\n",
    "Temporal processes can be run by multi-timestepping, with each process updating the state according to its `sync_state` state, which it can update adaptively."
   ]
  },
  {
   "cell_type": "code",
   "id": "f0eff7b0-f0e2-4d56-8b16-ef0fc1ce5aba",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.638106Z",
     "start_time": "2025-02-12T16:39:55.636671Z"
    }
   },
   "source": [
    "# multitimestep_spec = {\n",
    "#     'temporal process2': {\n",
    "#         '_ports': {'port1': 'Any'},\n",
    "#         'wires': {'port1': 'state'},\n",
    "#         '_sync_step': 1.0,\n",
    "#     },\n",
    "#     'temporal process1': {\n",
    "#         '_ports': {'port1': 'Any'},\n",
    "#         'wires': {'port1': 'state'},\n",
    "#         '_sync_step': 0.5,\n",
    "#     },\n",
    "# }\n",
    "# multitimestep_spec = replace_regex_recursive(multitimestep_spec)\n",
    "# plot_multitimestep(multitimestep_spec, total_time=2.0, **plot_settings2, filename='multitimestep')"
   ],
   "outputs": [],
   "execution_count": 51
  },
  {
   "cell_type": "markdown",
   "id": "a00cf677-ebac-4309-8ac9-a5b2855724af",
   "metadata": {},
   "source": [
    "### Flows\n",
    "\n",
    "A directed acyclic graph (DAG) – a type of workflow that declares the dependencies between processes. A flow run is not temporal, and the order is entirely based on dependencies. It can triggered between time steps, running the steps in the order as determined by their dependencies."
   ]
  },
  {
   "cell_type": "code",
   "id": "db45c99b-b97b-47bc-a73e-c669882f3969",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.644165Z",
     "start_time": "2025-02-12T16:39:55.642744Z"
    }
   },
   "source": [
    "# flow = {\n",
    "#     'step1': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': [], \n",
    "#     },\n",
    "#     'step2': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': 'step1', \n",
    "#     },\n",
    "#     'step3': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': [], \n",
    "#     },\n",
    "#     'step4': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': ['step2', 'step3'], \n",
    "#     },\n",
    "# }\n",
    "\n",
    "# plot_flow(flow, **plot_settings2, filename='flow')"
   ],
   "outputs": [],
   "execution_count": 52
  },
  {
   "cell_type": "markdown",
   "id": "5882581a-c5ef-40b5-acbd-efddad269807",
   "metadata": {},
   "source": [
    "## Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff81178f-36e4-4200-982a-ed8155b87934",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Multiscale map of a cell\n",
    "\n",
    "Bigraphs can represent the hierarchical nesting of cellular structures. Below is a multiscale map of a cell from [Qin et al. A multi-scale map of cell structure fusing protein images and interactions. Nature. 2021.](https://pubmed.ncbi.nlm.nih.gov/34819669/)"
   ]
  },
  {
   "cell_type": "code",
   "id": "47bdf8e5-0bdd-4cac-b57e-2449c75fa178",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.735820Z",
     "start_time": "2025-02-12T16:39:55.649145Z"
    }
   },
   "source": [
    "music_map = {\n",
    "    'cell': {\n",
    "        'cytoplasm': {\n",
    "            'secretory organelles': {\n",
    "                'transmembrane transport systems': {\n",
    "                    'ion transmembrane transport systems': {}}},\n",
    "            'cytoplasmic organelles': {\n",
    "                'subgroup of cytoplasmic organelles': {\n",
    "                    'metabolic organelles': {\n",
    "                        'subgroup of metabolic organelles': {},\n",
    "                        'mitochondrion': {}},\n",
    "                    'ribonucleoprotein complex family': {\n",
    "                        'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
    "                        'ribosome biogenesis community': {\n",
    "                            'ribosome': {\n",
    "                                'ribosomal subunit': {\n",
    "                                    'mitochondirial large ribosomal subunit': {\n",
    "                                        # 'large ribosomal subunit subcomplex 1': {}\n",
    "                                    },\n",
    "                                    'mito-cyto ribosomal cluster': {},\n",
    "                                    'cotranslational protein targeting to membrane system': {},\n",
    "                                    'ribosomal complex 5': {},\n",
    "                                    'ribosomal complex 2': {}}},\n",
    "                            'nucleic acid binding complex': {}}}}},\n",
    "            'transcriptional regulation complex family': {\n",
    "                'transcriptional regulation complexes': {}}},\n",
    "        'nucleus': {\n",
    "            'DNA metabolic assembly': {},\n",
    "            'nuclear lumen': {\n",
    "                'nucleolus': {\n",
    "                    'ribonucleprotein complex family': {},\n",
    "                    'RNA processing complex family': {\n",
    "                        'RNA processing complex 1': {},\n",
    "                        'RNA splicing complex family': {}},\n",
    "                    'splicosomal complex family': {}},\n",
    "                'nuclear splicing speckle': {}},\n",
    "            'nuclear body': {},\n",
    "            'nucleo-plasm 1': {\n",
    "                'chromosome organization complex family': {\n",
    "                    'chromatin organization complex family': {\n",
    "                        'HAT complex family': {\n",
    "                            'NuA4 HAT complex': {}}}}},\n",
    "            'nucleo-plasm 2': {\n",
    "                'chromatin regulation complex': {}},\n",
    "            'nuclear transcriptional speckle': {\n",
    "                'negative regulation of RNA biosynthesis process': {}}}}}\n",
    "music_map1 = replace_regex_recursive(music_map)\n",
    "plot_bigraph(music_map1, **plot_settings, filename='cell_hierarchy')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell_hierarchy\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"705pt\" height=\"700pt\"\n viewBox=\"0.00 0.00 724.95 720.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.629041 0.629041) rotate(0) translate(4 1108.81)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-1108.81 1116.45,-1108.81 1116.45,4 -4,4\"/>\n<!-- (&#39;cell&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;cell&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"482.39\" cy=\"-1086.81\" rx=\"18\" ry=\"18\"/>\n<text text-anchor=\"start\" x=\"473.73\" y=\"-1083.21\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"325.39\" cy=\"-999.39\" rx=\"33.41\" ry=\"33.41\"/>\n<text text-anchor=\"start\" x=\"300.73\" y=\"-995.79\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n</g>\n<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M466.18,-1076.99C440.02,-1062.75 388.08,-1034.49 355.12,-1016.57\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;) -->\n<g id=\"node25\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"737.39\" cy=\"-999.39\" rx=\"25.92\" ry=\"25.92\"/>\n<text text-anchor=\"start\" x=\"719.07\" y=\"-995.79\" font-family=\"Times,serif\" font-size=\"12.00\">nucleus</text>\n</g>\n<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;) -->\n<g id=\"edge24\" class=\"edge\">\n<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M499.92,-1079.93C544.26,-1065.09 660.98,-1025.99 712.43,-1008.76\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"154.39\" cy=\"-875.88\" rx=\"39.81\" ry=\"39.81\"/>\n<text text-anchor=\"start\" x=\"132.41\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">secretory</text>\n<text text-anchor=\"start\" x=\"129.84\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M295.74,-982.85C272.07,-969.92 238.5,-950.39 211.39,-929.98 201.45,-922.5 191.3,-913.54 182.36,-905.13\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"266.39\" cy=\"-875.88\" rx=\"46.09\" ry=\"46.09\"/>\n<text text-anchor=\"start\" x=\"237.4\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n<text text-anchor=\"start\" x=\"241.84\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M310.81,-968.36C303.44,-953.18 294.38,-934.53 286.37,-918.03\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node23\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"384.39\" cy=\"-875.88\" rx=\"54.1\" ry=\"54.1\"/>\n<text text-anchor=\"start\" x=\"349.74\" y=\"-890.28\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n<text text-anchor=\"start\" x=\"360.07\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"363.73\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"368.73\" y=\"-854.28\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge22\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M339.98,-968.36C346.26,-955.43 353.76,-939.98 360.8,-925.49\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"87.39\" cy=\"-728.39\" rx=\"57.39\" ry=\"57.39\"/>\n<text text-anchor=\"start\" x=\"50.41\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n<text text-anchor=\"start\" x=\"66.07\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n<text text-anchor=\"start\" x=\"68.39\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M137.83,-838.92C129.85,-821.58 120.13,-800.47 111.35,-781.42\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"57.39\" cy=\"-569.58\" rx=\"57.39\" ry=\"57.39\"/>\n<text text-anchor=\"start\" x=\"49.73\" y=\"-583.98\" font-family=\"Times,serif\" font-size=\"12.00\">ion</text>\n<text text-anchor=\"start\" x=\"20.41\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n<text text-anchor=\"start\" x=\"36.07\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n<text text-anchor=\"start\" x=\"38.39\" y=\"-547.98\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M76.65,-671.22C73.89,-656.81 70.93,-641.3 68.17,-626.88\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node7\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"266.39\" cy=\"-728.39\" rx=\"46.09\" ry=\"46.09\"/>\n<text text-anchor=\"start\" x=\"244.06\" y=\"-742.79\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n<text text-anchor=\"start\" x=\"261.4\" y=\"-730.79\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n<text text-anchor=\"start\" x=\"237.4\" y=\"-718.79\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n<text text-anchor=\"start\" x=\"241.84\" y=\"-706.79\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M266.39,-828.85C266.39,-811.7 266.39,-792.25 266.39,-775.12\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node8\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"172.39\" cy=\"-569.58\" rx=\"39.81\" ry=\"39.81\"/>\n<text text-anchor=\"start\" x=\"148.74\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n<text text-anchor=\"start\" x=\"147.84\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge7\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M242.67,-687.82C227.34,-662.23 207.57,-629.27 192.91,-604.81\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node11\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"295.39\" cy=\"-569.58\" rx=\"65.41\" ry=\"65.41\"/>\n<text text-anchor=\"start\" x=\"252.74\" y=\"-577.98\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleoprotein</text>\n<text text-anchor=\"start\" x=\"274.73\" y=\"-565.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"279.73\" y=\"-553.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge10\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M274.71,-682.45C277.45,-667.61 280.56,-650.8 283.52,-634.8\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node9\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"73.39\" cy=\"-422.15\" rx=\"39.81\" ry=\"39.81\"/>\n<text text-anchor=\"start\" x=\"51.06\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n<text text-anchor=\"start\" x=\"68.4\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n<text text-anchor=\"start\" x=\"49.74\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n<text text-anchor=\"start\" x=\"48.84\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge8\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M149.93,-535.58C133.73,-511.78 111.95,-479.79 95.77,-456.02\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n<g id=\"node10\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"177.39\" cy=\"-422.15\" rx=\"46.02\" ry=\"46.02\"/>\n<text text-anchor=\"start\" x=\"142.07\" y=\"-418.55\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondrion</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n<g id=\"edge9\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M173.76,-528.84C174.4,-510.34 175.16,-488.22 175.82,-469.01\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"node12\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"291.39\" cy=\"-422.15\" rx=\"41.38\" ry=\"41.38\"/>\n<text text-anchor=\"start\" x=\"278.73\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"265.73\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n<text text-anchor=\"start\" x=\"270.73\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"288.39\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"edge11\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M293.61,-503.52C293.24,-490.21 292.86,-476.55 292.53,-464.36\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n<g id=\"node14\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"394.39\" cy=\"-422.15\" rx=\"43.74\" ry=\"43.74\"/>\n<text text-anchor=\"start\" x=\"372.06\" y=\"-430.55\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n<text text-anchor=\"start\" x=\"369.06\" y=\"-418.55\" font-family=\"Times,serif\" font-size=\"12.00\">biogenesis</text>\n<text text-anchor=\"start\" x=\"367.06\" y=\"-406.55\" font-family=\"Times,serif\" font-size=\"12.00\">community</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n<g id=\"edge13\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M332.26,-514.43C344.79,-496.01 358.49,-475.9 369.82,-459.25\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"node13\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"291.39\" cy=\"-301.09\" rx=\"39.03\" ry=\"39.03\"/>\n<text text-anchor=\"start\" x=\"278.73\" y=\"-315.49\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"272.4\" y=\"-303.49\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n<text text-anchor=\"start\" x=\"270.73\" y=\"-291.49\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"288.39\" y=\"-279.49\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"edge12\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M291.39,-379.84C291.39,-367.22 291.39,-353.38 291.39,-340.92\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n<g id=\"node15\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"387.39\" cy=\"-301.09\" rx=\"30.65\" ry=\"30.65\"/>\n<text text-anchor=\"start\" x=\"365.06\" y=\"-297.49\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n<g id=\"edge14\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M391.83,-377.49C390.95,-362.49 389.98,-346.05 389.17,-332.35\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n<g id=\"node22\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"475.39\" cy=\"-301.09\" rx=\"39.03\" ry=\"39.03\"/>\n<text text-anchor=\"start\" x=\"458.07\" y=\"-315.49\" font-family=\"Times,serif\" font-size=\"12.00\">nucleic</text>\n<text text-anchor=\"start\" x=\"465.4\" y=\"-303.49\" font-family=\"Times,serif\" font-size=\"12.00\">acid</text>\n<text text-anchor=\"start\" x=\"457.06\" y=\"-291.49\" font-family=\"Times,serif\" font-size=\"12.00\">binding</text>\n<text text-anchor=\"start\" x=\"454.73\" y=\"-279.49\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n<g id=\"edge21\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M418.92,-385.1C429.91,-368.95 442.76,-350.05 453.43,-334.37\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"node16\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"387.39\" cy=\"-187.03\" rx=\"39.03\" ry=\"39.03\"/>\n<text text-anchor=\"start\" x=\"363.4\" y=\"-189.43\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"369.73\" y=\"-177.43\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"edge15\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M387.39,-269.47C387.39,-256.38 387.39,-240.95 387.39,-227.05\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"node17\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"162.39\" cy=\"-56\" rx=\"55.05\" ry=\"55.05\"/>\n<text text-anchor=\"start\" x=\"127.07\" y=\"-70.4\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondirial</text>\n<text text-anchor=\"start\" x=\"150.51\" y=\"-58.4\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n<text text-anchor=\"start\" x=\"138.4\" y=\"-46.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"144.73\" y=\"-34.4\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"edge16\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M350.4,-172.46C316.76,-159.32 266.4,-137.67 226.39,-112.01 218.38,-106.86 210.31,-100.79 202.73,-94.58\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n<g id=\"node18\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"274.39\" cy=\"-56\" rx=\"39.03\" ry=\"39.03\"/>\n<text text-anchor=\"start\" x=\"251.07\" y=\"-64.4\" font-family=\"Times,serif\" font-size=\"12.00\">mito&#45;cyto</text>\n<text text-anchor=\"start\" x=\"250.4\" y=\"-52.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"258.4\" y=\"-40.4\" font-family=\"Times,serif\" font-size=\"12.00\">cluster</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n<g id=\"edge17\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M361.75,-156.87C349.78,-143.25 335.35,-126.81 322.39,-112.01 315.13,-103.71 307.26,-94.7 300.04,-86.42\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n<g id=\"node19\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"387.39\" cy=\"-56\" rx=\"56\" ry=\"56\"/>\n<text text-anchor=\"start\" x=\"352.07\" y=\"-82.4\" font-family=\"Times,serif\" font-size=\"12.00\">cotranslational</text>\n<text text-anchor=\"start\" x=\"370.4\" y=\"-70.4\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n<text text-anchor=\"start\" x=\"366.18\" y=\"-58.4\" font-family=\"Times,serif\" font-size=\"12.00\">targeting</text>\n<text text-anchor=\"start\" x=\"382.73\" y=\"-46.4\" font-family=\"Times,serif\" font-size=\"12.00\">to</text>\n<text text-anchor=\"start\" x=\"362.07\" y=\"-34.4\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n<text text-anchor=\"start\" x=\"370.73\" y=\"-22.4\" font-family=\"Times,serif\" font-size=\"12.00\">system</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n<g id=\"edge18\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M387.39,-147.32C387.39,-136.48 387.39,-124.48 387.39,-112.86\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n<g id=\"node20\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"500.39\" cy=\"-56\" rx=\"39.03\" ry=\"39.03\"/>\n<text text-anchor=\"start\" x=\"476.4\" y=\"-64.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"479.73\" y=\"-52.4\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"497.39\" y=\"-40.4\" font-family=\"Times,serif\" font-size=\"12.00\">5</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n<g id=\"edge19\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M413.04,-156.87C425.01,-143.25 439.44,-126.81 452.39,-112.01 459.66,-103.71 467.53,-94.7 474.75,-86.42\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n<g id=\"node21\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"596.39\" cy=\"-56\" rx=\"39.03\" ry=\"39.03\"/>\n<text text-anchor=\"start\" x=\"572.4\" y=\"-64.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"575.73\" y=\"-52.4\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"593.39\" y=\"-40.4\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n<g id=\"edge20\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M424.97,-173.89C459.24,-161.68 510.21,-140.56 548.39,-112.01 557.33,-105.33 565.79,-96.74 573.05,-88.35\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n<g id=\"node24\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"384.39\" cy=\"-728.39\" rx=\"54.1\" ry=\"54.1\"/>\n<text text-anchor=\"start\" x=\"349.74\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n<text text-anchor=\"start\" x=\"360.07\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"358.74\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n<g id=\"edge23\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M384.39,-821.12C384.39,-808.81 384.39,-795.73 384.39,-783.42\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n<g id=\"node26\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"551.39\" cy=\"-875.88\" rx=\"38.55\" ry=\"38.55\"/>\n<text text-anchor=\"start\" x=\"538.4\" y=\"-884.28\" font-family=\"Times,serif\" font-size=\"12.00\">DNA</text>\n<text text-anchor=\"start\" x=\"527.74\" y=\"-872.28\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n<text text-anchor=\"start\" x=\"529.06\" y=\"-860.28\" font-family=\"Times,serif\" font-size=\"12.00\">assembly</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n<g id=\"edge25\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M712.64,-989.41C683.74,-978.21 635.32,-957.18 599.39,-929.98 590.78,-923.46 582.47,-915.24 575.26,-907.22\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n<g id=\"node27\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"638.39\" cy=\"-875.88\" rx=\"30.06\" ry=\"30.06\"/>\n<text text-anchor=\"start\" x=\"620.74\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"623.4\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">lumen</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n<g id=\"edge26\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M719.87,-979.32C707.65,-965.8 691.15,-947.1 677.39,-929.98 670.09,-920.89 662.41,-910.61 655.82,-901.54\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n<g id=\"node35\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"737.39\" cy=\"-875.88\" rx=\"50.8\" ry=\"50.8\"/>\n<text text-anchor=\"start\" x=\"705.07\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n<text text-anchor=\"start\" x=\"734.39\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n<g id=\"edge34\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M737.39,-972.63C737.39,-959.53 737.39,-943.14 737.39,-927.64\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n<g id=\"node40\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"857.39\" cy=\"-875.88\" rx=\"50.8\" ry=\"50.8\"/>\n<text text-anchor=\"start\" x=\"825.07\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n<text text-anchor=\"start\" x=\"854.39\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n<g id=\"edge39\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M755.62,-979.94C773.06,-962.28 799.85,-935.15 821.57,-913.16\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n<g id=\"node42\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"980.39\" cy=\"-875.88\" rx=\"54.1\" ry=\"54.1\"/>\n<text text-anchor=\"start\" x=\"962.74\" y=\"-884.28\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"945.74\" y=\"-872.28\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n<text text-anchor=\"start\" x=\"962.4\" y=\"-860.28\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n<g id=\"edge41\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M763.06,-991.62C799.12,-981.44 866.11,-960.16 917.39,-929.98 925.29,-925.34 933.17,-919.7 940.55,-913.85\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n<g id=\"node44\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1082.39\" cy=\"-875.88\" rx=\"30.06\" ry=\"30.06\"/>\n<text text-anchor=\"start\" x=\"1064.74\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"1070.39\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">body</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n<g id=\"edge43\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M763.93,-998.52C820.13,-997.57 953.38,-988.78 1043.39,-929.98 1053.55,-923.35 1061.98,-913.11 1068.37,-903.38\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n<g id=\"node28\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.39\" cy=\"-728.39\" rx=\"31.44\" ry=\"31.44\"/>\n<text text-anchor=\"start\" x=\"538.4\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">nucleolus</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n<g id=\"edge27\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M624.45,-848.53C610.8,-822.74 590.08,-783.59 576.15,-757.27\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n<g id=\"node34\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"642.39\" cy=\"-728.39\" rx=\"31.96\" ry=\"31.96\"/>\n<text text-anchor=\"start\" x=\"624.74\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"623.4\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n<text text-anchor=\"start\" x=\"624.4\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n<g id=\"edge33\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M639.21,-845.09C639.89,-820.65 640.84,-786.12 641.52,-761.16\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node29\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"440.39\" cy=\"-569.58\" rx=\"61.17\" ry=\"61.17\"/>\n<text text-anchor=\"start\" x=\"400.74\" y=\"-577.98\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleprotein</text>\n<text text-anchor=\"start\" x=\"419.73\" y=\"-565.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"424.73\" y=\"-553.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge28\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M542.22,-702.54C524.96,-680.17 499.11,-646.67 477.77,-619.02\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node30\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.39\" cy=\"-569.58\" rx=\"41.38\" ry=\"41.38\"/>\n<text text-anchor=\"start\" x=\"548.73\" y=\"-583.98\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"535.73\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n<text text-anchor=\"start\" x=\"540.73\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"545.73\" y=\"-547.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge29\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M561.39,-696.03C561.39,-671.81 561.39,-638.08 561.39,-611.7\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node33\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"666.39\" cy=\"-569.58\" rx=\"45.63\" ry=\"45.63\"/>\n<text text-anchor=\"start\" x=\"637.73\" y=\"-577.98\" font-family=\"Times,serif\" font-size=\"12.00\">splicosomal</text>\n<text text-anchor=\"start\" x=\"645.73\" y=\"-565.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"650.73\" y=\"-553.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge32\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M578.97,-701.14C595.96,-675.77 621.88,-637.06 641.09,-608.38\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"node31\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"529.39\" cy=\"-422.15\" rx=\"41.38\" ry=\"41.38\"/>\n<text text-anchor=\"start\" x=\"516.73\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"503.73\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n<text text-anchor=\"start\" x=\"508.73\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"526.39\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"edge30\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M552.57,-528.45C548.11,-508.19 542.71,-483.67 538.25,-463.39\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node32\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"627.39\" cy=\"-422.15\" rx=\"39.03\" ry=\"39.03\"/>\n<text text-anchor=\"start\" x=\"614.73\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"608.4\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n<text text-anchor=\"start\" x=\"606.73\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"611.73\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge31\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M578.39,-531.13C588.55,-508.74 601.38,-480.47 611.36,-458.47\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node36\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"748.39\" cy=\"-728.39\" rx=\"48.92\" ry=\"48.92\"/>\n<text text-anchor=\"start\" x=\"717.4\" y=\"-742.79\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n<text text-anchor=\"start\" x=\"718.51\" y=\"-730.79\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n<text text-anchor=\"start\" x=\"727.73\" y=\"-718.79\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"732.73\" y=\"-706.79\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge35\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M741.21,-824.4C742.35,-809.3 743.6,-792.78 744.73,-777.82\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node37\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"777.39\" cy=\"-569.58\" rx=\"47.35\" ry=\"47.35\"/>\n<text text-anchor=\"start\" x=\"753.07\" y=\"-583.98\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n<text text-anchor=\"start\" x=\"747.51\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n<text text-anchor=\"start\" x=\"756.73\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"761.73\" y=\"-547.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge36\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M757.25,-679.49C760.92,-659.67 765.15,-636.78 768.79,-617.11\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node38\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"777.39\" cy=\"-422.15\" rx=\"34.31\" ry=\"34.31\"/>\n<text text-anchor=\"start\" x=\"765.73\" y=\"-430.55\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n<text text-anchor=\"start\" x=\"756.73\" y=\"-418.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"761.73\" y=\"-406.55\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge37\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M777.39,-521.36C777.39,-500.46 777.39,-476.34 777.39,-457.24\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n<g id=\"node39\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"777.39\" cy=\"-301.09\" rx=\"34.31\" ry=\"34.31\"/>\n<text text-anchor=\"start\" x=\"762.73\" y=\"-309.49\" font-family=\"Times,serif\" font-size=\"12.00\">NuA4</text>\n<text text-anchor=\"start\" x=\"765.73\" y=\"-297.49\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n<text text-anchor=\"start\" x=\"756.73\" y=\"-285.49\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n<g id=\"edge38\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M777.39,-387.03C777.39,-371.09 777.39,-352.16 777.39,-336.22\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n<g id=\"node41\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"857.39\" cy=\"-728.39\" rx=\"39.49\" ry=\"39.49\"/>\n<text text-anchor=\"start\" x=\"833.07\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n<text text-anchor=\"start\" x=\"833.07\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"836.73\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n<g id=\"edge40\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M857.39,-824.4C857.39,-806.18 857.39,-785.9 857.39,-768.77\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n<g id=\"node43\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"980.39\" cy=\"-728.39\" rx=\"56\" ry=\"56\"/>\n<text text-anchor=\"start\" x=\"960.07\" y=\"-754.79\" font-family=\"Times,serif\" font-size=\"12.00\">negative</text>\n<text text-anchor=\"start\" x=\"956.07\" y=\"-742.79\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"975.4\" y=\"-730.79\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n<text text-anchor=\"start\" x=\"967.73\" y=\"-718.79\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"950.73\" y=\"-706.79\" font-family=\"Times,serif\" font-size=\"12.00\">biosynthesis</text>\n<text text-anchor=\"start\" x=\"962.4\" y=\"-694.79\" font-family=\"Times,serif\" font-size=\"12.00\">process</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n<g id=\"edge42\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M980.39,-821.12C980.39,-809.46 980.39,-797.11 980.39,-785.37\"/>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x10fe256d0>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 53
  },
  {
   "cell_type": "code",
   "id": "ab431982-d156-4b7a-9918-384a5ea5132a",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.744452Z",
     "start_time": "2025-02-12T16:39:55.741801Z"
    }
   },
   "source": [
    "music_map"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cell': {'cytoplasm': {'secretory organelles': {'transmembrane transport systems': {'ion transmembrane transport systems': {}}},\n",
       "   'cytoplasmic organelles': {'subgroup of cytoplasmic organelles': {'metabolic organelles': {'subgroup of metabolic organelles': {},\n",
       "      'mitochondrion': {}},\n",
       "     'ribonucleoprotein complex family': {'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
       "      'ribosome biogenesis community': {'ribosome': {'ribosomal subunit': {'mitochondirial large ribosomal subunit': {},\n",
       "         'mito-cyto ribosomal cluster': {},\n",
       "         'cotranslational protein targeting to membrane system': {},\n",
       "         'ribosomal complex 5': {},\n",
       "         'ribosomal complex 2': {}}},\n",
       "       'nucleic acid binding complex': {}}}}},\n",
       "   'transcriptional regulation complex family': {'transcriptional regulation complexes': {}}},\n",
       "  'nucleus': {'DNA metabolic assembly': {},\n",
       "   'nuclear lumen': {'nucleolus': {'ribonucleprotein complex family': {},\n",
       "     'RNA processing complex family': {'RNA processing complex 1': {},\n",
       "      'RNA splicing complex family': {}},\n",
       "     'splicosomal complex family': {}},\n",
       "    'nuclear splicing speckle': {}},\n",
       "   'nuclear body': {},\n",
       "   'nucleo-plasm 1': {'chromosome organization complex family': {'chromatin organization complex family': {'HAT complex family': {'NuA4 HAT complex': {}}}}},\n",
       "   'nucleo-plasm 2': {'chromatin regulation complex': {}},\n",
       "   'nuclear transcriptional speckle': {'negative regulation of RNA biosynthesis process': {}}}}}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 54
  },
  {
   "cell_type": "code",
   "id": "5a76608a-fff0-4005-9f32-c02afef39d9e",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.875748Z",
     "start_time": "2025-02-12T16:39:55.750203Z"
    }
   },
   "source": [
    "import copy\n",
    "\n",
    "transcription_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'RNA processing': 'any',\n",
    "        'regulation': 'any',\n",
    "        'nucleic acids': 'any',\n",
    "        'negative regulation': 'any'\n",
    "    },\n",
    "    'outputs': {\n",
    "        'RNA processing': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "        'regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "        'nucleic acids': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', \n",
    "            'ribonucleoprotein complex family', 'ribosome biogenesis community', 'nucleic acid binding complex'],\n",
    "        'negative regulation': [\n",
    "            'nucleus', 'nuclear transcriptional speckle', 'negative regulation of RNA biosynthesis process'],\n",
    "    }\n",
    "}\n",
    "\n",
    "translation_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'ribosomes': 'any',\n",
    "        'membrane': 'any'\n",
    "    },\n",
    "    'outputs': {\n",
    "        'ribosomes': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', 'ribosome biogenesis community', 'ribosome'],\n",
    "        # 'membrane': ['],\n",
    "    }\n",
    "}\n",
    "\n",
    "metabolism_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'organelles': 'any',\n",
    "        'mitochondrion': 'any',\n",
    "        'ion transport': 'any',\n",
    "    },\n",
    "    'outputs': {\n",
    "        'organelles': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', ],\n",
    "        'mitochondrion': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', 'mitochondrion'],\n",
    "        'ion transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems', 'ion transmembrane transport systems'],\n",
    "    }\n",
    "}\n",
    "\n",
    "cell_cycle_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'DNA': 'any',\n",
    "        'chromosome': 'any',\n",
    "        'chromatin': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'DNA': ['nucleus', 'DNA metabolic assembly'],\n",
    "        'chromosome': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family',],\n",
    "        'chromatin': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family', 'chromatin organization complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "signalling_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'transport': 'any',\n",
    "        'trancript regulation': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'trancript regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "    }\n",
    "}\n",
    "\n",
    "protein_transport_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'transmembrane transport': 'any',\n",
    "        'secretory': 'any',\n",
    "        'cotranslation': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'transmembrane transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'secretory': [\n",
    "            'cytoplasm', 'secretory organelles'],\n",
    "        'cotranslation': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', \n",
    "            'ribosome biogenesis community', 'ribosome', 'ribosomal subunit', 'cotranslational protein targeting to membrane system']\n",
    "    }\n",
    "}\n",
    "\n",
    "rna_processing_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'processing complex': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'processing complex': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "# compose schema\n",
    "music_map['cell']['transcription'] = transcription_process\n",
    "music_map['cell']['translation'] = translation_process\n",
    "music_map['cell']['metabolism'] = metabolism_process\n",
    "music_map['cell']['cell cycle'] = cell_cycle_process\n",
    "music_map['cell']['signalling'] = signalling_process\n",
    "music_map['cell']['protein transport'] = protein_transport_process\n",
    "music_map['cell']['RNA processing'] = rna_processing_process\n",
    "\n",
    "music_map2 = replace_regex_recursive(music_map)\n",
    "\n",
    "# plot\n",
    "plot_settingsx = copy.deepcopy(plot_settings)\n",
    "plot_settingsx['label_margin'] = '0.02'\n",
    "plot_settingsx['node_border_colors'] = {\n",
    "    ('cell', 'translation'): 'blue',\n",
    "    ('cell', 'transcription'): 'blue',\n",
    "    ('cell', 'metabolism'): 'blue',\n",
    "    ('cell', 'cell<br/>cycle'): 'blue',\n",
    "    ('cell', 'signalling'): 'blue',\n",
    "    ('cell', 'protein<br/>transport'): 'blue',\n",
    "    ('cell', 'RNA<br/>processing'): 'blue',\n",
    "}\n",
    "# plot_settingsx.update({'out_dir': 'out','dpi': '250'})\n",
    "plot_bigraph(music_map2, **plot_settingsx, filename='cell_hierarchy_functions')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell_hierarchy_functions\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"749pt\" height=\"700pt\"\n viewBox=\"0.00 0.00 770.64 720.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.473411 0.473411) rotate(0) translate(4 1474.63)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-1474.63 1578.62,-1474.63 1578.62,4 -4,4\"/>\n<!-- (&#39;cell&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;cell&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"821.98\" cy=\"-1452.63\" rx=\"18\" ry=\"18\"/>\n<text text-anchor=\"start\" x=\"813.32\" y=\"-1449.03\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"608.98\" cy=\"-1036.09\" rx=\"21.71\" ry=\"21.71\"/>\n<text text-anchor=\"start\" x=\"590.65\" y=\"-1032.49\" font-family=\"Times,serif\" font-size=\"12.00\">nucleus</text>\n</g>\n<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M813.76,-1435.63C780.35,-1370.61 654.79,-1126.25 618.72,-1056.06\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"node22\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1110.98\" cy=\"-1368.97\" rx=\"28.67\" ry=\"28.67\"/>\n<text text-anchor=\"start\" x=\"1086.32\" y=\"-1365.37\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n</g>\n<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"edge21\" class=\"edge\">\n<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M839.62,-1446.65C888.15,-1432.93 1023.94,-1394.56 1082.98,-1377.88\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"334.98\" cy=\"-893.7\" rx=\"27\" ry=\"27\"/>\n<text text-anchor=\"start\" x=\"317.32\" y=\"-896.1\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"319.98\" y=\"-884.1\" font-family=\"Times,serif\" font-size=\"12.00\">lumen</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M586.58,-1032.58C547.29,-1027.32 464.56,-1012.39 405.98,-975.06 382.59,-960.15 362.37,-935.37 349.56,-917.27\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n<g id=\"node11\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"725.98\" cy=\"-893.7\" rx=\"51.05\" ry=\"51.05\"/>\n<text text-anchor=\"start\" x=\"708.32\" y=\"-902.1\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"691.32\" y=\"-890.1\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n<text text-anchor=\"start\" x=\"707.99\" y=\"-878.1\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n<g id=\"edge10\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M622.86,-1018.44C640.11,-997.74 670.11,-961.74 693.31,-933.89\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n<g id=\"node13\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"495.98\" cy=\"-893.7\" rx=\"47.75\" ry=\"47.75\"/>\n<text text-anchor=\"start\" x=\"463.65\" y=\"-896.1\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n<text text-anchor=\"start\" x=\"492.98\" y=\"-884.1\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n<g id=\"edge12\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M595.35,-1018.16C578.27,-996.94 548.5,-959.96 525.99,-931.99\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n<g id=\"node18\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"608.98\" cy=\"-893.7\" rx=\"47.75\" ry=\"47.75\"/>\n<text text-anchor=\"start\" x=\"576.65\" y=\"-896.1\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n<text text-anchor=\"start\" x=\"605.98\" y=\"-884.1\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n<g id=\"edge17\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M608.98,-1013.47C608.98,-994.54 608.98,-966.39 608.98,-942.24\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n<g id=\"node20\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"425.98\" cy=\"-777.85\" rx=\"35.49\" ry=\"35.49\"/>\n<text text-anchor=\"start\" x=\"412.98\" y=\"-786.25\" font-family=\"Times,serif\" font-size=\"12.00\">DNA</text>\n<text text-anchor=\"start\" x=\"402.32\" y=\"-774.25\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n<text text-anchor=\"start\" x=\"403.65\" y=\"-762.25\" font-family=\"Times,serif\" font-size=\"12.00\">assembly</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n<g id=\"edge19\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M586.73,-1031.98C549.13,-1025.43 473.75,-1006.41 438.98,-956.06 409.88,-913.92 413.05,-851.92 418.88,-813.53\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n<g id=\"node21\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"821.98\" cy=\"-893.7\" rx=\"27\" ry=\"27\"/>\n<text text-anchor=\"start\" x=\"804.32\" y=\"-896.1\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"809.98\" y=\"-884.1\" font-family=\"Times,serif\" font-size=\"12.00\">body</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n<g id=\"edge20\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M630.69,-1030.04C666.64,-1020.73 739.21,-997.92 785.98,-956.06 797.11,-946.09 805.76,-931.84 811.79,-919.58\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"245.98\" cy=\"-670.02\" rx=\"26.83\" ry=\"26.83\"/>\n<text text-anchor=\"start\" x=\"222.98\" y=\"-666.42\" font-family=\"Times,serif\" font-size=\"12.00\">nucleolus</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M324.94,-867.7C307.8,-825.01 273.09,-738.55 255.98,-695.92\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n<g id=\"node10\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"334.98\" cy=\"-670.02\" rx=\"28.91\" ry=\"28.91\"/>\n<text text-anchor=\"start\" x=\"317.32\" y=\"-678.42\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n<text text-anchor=\"start\" x=\"315.98\" y=\"-666.42\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n<text text-anchor=\"start\" x=\"316.99\" y=\"-654.42\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n<g id=\"edge9\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M334.98,-865.96C334.98,-823.93 334.98,-742.75 334.98,-699.6\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"127.98\" cy=\"-415.66\" rx=\"38.33\" ry=\"38.33\"/>\n<text text-anchor=\"start\" x=\"115.31\" y=\"-430.06\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"102.32\" y=\"-418.06\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n<text text-anchor=\"start\" x=\"107.32\" y=\"-406.06\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"112.32\" y=\"-394.06\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M234.66,-644.82C213.61,-599.79 168.35,-503.01 144.1,-451.15\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node8\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"345.98\" cy=\"-415.66\" rx=\"58.11\" ry=\"58.11\"/>\n<text text-anchor=\"start\" x=\"306.32\" y=\"-424.06\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleprotein</text>\n<text text-anchor=\"start\" x=\"325.32\" y=\"-412.06\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"330.32\" y=\"-400.06\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge7\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M255.75,-644.36C271.62,-604.3 303.38,-524.17 324.66,-470.46\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node9\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"226.98\" cy=\"-415.66\" rx=\"42.57\" ry=\"42.57\"/>\n<text text-anchor=\"start\" x=\"198.31\" y=\"-424.06\" font-family=\"Times,serif\" font-size=\"12.00\">splicosomal</text>\n<text text-anchor=\"start\" x=\"206.32\" y=\"-412.06\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"211.32\" y=\"-400.06\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge8\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M243.98,-642.49C240.68,-598.7 234.08,-510.97 230.14,-458.68\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"127.98\" cy=\"-263.22\" rx=\"38.33\" ry=\"38.33\"/>\n<text text-anchor=\"start\" x=\"115.31\" y=\"-277.62\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"102.32\" y=\"-265.62\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n<text text-anchor=\"start\" x=\"107.32\" y=\"-253.62\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"124.98\" y=\"-241.62\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M127.98,-376.7C127.98,-353.98 127.98,-325.2 127.98,-302.44\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node7\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"35.98\" cy=\"-263.22\" rx=\"35.98\" ry=\"35.98\"/>\n<text text-anchor=\"start\" x=\"23.31\" y=\"-277.62\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"16.98\" y=\"-265.62\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n<text text-anchor=\"start\" x=\"15.32\" y=\"-253.62\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"20.32\" y=\"-241.62\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M108.01,-382.02C92.21,-356.18 70.3,-320.35 54.8,-295.01\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;transcription&#39;) -->\n<g id=\"node45\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;transcription&#39;)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"776.74,-281.22 713.21,-281.22 713.21,-245.22 776.74,-245.22 776.74,-281.22\"/>\n<text text-anchor=\"start\" x=\"714.65\" y=\"-259.62\" font-family=\"Times,serif\" font-size=\"12.00\">transcription</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n<g id=\"edge45\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M155.41,-373.51C161.2,-367.33 167.76,-361.65 174.98,-357.55 210.64,-337.26 499.51,-308.34 539.98,-301.55 600.99,-291.32 671.96,-278.1 712.36,-270.45\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"152.83,-371.14 149,-381.02 158.16,-375.69 152.83,-371.14\"/>\n<text text-anchor=\"start\" x=\"396.81\" y=\"-331.55\" font-family=\"Times,serif\" font-size=\"10.00\">RNA</text>\n<text text-anchor=\"start\" x=\"385.98\" y=\"-321.55\" font-family=\"Times,serif\" font-size=\"10.00\">processing</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;) -->\n<g id=\"node51\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"238.08,-281.22 183.88,-281.22 183.88,-245.22 238.08,-245.22 238.08,-281.22\"/>\n<text text-anchor=\"start\" x=\"198.31\" y=\"-265.62\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"185.32\" y=\"-253.62\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;) -->\n<g id=\"edge60\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;RNA&lt;br/&gt;processing&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M151.76,-371.55C168.15,-341.85 189.07,-303.93 201.15,-282.03\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"148.9,-369.5 147.13,-379.95 155.02,-372.89 148.9,-369.5\"/>\n<text text-anchor=\"start\" x=\"179.98\" y=\"-331.55\" font-family=\"Times,serif\" font-size=\"10.00\">processing</text>\n<text text-anchor=\"start\" x=\"184.15\" y=\"-321.55\" font-family=\"Times,serif\" font-size=\"10.00\">complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n<g id=\"node12\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"703.98\" cy=\"-544.73\" rx=\"52.95\" ry=\"52.95\"/>\n<text text-anchor=\"start\" x=\"683.65\" y=\"-571.13\" font-family=\"Times,serif\" font-size=\"12.00\">negative</text>\n<text text-anchor=\"start\" x=\"679.65\" y=\"-559.13\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"698.98\" y=\"-547.13\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n<text text-anchor=\"start\" x=\"691.31\" y=\"-535.13\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"674.31\" y=\"-523.13\" font-family=\"Times,serif\" font-size=\"12.00\">biosynthesis</text>\n<text text-anchor=\"start\" x=\"685.98\" y=\"-511.13\" font-family=\"Times,serif\" font-size=\"12.00\">process</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n<g id=\"edge11\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M722.76,-841.87C718.63,-776.75 711.5,-664.4 707.32,-598.39\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n<g id=\"edge46\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M699.48,-479.62C698.35,-443.43 699.3,-397.57 707.44,-357.55 713,-330.17 726.27,-300.56 735.45,-282.2\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"695.98,-479.64 699.85,-489.5 702.97,-479.37 695.98,-479.64\"/>\n<text text-anchor=\"start\" x=\"711.31\" y=\"-417.66\" font-family=\"Times,serif\" font-size=\"10.00\">negative</text>\n<text text-anchor=\"start\" x=\"707.98\" y=\"-407.66\" font-family=\"Times,serif\" font-size=\"10.00\">regulation</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node14\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"505.98\" cy=\"-670.02\" rx=\"45.87\" ry=\"45.87\"/>\n<text text-anchor=\"start\" x=\"474.98\" y=\"-684.42\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n<text text-anchor=\"start\" x=\"476.09\" y=\"-672.42\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n<text text-anchor=\"start\" x=\"485.32\" y=\"-660.42\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"490.32\" y=\"-648.42\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge13\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M498.12,-845.14C499.84,-807.1 502.23,-754.11 503.92,-716.6\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node15\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"503.98\" cy=\"-415.66\" rx=\"44.3\" ry=\"44.3\"/>\n<text text-anchor=\"start\" x=\"479.65\" y=\"-430.06\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n<text text-anchor=\"start\" x=\"474.09\" y=\"-418.06\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n<text text-anchor=\"start\" x=\"483.32\" y=\"-406.06\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"488.32\" y=\"-394.06\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge14\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M505.62,-623.32C505.25,-577.16 504.69,-506.33 504.33,-460.76\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n<g id=\"node48\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"530.98,-281.22 476.98,-281.22 476.98,-245.22 530.98,-245.22 530.98,-281.22\"/>\n<text text-anchor=\"start\" x=\"495.32\" y=\"-265.62\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n<text text-anchor=\"start\" x=\"491.32\" y=\"-253.62\" font-family=\"Times,serif\" font-size=\"12.00\">cycle</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n<g id=\"edge52\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M523.97,-615.31C536,-576.49 550.71,-522.53 556.98,-473.78 563.56,-422.54 562.33,-408.93 556.98,-357.55 555.19,-340.38 556.41,-335.13 548.98,-319.55 542.37,-305.69 531.6,-292.28 522.26,-282.13\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"520.78,-613.8 521.12,-624.39 527.46,-615.9 520.78,-613.8\"/>\n<text text-anchor=\"start\" x=\"561.98\" y=\"-412.66\" font-family=\"Times,serif\" font-size=\"10.00\">chromosome</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node16\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"406.98\" cy=\"-263.22\" rx=\"31.25\" ry=\"31.25\"/>\n<text text-anchor=\"start\" x=\"395.31\" y=\"-271.62\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n<text text-anchor=\"start\" x=\"386.32\" y=\"-259.62\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"391.32\" y=\"-247.62\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge15\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M480,-377.48C462.61,-350.5 439.51,-314.67 423.99,-290.6\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n<g id=\"edge53\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M503.98,-359.3C503.98,-331.94 503.98,-300.81 503.98,-281.85\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"500.48,-359.23 503.98,-369.23 507.48,-359.23 500.48,-359.23\"/>\n<text text-anchor=\"start\" x=\"503.98\" y=\"-326.55\" font-family=\"Times,serif\" font-size=\"10.00\">chromatin</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n<g id=\"node17\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"406.98\" cy=\"-134.95\" rx=\"31.25\" ry=\"31.25\"/>\n<text text-anchor=\"start\" x=\"392.31\" y=\"-143.35\" font-family=\"Times,serif\" font-size=\"12.00\">NuA4</text>\n<text text-anchor=\"start\" x=\"395.31\" y=\"-131.35\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n<text text-anchor=\"start\" x=\"386.32\" y=\"-119.35\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n<g id=\"edge16\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M406.98,-231.01C406.98,-211.44 406.98,-186.31 406.98,-166.8\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n<g id=\"node19\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"608.98\" cy=\"-670.02\" rx=\"36.44\" ry=\"36.44\"/>\n<text text-anchor=\"start\" x=\"584.65\" y=\"-678.42\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n<text text-anchor=\"start\" x=\"584.65\" y=\"-666.42\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"588.32\" y=\"-654.42\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n<g id=\"edge18\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M608.98,-845.14C608.98,-803.8 608.98,-744.78 608.98,-707.19\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;) -->\n<g id=\"edge54\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cell&lt;br/&gt;cycle&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M422,-730.26C416.75,-651.6 412.26,-488.23 450.98,-357.55 459.39,-329.14 477.8,-300.02 490.54,-282.03\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"418.5,-730.45 422.7,-740.18 425.48,-729.96 418.5,-730.45\"/>\n<text text-anchor=\"start\" x=\"424.98\" y=\"-541.73\" font-family=\"Times,serif\" font-size=\"10.00\">DNA</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node23\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1159.98\" cy=\"-1148.18\" rx=\"51.05\" ry=\"51.05\"/>\n<text text-anchor=\"start\" x=\"1125.32\" y=\"-1162.58\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n<text text-anchor=\"start\" x=\"1135.65\" y=\"-1150.58\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"1139.32\" y=\"-1138.58\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"1144.32\" y=\"-1126.58\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge22\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1117.18,-1340.26C1125.14,-1304.73 1139.08,-1242.48 1148.88,-1198.72\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node25\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1063.98\" cy=\"-1260.26\" rx=\"43.04\" ry=\"43.04\"/>\n<text text-anchor=\"start\" x=\"1034.99\" y=\"-1262.66\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n<text text-anchor=\"start\" x=\"1039.43\" y=\"-1250.66\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge24\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1099.48,-1341.86C1093.99,-1329.4 1087.31,-1314.23 1081.22,-1300.42\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n<g id=\"node42\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1378.98\" cy=\"-1036.09\" rx=\"36.76\" ry=\"36.76\"/>\n<text text-anchor=\"start\" x=\"1356.99\" y=\"-1038.49\" font-family=\"Times,serif\" font-size=\"12.00\">secretory</text>\n<text text-anchor=\"start\" x=\"1354.43\" y=\"-1026.49\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge41\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1128.98,-1345.74C1176.16,-1287.5 1303.37,-1130.44 1355.85,-1065.65\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n<g id=\"node24\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1224.98\" cy=\"-893.7\" rx=\"51.05\" ry=\"51.05\"/>\n<text text-anchor=\"start\" x=\"1190.32\" y=\"-902.1\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n<text text-anchor=\"start\" x=\"1200.65\" y=\"-890.1\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"1199.32\" y=\"-878.1\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n<g id=\"edge23\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1172.66,-1097.9C1184.12,-1053.41 1200.88,-988.3 1212.32,-943.85\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n<g id=\"edge47\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1230.69,-830.59C1236.04,-750.38 1238.58,-607.34 1202.98,-491.78 1182.46,-425.19 1178.06,-398.9 1121.98,-357.55 1014.52,-278.33 847.76,-265.82 777.72,-264.22\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1227.21,-830.23 1230,-840.45 1234.19,-830.72 1227.21,-830.23\"/>\n<text text-anchor=\"start\" x=\"1223.98\" y=\"-541.73\" font-family=\"Times,serif\" font-size=\"10.00\">regulation</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;signalling&#39;) -->\n<g id=\"node49\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;signalling&#39;)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"1306.98,-688.02 1252.98,-688.02 1252.98,-652.02 1306.98,-652.02 1306.98,-688.02\"/>\n<text text-anchor=\"start\" x=\"1256.31\" y=\"-666.42\" font-family=\"Times,serif\" font-size=\"12.00\">signalling</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;) -->\n<g id=\"edge55\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1239.97,-832.29C1251.91,-784.13 1267.87,-719.83 1275.52,-689\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1236.63,-831.21 1237.62,-841.76 1243.42,-832.89 1236.63,-831.21\"/>\n<text text-anchor=\"start\" x=\"1265.31\" y=\"-779.85\" font-family=\"Times,serif\" font-size=\"10.00\">trancript</text>\n<text text-anchor=\"start\" x=\"1261.98\" y=\"-769.85\" font-family=\"Times,serif\" font-size=\"10.00\">regulation</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node26\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1050.98\" cy=\"-1036.09\" rx=\"43.04\" ry=\"43.04\"/>\n<text text-anchor=\"start\" x=\"1028.64\" y=\"-1050.49\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n<text text-anchor=\"start\" x=\"1045.98\" y=\"-1038.49\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n<text text-anchor=\"start\" x=\"1021.99\" y=\"-1026.49\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n<text text-anchor=\"start\" x=\"1026.43\" y=\"-1014.49\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge25\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1061.47,-1216.43C1059.17,-1177.09 1055.78,-1119.1 1053.48,-1079.8\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"node27\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"928.98\" cy=\"-893.7\" rx=\"62.36\" ry=\"62.36\"/>\n<text text-anchor=\"start\" x=\"886.32\" y=\"-902.1\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleoprotein</text>\n<text text-anchor=\"start\" x=\"908.32\" y=\"-890.1\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"913.32\" y=\"-878.1\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n<g id=\"edge26\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1022.68,-1002.53C1006.98,-984.47 987.19,-961.68 969.84,-941.72\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node39\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1050.98\" cy=\"-893.7\" rx=\"36.76\" ry=\"36.76\"/>\n<text text-anchor=\"start\" x=\"1027.32\" y=\"-896.1\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n<text text-anchor=\"start\" x=\"1026.43\" y=\"-884.1\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge38\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1050.98,-992.21C1050.98,-972.69 1050.98,-949.84 1050.98,-931.17\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n<g id=\"node28\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"882.98\" cy=\"-670.02\" rx=\"40.69\" ry=\"40.69\"/>\n<text text-anchor=\"start\" x=\"860.65\" y=\"-678.42\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n<text text-anchor=\"start\" x=\"857.65\" y=\"-666.42\" font-family=\"Times,serif\" font-size=\"12.00\">biogenesis</text>\n<text text-anchor=\"start\" x=\"855.65\" y=\"-654.42\" font-family=\"Times,serif\" font-size=\"12.00\">community</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n<g id=\"edge27\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M916.35,-831.83C908.34,-793.25 898.22,-744.45 891.23,-710.78\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"node37\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"785.98\" cy=\"-670.02\" rx=\"38.33\" ry=\"38.33\"/>\n<text text-anchor=\"start\" x=\"773.31\" y=\"-684.42\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"760.32\" y=\"-672.42\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n<text text-anchor=\"start\" x=\"765.32\" y=\"-660.42\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"782.98\" y=\"-648.42\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"edge36\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M895.27,-840.44C867.77,-797.82 829.81,-738.97 806.63,-703.03\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n<g id=\"node29\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"882.98\" cy=\"-415.66\" rx=\"35.98\" ry=\"35.98\"/>\n<text text-anchor=\"start\" x=\"865.65\" y=\"-430.06\" font-family=\"Times,serif\" font-size=\"12.00\">nucleic</text>\n<text text-anchor=\"start\" x=\"872.98\" y=\"-418.06\" font-family=\"Times,serif\" font-size=\"12.00\">acid</text>\n<text text-anchor=\"start\" x=\"864.64\" y=\"-406.06\" font-family=\"Times,serif\" font-size=\"12.00\">binding</text>\n<text text-anchor=\"start\" x=\"862.32\" y=\"-394.06\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n<g id=\"edge28\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M882.98,-628.47C882.98,-579.85 882.98,-499.1 882.98,-452.61\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n<g id=\"node30\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1278.98\" cy=\"-415.66\" rx=\"26.1\" ry=\"26.1\"/>\n<text text-anchor=\"start\" x=\"1256.65\" y=\"-412.06\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n<g id=\"edge29\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M907.86,-637.08C940.36,-597.46 1000.94,-530.42 1066.98,-491.78 1089.18,-478.79 1097.67,-482.18 1121.98,-473.78 1168.86,-457.59 1223.16,-437.54 1254.05,-426.02\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;) -->\n<g id=\"edge44\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;transcription&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M854.38,-377.07C849.06,-370.46 843.45,-363.72 837.98,-357.55 813.7,-330.17 783.44,-300.56 764.16,-282.2\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"851.52,-379.1 860.47,-384.77 857.01,-374.76 851.52,-379.1\"/>\n<text text-anchor=\"start\" x=\"820.98\" y=\"-331.55\" font-family=\"Times,serif\" font-size=\"10.00\">nucleic</text>\n<text text-anchor=\"start\" x=\"825.14\" y=\"-321.55\" font-family=\"Times,serif\" font-size=\"10.00\">acids</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"node31\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1200.98\" cy=\"-263.22\" rx=\"35.97\" ry=\"35.97\"/>\n<text text-anchor=\"start\" x=\"1176.98\" y=\"-265.62\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"1183.31\" y=\"-253.62\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"edge30\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1269.27,-390.77C1261.09,-371.26 1248.7,-343.16 1235.98,-319.55 1231.45,-311.15 1226.12,-302.3 1221.01,-294.21\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;translation&#39;) -->\n<g id=\"node46\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;translation&#39;)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"1341.98,-281.22 1287.98,-281.22 1287.98,-245.22 1341.98,-245.22 1341.98,-281.22\"/>\n<text text-anchor=\"start\" x=\"1289.65\" y=\"-259.62\" font-family=\"Times,serif\" font-size=\"12.00\">translation</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;) -->\n<g id=\"edge48\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1275.81,-377.66C1275.36,-359.77 1276.41,-338.19 1281.87,-319.55 1285.83,-306.01 1293.91,-292.4 1301.1,-282.06\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1272.31,-377.74 1276.25,-387.58 1279.31,-377.43 1272.31,-377.74\"/>\n<text text-anchor=\"start\" x=\"1280.98\" y=\"-326.55\" font-family=\"Times,serif\" font-size=\"10.00\">ribosomes</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n<g id=\"node32\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1257.98\" cy=\"-134.95\" rx=\"52.95\" ry=\"52.95\"/>\n<text text-anchor=\"start\" x=\"1222.66\" y=\"-161.35\" font-family=\"Times,serif\" font-size=\"12.00\">cotranslational</text>\n<text text-anchor=\"start\" x=\"1240.98\" y=\"-149.35\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n<text text-anchor=\"start\" x=\"1236.76\" y=\"-137.35\" font-family=\"Times,serif\" font-size=\"12.00\">targeting</text>\n<text text-anchor=\"start\" x=\"1253.31\" y=\"-125.35\" font-family=\"Times,serif\" font-size=\"12.00\">to</text>\n<text text-anchor=\"start\" x=\"1232.66\" y=\"-113.35\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n<text text-anchor=\"start\" x=\"1241.31\" y=\"-101.35\" font-family=\"Times,serif\" font-size=\"12.00\">system</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n<g id=\"edge31\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1215.8,-229.38C1222.07,-215.49 1229.5,-199.05 1236.35,-183.85\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"node33\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1380.98\" cy=\"-134.95\" rx=\"52\" ry=\"52\"/>\n<text text-anchor=\"start\" x=\"1345.65\" y=\"-149.35\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondirial</text>\n<text text-anchor=\"start\" x=\"1369.09\" y=\"-137.35\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n<text text-anchor=\"start\" x=\"1356.98\" y=\"-125.35\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"1363.31\" y=\"-113.35\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n<g id=\"edge32\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1232.62,-244.19C1257.09,-229.83 1291.48,-208.76 1319.98,-187.9 1327.27,-182.56 1334.8,-176.58 1341.96,-170.63\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n<g id=\"node34\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"970.98\" cy=\"-134.95\" rx=\"35.97\" ry=\"35.97\"/>\n<text text-anchor=\"start\" x=\"947.65\" y=\"-143.35\" font-family=\"Times,serif\" font-size=\"12.00\">mito&#45;cyto</text>\n<text text-anchor=\"start\" x=\"946.98\" y=\"-131.35\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"954.98\" y=\"-119.35\" font-family=\"Times,serif\" font-size=\"12.00\">cluster</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n<g id=\"edge33\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1165.45,-253.39C1126.27,-242.48 1062.62,-221.16 1015.98,-187.9 1007.14,-181.59 998.9,-173.21 991.94,-165.01\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n<g id=\"node35\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1060.98\" cy=\"-134.95\" rx=\"35.97\" ry=\"35.97\"/>\n<text text-anchor=\"start\" x=\"1036.98\" y=\"-143.35\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"1040.32\" y=\"-131.35\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"1057.98\" y=\"-119.35\" font-family=\"Times,serif\" font-size=\"12.00\">5</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n<g id=\"edge34\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1171.53,-241.48C1152.13,-227.28 1126.61,-207.53 1105.98,-187.9 1098.28,-180.58 1090.5,-172.07 1083.62,-164.08\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n<g id=\"node36\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1150.98\" cy=\"-134.95\" rx=\"35.97\" ry=\"35.97\"/>\n<text text-anchor=\"start\" x=\"1126.98\" y=\"-143.35\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"1130.32\" y=\"-131.35\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"1147.98\" y=\"-119.35\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n<g id=\"edge35\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1187.71,-228.72C1180.42,-210.3 1171.41,-187.54 1164.13,-169.17\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n<g id=\"node50\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"1487.98,-36 1433.98,-36 1433.98,0 1487.98,0 1487.98,-36\"/>\n<text text-anchor=\"start\" x=\"1443.98\" y=\"-20.4\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n<text text-anchor=\"start\" x=\"1439.65\" y=\"-8.4\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n<g id=\"edge59\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1306.44,-91.25C1310.91,-87.96 1315.45,-84.83 1319.98,-82 1356.92,-58.9 1403.77,-39.74 1433.07,-28.83\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1304.44,-88.37 1298.63,-97.24 1308.7,-93.93 1304.44,-88.37\"/>\n<text text-anchor=\"start\" x=\"1371.98\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">cotranslation</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"node38\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"792.98\" cy=\"-415.66\" rx=\"35.98\" ry=\"35.98\"/>\n<text text-anchor=\"start\" x=\"780.31\" y=\"-430.06\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n<text text-anchor=\"start\" x=\"773.98\" y=\"-418.06\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n<text text-anchor=\"start\" x=\"772.32\" y=\"-406.06\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n<text text-anchor=\"start\" x=\"789.98\" y=\"-394.06\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n<g id=\"edge37\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M787.03,-630.96C788.37,-582.63 790.67,-499.85 791.98,-452.57\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n<g id=\"node40\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1149.98\" cy=\"-670.02\" rx=\"40.38\" ry=\"40.38\"/>\n<text text-anchor=\"start\" x=\"1114.65\" y=\"-666.42\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondrion</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n<g id=\"edge39\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1065.8,-859.51C1083.98,-818.81 1114.69,-750.02 1133.57,-707.76\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"node41\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"977.98\" cy=\"-670.02\" rx=\"36.76\" ry=\"36.76\"/>\n<text text-anchor=\"start\" x=\"955.64\" y=\"-684.42\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n<text text-anchor=\"start\" x=\"972.98\" y=\"-672.42\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n<text text-anchor=\"start\" x=\"954.32\" y=\"-660.42\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n<text text-anchor=\"start\" x=\"953.43\" y=\"-648.42\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n<g id=\"edge40\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1039.57,-858.06C1025.9,-816.54 1003.12,-747.35 989.42,-705.78\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;metabolism&#39;) -->\n<g id=\"node47\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"1113.41,-433.66 1054.54,-433.66 1054.54,-397.66 1113.41,-397.66 1113.41,-433.66\"/>\n<text text-anchor=\"start\" x=\"1055.98\" y=\"-412.06\" font-family=\"Times,serif\" font-size=\"12.00\">metabolism</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;) -->\n<g id=\"edge49\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1049.83,-844.53C1048.71,-768.98 1049.27,-618.16 1066.98,-491.78 1069.76,-471.92 1075.12,-449.54 1079.08,-434.42\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1046.33,-844.54 1049.99,-854.49 1053.33,-844.43 1046.33,-844.54\"/>\n<text text-anchor=\"start\" x=\"1053.98\" y=\"-667.02\" font-family=\"Times,serif\" font-size=\"10.00\">organelles</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;) -->\n<g id=\"edge50\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1112.87,-633.06C1100.18,-619.81 1088.35,-606.01 1085.1,-597.68 1063.16,-541.47 1073.5,-467.75 1080.04,-434.41\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1110.27,-635.4 1119.76,-640.11 1115.27,-630.51 1110.27,-635.4\"/>\n<text text-anchor=\"start\" x=\"1084.98\" y=\"-541.73\" font-family=\"Times,serif\" font-size=\"10.00\">mitochondrion</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"node43\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1378.98\" cy=\"-893.7\" rx=\"54.34\" ry=\"54.34\"/>\n<text text-anchor=\"start\" x=\"1341.99\" y=\"-902.1\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n<text text-anchor=\"start\" x=\"1357.65\" y=\"-890.1\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n<text text-anchor=\"start\" x=\"1359.98\" y=\"-878.1\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"edge42\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1378.98,-998.57C1378.98,-983.46 1378.98,-965.62 1378.98,-949.03\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n<g id=\"edge57\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1424.16,-1017.88C1471.16,-996.69 1537.98,-955.81 1537.98,-894.7 1537.98,-894.7 1537.98,-894.7 1537.98,-133.95 1537.98,-94.08 1505.6,-57.58 1482.84,-36.85\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1423.03,-1014.56 1415.26,-1021.75 1425.83,-1020.98 1423.03,-1014.56\"/>\n<text text-anchor=\"start\" x=\"1537.98\" y=\"-412.66\" font-family=\"Times,serif\" font-size=\"10.00\">secretory</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"node44\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"1378.98\" cy=\"-670.02\" rx=\"54.34\" ry=\"54.34\"/>\n<text text-anchor=\"start\" x=\"1371.31\" y=\"-684.42\" font-family=\"Times,serif\" font-size=\"12.00\">ion</text>\n<text text-anchor=\"start\" x=\"1341.99\" y=\"-672.42\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n<text text-anchor=\"start\" x=\"1357.65\" y=\"-660.42\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n<text text-anchor=\"start\" x=\"1359.98\" y=\"-648.42\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n<g id=\"edge43\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M1378.98,-838.47C1378.98,-803.99 1378.98,-759.52 1378.98,-725.07\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;) -->\n<g id=\"edge56\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;signalling&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1349.25,-834.13C1345.87,-827.16 1342.53,-820.11 1339.43,-813.34 1319.16,-769.02 1297.66,-715.78 1286.93,-688.71\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1346,-835.46 1353.54,-842.9 1352.28,-832.38 1346,-835.46\"/>\n<text text-anchor=\"start\" x=\"1338.98\" y=\"-774.85\" font-family=\"Times,serif\" font-size=\"10.00\">transport</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;) -->\n<g id=\"edge58\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;protein&lt;br/&gt;transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1412.57,-836C1435.21,-792.42 1460.98,-729.9 1460.98,-671.02 1460.98,-671.02 1460.98,-671.02 1460.98,-133.95 1460.98,-99.62 1460.98,-59.53 1460.98,-36.92\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1409.51,-834.3 1407.91,-844.77 1415.69,-837.59 1409.51,-834.3\"/>\n<text text-anchor=\"start\" x=\"1460.98\" y=\"-331.55\" font-family=\"Times,serif\" font-size=\"10.00\">transmembrane</text>\n<text text-anchor=\"start\" x=\"1474.02\" y=\"-321.55\" font-family=\"Times,serif\" font-size=\"10.00\">transport</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;) -->\n<g id=\"edge51\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;metabolism&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1331.19,-623.98C1326.28,-620.79 1321.18,-617.95 1315.98,-615.68 1285.5,-602.37 1192.06,-620.02 1167.43,-597.68 1131.73,-565.29 1164.77,-535.27 1143.98,-491.78 1133.54,-469.95 1115.68,-448.75 1102.15,-434.47\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1328.94,-626.67 1339.12,-629.59 1332.98,-620.96 1328.94,-626.67\"/>\n<text text-anchor=\"start\" x=\"1178.36\" y=\"-546.73\" font-family=\"Times,serif\" font-size=\"10.00\">ion</text>\n<text text-anchor=\"start\" x=\"1166.98\" y=\"-536.73\" font-family=\"Times,serif\" font-size=\"10.00\">transport</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;translation&#39;, &#39;m&#39;, &#39;e&#39;, &#39;m&#39;, &#39;b&#39;, &#39;r&#39;, &#39;a&#39;, &#39;n&#39;, &#39;e&#39;)_output -->\n<!-- (&#39;cell&#39;, &#39;translation&#39;, &#39;m&#39;, &#39;e&#39;, &#39;m&#39;, &#39;b&#39;, &#39;r&#39;, &#39;a&#39;, &#39;n&#39;, &#39;e&#39;)_output&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;) -->\n<g id=\"edge61\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;translation&#39;, &#39;m&#39;, &#39;e&#39;, &#39;m&#39;, &#39;b&#39;, &#39;r&#39;, &#39;a&#39;, &#39;n&#39;, &#39;e&#39;)_output&#45;&gt;(&#39;cell&#39;, &#39;translation&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1342.51,-393.65C1336.17,-363.87 1324.69,-309.88 1318.75,-281.98\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"1339.06,-394.26 1344.56,-403.31 1345.91,-392.8 1339.06,-394.26\"/>\n<text text-anchor=\"start\" x=\"1330.98\" y=\"-326.55\" font-family=\"Times,serif\" font-size=\"10.00\">membrane</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x119569e50>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 55
  },
  {
   "cell_type": "code",
   "id": "fe3008ad-498e-449f-8521-6ea29fd903f1",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.920443Z",
     "start_time": "2025-02-12T16:39:55.881857Z"
    }
   },
   "source": [
    "new_process = {\n",
    "    'hypothesized mechanism': {\n",
    "        '_type': 'step',\n",
    "        '_inputs': {\n",
    "            '1': 'Any',\n",
    "            '2': 'Any',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "new_process = replace_regex_recursive(new_process)\n",
    "\n",
    "# plot\n",
    "plot_settingsx = copy.deepcopy(plot_settings)\n",
    "plot_settingsx['label_margin'] = '0.03'\n",
    "plot_settingsx['node_border_colors'] = {\n",
    "    ('hypothesized<br/>mechanism',): 'blue',\n",
    "}\n",
    "plot_settingsx['port_labels'] = False\n",
    "plot_bigraph(new_process, **plot_settingsx, filename='hypothesized_mechanism')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/hypothesized_mechanism\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"74pt\" height=\"100pt\"\n viewBox=\"0.00 0.00 75.64 103.08\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 99.08)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-99.08 71.64,-99.08 71.64,4 -4,4\"/>\n<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;hypothesized&lt;br/&gt;mechanism&#39;,)</title>\n<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"67.64,-36 0,-36 0,0 67.64,0 67.64,-36\"/>\n<text text-anchor=\"start\" x=\"2.16\" y=\"-20.4\" font-family=\"Times,serif\" font-size=\"12.00\">hypothesized</text>\n<text text-anchor=\"start\" x=\"6.49\" y=\"-8.4\" font-family=\"Times,serif\" font-size=\"12.00\">mechanism</text>\n</g>\n<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M17.21,-71.76C19.3,-65.11 22.09,-56.25 24.79,-47.68\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"28.09,-48.84 27.76,-38.26 21.42,-46.74 28.09,-48.84\"/>\n</g>\n<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;2&#39;)_input -->\n<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;2&#39;)_input&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;2&#39;)_input&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M51.26,-71.76C49.06,-65.11 46.13,-56.25 43.3,-47.68\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"46.64,-46.65 40.18,-38.25 39.99,-48.84 46.64,-46.65\"/>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1194458e0>"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 56
  },
  {
   "cell_type": "markdown",
   "id": "ae52f73c-e384-4880-b18f-47b4c25ce203",
   "metadata": {},
   "source": [
    "### Cell structure and function\n",
    "Processes represent their functions, we show a more minimal cell structure with added processes."
   ]
  },
  {
   "cell_type": "code",
   "id": "b39417ce-d9a1-4985-abce-ce22df816897",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:55.972680Z",
     "start_time": "2025-02-12T16:39:55.926223Z"
    }
   },
   "source": [
    "core = VisualizeTypes()\n",
    "core.register(key='concentrations', schema='float')\n",
    "core.register(key='sequences', schema='float')\n",
    "core.register(key='membrane', schema={\n",
    "    'transporters': 'concentrations',\n",
    "    'lipids': 'concentrations',\n",
    "    'transmembrane transport': {\n",
    "        '_type': 'process',\n",
    "        '_inputs': {},\n",
    "        '_outputs': {\n",
    "            'transporters': 'concentrations',\n",
    "            'internal': 'concentrations',\n",
    "            'external': 'concentrations'\n",
    "        }\n",
    "    }\n",
    "})\n",
    "\n",
    "core.register(key='cytoplasm', schema={\n",
    "    'metabolites': 'concentrations',\n",
    "    'ribosomal complexes': 'concentrations',\n",
    "    'transcript regulation complex': {\n",
    "        'transcripts': 'concentrations'},\n",
    "    'translation': {\n",
    "        '_type': 'process',\n",
    "        '_outputs': {\n",
    "            'p1': 'concentrations',\n",
    "            'p2': 'concentrations'}}})\n",
    "\n",
    "core.register(\n",
    "    key='nucleoid',\n",
    "    schema={\n",
    "        'chromosome': {\n",
    "            'genes': 'sequences'}})\n",
    "\n",
    "core.register(\n",
    "    key='cell',\n",
    "    schema={\n",
    "        'membrane': 'membrane',\n",
    "        'cytoplasm': 'cytoplasm',\n",
    "        'nucleoid': 'nucleoid'})\n",
    "\n",
    "# state\n",
    "cell_struct_state = {\n",
    "    'cell': {\n",
    "        'membrane': {\n",
    "            'transmembrane transport': {\n",
    "                'outputs': {\n",
    "                    'transporters': ['transporters'],\n",
    "                    'internal': ['..', 'cytoplasm', 'metabolites']}}},\n",
    "        'cytoplasm': {\n",
    "            'translation': {\n",
    "                'outputs': {\n",
    "                    'p1': ['ribosomal complexes'],\n",
    "                    'p2': ['transcript regulation complex', 'transcripts']}}}}}\n",
    "\n",
    "plot_bigraph(\n",
    "    cell_struct_state,\n",
    "    schema={'cell': 'cell'},\n",
    "    core=core,\n",
    "    remove_process_place_edges=True,\n",
    "    out_dir='out',\n",
    "    filename='cell')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"599pt\" height=\"658pt\"\n viewBox=\"0.00 0.00 615.92 676.98\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 672.98)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-672.98 611.92,-672.98 611.92,4 -4,4\"/>\n<!-- (&#39;cell&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;cell&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"304.02\" cy=\"-650.98\" rx=\"18\" ry=\"18\"/>\n<text text-anchor=\"start\" x=\"295.36\" y=\"-647.38\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"96.02\" cy=\"-561.79\" rx=\"34.19\" ry=\"34.19\"/>\n<text text-anchor=\"start\" x=\"70.7\" y=\"-558.19\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n</g>\n<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M287,-642.84C252.26,-628.28 172.73,-594.95 128.03,-576.21\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"304.02\" cy=\"-561.79\" rx=\"33.41\" ry=\"33.41\"/>\n<text text-anchor=\"start\" x=\"279.36\" y=\"-558.19\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n</g>\n<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M304.02,-632.34C304.02,-621.89 304.02,-608.31 304.02,-595.93\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleoid&#39;) -->\n<g id=\"node10\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"422.02\" cy=\"-561.79\" rx=\"28.68\" ry=\"28.68\"/>\n<text text-anchor=\"start\" x=\"401.36\" y=\"-558.19\" font-family=\"Times,serif\" font-size=\"12.00\">nucleoid</text>\n</g>\n<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;) -->\n<g id=\"edge9\" class=\"edge\">\n<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M318.49,-639.29C338.55,-624.46 375.04,-597.5 398.9,-579.87\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"96.02\" cy=\"-402.83\" rx=\"37.74\" ry=\"37.74\"/>\n<text text-anchor=\"start\" x=\"67.7\" y=\"-399.23\" font-family=\"Times,serif\" font-size=\"12.00\">transporters</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M96.02,-526.73C96.02,-501.58 96.02,-467.34 96.02,-441.39\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"20.02\" cy=\"-402.83\" rx=\"20.02\" ry=\"20.02\"/>\n<text text-anchor=\"start\" x=\"6.69\" y=\"-399.23\" font-family=\"Times,serif\" font-size=\"12.00\">lipids</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M75.24,-533.48C66.34,-520.9 56.32,-505.48 49.02,-490.6 38.1,-468.36 29.72,-441 24.84,-423.03\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n<g id=\"node13\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"249.43,-133.8 122.61,-133.8 122.61,-97.8 249.43,-97.8 249.43,-133.8\"/>\n<text text-anchor=\"start\" x=\"126.21\" y=\"-112.2\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane transport</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n<g id=\"edge12\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M95.89,-353.07C97.62,-303.74 105.58,-226.56 135.81,-167.6 142.27,-155.03 152.93,-143.58 162.77,-134.74\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"92.39,-352.77 95.62,-362.86 99.39,-352.96 92.39,-352.77\"/>\n<text text-anchor=\"start\" x=\"136.02\" y=\"-229.33\" font-family=\"Times,serif\" font-size=\"10.00\">transporters</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"189.02\" cy=\"-402.83\" rx=\"36.95\" ry=\"36.95\"/>\n<text text-anchor=\"start\" x=\"161.36\" y=\"-399.23\" font-family=\"Times,serif\" font-size=\"12.00\">metabolites</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M278.89,-538.62C264.93,-525.52 247.84,-508.11 235.02,-490.6 222.93,-474.09 212.07,-454 203.94,-437.3\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;) -->\n<g id=\"node7\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"342.02\" cy=\"-232.33\" rx=\"64.73\" ry=\"64.73\"/>\n<text text-anchor=\"start\" x=\"290.87\" y=\"-228.73\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal complexes</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M304.99,-527.62C306.75,-480.24 311.4,-390.55 323.02,-315.06 323.97,-308.87 325.14,-302.46 326.43,-296.06\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;) -->\n<g id=\"node8\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"420.02\" cy=\"-402.83\" rx=\"87.77\" ry=\"87.77\"/>\n<text text-anchor=\"start\" x=\"349.38\" y=\"-399.23\" font-family=\"Times,serif\" font-size=\"12.00\">transcript regulation complex</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;) -->\n<g id=\"edge7\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M323.96,-533.81C336.14,-517.33 352.33,-495.43 367.94,-474.31\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n<g id=\"edge13\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M188.25,-353.7C187.98,-335.91 187.69,-315.57 187.48,-297.06 186.8,-237.6 186.33,-167.11 186.13,-134.71\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"184.74,-353.52 188.4,-363.47 191.74,-353.42 184.74,-353.52\"/>\n<text text-anchor=\"start\" x=\"188.02\" y=\"-229.33\" font-family=\"Times,serif\" font-size=\"10.00\">internal</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n<g id=\"node14\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"390.95,-36 333.09,-36 333.09,0 390.95,0 390.95,-36\"/>\n<text text-anchor=\"start\" x=\"336.69\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">translation</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n<g id=\"edge15\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M349.14,-155.78C353.17,-113 357.88,-62.97 360.34,-36.89\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"345.67,-155.26 348.22,-165.54 352.64,-155.92 345.67,-155.26\"/>\n<text text-anchor=\"start\" x=\"356.02\" y=\"-112.8\" font-family=\"Times,serif\" font-size=\"10.00\">p1</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;) -->\n<g id=\"node9\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"413.02\" cy=\"-115.8\" rx=\"33.8\" ry=\"33.8\"/>\n<text text-anchor=\"start\" x=\"388.03\" y=\"-112.2\" font-family=\"Times,serif\" font-size=\"12.00\">transcripts</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;) -->\n<g id=\"edge8\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M417.87,-314.38C416.51,-259.06 414.84,-190.68 413.85,-150.58\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n<g id=\"edge14\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M391.87,-75.06C384.77,-61.74 377.21,-47.54 371.48,-36.77\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"388.66,-76.5 396.45,-83.68 394.84,-73.21 388.66,-76.5\"/>\n<text text-anchor=\"start\" x=\"385.02\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">p2</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n<g id=\"node11\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"567.02\" cy=\"-402.83\" rx=\"40.89\" ry=\"40.89\"/>\n<text text-anchor=\"start\" x=\"536.03\" y=\"-399.23\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n<g id=\"edge10\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M447.27,-546.2C467.83,-533.43 496.68,-513.4 517.02,-490.6 530.22,-475.81 541.58,-456.99 550.12,-440.66\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n<g id=\"node12\" class=\"node\">\n<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"567.02\" cy=\"-115.8\" rx=\"20.41\" ry=\"20.41\"/>\n<text text-anchor=\"start\" x=\"553.36\" y=\"-112.2\" font-family=\"Times,serif\" font-size=\"12.00\">genes</text>\n</g>\n<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n<g id=\"edge11\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M567.02,-360.99C567.02,-299.11 567.02,-183.87 567.02,-137.2\"/>\n</g>\n<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;e&#39;, &#39;x&#39;, &#39;t&#39;, &#39;e&#39;, &#39;r&#39;, &#39;n&#39;, &#39;a&#39;, &#39;l&#39;)_output -->\n<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;e&#39;, &#39;x&#39;, &#39;t&#39;, &#39;e&#39;, &#39;r&#39;, &#39;n&#39;, &#39;a&#39;, &#39;l&#39;)_output&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n<g id=\"edge16\" class=\"edge\">\n<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;e&#39;, &#39;x&#39;, &#39;t&#39;, &#39;e&#39;, &#39;r&#39;, &#39;n&#39;, &#39;a&#39;, &#39;l&#39;)_output&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M252.34,-377.5C247.53,-356.33 240.65,-324.78 235.82,-297.06 225.84,-239.85 237.9,-222.17 218.02,-167.6 213.74,-155.85 206.44,-143.95 199.89,-134.63\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"248.92,-378.22 254.57,-387.18 255.74,-376.65 248.92,-378.22\"/>\n<text text-anchor=\"start\" x=\"236.02\" y=\"-229.33\" font-family=\"Times,serif\" font-size=\"10.00\">external</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x10fe148b0>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 57
  },
  {
   "cell_type": "markdown",
   "id": "188eb4a6-c526-4b38-a22d-97f7fb8b7244",
   "metadata": {},
   "source": [
    "### Whole-cell *E. coli* model\n",
    "\n",
    "We here plot the processes and wiring of a whole-cell model of *E. coli* from [Macklin et al. Simultaneous cross-evaluation of heterogeneous E. coli datasets via mechanistic simulation. Science. 2020.](https://pubmed.ncbi.nlm.nih.gov/32703847/)"
   ]
  },
  {
   "cell_type": "code",
   "id": "16960d1b-6e35-421e-bb7d-3efe5052e4fc",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-02-12T16:39:56.041102Z",
     "start_time": "2025-02-12T16:39:55.978968Z"
    }
   },
   "source": [
    "ecoli = { \n",
    "    'chromosome-structure': { \n",
    "        'wires': {\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'DnaA_boxes': ('unique', 'DnaA_box')\n",
    "        }\n",
    "    },\n",
    "    'metabolism': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'catalysts': ('bulk',),\n",
    "            'kinetics_enzymes': ('bulk',),\n",
    "            'kinetics_substrates': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'environment': ('environment',),\n",
    "            'amino_acids_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'tf-binding': { \n",
    "        'wires': {\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'active_tfs_total': ('bulk',),\n",
    "            'inactive_tfs_total': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'molecules': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'molecules': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'ntps': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'rna-degradation': { \n",
    "        'wires': {\n",
    "            'charged_trna': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'nmps': ('bulk',),\n",
    "            'fragmentMetabolites': ('bulk',),\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'endoRnases': ('bulk',),\n",
    "            'exoRnases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'RNA': ('unique', 'RNA'),\n",
    "            'subunits': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'molecules': ('bulk',),\n",
    "            'monomers': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'ppgpp_reaction_metabolites': ('bulk',),\n",
    "            'uncharged_trna': ('bulk',),\n",
    "            'charged_trna': ('bulk',),\n",
    "            'charging_molecules': ('bulk',),\n",
    "            'synthetases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules_total': ('bulk',),\n",
    "            'amino_acids_total': ('bulk',),\n",
    "            'charged_trna_total': ('bulk',),\n",
    "            'uncharged_trna_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'complexation': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',), \n",
    "        }\n",
    "    },\n",
    "    'two-component-system': { \n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'equilibrium': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'protein-degradation': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'proteins': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'chromosome-replication': { \n",
    "        'wires': {\n",
    "            'replisome_trimers': ('bulk',),\n",
    "            'replisome_monomers': ('bulk',),\n",
    "            'dntps': ('bulk',),\n",
    "            'ppi': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'environment': ('environment',)\n",
    "        }\n",
    "    },\n",
    "    'unique': {\n",
    "        'chromosome_domain': {},\n",
    "        'full_chromosome': {},\n",
    "        'oriC': {},\n",
    "        'active_replisome': {},\n",
    "        'RNA': {},\n",
    "        'active_ribosome': {},\n",
    "        'DnaA_box': {},\n",
    "        'promoter': {},\n",
    "    },\n",
    "    'bulk': {},\n",
    "    'environment': {},\n",
    "}\n",
    "node_groups = [[ #line up the top nodes for visual effect\n",
    "        ('unique',), ('bulk',), ('environment',), \n",
    "    ],\n",
    "]\n",
    "# plot settings\n",
    "plot_settings_ecoli = {\n",
    "    'remove_process_place_edges': True,\n",
    "    'rankdir': 'RL',\n",
    "    'dpi': '100',\n",
    "    'out_dir': 'out',\n",
    "}\n",
    "fig = plot_bigraph(ecoli, **plot_settings_ecoli, node_groups=node_groups, filename='ecoli')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/ecoli\n"
     ]
    }
   ],
   "execution_count": 58
  },
  {
   "cell_type": "markdown",
   "id": "54e1102d-521c-49cc-8ed2-be26c2b84c3d",
   "metadata": {
    "tags": []
   },
   "source": [
    "[See E. coli wiring diagram](https://github.com/vivarium-collective/bigraph-viz/blob/main/doc/_static/ecoli.png?raw=true)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
