{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5338cfab-cf15-4c63-a091-c1cbb51bf8dd",
   "metadata": {},
   "source": [
    "# Process Bigraph Schema\n",
    "This tutorial covers Process Bigraph Schema, starting with single `store` and increasing in detail as you scroll towards the bottom. It uses the libraries [bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) for the schema validation and type registry and [bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) for visualization.\n",
    "\n",
    "## Bigraph-schema and Bigraph-viz\n",
    "\n",
    "[Bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) is a serializable type schema for Process Bigraph simulations. From this library, we will use the `validate_schema` method to check our schemas, and `fill` to fill in missing details to prepare the schema for simulation. We will also use the `type_registry` to access different schemas, and register new ones.\n",
    "\n",
    "[Bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) is an easy-to-use plotting tool for compositional bigraph schema, built on top of Graphviz. It reads raw Compositional Bigraph Schemas and generates a figure that can be displayed in a notebook or saved to file. Bigraph-viz's main plotting function is `plot_bigraph` for the bigraph structure. `plot_flow` and `plot_multitimestep` make plots of process orchestration."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df69e36f-3da8-4bfb-8bc0-0aeb04a6f7d7",
   "metadata": {},
   "source": [
    "### Installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63fb7d1a-cd49-4aac-8d4f-31bb54496a51",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-e git+https://github.com/vivarium-collective/bigraph-schema.git@0c90a88ffe83c4b88d057e28a4f30fb19ef0263e#egg=bigraph_schema\n",
      "bigraph-viz==0.0.31\n",
      "-e git+https://github.com/vivarium-collective/process-bigraph.git@6517fc88103bc3cfaad4d5a1fdbd9131015360d4#egg=process_bigraph\n"
     ]
    }
   ],
   "source": [
    "# !pip install bigraph-schema\n",
    "# !pip install bigraph-viz\n",
    "!pip freeze | grep bigraph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f700afa2-bfee-49bf-a4b1-e23f0540461d",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2ab3cb32-e007-45c4-807e-b1b442ac1608",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from bigraph_viz import plot_bigraph, plot_flow, plot_multitimestep, pf\n",
    "from bigraph_viz.dict_utils import replace_regex_recursive\n",
    "from bigraph_viz.diagram import plot_bigraph\n",
    "\n",
    "plot_settings = {\n",
    "    # 'remove_process_place_edges': True\n",
    "}\n",
    "save_images = False\n",
    "if save_images:\n",
    "    plot_settings.update({'out_dir': 'out','dpi': '250'})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e716c23-5fbe-4c08-a868-09611ca40ec4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Stores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a6e9cc2-e0df-4f73-b4c3-eddb1055b088",
   "metadata": {},
   "source": [
    "### Simple stores\n",
    "A store can be declared in a JSON dictionary with the store's name mapped to its `value`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d711b952-089a-4938-817d-d1ccad9de56e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/simple_store\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"74pt\" height=\"74pt\"\n",
       " viewBox=\"0.00 0.00 76.40 76.40\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 72.4)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-72.4 72.4,-72.4 72.4,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"34.2\" cy=\"-34.2\" rx=\"34.2\" ry=\"34.2\"/>\n",
       "<text text-anchor=\"start\" x=\"8.87\" y=\"-30.6\" font-family=\"Times,serif\" font-size=\"12.00\">store1: 1.0</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x105b709a0>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simple_store_state = {\n",
    "    'store1': 1.0,\n",
    "}\n",
    "plot_bigraph(simple_store_state, **plot_settings, show_values=True, filename='simple_store')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db1cead7-e691-4291-920e-827d69c00727",
   "metadata": {},
   "source": [
    "To include values *and* types in a node label, you can instead have the store name map to a dict with `_value` and `_type` keys."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f3e5fa71-7ffc-4a8f-a1a0-dc547c886f84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/store\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"58pt\" height=\"58pt\"\n",
       " viewBox=\"0.00 0.00 59.66 59.66\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 55.66)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-55.66 55.66,-55.66 55.66,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"25.83\" cy=\"-25.83\" rx=\"25.83\" ry=\"25.83\"/>\n",
       "<text text-anchor=\"start\" x=\"11.16\" y=\"-28.23\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "<text text-anchor=\"start\" x=\"11.16\" y=\"-16.23\" font-family=\"Times,serif\" font-size=\"12.00\">[float]</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x105ca8e80>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "typed_store_spec = {\n",
    "    'store1': {\n",
    "        '_value': 1.0, \n",
    "        '_type': 'float',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(typed_store_spec, **plot_settings, show_values=True, show_types=True, filename='store')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de1011cb-903b-4c15-baed-a62f4afdd6b9",
   "metadata": {},
   "source": [
    "### Hierarchy\n",
    "A hierarchy is a place graph of nested stores. Stores can be placed within stores using JSON dictionary nesting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0f3f0824-f89d-4c6e-830e-2c5a321f7b33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/hierarchy\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"235pt\" height=\"343pt\"\n",
       " viewBox=\"0.00 0.00 241.78 352.56\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 348.56)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-348.56 237.78,-348.56 237.78,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"150.23\" cy=\"-322.97\" rx=\"21.59\" ry=\"21.59\"/>\n",
       "<text text-anchor=\"start\" x=\"135.57\" y=\"-319.37\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"114.23\" cy=\"-238.47\" rx=\"26.91\" ry=\"26.91\"/>\n",
       "<text text-anchor=\"start\" x=\"95.07\" y=\"-234.87\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M141.7,-302.42C136.67,-290.88 130.25,-276.19 124.89,-263.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"186.23\" cy=\"-238.47\" rx=\"26.91\" ry=\"26.91\"/>\n",
       "<text text-anchor=\"start\" x=\"167.07\" y=\"-234.87\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M158.76,-302.42C163.79,-290.88 170.2,-276.19 175.57,-263.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"32.23\" cy=\"-143.33\" rx=\"32.23\" ry=\"32.23\"/>\n",
       "<text text-anchor=\"start\" x=\"8.57\" y=\"-139.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M96.44,-217.26C83.81,-202.92 66.81,-183.61 53.48,-168.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"114.23\" cy=\"-143.33\" rx=\"32.23\" ry=\"32.23\"/>\n",
       "<text text-anchor=\"start\" x=\"90.57\" y=\"-139.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M114.23,-211.03C114.23,-200.27 114.23,-187.74 114.23,-176.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"196.23\" cy=\"-143.33\" rx=\"32.23\" ry=\"32.23\"/>\n",
       "<text text-anchor=\"start\" x=\"172.57\" y=\"-139.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M132.02,-217.26C144.65,-202.92 161.65,-183.61 174.98,-168.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"196.23\" cy=\"-37.55\" rx=\"37.55\" ry=\"37.55\"/>\n",
       "<text text-anchor=\"start\" x=\"168.07\" y=\"-33.95\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M196.23,-110.34C196.23,-99.5 196.23,-87.3 196.23,-76.07\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x102e5e670>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hierarchy_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': {\n",
    "            'store1.1.1': 'Any',\n",
    "            'store1.1.2': 'Any',\n",
    "            'store1.1.3': {\n",
    "                'store1.1.3.1': 'Any',\n",
    "            },\n",
    "        },\n",
    "        'store1.2': 'Any',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(hierarchy_spec, **plot_settings, filename='hierarchy')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c825e07c-8cd5-4df1-8958-aa0f36d61045",
   "metadata": {},
   "source": [
    "## Processes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ea1a83-aeac-4048-9593-7b30588b0ad8",
   "metadata": {},
   "source": [
    "### Single process\n",
    "Processes require ports, which are declared with the `_ports` key mapped to a dict that has port names matched to types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7c74237b-0d66-4e8f-a8ca-79174fadad8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/process\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"116pt\" height=\"95pt\"\n",
       " viewBox=\"0.00 0.00 119.83 98.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 94)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-94 115.83,-94 115.83,4 -4,4\"/>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-63 0,-63 0,-27 54,-27 54,-63\"/>\n",
       "<text text-anchor=\"start\" x=\"6.01\" y=\"-41.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M110.18,-71.7C105.62,-70.22 85.18,-63.58 65.82,-57.29\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"67.01,-53.99 56.42,-54.23 64.85,-60.65 67.01,-53.99\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-67\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M110.18,-18.3C105.62,-19.78 85.18,-26.42 65.82,-32.71\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"64.85,-29.35 56.42,-35.77 67.01,-36.01 64.85,-29.35\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-32\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x105b37280>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_spec = {\n",
    "    'process1': {\n",
    "        '_type': 'edge',\n",
    "        '_inputs': {\n",
    "            'port1': 'Any',\n",
    "            'port2': 'Any',\n",
    "        },\n",
    "    },\n",
    "}\n",
    "plot_bigraph(process_spec, **plot_settings, rankdir='RL', filename='process')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68a2c480-32fe-43d3-b618-8fc6c647b8ca",
   "metadata": {},
   "source": [
    "### Multiple processes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8c2457a3-fc47-4850-8b9e-9ee98e81567d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/multiple_processes\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"251pt\" height=\"123pt\"\n",
       " viewBox=\"0.00 0.00 258.11 126.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 122)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-122 254.11,-122 254.11,4 -4,4\"/>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-118 0,-118 0,-82 54,-82 54,-118\"/>\n",
       "<text text-anchor=\"start\" x=\"6.01\" y=\"-96.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"141,-118 87,-118 87,-82 141,-82 141,-118\"/>\n",
       "<text text-anchor=\"start\" x=\"93.01\" y=\"-96.4\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;process3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"243,-118 189,-118 189,-82 243,-82 243,-118\"/>\n",
       "<text text-anchor=\"start\" x=\"195.01\" y=\"-96.4\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M5.33,-31.02C6.53,-40.19 8.62,-53.06 11.89,-64 12.53,-66.14 13.27,-68.32 14.06,-70.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"10.79,-71.76 17.77,-79.73 17.28,-69.15 10.79,-71.76\"/>\n",
       "<text text-anchor=\"start\" x=\"12\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M46.75,-23.76C44.37,-32.82 39.04,-53.13 34.51,-70.37\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"31.23,-69.08 32.08,-79.64 38.01,-70.86 31.23,-69.08\"/>\n",
       "<text text-anchor=\"start\" x=\"39\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;) -->\n",
       "<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M105.61,-36.57C105.62,-44.91 105.9,-54.99 106.89,-64 107.12,-66.04 107.4,-68.13 107.73,-70.24\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"104.24,-70.6 109.48,-79.8 111.12,-69.34 104.24,-70.6\"/>\n",
       "<text text-anchor=\"start\" x=\"107\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;) -->\n",
       "<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M148.8,-21.66C145.38,-29.28 135.23,-51.81 126.78,-70.6\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"123.59,-69.16 122.68,-79.72 129.97,-72.03 123.59,-69.16\"/>\n",
       "<text text-anchor=\"start\" x=\"135\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;) -->\n",
       "<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M192.66,-29.96C191.89,-39.23 191.64,-52.74 194.89,-64 195.6,-66.44 196.51,-68.88 197.58,-71.27\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"194.41,-72.78 202.19,-79.98 200.6,-69.5 194.41,-72.78\"/>\n",
       "<text text-anchor=\"start\" x=\"195\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;) -->\n",
       "<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M236.69,-23.76C234.2,-32.82 228.61,-53.13 223.87,-70.37\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"220.6,-69.08 221.32,-79.65 227.35,-70.94 220.6,-69.08\"/>\n",
       "<text text-anchor=\"start\" x=\"229\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x105b37a60>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_schema = {\n",
    "    '_type': 'process',\n",
    "    '_inputs': {\n",
    "        'port1': 'Any',\n",
    "        'port2': 'Any'\n",
    "    }\n",
    "}\n",
    "\n",
    "processes_spec = {\n",
    "    'process1': process_schema,\n",
    "    'process2': process_schema,\n",
    "    'process3': process_schema,\n",
    "}\n",
    "plot_bigraph(processes_spec, **plot_settings, rankdir='BT', filename='multiple_processes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e5f5a6-8dd4-41c3-ba37-33ed9e3e0830",
   "metadata": {},
   "source": [
    "## Wires\n",
    "To connect a process's ports to stores, add wiring with a `_wires` key that maps port names to relative paths in the store hierarchy. The ports used by the wires need to match the ports in the schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a6ecad5d-8948-4636-910c-79284a3e8679",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/wires\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"168pt\" height=\"129pt\"\n",
       " viewBox=\"0.00 0.00 173.18 133.18\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 129.18)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-129.18 169.18,-129.18 169.18,4 -4,4\"/>\n",
       "<!-- (&#39;node1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;node1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"21.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n",
       "<text text-anchor=\"start\" x=\"6.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"109.59,-36 55.59,-36 55.59,0 109.59,0 109.59,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"61.6\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M34.23,-85.27C42.54,-73.88 53.59,-58.75 63.02,-45.81\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"65.59,-48.24 68.65,-38.1 59.93,-44.11 65.59,-48.24\"/>\n",
       "<text text-anchor=\"start\" x=\"55.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;node2&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"82.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n",
       "<text text-anchor=\"start\" x=\"67.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node2</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M82.59,-81.06C82.59,-71.08 82.59,-59.07 82.59,-48.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"86.09,-48.3 82.59,-38.3 79.09,-48.3 86.09,-48.3\"/>\n",
       "<text text-anchor=\"start\" x=\"82.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;node3&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n",
       "<text text-anchor=\"start\" x=\"128.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node3</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M130.95,-85.27C122.64,-73.88 111.59,-58.75 102.16,-45.81\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"105.25,-44.11 96.53,-38.1 99.59,-48.24 105.25,-44.11\"/>\n",
       "<text text-anchor=\"start\" x=\"114.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port3</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x11a687d00>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "connected_process_spec = {\n",
    "    'process1': {\n",
    "        '_type': 'edge',\n",
    "        '_inputs': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'int',\n",
    "            'port3': 'int',\n",
    "        },\n",
    "        'inputs': {\n",
    "            'port1': ['node1'],\n",
    "            'port2': ['node2'],\n",
    "            'port3': ['node3'],\n",
    "        }\n",
    "    },\n",
    "    'node1': 'float',  # TODO -- shouldn't these fill?\n",
    "    'node2': 'int',\n",
    "    'node3': 'int',\n",
    "}\n",
    "plot_bigraph(connected_process_spec, **plot_settings, filename='wires')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e037285d-116a-404d-96ec-ba47318a2e8b",
   "metadata": {},
   "source": [
    "### Advanced wiring\n",
    "**TODO**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b17201-2489-45c8-a06b-9fdc29a714a7",
   "metadata": {},
   "source": [
    "## Composites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf5d428-db23-4985-80ff-6b93d22973aa",
   "metadata": {},
   "source": [
    "### Flat composite\n",
    "A flat composite connects stores with multiple connected processes, without nesting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "781a84e6-bbe2-4708-b441-47efc611b46f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/flat_composite\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"168pt\" height=\"137pt\"\n",
       " viewBox=\"0.00 0.00 172.93 140.82\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 136.82)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-136.82 168.93,-136.82 168.93,4 -4,4\"/>\n",
       "<!-- (&#39;store1.1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1.1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"138.02\" cy=\"-105.91\" rx=\"26.91\" ry=\"26.91\"/>\n",
       "<text text-anchor=\"start\" x=\"118.86\" y=\"-102.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-118.91 0,-118.91 0,-82.91 54,-82.91 54,-118.91\"/>\n",
       "<text text-anchor=\"start\" x=\"6.01\" y=\"-97.31\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M99.1,-104.17C84.49,-103.5 68.12,-102.75 54.77,-102.14\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"98.69,-107.66 108.84,-104.62 99.01,-100.66 98.69,-107.66\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-104.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-53.91 0,-53.91 0,-17.91 54,-17.91 54,-53.91\"/>\n",
       "<text text-anchor=\"start\" x=\"6.01\" y=\"-32.31\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M104.96,-85.33C89.01,-75.09 69.95,-62.85 54.79,-53.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"102.85,-88.13 113.16,-90.59 106.63,-82.24 102.85,-88.13\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-78.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1.2&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"138.02\" cy=\"-26.91\" rx=\"26.91\" ry=\"26.91\"/>\n",
       "<text text-anchor=\"start\" x=\"118.86\" y=\"-23.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M100.28,-36.89C90.61,-40.45 80.51,-45.07 72,-50.91 59.8,-59.28 48.8,-71.78 40.85,-82.19\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"101.12,-40.3 109.48,-33.79 98.88,-33.67 101.12,-40.3\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-52.91\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M100.86,-15C91.44,-13.33 81.31,-12.8 72,-14.91 66.22,-16.22 60.31,-18.38 54.78,-20.83\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"99.91,-18.38 110.44,-17.18 101.46,-11.55 99.91,-18.38\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-16.91\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x11a5fa280>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flat_composite_spec = {\n",
    "    'store1.1': 'float',\n",
    "    'store1.2': 'int',\n",
    "    'process1': {\n",
    "        '_type': 'process',\n",
    "        'outputs': {\n",
    "            'port1': ['store1.1'],\n",
    "            'port2': ['store1.2'],\n",
    "        }\n",
    "    },\n",
    "    'process2': {\n",
    "        '_type': 'process',\n",
    "        'inputs': {\n",
    "            'port1': ['store1.1'],\n",
    "            'port2': ['store1.2'],\n",
    "        }\n",
    "    },\n",
    "}\n",
    "plot_bigraph(flat_composite_spec, **plot_settings, rankdir='RL', filename='flat_composite')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398509c7-cf5c-4cce-a179-3a39752b0dde",
   "metadata": {},
   "source": [
    "### Nested composite\n",
    "A nested composite has store hierarchies with multiple connected processes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "701e759d-7f9b-41bb-911c-a8208c892462",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "argument of type 'NoneType' is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[24], line 27\u001b[0m\n\u001b[1;32m      1\u001b[0m nested_composite_spec \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstore1\u001b[39m\u001b[38;5;124m'\u001b[39m: {\n\u001b[1;32m      3\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstore1.1\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfloat\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     25\u001b[0m     }\n\u001b[1;32m     26\u001b[0m }\n\u001b[0;32m---> 27\u001b[0m \u001b[43mplot_bigraph\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnested_composite_spec\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mplot_settings\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfilename\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mnested_composite\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/bigraph-viz/venv/lib/python3.9/site-packages/bigraph_viz/diagram.py:336\u001b[0m, in \u001b[0;36mplot_bigraph\u001b[0;34m(state, schema, core, out_dir, filename, file_format, size, node_label_size, show_values, show_types, port_labels, port_label_size, rankdir, print_source, dpi, label_margin, remove_nodes)\u001b[0m\n\u001b[1;32m    333\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m core\u001b[38;5;241m.\u001b[39mexists(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprocess\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m    334\u001b[0m     core\u001b[38;5;241m.\u001b[39mregister(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprocess\u001b[39m\u001b[38;5;124m'\u001b[39m, process_type)\n\u001b[0;32m--> 336\u001b[0m schema, state \u001b[38;5;241m=\u001b[39m \u001b[43mcore\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcomplete\u001b[49m\u001b[43m(\u001b[49m\u001b[43mschema\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstate\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    338\u001b[0m \u001b[38;5;66;03m# parse out the network\u001b[39;00m\n\u001b[1;32m    339\u001b[0m graph_dict \u001b[38;5;241m=\u001b[39m get_graph_dict(\n\u001b[1;32m    340\u001b[0m     schema\u001b[38;5;241m=\u001b[39mschema,\n\u001b[1;32m    341\u001b[0m     state\u001b[38;5;241m=\u001b[39mstate,\n\u001b[1;32m    342\u001b[0m     core\u001b[38;5;241m=\u001b[39mcore,\n\u001b[1;32m    343\u001b[0m     remove_nodes\u001b[38;5;241m=\u001b[39mremove_nodes,\n\u001b[1;32m    344\u001b[0m )\n",
      "File \u001b[0;32m~/code/bigraph-schema/bigraph_schema/type_system.py:1251\u001b[0m, in \u001b[0;36mTypeSystem.complete\u001b[0;34m(self, initial_schema, initial_state)\u001b[0m\n\u001b[1;32m   1245\u001b[0m state \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhydrate(\n\u001b[1;32m   1246\u001b[0m     full_schema,\n\u001b[1;32m   1247\u001b[0m     initial_state)\n\u001b[1;32m   1249\u001b[0m \u001b[38;5;66;03m# fill in the parts of the composition schema\u001b[39;00m\n\u001b[1;32m   1250\u001b[0m \u001b[38;5;66;03m# determined by the state\u001b[39;00m\n\u001b[0;32m-> 1251\u001b[0m schema, state \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minfer_schema\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1252\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfull_schema\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1253\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstate\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1255\u001b[0m final_state \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfill(schema, state)\n\u001b[1;32m   1257\u001b[0m \u001b[38;5;66;03m# TODO: add flag to types.access(copy=True)\u001b[39;00m\n",
      "File \u001b[0;32m~/code/bigraph-schema/bigraph_schema/type_system.py:1177\u001b[0m, in \u001b[0;36mTypeSystem.infer_schema\u001b[0;34m(self, schema, state, top_state, path)\u001b[0m\n\u001b[1;32m   1174\u001b[0m             inner_path \u001b[38;5;241m=\u001b[39m path \u001b[38;5;241m+\u001b[39m (key,)\n\u001b[1;32m   1175\u001b[0m             \u001b[38;5;66;03m# if get_path(schema, inner_path) is None or get_path(state, inner_path) is None:\u001b[39;00m\n\u001b[0;32m-> 1177\u001b[0m             schema, top_state \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minfer_schema\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1178\u001b[0m \u001b[43m                \u001b[49m\u001b[43mschema\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1179\u001b[0m \u001b[43m                \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1180\u001b[0m \u001b[43m                \u001b[49m\u001b[43mtop_state\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtop_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1181\u001b[0m \u001b[43m                \u001b[49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minner_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1183\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(state, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m   1184\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n",
      "File \u001b[0;32m~/code/bigraph-schema/bigraph_schema/type_system.py:1177\u001b[0m, in \u001b[0;36mTypeSystem.infer_schema\u001b[0;34m(self, schema, state, top_state, path)\u001b[0m\n\u001b[1;32m   1174\u001b[0m             inner_path \u001b[38;5;241m=\u001b[39m path \u001b[38;5;241m+\u001b[39m (key,)\n\u001b[1;32m   1175\u001b[0m             \u001b[38;5;66;03m# if get_path(schema, inner_path) is None or get_path(state, inner_path) is None:\u001b[39;00m\n\u001b[0;32m-> 1177\u001b[0m             schema, top_state \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minfer_schema\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1178\u001b[0m \u001b[43m                \u001b[49m\u001b[43mschema\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1179\u001b[0m \u001b[43m                \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1180\u001b[0m \u001b[43m                \u001b[49m\u001b[43mtop_state\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtop_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1181\u001b[0m \u001b[43m                \u001b[49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minner_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1183\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(state, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m   1184\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n",
      "File \u001b[0;32m~/code/bigraph-schema/bigraph_schema/type_system.py:1135\u001b[0m, in \u001b[0;36mTypeSystem.infer_schema\u001b[0;34m(self, schema, state, top_state, path)\u001b[0m\n\u001b[1;32m   1128\u001b[0m state_type \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m   1129\u001b[0m     key: value\n\u001b[1;32m   1130\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m key, value \u001b[38;5;129;01min\u001b[39;00m state\u001b[38;5;241m.\u001b[39mitems()\n\u001b[1;32m   1131\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m key\u001b[38;5;241m.\u001b[39mstartswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m'\u001b[39m)}\n\u001b[1;32m   1132\u001b[0m state_schema \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maccess(\n\u001b[1;32m   1133\u001b[0m     state_type)\n\u001b[0;32m-> 1135\u001b[0m hydrated_state \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdeserialize\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstate_schema\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstate\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1136\u001b[0m top_state \u001b[38;5;241m=\u001b[39m set_path(\n\u001b[1;32m   1137\u001b[0m     top_state,\n\u001b[1;32m   1138\u001b[0m     path,\n\u001b[1;32m   1139\u001b[0m     hydrated_state)\n\u001b[1;32m   1141\u001b[0m update \u001b[38;5;241m=\u001b[39m state_type\n",
      "File \u001b[0;32m~/code/bigraph-schema/bigraph_schema/type_system.py:614\u001b[0m, in \u001b[0;36mTypeSystem.deserialize\u001b[0;34m(self, schema, encoded)\u001b[0m\n\u001b[1;32m    612\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m key, branch \u001b[38;5;129;01min\u001b[39;00m encoded\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m    613\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m schema:\n\u001b[0;32m--> 614\u001b[0m             result[key] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdeserialize\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    615\u001b[0m \u001b[43m                \u001b[49m\u001b[43mschema\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    616\u001b[0m \u001b[43m                \u001b[49m\u001b[43mbranch\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    617\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[1;32m    619\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/code/bigraph-schema/bigraph_schema/type_system.py:590\u001b[0m, in \u001b[0;36mTypeSystem.deserialize\u001b[0;34m(self, schema, encoded)\u001b[0m\n\u001b[1;32m    587\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdeserialize\u001b[39m(\u001b[38;5;28mself\u001b[39m, schema, encoded):\n\u001b[1;32m    588\u001b[0m     found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maccess(schema)\n\u001b[0;32m--> 590\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_deserialize\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mfound\u001b[49m:\n\u001b[1;32m    591\u001b[0m         deserialize \u001b[38;5;241m=\u001b[39m found[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_deserialize\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m    592\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(deserialize, \u001b[38;5;28mstr\u001b[39m):\n",
      "\u001b[0;31mTypeError\u001b[0m: argument of type 'NoneType' is not iterable"
     ]
    }
   ],
   "source": [
    "nested_composite_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': 'float',\n",
    "        'store1.2': 'int',\n",
    "        'process1': {\n",
    "            '_type': 'process',\n",
    "            'inputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'proces',\n",
    "            'outputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "    },\n",
    "    'process3': {\n",
    "        '_type': 'process',\n",
    "        'inputs': {\n",
    "            'port1': ['store1'],\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(nested_composite_spec, **plot_settings, filename='nested_composite')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c8c1704-2331-44eb-bd05-72e36ccb6858",
   "metadata": {},
   "source": [
    "### Composite process\n",
    "Composite processes are processes with internal stores and internal processes, which can run as their own simulation. The composite process has schema, allowing it to connect to a super-simulation. This allows for improved distributed computation, with composite processes able to run on their own computer and syncronize to the super-simulation via message passing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "638c5dc7-4cfd-46e4-b66a-8c42ca98ff1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 8.1.0 (20230707.0739)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"198pt\" height=\"312pt\"\n",
       " viewBox=\"0.00 0.00 204.16 321.41\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 317.41)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-317.41 200.16,-317.41 200.16,4 -4,4\"/>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"123.33\" cy=\"-111.08\" rx=\"27.83\" ry=\"27.83\"/>\n",
       "<text text-anchor=\"start\" x=\"104.58\" y=\"-106.8\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"191.33,-36 137.33,-36 137.33,0 191.33,0 191.33,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"143.71\" y=\"-13.72\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M134.75,-84.71C140.93,-70.99 148.48,-54.22 154.4,-41.06\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"155.32\" cy=\"-39.02\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"147.33\" y=\"-55.75\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"56.33,-36 2.33,-36 2.33,0 56.33,0 56.33,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"8.71\" y=\"-13.72\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M115.24,-83.62C111.13,-73.44 105.28,-62.29 97.33,-54 87.51,-43.76 74.03,-35.97 61.67,-30.37\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"59.47\" cy=\"-29.41\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"105.33\" y=\"-55.75\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"49.33\" cy=\"-111.08\" rx=\"27.83\" ry=\"27.83\"/>\n",
       "<text text-anchor=\"start\" x=\"30.58\" y=\"-106.8\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M55.14,-82.9C58.49,-72.74 63.66,-61.78 71.58,-54 88.01,-37.85 112.62,-29.02 132.36,-24.27\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"134.32\" cy=\"-23.82\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72.33\" y=\"-55.75\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M37.13,-85.48C34.48,-79.04 32.03,-72.01 30.58,-65.25 28.93,-57.58 28.36,-49.02 28.28,-41.38\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"28.29\" cy=\"-39.25\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"31.33\" y=\"-55.75\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"148.56,-226.16 52.11,-226.16 52.11,-190.16 148.56,-190.16 148.56,-226.16\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"152.56,-230.16 48.11,-230.16 48.11,-186.16 152.56,-186.16 152.56,-230.16\"/>\n",
       "<text text-anchor=\"start\" x=\"55.71\" y=\"-203.88\" font-family=\"Times,serif\" font-size=\"12.00\">composite_process</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M152.67,-185.27C158.92,-180.53 164.44,-174.87 168.33,-168.16 177.22,-152.82 162.4,-137.24 147.44,-126.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M106.68,-180.91C109.83,-167.87 113.64,-152.14 116.82,-139.01\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"106.14\" cy=\"-183.15\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"111.33\" y=\"-158.66\" font-family=\"Times,serif\" font-size=\"10.00\">tunnel port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M88.76,-185.58C80.94,-171.01 70.61,-151.74 62.48,-136.59\"/>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M43.55,-189.13C34.62,-183.79 26.54,-176.93 21.08,-168.16 13.82,-156.5 20.55,-142.73 29.26,-131.75\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"45.25\" cy=\"-190.09\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"21.33\" y=\"-158.66\" font-family=\"Times,serif\" font-size=\"10.00\">tunnel port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port1&#39;)&#45;&gt;(&#39;composite_process&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;port1&#39;)&#45;&gt;(&#39;composite_process&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M76.97,-278.31C76.66,-269.3 77,-257.94 79.58,-248.16 80.7,-243.91 82.36,-239.61 84.26,-235.5\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"85.3\" cy=\"-233.36\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"80.33\" y=\"-249.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port2&#39;)&#45;&gt;(&#39;composite_process&#39;,) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;port2&#39;)&#45;&gt;(&#39;composite_process&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M122.1,-289.83C119.39,-279.81 112.79,-255.34 107.56,-235.95\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"106.92\" cy=\"-233.56\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"112.33\" y=\"-249.91\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x110bd51f0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "composite_process_spec = {\n",
    "    'composite_process': {\n",
    "        'store1.1': 'Any',\n",
    "        'store1.2': 'Any',\n",
    "        'process1': {\n",
    "            '_type': 'python_process:temporal',\n",
    "            'wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'python_process:temporal',\n",
    "            'wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        '_ports': {\n",
    "            'port1': 'Any', \n",
    "            'port2': 'Any', \n",
    "        },\n",
    "        '_tunnels': {\n",
    "            'port1': 'store1.1',\n",
    "            'port2': 'store1.2',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(composite_process_spec, **plot_settings, filename='composite_process')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "847cef26-65a1-4da6-9dd3-eb5839277970",
   "metadata": {},
   "source": [
    "## Orchestration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e76c97a8-3037-464a-929e-ba4414c73c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_settings2={}\n",
    "if save_images:\n",
    "    plot_settings2.update({\n",
    "        'out_dir': 'out',\n",
    "        'dpi': '250'\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48975890-06a4-462a-9b1d-6bddd51779fe",
   "metadata": {},
   "source": [
    "### Multi-timestepping\n",
    "\n",
    "Temporal processes can be run by multi-timestepping, with each process updating the state according to its `sync_state` state, which it can update adaptively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f0eff7b0-f0e2-4d56-8b16-ef0fc1ce5aba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 8.1.0 (20230707.0739)\n",
       " -->\n",
       "<!-- Title: flow Pages: 1 -->\n",
       "<svg width=\"226pt\" height=\"164pt\"\n",
       " viewBox=\"0.00 0.00 232.00 169.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 165)\">\n",
       "<title>flow</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-165 228,-165 228,4 -4,4\"/>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"64.42,-41.5 -0.33,-41.5 -0.33,-0.5 64.42,-0.5 64.42,-41.5\"/>\n",
       "<text text-anchor=\"start\" x=\"7.67\" y=\"-24.2\" font-family=\"Times,serif\" font-size=\"14.00\">temporal</text>\n",
       "<text text-anchor=\"start\" x=\"8.42\" y=\"-7.7\" font-family=\"Times,serif\" font-size=\"14.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"127.78\" cy=\"-21\" rx=\"7.95\" ry=\"7.95\"/>\n",
       "<text text-anchor=\"middle\" x=\"127.78\" y=\"-17.7\" font-family=\"Times,serif\" font-size=\"9.00\">1.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M65.26,-21C84.27,-21 106.69,-21 118.85,-21\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"207.57\" cy=\"-21\" rx=\"7.95\" ry=\"7.95\"/>\n",
       "<text text-anchor=\"middle\" x=\"207.57\" y=\"-17.7\" font-family=\"Times,serif\" font-size=\"9.00\">2.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M136.63,-21C152.04,-21 183.65,-21 198.93,-21\"/>\n",
       "</g>\n",
       "<!-- (&#39;state&#39;,) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;state&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"156.92\" cy=\"-90.31\" rx=\"21.22\" ry=\"21.22\"/>\n",
       "<text text-anchor=\"start\" x=\"144.54\" y=\"-85.26\" font-family=\"Times,serif\" font-size=\"14.00\">state</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M133.17,-33.82C136.9,-42.68 141.97,-54.74 146.47,-65.45\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"134.79,-33.15 131.24,-29.22 131.56,-34.51 134.79,-33.15\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"144.8,-66.01 148.36,-69.94 148.03,-64.65 144.8,-66.01\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) end -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) end</title>\n",
       "<ellipse fill=\"none\" stroke=\"none\" stroke-width=\"2\" cx=\"223.52\" cy=\"-21\" rx=\"0\" ry=\"0\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) end -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) end</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M216.29,-21C216.45,-21 216.61,-21 216.77,-21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"214.6,-22.75 219.6,-21 214.6,-19.25 214.6,-22.75\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M199.41,-32.16C192.26,-41.94 181.67,-56.44 172.85,-68.51\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"200.8,-33.23 202.34,-28.16 197.97,-31.16 200.8,-33.23\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"171.47,-67.43 169.93,-72.5 174.29,-69.5 171.47,-67.43\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"64.42,-161.17 -0.33,-161.17 -0.33,-120.17 64.42,-120.17 64.42,-161.17\"/>\n",
       "<text text-anchor=\"start\" x=\"7.67\" y=\"-143.87\" font-family=\"Times,serif\" font-size=\"14.00\">temporal</text>\n",
       "<text text-anchor=\"start\" x=\"8.42\" y=\"-127.37\" font-family=\"Times,serif\" font-size=\"14.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"87.89\" cy=\"-140.67\" rx=\"7.95\" ry=\"7.95\"/>\n",
       "<text text-anchor=\"middle\" x=\"87.89\" y=\"-137.37\" font-family=\"Times,serif\" font-size=\"9.00\">0.5</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M65.18,-140.67C70.25,-140.67 75.08,-140.67 78.98,-140.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"127.78\" cy=\"-140.67\" rx=\"7.95\" ry=\"7.95\"/>\n",
       "<text text-anchor=\"middle\" x=\"127.78\" y=\"-137.37\" font-family=\"Times,serif\" font-size=\"9.00\">1.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M96.74,-140.67C103.36,-140.67 112.35,-140.67 118.95,-140.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M99,-132.57C108.75,-125.46 123.18,-114.92 135.21,-106.15\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"98.03,-131.11 95.02,-135.47 100.09,-133.94 98.03,-131.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"136.18,-107.61 139.18,-103.25 134.11,-104.78 136.18,-107.61\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"167.67\" cy=\"-140.67\" rx=\"7.95\" ry=\"7.95\"/>\n",
       "<text text-anchor=\"middle\" x=\"167.67\" y=\"-137.37\" font-family=\"Times,serif\" font-size=\"9.00\">1.5</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M136.64,-140.67C143.25,-140.67 152.24,-140.67 158.85,-140.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M134.47,-129.11C137.15,-124.48 140.33,-118.99 143.44,-113.6\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"133.07,-128.04 132.08,-133.24 136.1,-129.79 133.07,-128.04\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"144.81,-114.73 145.8,-109.53 141.78,-112.98 144.81,-114.73\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"207.57\" cy=\"-140.67\" rx=\"7.95\" ry=\"7.95\"/>\n",
       "<text text-anchor=\"middle\" x=\"207.57\" y=\"-137.37\" font-family=\"Times,serif\" font-size=\"9.00\">2.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M176.53,-140.67C183.14,-140.67 192.13,-140.67 198.74,-140.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M164.86,-127.52C164.15,-124.16 163.35,-120.43 162.55,-116.66\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"163.14,-127.83 165.9,-132.36 166.56,-127.1 163.14,-127.83\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"164.28,-116.42 161.53,-111.9 160.86,-117.15 164.28,-116.42\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) end -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) end</title>\n",
       "<ellipse fill=\"none\" stroke=\"none\" stroke-width=\"2\" cx=\"223.52\" cy=\"-140.67\" rx=\"0\" ry=\"0\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) end -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) end</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M216.29,-140.67C216.45,-140.67 216.61,-140.67 216.77,-140.67\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"214.6,-142.42 219.6,-140.67 214.6,-138.92 214.6,-142.42\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M198.05,-131.22C191.72,-124.92 183.19,-116.43 175.54,-108.83\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"196.58,-132.21 201.36,-134.5 199.04,-129.73 196.58,-132.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"177.1,-107.91 172.32,-105.63 174.63,-110.39 177.1,-107.91\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x110b1a520>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multitimestep_spec = {\n",
    "    'temporal process2': {\n",
    "        '_ports': {'port1': 'Any'},\n",
    "        'wires': {'port1': 'state'},\n",
    "        '_sync_step': 1.0,\n",
    "    },\n",
    "    'temporal process1': {\n",
    "        '_ports': {'port1': 'Any'},\n",
    "        'wires': {'port1': 'state'},\n",
    "        '_sync_step': 0.5,\n",
    "    },\n",
    "}\n",
    "multitimestep_spec = replace_regex_recursive(multitimestep_spec)\n",
    "plot_multitimestep(multitimestep_spec, total_time=2.0, **plot_settings2, filename='multitimestep')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a00cf677-ebac-4309-8ac9-a5b2855724af",
   "metadata": {},
   "source": [
    "### Flows\n",
    "\n",
    "A directed acyclic graph (DAG) – a type of workflow that declares the dependencies between processes. A flow run is not temporal, and the order is entirely based on dependencies. It can triggered between time steps, running the steps in the order as determined by their dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "db45c99b-b97b-47bc-a73e-c669882f3969",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 8.1.0 (20230707.0739)\n",
       " -->\n",
       "<!-- Title: flow Pages: 1 -->\n",
       "<svg width=\"130pt\" height=\"183pt\"\n",
       " viewBox=\"0.00 0.00 134.00 188.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 184)\">\n",
       "<title>flow</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-184 130,-184 130,4 -4,4\"/>\n",
       "<!-- (&#39;step1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;step1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-180 0,-180 0,-144 54,-144 54,-180\"/>\n",
       "<text text-anchor=\"start\" x=\"12.75\" y=\"-156.95\" font-family=\"Times,serif\" font-size=\"14.00\">step1</text>\n",
       "</g>\n",
       "<!-- (&#39;step2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;step2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-108 0,-108 0,-72 54,-72 54,-108\"/>\n",
       "<text text-anchor=\"start\" x=\"12.75\" y=\"-84.95\" font-family=\"Times,serif\" font-size=\"14.00\">step2</text>\n",
       "</g>\n",
       "<!-- (&#39;step1&#39;,)&#45;&gt;(&#39;step2&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;step1&#39;,)&#45;&gt;(&#39;step2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M27,-143.34C27,-136.75 27,-129.08 27,-121.67\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"30.5,-121.9 27,-111.9 23.5,-121.9 30.5,-121.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;step4&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;step4&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"90,-36 36,-36 36,0 90,0 90,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"48.75\" y=\"-12.95\" font-family=\"Times,serif\" font-size=\"14.00\">step4</text>\n",
       "</g>\n",
       "<!-- (&#39;step2&#39;,)&#45;&gt;(&#39;step4&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;step2&#39;,)&#45;&gt;(&#39;step4&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M36.08,-71.34C39.68,-64.34 43.91,-56.12 47.94,-48.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"51.33,-49.4 52.79,-38.9 45.11,-46.19 51.33,-49.4\"/>\n",
       "</g>\n",
       "<!-- (&#39;step3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;step3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"126,-108 72,-108 72,-72 126,-72 126,-108\"/>\n",
       "<text text-anchor=\"start\" x=\"84.75\" y=\"-84.95\" font-family=\"Times,serif\" font-size=\"14.00\">step3</text>\n",
       "</g>\n",
       "<!-- (&#39;step3&#39;,)&#45;&gt;(&#39;step4&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;step3&#39;,)&#45;&gt;(&#39;step4&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M89.92,-71.34C86.32,-64.34 82.09,-56.12 78.06,-48.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"80.89,-46.19 73.21,-38.9 74.67,-49.4 80.89,-46.19\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x110bc11f0>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flow = {\n",
    "    'step1': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': [], \n",
    "    },\n",
    "    'step2': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': 'step1', \n",
    "    },\n",
    "    'step3': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': [], \n",
    "    },\n",
    "    'step4': {\n",
    "        '_type': 'step_process',\n",
    "        '_ports': {},\n",
    "        '_depends_on': ['step2', 'step3'], \n",
    "    },\n",
    "}\n",
    "\n",
    "plot_flow(flow, **plot_settings2, filename='flow')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5882581a-c5ef-40b5-acbd-efddad269807",
   "metadata": {},
   "source": [
    "## Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff81178f-36e4-4200-982a-ed8155b87934",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Multiscale map of a cell\n",
    "\n",
    "Bigraphs can represent the hierarchical nesting of cellular structures. Below is a multiscale map of a cell from [Qin et al. A multi-scale map of cell structure fusing protein images and interactions. Nature. 2021.](https://pubmed.ncbi.nlm.nih.gov/34819669/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47bdf8e5-0bdd-4cac-b57e-2449c75fa178",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell_hierarchy\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 9.0.0 (0)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"667pt\" height=\"700pt\"\n",
       " viewBox=\"0.00 0.00 686.09 720.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.629041 0.629041) rotate(0) translate(4 1108.81)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-1108.81 1056.4,-1108.81 1056.4,4 -4,4\"/>\n",
       "<!-- (&#39;cell&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;cell&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"424.39\" cy=\"-1086.81\" rx=\"18\" ry=\"18\"/>\n",
       "<text text-anchor=\"start\" x=\"415.73\" y=\"-1083.21\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"286.39\" cy=\"-999.39\" rx=\"33.41\" ry=\"33.41\"/>\n",
       "<text text-anchor=\"start\" x=\"261.73\" y=\"-995.79\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M408.83,-1076.17C386.14,-1062.13 343.56,-1035.78 315.09,-1018.15\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"654.39\" cy=\"-999.39\" rx=\"25.92\" ry=\"25.92\"/>\n",
       "<text text-anchor=\"start\" x=\"636.07\" y=\"-995.79\" font-family=\"Times,serif\" font-size=\"12.00\">nucleus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M441.68,-1079.39C482.11,-1064.37 582.95,-1026.93 629.78,-1009.54\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"115.39\" cy=\"-875.88\" rx=\"39.81\" ry=\"39.81\"/>\n",
       "<text text-anchor=\"start\" x=\"93.41\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">secretory</text>\n",
       "<text text-anchor=\"start\" x=\"90.84\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M256.74,-982.85C233.07,-969.92 199.5,-950.39 172.39,-929.98 162.45,-922.5 152.3,-913.54 143.36,-905.13\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"227.39\" cy=\"-875.88\" rx=\"46.09\" ry=\"46.09\"/>\n",
       "<text text-anchor=\"start\" x=\"198.4\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"202.84\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M271.81,-968.36C264.44,-953.18 255.38,-934.53 247.37,-918.03\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"345.39\" cy=\"-875.88\" rx=\"54.1\" ry=\"54.1\"/>\n",
       "<text text-anchor=\"start\" x=\"310.74\" y=\"-890.28\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"321.07\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"324.73\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"329.73\" y=\"-854.28\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M300.98,-968.36C307.26,-955.43 314.76,-939.98 321.8,-925.49\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"89.39\" cy=\"-728.39\" rx=\"57.39\" ry=\"57.39\"/>\n",
       "<text text-anchor=\"start\" x=\"52.41\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"68.07\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"70.39\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M108.43,-835.89C105.67,-820.49 102.46,-802.49 99.46,-785.72\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"57.39\" cy=\"-569.58\" rx=\"57.39\" ry=\"57.39\"/>\n",
       "<text text-anchor=\"start\" x=\"49.73\" y=\"-583.98\" font-family=\"Times,serif\" font-size=\"12.00\">ion</text>\n",
       "<text text-anchor=\"start\" x=\"20.41\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"36.07\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"38.39\" y=\"-547.98\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M77.93,-671.22C74.97,-656.72 71.79,-641.11 68.83,-626.62\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"227.39\" cy=\"-728.39\" rx=\"46.09\" ry=\"46.09\"/>\n",
       "<text text-anchor=\"start\" x=\"205.06\" y=\"-742.79\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"222.4\" y=\"-730.79\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"198.4\" y=\"-718.79\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"202.84\" y=\"-706.79\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M227.39,-828.85C227.39,-811.7 227.39,-792.25 227.39,-775.12\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"172.39\" cy=\"-569.58\" rx=\"39.81\" ry=\"39.81\"/>\n",
       "<text text-anchor=\"start\" x=\"148.74\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"147.84\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M212.22,-684.12C203.86,-660.28 193.58,-630.97 185.56,-608.11\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"295.39\" cy=\"-569.58\" rx=\"65.41\" ry=\"65.41\"/>\n",
       "<text text-anchor=\"start\" x=\"252.74\" y=\"-577.98\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleoprotein</text>\n",
       "<text text-anchor=\"start\" x=\"274.73\" y=\"-565.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"279.73\" y=\"-553.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M245.62,-685.37C252.94,-668.49 261.51,-648.73 269.45,-630.4\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"73.39\" cy=\"-422.15\" rx=\"39.81\" ry=\"39.81\"/>\n",
       "<text text-anchor=\"start\" x=\"51.06\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"68.4\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"49.74\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"48.84\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M149.93,-535.58C133.73,-511.78 111.95,-479.79 95.77,-456.02\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"177.39\" cy=\"-422.15\" rx=\"46.02\" ry=\"46.02\"/>\n",
       "<text text-anchor=\"start\" x=\"142.07\" y=\"-418.55\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondrion</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M173.76,-528.84C174.4,-510.34 175.16,-488.22 175.82,-469.01\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"291.39\" cy=\"-422.15\" rx=\"41.38\" ry=\"41.38\"/>\n",
       "<text text-anchor=\"start\" x=\"278.73\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"265.73\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"270.73\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"288.39\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M293.61,-503.52C293.24,-490.21 292.86,-476.55 292.53,-464.36\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"394.39\" cy=\"-422.15\" rx=\"43.74\" ry=\"43.74\"/>\n",
       "<text text-anchor=\"start\" x=\"372.06\" y=\"-430.55\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "<text text-anchor=\"start\" x=\"369.06\" y=\"-418.55\" font-family=\"Times,serif\" font-size=\"12.00\">biogenesis</text>\n",
       "<text text-anchor=\"start\" x=\"367.06\" y=\"-406.55\" font-family=\"Times,serif\" font-size=\"12.00\">community</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M332.26,-514.43C344.79,-496.01 358.49,-475.9 369.82,-459.25\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"291.39\" cy=\"-301.09\" rx=\"39.03\" ry=\"39.03\"/>\n",
       "<text text-anchor=\"start\" x=\"278.73\" y=\"-315.49\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"272.4\" y=\"-303.49\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"270.73\" y=\"-291.49\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"288.39\" y=\"-279.49\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M291.39,-379.84C291.39,-367.22 291.39,-353.38 291.39,-340.92\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"387.39\" cy=\"-301.09\" rx=\"30.65\" ry=\"30.65\"/>\n",
       "<text text-anchor=\"start\" x=\"365.06\" y=\"-297.49\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M391.83,-377.49C390.95,-362.49 389.98,-346.05 389.17,-332.35\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"475.39\" cy=\"-301.09\" rx=\"39.03\" ry=\"39.03\"/>\n",
       "<text text-anchor=\"start\" x=\"458.07\" y=\"-315.49\" font-family=\"Times,serif\" font-size=\"12.00\">nucleic</text>\n",
       "<text text-anchor=\"start\" x=\"465.4\" y=\"-303.49\" font-family=\"Times,serif\" font-size=\"12.00\">acid</text>\n",
       "<text text-anchor=\"start\" x=\"457.06\" y=\"-291.49\" font-family=\"Times,serif\" font-size=\"12.00\">binding</text>\n",
       "<text text-anchor=\"start\" x=\"454.73\" y=\"-279.49\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M418.92,-385.1C429.91,-368.95 442.76,-350.05 453.43,-334.37\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"387.39\" cy=\"-187.03\" rx=\"39.03\" ry=\"39.03\"/>\n",
       "<text text-anchor=\"start\" x=\"363.4\" y=\"-189.43\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"369.73\" y=\"-177.43\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M387.39,-269.47C387.39,-256.38 387.39,-240.95 387.39,-227.05\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"162.39\" cy=\"-56\" rx=\"55.05\" ry=\"55.05\"/>\n",
       "<text text-anchor=\"start\" x=\"127.07\" y=\"-70.4\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondirial</text>\n",
       "<text text-anchor=\"start\" x=\"150.51\" y=\"-58.4\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n",
       "<text text-anchor=\"start\" x=\"138.4\" y=\"-46.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"144.73\" y=\"-34.4\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondirial&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M350.4,-172.46C316.76,-159.32 266.4,-137.67 226.39,-112.01 218.38,-106.86 210.31,-100.79 202.73,-94.58\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"274.39\" cy=\"-56\" rx=\"39.03\" ry=\"39.03\"/>\n",
       "<text text-anchor=\"start\" x=\"251.07\" y=\"-64.4\" font-family=\"Times,serif\" font-size=\"12.00\">mito&#45;cyto</text>\n",
       "<text text-anchor=\"start\" x=\"250.4\" y=\"-52.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"258.4\" y=\"-40.4\" font-family=\"Times,serif\" font-size=\"12.00\">cluster</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M361.75,-156.87C349.78,-143.25 335.35,-126.81 322.39,-112.01 315.13,-103.71 307.26,-94.7 300.04,-86.42\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"387.39\" cy=\"-56\" rx=\"56\" ry=\"56\"/>\n",
       "<text text-anchor=\"start\" x=\"352.07\" y=\"-82.4\" font-family=\"Times,serif\" font-size=\"12.00\">cotranslational</text>\n",
       "<text text-anchor=\"start\" x=\"370.4\" y=\"-70.4\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n",
       "<text text-anchor=\"start\" x=\"366.18\" y=\"-58.4\" font-family=\"Times,serif\" font-size=\"12.00\">targeting</text>\n",
       "<text text-anchor=\"start\" x=\"382.73\" y=\"-46.4\" font-family=\"Times,serif\" font-size=\"12.00\">to</text>\n",
       "<text text-anchor=\"start\" x=\"362.07\" y=\"-34.4\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "<text text-anchor=\"start\" x=\"370.73\" y=\"-22.4\" font-family=\"Times,serif\" font-size=\"12.00\">system</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M387.39,-147.32C387.39,-136.48 387.39,-124.48 387.39,-112.86\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"500.39\" cy=\"-56\" rx=\"39.03\" ry=\"39.03\"/>\n",
       "<text text-anchor=\"start\" x=\"476.4\" y=\"-64.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"479.73\" y=\"-52.4\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"497.39\" y=\"-40.4\" font-family=\"Times,serif\" font-size=\"12.00\">5</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M413.04,-156.87C425.01,-143.25 439.44,-126.81 452.39,-112.01 459.66,-103.71 467.53,-94.7 474.75,-86.42\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"596.39\" cy=\"-56\" rx=\"39.03\" ry=\"39.03\"/>\n",
       "<text text-anchor=\"start\" x=\"572.4\" y=\"-64.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"575.73\" y=\"-52.4\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"593.39\" y=\"-40.4\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M424.97,-173.89C459.24,-161.68 510.21,-140.56 548.39,-112.01 557.33,-105.33 565.79,-96.74 573.05,-88.35\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"345.39\" cy=\"-728.39\" rx=\"54.1\" ry=\"54.1\"/>\n",
       "<text text-anchor=\"start\" x=\"310.74\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"321.07\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"319.74\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M345.39,-821.12C345.39,-808.81 345.39,-795.73 345.39,-783.42\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"489.39\" cy=\"-875.88\" rx=\"38.55\" ry=\"38.55\"/>\n",
       "<text text-anchor=\"start\" x=\"476.4\" y=\"-884.28\" font-family=\"Times,serif\" font-size=\"12.00\">DNA</text>\n",
       "<text text-anchor=\"start\" x=\"465.74\" y=\"-872.28\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"467.06\" y=\"-860.28\" font-family=\"Times,serif\" font-size=\"12.00\">assembly</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M630.62,-987.22C605.97,-975.03 566.98,-954.03 537.39,-929.98 529,-923.17 520.77,-914.85 513.56,-906.84\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"576.39\" cy=\"-875.88\" rx=\"30.06\" ry=\"30.06\"/>\n",
       "<text text-anchor=\"start\" x=\"558.74\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"561.4\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">lumen</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M640.45,-976.67C627.04,-955.78 606.78,-924.22 592.65,-902.2\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"654.39\" cy=\"-875.88\" rx=\"30.06\" ry=\"30.06\"/>\n",
       "<text text-anchor=\"start\" x=\"636.74\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"642.39\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">body</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M654.39,-972.63C654.39,-953.4 654.39,-927.08 654.39,-906.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"753.39\" cy=\"-875.88\" rx=\"50.8\" ry=\"50.8\"/>\n",
       "<text text-anchor=\"start\" x=\"721.07\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"750.39\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M670.74,-978.33C684.47,-961.48 704.5,-936.9 721.4,-916.15\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"873.39\" cy=\"-875.88\" rx=\"50.8\" ry=\"50.8\"/>\n",
       "<text text-anchor=\"start\" x=\"841.07\" y=\"-878.28\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"870.39\" y=\"-866.28\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M679.57,-990.36C711.99,-979.41 769.37,-957.92 813.39,-929.98 821.5,-924.84 829.59,-918.61 837.09,-912.26\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"996.39\" cy=\"-875.88\" rx=\"54.1\" ry=\"54.1\"/>\n",
       "<text text-anchor=\"start\" x=\"978.74\" y=\"-884.28\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"961.74\" y=\"-872.28\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"978.4\" y=\"-860.28\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M680.78,-995.97C731.99,-990.43 847.37,-973.72 933.39,-929.98 941.68,-925.77 949.82,-920.27 957.36,-914.42\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.39\" cy=\"-728.39\" rx=\"31.44\" ry=\"31.44\"/>\n",
       "<text text-anchor=\"start\" x=\"538.4\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">nucleolus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M573.32,-845.09C570.79,-820.48 567.2,-785.65 564.62,-760.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"642.39\" cy=\"-728.39\" rx=\"31.96\" ry=\"31.96\"/>\n",
       "<text text-anchor=\"start\" x=\"624.74\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"623.4\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"624.4\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M588.81,-847.52C600.35,-822.08 617.48,-784.32 629.25,-758.38\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"440.39\" cy=\"-569.58\" rx=\"61.17\" ry=\"61.17\"/>\n",
       "<text text-anchor=\"start\" x=\"400.74\" y=\"-577.98\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleprotein</text>\n",
       "<text text-anchor=\"start\" x=\"419.73\" y=\"-565.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"424.73\" y=\"-553.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M542.22,-702.54C524.96,-680.17 499.11,-646.67 477.77,-619.02\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.39\" cy=\"-569.58\" rx=\"41.38\" ry=\"41.38\"/>\n",
       "<text text-anchor=\"start\" x=\"548.73\" y=\"-583.98\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"535.73\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"540.73\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"545.73\" y=\"-547.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M561.39,-696.03C561.39,-671.81 561.39,-638.08 561.39,-611.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"666.39\" cy=\"-569.58\" rx=\"45.63\" ry=\"45.63\"/>\n",
       "<text text-anchor=\"start\" x=\"637.73\" y=\"-577.98\" font-family=\"Times,serif\" font-size=\"12.00\">splicosomal</text>\n",
       "<text text-anchor=\"start\" x=\"645.73\" y=\"-565.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"650.73\" y=\"-553.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M578.97,-701.14C595.96,-675.77 621.88,-637.06 641.09,-608.38\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"529.39\" cy=\"-422.15\" rx=\"41.38\" ry=\"41.38\"/>\n",
       "<text text-anchor=\"start\" x=\"516.73\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"503.73\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"508.73\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"526.39\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M552.57,-528.45C548.11,-508.19 542.71,-483.67 538.25,-463.39\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"627.39\" cy=\"-422.15\" rx=\"39.03\" ry=\"39.03\"/>\n",
       "<text text-anchor=\"start\" x=\"614.73\" y=\"-436.55\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"608.4\" y=\"-424.55\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"606.73\" y=\"-412.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"611.73\" y=\"-400.55\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M578.39,-531.13C588.55,-508.74 601.38,-480.47 611.36,-458.47\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"762.39\" cy=\"-728.39\" rx=\"48.92\" ry=\"48.92\"/>\n",
       "<text text-anchor=\"start\" x=\"731.4\" y=\"-742.79\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n",
       "<text text-anchor=\"start\" x=\"732.51\" y=\"-730.79\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"741.73\" y=\"-718.79\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"746.73\" y=\"-706.79\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M756.52,-824.4C757.45,-809.3 758.47,-792.78 759.4,-777.82\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"777.39\" cy=\"-569.58\" rx=\"47.35\" ry=\"47.35\"/>\n",
       "<text text-anchor=\"start\" x=\"753.07\" y=\"-583.98\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"747.51\" y=\"-571.98\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"756.73\" y=\"-559.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"761.73\" y=\"-547.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M767.02,-679.06C768.89,-659.53 771.03,-637.07 772.89,-617.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"777.39\" cy=\"-422.15\" rx=\"34.31\" ry=\"34.31\"/>\n",
       "<text text-anchor=\"start\" x=\"765.73\" y=\"-430.55\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"756.73\" y=\"-418.55\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"761.73\" y=\"-406.55\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M777.39,-521.36C777.39,-500.46 777.39,-476.34 777.39,-457.24\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"777.39\" cy=\"-301.09\" rx=\"34.31\" ry=\"34.31\"/>\n",
       "<text text-anchor=\"start\" x=\"762.73\" y=\"-309.49\" font-family=\"Times,serif\" font-size=\"12.00\">NuA4</text>\n",
       "<text text-anchor=\"start\" x=\"765.73\" y=\"-297.49\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"756.73\" y=\"-285.49\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M777.39,-387.03C777.39,-371.09 777.39,-352.16 777.39,-336.22\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"873.39\" cy=\"-728.39\" rx=\"39.49\" ry=\"39.49\"/>\n",
       "<text text-anchor=\"start\" x=\"849.07\" y=\"-736.79\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"849.07\" y=\"-724.79\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"852.73\" y=\"-712.79\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M873.39,-824.4C873.39,-806.18 873.39,-785.9 873.39,-768.77\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"996.39\" cy=\"-728.39\" rx=\"56\" ry=\"56\"/>\n",
       "<text text-anchor=\"start\" x=\"976.07\" y=\"-754.79\" font-family=\"Times,serif\" font-size=\"12.00\">negative</text>\n",
       "<text text-anchor=\"start\" x=\"972.07\" y=\"-742.79\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"991.4\" y=\"-730.79\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"983.73\" y=\"-718.79\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"966.73\" y=\"-706.79\" font-family=\"Times,serif\" font-size=\"12.00\">biosynthesis</text>\n",
       "<text text-anchor=\"start\" x=\"978.4\" y=\"-694.79\" font-family=\"Times,serif\" font-size=\"12.00\">process</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M996.39,-821.12C996.39,-809.46 996.39,-797.11 996.39,-785.37\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x11a767520>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "music_map = {\n",
    "    'cell': {\n",
    "        'cytoplasm': {\n",
    "            'secretory organelles': {\n",
    "                'transmembrane transport systems': {\n",
    "                    'ion transmembrane transport systems': {}}},\n",
    "            'cytoplasmic organelles': {\n",
    "                'subgroup of cytoplasmic organelles': {\n",
    "                    'metabolic organelles': {\n",
    "                        'subgroup of metabolic organelles': {},\n",
    "                        'mitochondrion': {}},\n",
    "                    'ribonucleoprotein complex family': {\n",
    "                        'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
    "                        'ribosome biogenesis community': {\n",
    "                            'ribosome': {\n",
    "                                'ribosomal subunit': {\n",
    "                                    'mitochondirial large ribosomal subunit': {\n",
    "                                        # 'large ribosomal subunit subcomplex 1': {}\n",
    "                                    },\n",
    "                                    'mito-cyto ribosomal cluster': {},\n",
    "                                    'cotranslational protein targeting to membrane system': {},\n",
    "                                    'ribosomal complex 5': {},\n",
    "                                    'ribosomal complex 2': {}}},\n",
    "                            'nucleic acid binding complex': {}}}}},\n",
    "            'transcriptional regulation complex family': {\n",
    "                'transcriptional regulation complexes': {}}},\n",
    "        'nucleus': {\n",
    "            'DNA metabolic assembly': {},\n",
    "            'nuclear lumen': {\n",
    "                'nucleolus': {\n",
    "                    'ribonucleprotein complex family': {},\n",
    "                    'RNA processing complex family': {\n",
    "                        'RNA processing complex 1': {},\n",
    "                        'RNA splicing complex family': {}},\n",
    "                    'splicosomal complex family': {}},\n",
    "                'nuclear splicing speckle': {}},\n",
    "            'nuclear body': {},\n",
    "            'nucleo-plasm 1': {\n",
    "                'chromosome organization complex family': {\n",
    "                    'chromatin organization complex family': {\n",
    "                        'HAT complex family': {\n",
    "                            'NuA4 HAT complex': {}}}}},\n",
    "            'nucleo-plasm 2': {\n",
    "                'chromatin regulation complex': {}},\n",
    "            'nuclear transcriptional speckle': {\n",
    "                'negative regulation of RNA biosynthesis process': {}}}}}\n",
    "music_map1 = replace_regex_recursive(music_map)\n",
    "plot_bigraph(music_map1, **plot_settings, filename='cell_hierarchy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ab431982-d156-4b7a-9918-384a5ea5132a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cell': {'cytoplasm': {'secretory organelles': {'transmembrane transport systems': {'ion transmembrane transport systems': {}}},\n",
       "   'cytoplasmic organelles': {'subgroup of cytoplasmic organelles': {'metabolic organelles': {'subgroup of metabolic organelles': {},\n",
       "      'mitochondrion': {}},\n",
       "     'ribonucleoprotein complex family': {'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
       "      'ribosome biogenesis community': {'ribosome': {'ribosomal subunit': {'mitochondirial large ribosomal subunit': {},\n",
       "         'mito-cyto ribosomal cluster': {},\n",
       "         'cotranslational protein targeting to membrane system': {},\n",
       "         'ribosomal complex 5': {},\n",
       "         'ribosomal complex 2': {}}},\n",
       "       'nucleic acid binding complex': {}}}}},\n",
       "   'transcriptional regulation complex family': {'transcriptional regulation complexes': {}}},\n",
       "  'nucleus': {'DNA metabolic assembly': {},\n",
       "   'nuclear lumen': {'nucleolus': {'ribonucleprotein complex family': {},\n",
       "     'RNA processing complex family': {'RNA processing complex 1': {},\n",
       "      'RNA splicing complex family': {}},\n",
       "     'splicosomal complex family': {}},\n",
       "    'nuclear splicing speckle': {}},\n",
       "   'nuclear body': {},\n",
       "   'nucleo-plasm 1': {'chromosome organization complex family': {'chromatin organization complex family': {'HAT complex family': {'NuA4 HAT complex': {}}}}},\n",
       "   'nucleo-plasm 2': {'chromatin regulation complex': {}},\n",
       "   'nuclear transcriptional speckle': {'negative regulation of RNA biosynthesis process': {}}}}}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "music_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5a76608a-fff0-4005-9f32-c02afef39d9e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "plot_bigraph() got an unexpected keyword argument 'node_border_colors'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[27], line 128\u001b[0m\n\u001b[1;32m    118\u001b[0m plot_settingsx[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnode_border_colors\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    119\u001b[0m     (\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcell\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtranslation\u001b[39m\u001b[38;5;124m'\u001b[39m): \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mblue\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m    120\u001b[0m     (\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcell\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtranscription\u001b[39m\u001b[38;5;124m'\u001b[39m): \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mblue\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    125\u001b[0m     (\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcell\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRNA<br/>processing\u001b[39m\u001b[38;5;124m'\u001b[39m): \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mblue\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m    126\u001b[0m }\n\u001b[1;32m    127\u001b[0m \u001b[38;5;66;03m# plot_settingsx.update({'out_dir': 'out','dpi': '250'})\u001b[39;00m\n\u001b[0;32m--> 128\u001b[0m \u001b[43mplot_bigraph\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmusic_map2\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mplot_settingsx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfilename\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcell_hierarchy_functions\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mTypeError\u001b[0m: plot_bigraph() got an unexpected keyword argument 'node_border_colors'"
     ]
    }
   ],
   "source": [
    "import copy\n",
    "\n",
    "transcription_process = {\n",
    "    '_ports': {\n",
    "        'RNA processing': 'Any',\n",
    "        'regulation': 'Any',\n",
    "        'nucleic acids': 'Any',\n",
    "        'negative regulation': 'Any'\n",
    "    },\n",
    "    'wires': {\n",
    "        'RNA processing': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "        'regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "        'nucleic acids': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', \n",
    "            'ribonucleoprotein complex family', 'ribosome biogenesis community', 'nucleic acid binding complex'],\n",
    "        'negative regulation': [\n",
    "            'nucleus', 'nuclear transcriptional speckle', 'negative regulation of RNA biosynthesis process'],\n",
    "    }\n",
    "}\n",
    "\n",
    "translation_process = {\n",
    "    '_ports': {\n",
    "        'ribosomes': 'Any',\n",
    "        'membrane': 'Any'\n",
    "    },\n",
    "    'wires': {\n",
    "        'ribosomes': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', 'ribosome biogenesis community', 'ribosome'],\n",
    "        # 'membrane': ['],\n",
    "    }\n",
    "}\n",
    "\n",
    "metabolism_process = {\n",
    "    '_ports': {\n",
    "        'organelles': 'Any',\n",
    "        'mitochondrion': 'Any',\n",
    "        'ion transport': 'Any',\n",
    "    },\n",
    "    'wires': {\n",
    "        'organelles': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', ],\n",
    "        'mitochondrion': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', 'mitochondrion'],\n",
    "        'ion transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems', 'ion transmembrane transport systems'],\n",
    "    }\n",
    "}\n",
    "\n",
    "cell_cycle_process = {\n",
    "    '_ports': {\n",
    "        'DNA': 'Any',\n",
    "        'chromosome': 'Any',\n",
    "        'chromatin': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'DNA': ['nucleus', 'DNA metabolic assembly'],\n",
    "        'chromosome': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family',],\n",
    "        'chromatin': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family', 'chromatin organization complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "signalling_process = {\n",
    "    '_ports': {\n",
    "        'transport': 'Any',\n",
    "        'trancript regulation': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'trancript regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "    }\n",
    "}\n",
    "\n",
    "protein_transport_process = {\n",
    "    '_ports': {\n",
    "        'transmembrane transport': 'Any',\n",
    "        'secretory': 'Any',\n",
    "        'cotranslation': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'transmembrane transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'secretory': [\n",
    "            'cytoplasm', 'secretory organelles'],\n",
    "        'cotranslation': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', \n",
    "            'ribosome biogenesis community', 'ribosome', 'ribosomal subunit', 'cotranslational protein targeting to membrane system']\n",
    "    }\n",
    "}\n",
    "\n",
    "rna_processing_process = {\n",
    "    '_ports': {\n",
    "        'processing complex': 'Any',\n",
    "    },\n",
    "    'wires': {        \n",
    "        'processing complex': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "# compose schema\n",
    "music_map['cell']['transcription'] = transcription_process\n",
    "music_map['cell']['translation'] = translation_process\n",
    "music_map['cell']['metabolism'] = metabolism_process\n",
    "music_map['cell']['cell cycle'] = cell_cycle_process\n",
    "music_map['cell']['signalling'] = signalling_process\n",
    "music_map['cell']['protein transport'] = protein_transport_process\n",
    "music_map['cell']['RNA processing'] = rna_processing_process\n",
    "\n",
    "music_map2 = replace_regex_recursive(music_map)\n",
    "\n",
    "# plot\n",
    "plot_settingsx = copy.deepcopy(plot_settings)\n",
    "plot_settingsx['label_margin'] = '0.02'\n",
    "plot_settingsx['node_border_colors'] = {\n",
    "    ('cell', 'translation'): 'blue',\n",
    "    ('cell', 'transcription'): 'blue',\n",
    "    ('cell', 'metabolism'): 'blue',\n",
    "    ('cell', 'cell<br/>cycle'): 'blue',\n",
    "    ('cell', 'signalling'): 'blue',\n",
    "    ('cell', 'protein<br/>transport'): 'blue',\n",
    "    ('cell', 'RNA<br/>processing'): 'blue',\n",
    "}\n",
    "# plot_settingsx.update({'out_dir': 'out','dpi': '250'})\n",
    "plot_bigraph(music_map2, **plot_settingsx, filename='cell_hierarchy_functions')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fe3008ad-498e-449f-8521-6ea29fd903f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 8.1.0 (20230707.0739)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"72pt\" height=\"113pt\"\n",
       " viewBox=\"0.00 0.00 74.57 116.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 112)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-112 70.57,-112 70.57,4 -4,4\"/>\n",
       "<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;hypothesized&lt;br/&gt;mechanism&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"blue\" stroke-width=\"2\" points=\"66.57,-36 0,-36 0,0 66.57,0 66.57,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"2.16\" y=\"-20.85\" font-family=\"Times,serif\" font-size=\"12.00\">hypothesized</text>\n",
       "<text text-anchor=\"start\" x=\"6.66\" y=\"-6.6\" font-family=\"Times,serif\" font-size=\"12.00\">mechanism</text>\n",
       "</g>\n",
       "<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;1&#39;) -->\n",
       "<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;1&#39;)&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;1&#39;)&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M25.61,-78.67C26.86,-69 28.76,-54.22 30.35,-41.79\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"30.65\" cy=\"-39.46\" rx=\"2\" ry=\"2\"/>\n",
       "</g>\n",
       "<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;2&#39;) -->\n",
       "<!-- (&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;2&#39;)&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;hypothesized&lt;br/&gt;mechanism&#39;, &#39;2&#39;)&#45;&gt;(&#39;hypothesized&lt;br/&gt;mechanism&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M41.94,-79.56C40.55,-69.85 38.34,-54.38 36.5,-41.51\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"36.2\" cy=\"-39.37\" rx=\"2\" ry=\"2\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x110c081f0>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_process = {\n",
    "    'hypothesized mechanism': {\n",
    "        '_ports': {\n",
    "            '1': 'Any',\n",
    "            '2': 'Any',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "new_process = replace_regex_recursive(new_process)\n",
    "\n",
    "# plot\n",
    "plot_settingsx = copy.deepcopy(plot_settings)\n",
    "plot_settingsx['label_margin'] = '0.03'\n",
    "plot_settingsx['node_border_colors'] = {\n",
    "    ('hypothesized<br/>mechanism',): 'blue',\n",
    "}\n",
    "plot_settingsx['port_labels'] = False\n",
    "plot_bigraph(new_process, **plot_settingsx, filename='hypothesized_mechanism')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae52f73c-e384-4880-b18f-47b4c25ce203",
   "metadata": {},
   "source": [
    "### Cell structure and function\n",
    "Processes represent their functions, we show a more minimal cell structure with added processes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b39417ce-d9a1-4985-abce-ce22df816897",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 8.1.0 (20230707.0739)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"651pt\" height=\"670pt\"\n",
       " viewBox=\"0.00 0.00 669.87 689.62\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 685.62)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-685.62 665.87,-685.62 665.87,4 -4,4\"/>\n",
       "<!-- (&#39;cell&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;cell&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"349.57\" cy=\"-663.62\" rx=\"18\" ry=\"18\"/>\n",
       "<text text-anchor=\"start\" x=\"341.32\" y=\"-659.34\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"139.57\" cy=\"-573.32\" rx=\"35.3\" ry=\"35.3\"/>\n",
       "<text text-anchor=\"start\" x=\"114.82\" y=\"-569.04\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M332.74,-655.54C297.99,-640.93 217.79,-607.21 172.41,-588.13\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"349.57\" cy=\"-573.32\" rx=\"34.37\" ry=\"34.37\"/>\n",
       "<text text-anchor=\"start\" x=\"325.57\" y=\"-569.04\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M349.57,-644.75C349.57,-634.37 349.57,-620.94 349.57,-608.59\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;) -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"520.57\" cy=\"-573.32\" rx=\"29.7\" ry=\"29.7\"/>\n",
       "<text text-anchor=\"start\" x=\"500.32\" y=\"-569.04\" font-family=\"Times,serif\" font-size=\"12.00\">nucleoid</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M365.97,-654.15C395.56,-638.87 458.14,-606.56 493.94,-588.07\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"38.57\" cy=\"-411.08\" rx=\"38.57\" ry=\"38.57\"/>\n",
       "<text text-anchor=\"start\" x=\"11.19\" y=\"-406.81\" font-family=\"Times,serif\" font-size=\"12.00\">transporters</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M117.46,-544.53C107.56,-531.61 96.01,-515.85 86.57,-501.02 75.45,-483.55 64.38,-463.3 55.68,-446.53\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"115.57\" cy=\"-411.08\" rx=\"20.36\" ry=\"20.36\"/>\n",
       "<text text-anchor=\"start\" x=\"102.82\" y=\"-406.81\" font-family=\"Times,serif\" font-size=\"12.00\">lipids</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M134.36,-537.55C129.54,-505.33 122.53,-458.53 118.57,-432.14\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"250.29,-135.62 126.84,-135.62 126.84,-99.62 250.29,-99.62 250.29,-135.62\"/>\n",
       "<text text-anchor=\"start\" x=\"130.44\" y=\"-113.34\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane transport</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M39.99,-371.71C43.68,-320.82 56.72,-230.4 102.32,-169.98 112.02,-157.12 125.97,-146.85 139.82,-138.98\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"141.92\" cy=\"-137.82\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"102.57\" y=\"-232.69\" font-family=\"Times,serif\" font-size=\"10.00\">transporters</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"229.57\" cy=\"-411.08\" rx=\"37.64\" ry=\"37.64\"/>\n",
       "<text text-anchor=\"start\" x=\"202.94\" y=\"-406.81\" font-family=\"Times,serif\" font-size=\"12.00\">metabolites</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M323.25,-550.29C308.39,-537.01 290.14,-519.18 276.57,-501.02 264.03,-484.23 252.91,-463.64 244.65,-446.5\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"390.57\" cy=\"-236.57\" rx=\"66.59\" ry=\"66.59\"/>\n",
       "<text text-anchor=\"start\" x=\"340.69\" y=\"-232.29\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal complexes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M339.96,-539.72C327.22,-490.94 308.94,-396.65 332.57,-321.15 336.05,-310.03 341.59,-299.08 347.9,-288.97\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;) -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"431.57\" cy=\"-411.08\" rx=\"89.93\" ry=\"89.93\"/>\n",
       "<text text-anchor=\"start\" x=\"362.94\" y=\"-406.81\" font-family=\"Times,serif\" font-size=\"12.00\">transcript regulation complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M365.19,-541.8C372.48,-527.55 381.54,-509.84 390.61,-492.13\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"500.54,-36 444.59,-36 444.59,0 500.54,0 500.54,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"448.19\" y=\"-13.72\" font-family=\"Times,serif\" font-size=\"12.00\">translation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M224.33,-372.87C215.71,-311.6 198.79,-191.26 191.75,-141.25\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"191.41\" cy=\"-138.83\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"214.57\" y=\"-232.69\" font-family=\"Times,serif\" font-size=\"10.00\">internal</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M395.57,-169.2C399.5,-141.63 406.53,-109.91 419.07,-83.25 426.57,-67.3 438.99,-52.02 449.98,-40.37\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"451.61\" cy=\"-38.68\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"419.57\" y=\"-113.74\" font-family=\"Times,serif\" font-size=\"10.00\">p1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;) -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"472.57\" cy=\"-117.62\" rx=\"34.37\" ry=\"34.37\"/>\n",
       "<text text-anchor=\"start\" x=\"448.57\" y=\"-113.34\" font-family=\"Times,serif\" font-size=\"12.00\">transcripts</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M454.78,-323.39C456.2,-316.57 457.49,-309.77 458.57,-303.15 467.04,-251.15 470.41,-190.04 471.73,-152.84\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript regulation complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M472.57,-82.47C472.57,-69 472.57,-53.89 472.57,-41.79\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"472.57\" cy=\"-39.31\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"472.57\" y=\"-55.75\" font-family=\"Times,serif\" font-size=\"10.00\">p2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"619.57\" cy=\"-411.08\" rx=\"42.3\" ry=\"42.3\"/>\n",
       "<text text-anchor=\"start\" x=\"589.19\" y=\"-406.81\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M540.49,-550.2C548.15,-541.17 556.71,-530.42 563.57,-520.02 578.1,-497.99 591.8,-471.74 602,-450.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"619.57\" cy=\"-117.62\" rx=\"21.29\" ry=\"21.29\"/>\n",
       "<text text-anchor=\"start\" x=\"606.07\" y=\"-113.34\" font-family=\"Times,serif\" font-size=\"12.00\">genes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M619.57,-368.01C619.57,-304.8 619.57,-187.62 619.57,-139.79\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;external&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;external&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;) -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;, &#39;external&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M284.58,-392.28C281.61,-343.37 271.55,-208.03 249.57,-169.98 242.72,-158.13 232.07,-147.77 221.57,-139.52\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"219.7\" cy=\"-138.1\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"277.57\" y=\"-232.69\" font-family=\"Times,serif\" font-size=\"10.00\">external</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x110bfd0a0>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_struct_func = {\n",
    "    'cell': {\n",
    "        'membrane': {\n",
    "            'transporters': 'concentrations',\n",
    "            'lipids': 'concentrations',\n",
    "            'transmembrane transport': {\n",
    "                '_value': {\n",
    "                    '_process': 'transport URI',\n",
    "                    '_config': {'parameter': 1}\n",
    "                },\n",
    "                'wires': {\n",
    "                    'transporters': 'transporters',\n",
    "                    'internal': ['..', 'cytoplasm', 'metabolites']},\n",
    "                '_ports': {\n",
    "                    'transporters': 'concentrations',\n",
    "                    'internal': 'concentrations',\n",
    "                    'external': 'concentrations'\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "        'cytoplasm': {\n",
    "            'metabolites': {\n",
    "                '_value': 1.1,\n",
    "                '_type': 'concentrations'\n",
    "            },\n",
    "            'ribosomal complexes': {\n",
    "                '_value': 2.2,\n",
    "                '_type': 'concentrations'\n",
    "            },\n",
    "            'transcript regulation complex': {\n",
    "                '_value': 0.01,\n",
    "                'transcripts': {\n",
    "                    '_value': 0.1,\n",
    "                    '_type': 'concentrations'\n",
    "                }\n",
    "            },\n",
    "            'translation': {\n",
    "                'wires': {\n",
    "                    'p1': 'ribosomal complexes',\n",
    "                    'p2': ['transcript regulation complex', 'transcripts']}}},\n",
    "        'nucleoid': {\n",
    "            'chromosome': {\n",
    "                'genes': 'sequences'\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(cell_struct_func, remove_process_place_edges=True, out_dir='out', filename='cell')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "188eb4a6-c526-4b38-a22d-97f7fb8b7244",
   "metadata": {},
   "source": [
    "### Whole-cell *E. coli* model\n",
    "\n",
    "We here plot the processes and wiring of a whole-cell model of *E. coli* from [Macklin et al. Simultaneous cross-evaluation of heterogeneous E. coli datasets via mechanistic simulation. Science. 2020.](https://pubmed.ncbi.nlm.nih.gov/32703847/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "16960d1b-6e35-421e-bb7d-3efe5052e4fc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/ecoli\n"
     ]
    }
   ],
   "source": [
    "ecoli = { \n",
    "    'chromosome-structure': { \n",
    "        'wires': {\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'DnaA_boxes': ('unique', 'DnaA_box')\n",
    "        }\n",
    "    },\n",
    "    'metabolism': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'catalysts': ('bulk',),\n",
    "            'kinetics_enzymes': ('bulk',),\n",
    "            'kinetics_substrates': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'environment': ('environment',),\n",
    "            'amino_acids_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'tf-binding': { \n",
    "        'wires': {\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'active_tfs_total': ('bulk',),\n",
    "            'inactive_tfs_total': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'molecules': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'molecules': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'ntps': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'rna-degradation': { \n",
    "        'wires': {\n",
    "            'charged_trna': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'nmps': ('bulk',),\n",
    "            'fragmentMetabolites': ('bulk',),\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'endoRnases': ('bulk',),\n",
    "            'exoRnases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'RNA': ('unique', 'RNA'),\n",
    "            'subunits': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'molecules': ('bulk',),\n",
    "            'monomers': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'ppgpp_reaction_metabolites': ('bulk',),\n",
    "            'uncharged_trna': ('bulk',),\n",
    "            'charged_trna': ('bulk',),\n",
    "            'charging_molecules': ('bulk',),\n",
    "            'synthetases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules_total': ('bulk',),\n",
    "            'amino_acids_total': ('bulk',),\n",
    "            'charged_trna_total': ('bulk',),\n",
    "            'uncharged_trna_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'complexation': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',), \n",
    "        }\n",
    "    },\n",
    "    'two-component-system': { \n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'equilibrium': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'protein-degradation': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'proteins': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'chromosome-replication': { \n",
    "        'wires': {\n",
    "            'replisome_trimers': ('bulk',),\n",
    "            'replisome_monomers': ('bulk',),\n",
    "            'dntps': ('bulk',),\n",
    "            'ppi': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'environment': ('environment',)\n",
    "        }\n",
    "    },\n",
    "    'unique': {\n",
    "        'chromosome_domain': {},\n",
    "        'full_chromosome': {},\n",
    "        'oriC': {},\n",
    "        'active_replisome': {},\n",
    "        'RNA': {},\n",
    "        'active_ribosome': {},\n",
    "        'DnaA_box': {},\n",
    "        'promoter': {},\n",
    "    },\n",
    "    'bulk': {},\n",
    "    'environment': {},\n",
    "}\n",
    "node_groups = [[ #line up the top nodes for visual effect\n",
    "        ('unique',), ('bulk',), ('environment',), \n",
    "    ],\n",
    "]\n",
    "# plot settings\n",
    "plot_settings_ecoli = {\n",
    "    'remove_process_place_edges': True,\n",
    "    'rankdir': 'RL',\n",
    "    'dpi': '100',\n",
    "    'out_dir': 'out',\n",
    "}\n",
    "fig = plot_bigraph(ecoli, **plot_settings_ecoli, node_groups=node_groups, filename='ecoli')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54e1102d-521c-49cc-8ed2-be26c2b84c3d",
   "metadata": {
    "tags": []
   },
   "source": [
    "[See E. coli wiring diagram](https://github.com/vivarium-collective/bigraph-viz/blob/main/doc/_static/ecoli.png?raw=true)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
