{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5338cfab-cf15-4c63-a091-c1cbb51bf8dd",
   "metadata": {},
   "source": [
    "# Process Bigraph Schema\n",
    "This tutorial covers Process Bigraph Schema, starting with single `store` and increasing in detail as you scroll towards the bottom. It uses the libraries [bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) for the schema validation and type registry and [bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) for visualization.\n",
    "\n",
    "## Bigraph-schema and Bigraph-viz\n",
    "\n",
    "[Bigraph-schema](https://github.com/vivarium-collective/bigraph-schema) is a serializable type schema for Process Bigraph simulations. From this library, we will use the `validate_schema` method to check our schemas, and `fill` to fill in missing details to prepare the schema for simulation. We will also use the `type_registry` to access different schemas, and register new ones.\n",
    "\n",
    "[Bigraph-viz](https://github.com/vivarium-collective/bigraph-viz) is an easy-to-use plotting tool for compositional bigraph schema, built on top of Graphviz. It reads raw Compositional Bigraph Schemas and generates a figure that can be displayed in a notebook or saved to file. Bigraph-viz's main plotting function is `plot_bigraph` for the bigraph structure. `plot_flow` and `plot_multitimestep` make plots of process orchestration."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df69e36f-3da8-4bfb-8bc0-0aeb04a6f7d7",
   "metadata": {},
   "source": [
    "### Installation"
   ]
  },
  {
   "cell_type": "code",
   "id": "63fb7d1a-cd49-4aac-8d4f-31bb54496a51",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:40.209207Z",
     "start_time": "2025-03-08T18:16:39.738676Z"
    }
   },
   "source": [
    "# !pip install bigraph-schema\n",
    "# !pip install bigraph-viz\n",
    "!pip freeze | grep bigraph"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-e git+https://github.com/vivarium-collective/bigraph-schema.git@651c3f52e4aee4f6de01178655610150b93b0ff1#egg=bigraph_schema\r\n",
      "-e git+https://github.com/vivarium-collective/bigraph-viz.git@0ad7e73d953acd43be7ac96ab58a01ed525cd4b3#egg=bigraph_viz\r\n",
      "-e git+https://github.com/vivarium-collective/process-bigraph.git@5cc7546f883b0d568891b999b1db00221e8098f1#egg=process_bigraph\r\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "id": "f700afa2-bfee-49bf-a4b1-e23f0540461d",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "id": "2ab3cb32-e007-45c4-807e-b1b442ac1608",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-04-04T18:48:12.539102Z",
     "start_time": "2025-04-04T18:48:11.698045Z"
    }
   },
   "source": [
    "# from bigraph_viz import plot_bigraph, plot_flow, plot_multitimestep, pf\n",
    "from bigraph_viz import plot_bigraph, replace_regex_recursive, VisualizeTypes\n",
    "\n",
    "plot_settings = {\n",
    "    'remove_process_place_edges': True\n",
    "}\n",
    "save_images = False\n",
    "if save_images:\n",
    "    plot_settings.update({'out_dir': 'out','dpi': '250'})"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "id": "0e716c23-5fbe-4c08-a868-09611ca40ec4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Stores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a6e9cc2-e0df-4f73-b4c3-eddb1055b088",
   "metadata": {},
   "source": [
    "### Simple stores\n",
    "A store can be declared in a JSON dictionary with the store's name mapped to its `value`. "
   ]
  },
  {
   "cell_type": "code",
   "id": "d711b952-089a-4938-817d-d1ccad9de56e",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:40.669326Z",
     "start_time": "2025-03-08T18:16:40.624306Z"
    }
   },
   "source": [
    "simple_store_state = {\n",
    "    'store': 1.0,\n",
    "}\n",
    "plot_bigraph(simple_store_state,\n",
    "             **plot_settings,\n",
    "             # show_values=True,\n",
    "             filename='simple_store')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/simple_store\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"153pt\" height=\"153pt\"\n viewBox=\"0.00 0.00 44.09 44.09\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 40.09)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-40.09 40.09,-40.09 40.09,4 -4,4\"/>\n<!-- (&#39;store&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"18.04\" cy=\"-18.04\" rx=\"18.04\" ry=\"18.04\"/>\n<text text-anchor=\"start\" x=\"6.38\" y=\"-14.44\" font-family=\"Times,serif\" font-size=\"12.00\">store</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192bac40>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "id": "db1cead7-e691-4291-920e-827d69c00727",
   "metadata": {},
   "source": [
    "To include values *and* types in a node label, you can instead have the store name map to a dict with `_value` and `_type` keys."
   ]
  },
  {
   "cell_type": "code",
   "id": "f3e5fa71-7ffc-4a8f-a1a0-dc547c886f84",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:40.775223Z",
     "start_time": "2025-03-08T18:16:40.731998Z"
    }
   },
   "source": [
    "typed_store_spec = {\n",
    "    'store': {\n",
    "        '_value': 1.0, \n",
    "        '_type': 'float',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(typed_store_spec,\n",
    "             **plot_settings,\n",
    "             show_values=True,\n",
    "             show_types=True,\n",
    "             filename='store')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/store\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"207pt\" height=\"207pt\"\n viewBox=\"0.00 0.00 59.66 59.66\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 55.66)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-55.66 55.66,-55.66 55.66,4 -4,4\"/>\n<!-- (&#39;store&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"25.83\" cy=\"-25.83\" rx=\"25.83\" ry=\"25.83\"/>\n<text text-anchor=\"start\" x=\"14.16\" y=\"-28.23\" font-family=\"Times,serif\" font-size=\"12.00\">store</text>\n<text text-anchor=\"start\" x=\"11.16\" y=\"-16.23\" font-family=\"Times,serif\" font-size=\"12.00\">[float]</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192a3ee0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "cell_type": "markdown",
   "id": "de1011cb-903b-4c15-baed-a62f4afdd6b9",
   "metadata": {},
   "source": [
    "### Hierarchy\n",
    "A hierarchy is a place graph of nested stores. Stores can be placed within stores using JSON dictionary nesting."
   ]
  },
  {
   "cell_type": "code",
   "id": "0f3f0824-f89d-4c6e-830e-2c5a321f7b33",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:40.838022Z",
     "start_time": "2025-03-08T18:16:40.783215Z"
    }
   },
   "source": [
    "hierarchy_spec = {\n",
    "    'object1': {\n",
    "        'object1.1': {\n",
    "            'object1.1.1': 'any',\n",
    "            'object1.1.2': 'any',\n",
    "            'object1.1.3': {\n",
    "                'object1.1.3.1': 'any',\n",
    "            },\n",
    "        },\n",
    "        'object1.2': 'any',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(hierarchy_spec,\n",
    "             **plot_settings,\n",
    "             filename='hierarchy')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/hierarchy\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"920pt\" height=\"1323pt\"\n viewBox=\"0.00 0.00 264.86 380.90\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 376.9)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-376.9 260.86,-376.9 260.86,4 -4,4\"/>\n<!-- (&#39;object1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;object1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"170.09\" cy=\"-347.77\" rx=\"25.13\" ry=\"25.13\"/>\n<text text-anchor=\"start\" x=\"152.43\" y=\"-344.17\" font-family=\"Times,serif\" font-size=\"12.00\">object1</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"131.09\" cy=\"-256.18\" rx=\"30.45\" ry=\"30.45\"/>\n<text text-anchor=\"start\" x=\"108.93\" y=\"-252.58\" font-family=\"Times,serif\" font-size=\"12.00\">object1.1</text>\n</g>\n<!-- (&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M160.05,-323.7C154.9,-311.88 148.6,-297.39 143.19,-284.96\"/>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.2&#39;) -->\n<g id=\"node7\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"210.09\" cy=\"-256.18\" rx=\"30.45\" ry=\"30.45\"/>\n<text text-anchor=\"start\" x=\"187.93\" y=\"-252.58\" font-family=\"Times,serif\" font-size=\"12.00\">object1.2</text>\n</g>\n<!-- (&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.2&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M180.19,-324.16C185.51,-312.24 192.08,-297.51 197.71,-284.91\"/>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"41.09\" cy=\"-153.96\" rx=\"35.77\" ry=\"35.77\"/>\n<text text-anchor=\"start\" x=\"14.43\" y=\"-150.36\" font-family=\"Times,serif\" font-size=\"12.00\">object1.1.3</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M110.67,-232.44C97.08,-217.31 79.18,-197.37 64.94,-181.51\"/>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.1&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"131.09\" cy=\"-153.96\" rx=\"35.77\" ry=\"35.77\"/>\n<text text-anchor=\"start\" x=\"104.43\" y=\"-150.36\" font-family=\"Times,serif\" font-size=\"12.00\">object1.1.1</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.1&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M131.09,-224.84C131.09,-214.01 131.09,-201.71 131.09,-190.48\"/>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.2&#39;) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"221.09\" cy=\"-153.96\" rx=\"35.77\" ry=\"35.77\"/>\n<text text-anchor=\"start\" x=\"194.43\" y=\"-150.36\" font-family=\"Times,serif\" font-size=\"12.00\">object1.1.2</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.2&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M151.51,-232.44C165.1,-217.31 183,-197.37 197.25,-181.51\"/>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;, &#39;object1.1.3.1&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;, &#39;object1.1.3.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"41.09\" cy=\"-41.09\" rx=\"41.09\" ry=\"41.09\"/>\n<text text-anchor=\"start\" x=\"9.93\" y=\"-37.49\" font-family=\"Times,serif\" font-size=\"12.00\">object1.1.3.1</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;, &#39;object1.1.3.1&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;, &#39;object1.1.3&#39;, &#39;object1.1.3.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M41.09,-117.27C41.09,-106.38 41.09,-94.32 41.09,-83.11\"/>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192bab20>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 5
  },
  {
   "cell_type": "markdown",
   "id": "c825e07c-8cd5-4df1-8958-aa0f36d61045",
   "metadata": {},
   "source": [
    "## Processes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ea1a83-aeac-4048-9593-7b30588b0ad8",
   "metadata": {},
   "source": [
    "### Single process\n",
    "Processes require ports, which are declared with the `_ports` key mapped to a dict that has port names matched to types."
   ]
  },
  {
   "cell_type": "code",
   "id": "7c74237b-0d66-4e8f-a8ca-79174fadad8b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:40.886230Z",
     "start_time": "2025-03-08T18:16:40.843113Z"
    }
   },
   "source": [
    "process_spec = {\n",
    "    'process': {\n",
    "        '_type': 'edge',\n",
    "        '_inputs': {\n",
    "            'port1': 'any',\n",
    "            'port2': 'any',\n",
    "        },\n",
    "    },\n",
    "}\n",
    "plot_bigraph(process_spec,\n",
    "             **plot_settings,\n",
    "             rankdir='RL',\n",
    "             filename='process')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/process\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"508pt\" height=\"278pt\"\n viewBox=\"0.00 0.00 146.26 80.15\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 76.15)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-76.15 142.26,-76.15 142.26,4 -4,4\"/>\n<!-- (&#39;process&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;process&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-53.58 0,-53.58 0,-17.58 54,-17.58 54,-53.58\"/>\n<text text-anchor=\"start\" x=\"9.01\" y=\"-31.98\" font-family=\"Times,serif\" font-size=\"12.00\">process</text>\n</g>\n<!-- (&#39;process&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;process&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M110.78,-55.47C99.18,-52.69 81.54,-48.45 65.52,-44.6\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"66.8,-41.3 56.26,-42.37 65.17,-48.11 66.8,-41.3\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-52.58\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_input -->\n<!-- (&#39;process&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_input&#45;&gt;(&#39;process&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;process&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_input&#45;&gt;(&#39;process&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M110.78,-16.54C99.29,-19.19 81.88,-23.19 65.99,-26.84\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"65.23,-23.43 56.26,-29.08 66.79,-30.25 65.23,-23.43\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-26.58\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192a3310>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "cell_type": "markdown",
   "id": "68a2c480-32fe-43d3-b618-8fc6c647b8ca",
   "metadata": {},
   "source": [
    "### Multiple processes"
   ]
  },
  {
   "cell_type": "code",
   "id": "8c2457a3-fc47-4850-8b9e-9ee98e81567d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:40.938048Z",
     "start_time": "2025-03-08T18:16:40.891724Z"
    }
   },
   "source": [
    "process_schema = {\n",
    "    '_type': 'process',\n",
    "    '_inputs': {\n",
    "        'port1': 'Any',\n",
    "    },\n",
    "    '_outputs': {\n",
    "        'port2': 'Any'\n",
    "    },\n",
    "}\n",
    "\n",
    "processes_spec = {\n",
    "    'process1': process_schema,\n",
    "    'process2': process_schema,\n",
    "    'process3': process_schema,\n",
    "}\n",
    "plot_bigraph(processes_spec,\n",
    "             **plot_settings,\n",
    "             rankdir='BT',\n",
    "             filename='multiple_processes')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/multiple_processes\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"948pt\" height=\"407pt\"\n viewBox=\"0.00 0.00 273.15 117.15\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 113.15)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-113.15 269.15,-113.15 269.15,4 -4,4\"/>\n<!-- (&#39;process1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;process1&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"80.58,-109.15 26.58,-109.15 26.58,-73.15 80.58,-73.15 80.58,-109.15\"/>\n<text text-anchor=\"start\" x=\"32.58\" y=\"-87.55\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;process2&#39;,) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;process2&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"170.58,-109.15 116.58,-109.15 116.58,-73.15 170.58,-73.15 170.58,-109.15\"/>\n<text text-anchor=\"start\" x=\"122.58\" y=\"-87.55\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;process3&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;process3&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"251.58,-109.15 197.58,-109.15 197.58,-73.15 251.58,-73.15 251.58,-109.15\"/>\n<text text-anchor=\"start\" x=\"203.58\" y=\"-87.55\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n</g>\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M18.16,-27.29C21.26,-35.43 25.66,-46.13 30.47,-55.15 31.81,-57.68 33.29,-60.26 34.83,-62.82\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"31.78,-64.55 40.1,-71.11 37.69,-60.79 31.78,-64.55\"/>\n<text text-anchor=\"start\" x=\"30.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M115.64,-27.61C117.72,-35.69 120.76,-46.2 124.47,-55.15 125.44,-57.49 126.52,-59.88 127.66,-62.25\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"124.49,-63.74 132.2,-71.01 130.71,-60.52 124.49,-63.74\"/>\n<text text-anchor=\"start\" x=\"123.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input -->\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process3&#39;,) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;1&#39;)_input&#45;&gt;(&#39;process3&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M205.92,-27.92C205.84,-35.92 206.3,-46.25 208.47,-55.15 209.02,-57.44 209.73,-59.76 210.54,-62.05\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"207.21,-63.16 214.26,-71.07 213.68,-60.49 207.21,-63.16\"/>\n<text text-anchor=\"start\" x=\"207.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output -->\n<!-- (&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;process1&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M56.94,-39.31C56.23,-50.04 55.41,-62.38 54.76,-72.21\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"60.41,-39.81 57.58,-29.6 53.43,-39.35 60.41,-39.81\"/>\n<text text-anchor=\"start\" x=\"55.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output -->\n<!-- (&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;process2&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M153.05,-39.04C151.04,-49.88 148.72,-62.39 146.88,-72.33\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"156.48,-39.7 154.86,-29.23 149.6,-38.42 156.48,-39.7\"/>\n<text text-anchor=\"start\" x=\"151.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output -->\n<!-- (&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process3&#39;,) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;process3&#39;, &#39;p&#39;, &#39;o&#39;, &#39;r&#39;, &#39;t&#39;, &#39;2&#39;)_output&#45;&gt;(&#39;process3&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M243.21,-37.99C239.28,-49 234.66,-61.93 231,-72.16\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"246.42,-39.42 246.49,-28.82 239.83,-37.06 246.42,-39.42\"/>\n<text text-anchor=\"start\" x=\"239.58\" y=\"-47.15\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192fb310>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 7
  },
  {
   "cell_type": "markdown",
   "id": "15e5f5a6-8dd4-41c3-ba37-33ed9e3e0830",
   "metadata": {},
   "source": [
    "## Wires\n",
    "To connect a process's ports to stores, add wiring with a `_wires` key that maps port names to relative paths in the store hierarchy. The ports used by the wires need to match the ports in the schema."
   ]
  },
  {
   "cell_type": "code",
   "id": "a6ecad5d-8948-4636-910c-79284a3e8679",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:40.990539Z",
     "start_time": "2025-03-08T18:16:40.942780Z"
    }
   },
   "source": [
    "connected_process_spec = {\n",
    "    'process1': {\n",
    "        '_type': 'edge',\n",
    "        '_inputs': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'integer',\n",
    "            'port3': 'integer',\n",
    "        },\n",
    "        'inputs': {\n",
    "            'port1': ['node1'],\n",
    "            'port2': ['node2'],\n",
    "            'port3': ['node3'],\n",
    "        }\n",
    "    },\n",
    "    'node1': 'float',  # TODO -- shouldn't these fill?\n",
    "    'node2': 'integer',\n",
    "    'node3': 'integer',\n",
    "}\n",
    "plot_bigraph(connected_process_spec,\n",
    "             **plot_settings,\n",
    "             filename='wires')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/wires\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"601pt\" height=\"462pt\"\n viewBox=\"0.00 0.00 173.18 133.18\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 129.18)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-129.18 169.18,-129.18 169.18,4 -4,4\"/>\n<!-- (&#39;node1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;node1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"21.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"6.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node1</text>\n</g>\n<!-- (&#39;process1&#39;,) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;process1&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"109.59,-36 55.59,-36 55.59,0 109.59,0 109.59,-36\"/>\n<text text-anchor=\"start\" x=\"61.6\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M34.23,-85.27C42.54,-73.88 53.59,-58.75 63.02,-45.81\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"65.59,-48.24 68.65,-38.1 59.93,-44.11 65.59,-48.24\"/>\n<text text-anchor=\"start\" x=\"55.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;node2&#39;,) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;node2&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"82.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"67.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node2</text>\n</g>\n<!-- (&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M82.59,-81.06C82.59,-71.08 82.59,-59.07 82.59,-48.28\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"86.09,-48.3 82.59,-38.3 79.09,-48.3 86.09,-48.3\"/>\n<text text-anchor=\"start\" x=\"82.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;node3&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;node3&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143.59\" cy=\"-103.59\" rx=\"21.59\" ry=\"21.59\"/>\n<text text-anchor=\"start\" x=\"128.93\" y=\"-99.99\" font-family=\"Times,serif\" font-size=\"12.00\">node3</text>\n</g>\n<!-- (&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M130.95,-85.27C122.64,-73.88 111.59,-58.75 102.16,-45.81\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"105.25,-44.11 96.53,-38.1 99.59,-48.24 105.25,-44.11\"/>\n<text text-anchor=\"start\" x=\"114.59\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port3</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192fbaf0>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 8
  },
  {
   "cell_type": "markdown",
   "id": "e037285d-116a-404d-96ec-ba47318a2e8b",
   "metadata": {},
   "source": [
    "### Advanced wiring\n",
    "**TODO**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b17201-2489-45c8-a06b-9fdc29a714a7",
   "metadata": {},
   "source": [
    "## Composites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf5d428-db23-4985-80ff-6b93d22973aa",
   "metadata": {},
   "source": [
    "### Flat composite\n",
    "A flat composite connects stores with multiple connected processes, without nesting."
   ]
  },
  {
   "cell_type": "code",
   "id": "781a84e6-bbe2-4708-b441-47efc611b46f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.044461Z",
     "start_time": "2025-03-08T18:16:40.998854Z"
    }
   },
   "source": [
    "flat_composite_spec = {\n",
    "    'store1.1': 'float',\n",
    "    'store1.2': 'integer',\n",
    "    'process1': {\n",
    "        '_type': 'process',\n",
    "        '_outputs': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'integer',\n",
    "        },\n",
    "        'outputs': {\n",
    "            'port1': ['store1.1'],\n",
    "            'port2': ['store1.2'],\n",
    "        }\n",
    "    },\n",
    "    'process2': {\n",
    "        '_type': 'process',\n",
    "        '_inputs': {\n",
    "            'port1': 'float',\n",
    "            'port2': 'integer',\n",
    "        },\n",
    "        'inputs': {\n",
    "            'port1': ['store1.1'],\n",
    "            'port2': ['store1.2'],\n",
    "        }\n",
    "    },\n",
    "}\n",
    "plot_bigraph(flat_composite_spec,\n",
    "             **plot_settings,\n",
    "             rankdir='RL',\n",
    "             filename='flat_composite')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/flat_composite\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"600pt\" height=\"489pt\"\n viewBox=\"0.00 0.00 172.93 140.82\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 136.82)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-136.82 168.93,-136.82 168.93,4 -4,4\"/>\n<!-- (&#39;store1.1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;store1.1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"138.02\" cy=\"-105.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"118.86\" y=\"-102.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n</g>\n<!-- (&#39;process1&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;process1&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-118.91 0,-118.91 0,-82.91 54,-82.91 54,-118.91\"/>\n<text text-anchor=\"start\" x=\"6.01\" y=\"-97.31\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M99.1,-104.17C84.49,-103.5 68.12,-102.75 54.77,-102.14\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"98.69,-107.66 108.84,-104.62 99.01,-100.66 98.69,-107.66\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-104.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;process2&#39;,) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;process2&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-53.91 0,-53.91 0,-17.91 54,-17.91 54,-53.91\"/>\n<text text-anchor=\"start\" x=\"6.01\" y=\"-32.31\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M114.43,-91.41C99.94,-82.1 80.79,-69.81 64.21,-59.16\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"66.37,-56.39 56.07,-53.93 62.59,-62.28 66.37,-56.39\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-78.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;store1.2&#39;,) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;store1.2&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"138.02\" cy=\"-26.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"118.86\" y=\"-23.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n</g>\n<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M100.28,-36.89C90.61,-40.45 80.51,-45.07 72,-50.91 59.8,-59.28 48.8,-71.78 40.85,-82.19\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"101.12,-40.3 109.48,-33.79 98.88,-33.67 101.12,-40.3\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-52.91\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.91,-17.52C99.8,-14.2 85.1,-11.94 72,-14.91 69.83,-15.4 67.65,-16.01 65.47,-16.71\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"64.26,-13.43 56.19,-20.29 66.78,-19.96 64.26,-13.43\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-16.91\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192a3a30>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 9
  },
  {
   "cell_type": "markdown",
   "id": "398509c7-cf5c-4cce-a179-3a39752b0dde",
   "metadata": {},
   "source": [
    "### Nested composite\n",
    "A nested composite has store hierarchies with multiple connected processes."
   ]
  },
  {
   "cell_type": "code",
   "id": "701e759d-7f9b-41bb-911c-a8208c892462",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.104555Z",
     "start_time": "2025-03-08T18:16:41.051233Z"
    }
   },
   "source": [
    "nested_composite_spec = {\n",
    "    'object1': {\n",
    "        'object1.1': 'float',\n",
    "        'object1.2': 'integer',\n",
    "        'process1': {\n",
    "            '_type': 'process',\n",
    "            '_inputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'inputs': {\n",
    "                'port1': ['object1.1'],\n",
    "                'port2': ['object1.2'],\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'process',\n",
    "            '_outputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'outputs': {\n",
    "                'port1': ['object1.1'],\n",
    "                'port2': ['object1.2'],\n",
    "            }\n",
    "        },\n",
    "    },\n",
    "    'process3': {\n",
    "        '_type': 'process',\n",
    "        '_inputs': {\n",
    "            'port1': 'tree',\n",
    "        },\n",
    "        'inputs': {\n",
    "            'port1': ['object1'],\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(nested_composite_spec, \n",
    "             **plot_settings,\n",
    "             filename='nested_composite')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/nested_composite\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"990pt\" height=\"858pt\"\n viewBox=\"0.00 0.00 285.00 247.17\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 243.17)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-243.17 281,-243.17 281,4 -4,4\"/>\n<!-- (&#39;object1&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;object1&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"137\" cy=\"-214.04\" rx=\"25.13\" ry=\"25.13\"/>\n<text text-anchor=\"start\" x=\"119.34\" y=\"-210.44\" font-family=\"Times,serif\" font-size=\"12.00\">object1</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"137\" cy=\"-112.45\" rx=\"30.45\" ry=\"30.45\"/>\n<text text-anchor=\"start\" x=\"114.84\" y=\"-108.85\" font-family=\"Times,serif\" font-size=\"12.00\">object1.1</text>\n</g>\n<!-- (&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M137,-187.91C137,-174.43 137,-157.71 137,-143.54\"/>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.2&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;object1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"58\" cy=\"-112.45\" rx=\"30.45\" ry=\"30.45\"/>\n<text text-anchor=\"start\" x=\"35.84\" y=\"-108.85\" font-family=\"Times,serif\" font-size=\"12.00\">object1.2</text>\n</g>\n<!-- (&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.2&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;object1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M121.39,-193.36C108.62,-177.27 90.51,-154.44 76.92,-137.3\"/>\n</g>\n<!-- (&#39;object1&#39;, &#39;process1&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;process1&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-36 0,-36 0,0 54,0 54,-36\"/>\n<text text-anchor=\"start\" x=\"6.01\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;process1&#39;) -->\n<!-- (&#39;object1&#39;, &#39;process2&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;object1&#39;, &#39;process2&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"186,-36 132,-36 132,0 186,0 186,-36\"/>\n<text text-anchor=\"start\" x=\"138.01\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;object1&#39;,)&#45;&gt;(&#39;object1&#39;, &#39;process2&#39;) -->\n<!-- (&#39;process3&#39;,) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;process3&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"277,-130.45 223,-130.45 223,-94.45 277,-94.45 277,-130.45\"/>\n<text text-anchor=\"start\" x=\"229.01\" y=\"-108.85\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n</g>\n<!-- (&#39;object1&#39;,)&#45;&gt;(&#39;process3&#39;,) -->\n<g id=\"edge7\" class=\"edge\">\n<title>(&#39;object1&#39;,)&#45;&gt;(&#39;process3&#39;,)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M160.37,-202.54C175.33,-195.09 194.56,-184.09 209,-170.91 218.53,-162.2 227.16,-150.93 234.01,-140.68\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"236.77,-142.87 239.2,-132.56 230.87,-139.11 236.77,-142.87\"/>\n<text text-anchor=\"start\" x=\"218\" y=\"-162.91\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process1&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M113.67,-91.85C96.98,-77.82 74.38,-58.82 56.45,-43.75\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"59.13,-41.43 49.22,-37.68 54.62,-46.79 59.13,-41.43\"/>\n<text text-anchor=\"start\" x=\"79\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process2&#39;) -->\n<g id=\"edge8\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.1&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M146.67,-70.83C149.49,-58.98 152.43,-46.62 154.73,-36.95\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"143.28,-69.93 144.37,-80.47 150.09,-71.56 143.28,-69.93\"/>\n<text text-anchor=\"start\" x=\"151\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.2&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process1&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.2&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M39.32,-87.59C34.7,-80.4 30.37,-72.22 27.89,-64 26.38,-58.98 25.6,-53.51 25.27,-48.15\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"28.77,-48.35 25.18,-38.38 21.77,-48.41 28.77,-48.35\"/>\n<text text-anchor=\"start\" x=\"28\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;object1&#39;, &#39;object1.2&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process2&#39;) -->\n<g id=\"edge9\" class=\"edge\">\n<title>(&#39;object1&#39;, &#39;object1.2&#39;)&#45;&gt;(&#39;object1&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M88.93,-83.14C105.4,-68.07 125.15,-49.99 139.47,-36.87\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"86.8,-80.35 81.78,-89.68 91.52,-85.51 86.8,-80.35\"/>\n<text text-anchor=\"start\" x=\"121\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192ea7f0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 10
  },
  {
   "cell_type": "markdown",
   "id": "0c8c1704-2331-44eb-bd05-72e36ccb6858",
   "metadata": {},
   "source": [
    "### Composite process\n",
    "Composite processes are processes with internal stores and internal processes, which can run as their own simulation. The composite process has schema, allowing it to connect to a super-simulation. This allows for improved distributed computation, with composite processes able to run on their own computer and syncronize to the super-simulation via message passing."
   ]
  },
  {
   "cell_type": "code",
   "id": "638c5dc7-4cfd-46e4-b66a-8c42ca98ff1d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.156910Z",
     "start_time": "2025-03-08T18:16:41.109839Z"
    }
   },
   "source": [
    "composite_process_spec = {\n",
    "    'composite_process': {\n",
    "        '_type': 'composite',\n",
    "        'store1.1': 'any',\n",
    "        'store1.2': 'any',\n",
    "        'process1': {\n",
    "            '_type': 'process',\n",
    "            '_outputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'outputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            '_type': 'process',\n",
    "            '_outputs': {\n",
    "                'port1': 'float',\n",
    "                'port2': 'integer',\n",
    "            },\n",
    "            'outputs': {\n",
    "                'port1': ['store1.1'],\n",
    "                'port2': ['store1.2'],\n",
    "            }\n",
    "        },\n",
    "        # '_inputs': {\n",
    "        #     'port1': 'Any', \n",
    "        #     'port2': 'Any', \n",
    "        # },\n",
    "        # '_tunnels': {\n",
    "        #     'port1': 'store1.1',\n",
    "        #     'port2': 'store1.2',\n",
    "        # }\n",
    "    }\n",
    "}\n",
    "\n",
    "plot_bigraph(\n",
    "    composite_process_spec, \n",
    "    **plot_settings, filename='composite_process')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/composite_process\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"674pt\" height=\"753pt\"\n viewBox=\"0.00 0.00 194.00 216.82\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(3.47222 3.47222) rotate(0) translate(4 212.82)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-212.82 190,-212.82 190,4 -4,4\"/>\n<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"130\" cy=\"-108.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"110.84\" y=\"-105.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;process1&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"186,-36 132,-36 132,0 186,0 186,-36\"/>\n<text text-anchor=\"start\" x=\"138.01\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M141.73,-71.96C145.65,-59.92 149.86,-47.02 153.14,-36.96\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"138.48,-70.64 138.7,-81.23 145.13,-72.81 138.48,-70.64\"/>\n<text text-anchor=\"start\" x=\"148\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;process2&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-36 0,-36 0,0 54,0 54,-36\"/>\n<text text-anchor=\"start\" x=\"6.01\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n<g id=\"edge7\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.08,-74.92C105.85,-67.47 99.73,-60 93,-54 81.87,-44.07 67.42,-36 54.88,-30.12\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"107.96,-76.55 116.39,-82.97 113.81,-72.69 107.96,-76.55\"/>\n<text text-anchor=\"start\" x=\"102\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"58\" cy=\"-108.91\" rx=\"26.91\" ry=\"26.91\"/>\n<text text-anchor=\"start\" x=\"38.84\" y=\"-105.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M62.8,-70.37C64.87,-64.41 67.79,-58.68 71.89,-54 87.05,-36.73 111.94,-27.87 131.24,-23.39\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"59.5,-69.15 60.31,-79.71 66.27,-70.95 59.5,-69.15\"/>\n<text text-anchor=\"start\" x=\"72\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n<g id=\"edge8\" class=\"edge\">\n<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M34.24,-78C31.67,-73.51 29.43,-68.78 27.89,-64 25.12,-55.37 24.68,-45.32 25.02,-36.91\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"31.09,-79.58 39.46,-86.08 36.98,-75.78 31.09,-79.58\"/>\n<text text-anchor=\"start\" x=\"28\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n</g>\n<!-- (&#39;composite_process&#39;,) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;composite_process&#39;,)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"152.26,-208.82 53.74,-208.82 53.74,-172.82 152.26,-172.82 152.26,-208.82\"/>\n<text text-anchor=\"start\" x=\"57.34\" y=\"-187.22\" font-family=\"Times,serif\" font-size=\"12.00\">composite_process</text>\n</g>\n<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M108.99,-172.08C112.66,-161.23 117.41,-147.18 121.46,-135.17\"/>\n</g>\n<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n<g id=\"edge4\" class=\"edge\">\n<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M93.01,-172.08C86.59,-160.68 78.18,-145.74 71.21,-133.36\"/>\n</g>\n<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1192b7af0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 11
  },
  {
   "cell_type": "markdown",
   "id": "847cef26-65a1-4da6-9dd3-eb5839277970",
   "metadata": {},
   "source": [
    "## Orchestration"
   ]
  },
  {
   "cell_type": "code",
   "id": "e76c97a8-3037-464a-929e-ba4414c73c34",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.163522Z",
     "start_time": "2025-03-08T18:16:41.162014Z"
    }
   },
   "source": [
    "plot_settings2={}\n",
    "if save_images:\n",
    "    plot_settings2.update({\n",
    "        'out_dir': 'out',\n",
    "        'dpi': '250'\n",
    "    })"
   ],
   "outputs": [],
   "execution_count": 12
  },
  {
   "cell_type": "markdown",
   "id": "48975890-06a4-462a-9b1d-6bddd51779fe",
   "metadata": {},
   "source": [
    "### Multi-timestepping\n",
    "\n",
    "Temporal processes can be run by multi-timestepping, with each process updating the state according to its `sync_state` state, which it can update adaptively."
   ]
  },
  {
   "cell_type": "code",
   "id": "f0eff7b0-f0e2-4d56-8b16-ef0fc1ce5aba",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.168875Z",
     "start_time": "2025-03-08T18:16:41.167517Z"
    }
   },
   "source": [
    "# multitimestep_spec = {\n",
    "#     'temporal process2': {\n",
    "#         '_ports': {'port1': 'Any'},\n",
    "#         'wires': {'port1': 'state'},\n",
    "#         '_sync_step': 1.0,\n",
    "#     },\n",
    "#     'temporal process1': {\n",
    "#         '_ports': {'port1': 'Any'},\n",
    "#         'wires': {'port1': 'state'},\n",
    "#         '_sync_step': 0.5,\n",
    "#     },\n",
    "# }\n",
    "# multitimestep_spec = replace_regex_recursive(multitimestep_spec)\n",
    "# plot_multitimestep(multitimestep_spec, total_time=2.0, **plot_settings2, filename='multitimestep')"
   ],
   "outputs": [],
   "execution_count": 13
  },
  {
   "cell_type": "markdown",
   "id": "a00cf677-ebac-4309-8ac9-a5b2855724af",
   "metadata": {},
   "source": [
    "### Flows\n",
    "\n",
    "A directed acyclic graph (DAG) – a type of workflow that declares the dependencies between processes. A flow run is not temporal, and the order is entirely based on dependencies. It can triggered between time steps, running the steps in the order as determined by their dependencies."
   ]
  },
  {
   "cell_type": "code",
   "id": "db45c99b-b97b-47bc-a73e-c669882f3969",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.174334Z",
     "start_time": "2025-03-08T18:16:41.172920Z"
    }
   },
   "source": [
    "# flow = {\n",
    "#     'step1': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': [], \n",
    "#     },\n",
    "#     'step2': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': 'step1', \n",
    "#     },\n",
    "#     'step3': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': [], \n",
    "#     },\n",
    "#     'step4': {\n",
    "#         '_type': 'step_process',\n",
    "#         '_ports': {},\n",
    "#         '_depends_on': ['step2', 'step3'], \n",
    "#     },\n",
    "# }\n",
    "\n",
    "# plot_flow(flow, **plot_settings2, filename='flow')"
   ],
   "outputs": [],
   "execution_count": 14
  },
  {
   "cell_type": "markdown",
   "id": "5882581a-c5ef-40b5-acbd-efddad269807",
   "metadata": {},
   "source": [
    "## Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff81178f-36e4-4200-982a-ed8155b87934",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Multiscale map of a cell\n",
    "\n",
    "Bigraphs can represent the hierarchical nesting of cellular structures. Below is a multiscale map of a cell from [Qin et al. A multi-scale map of cell structure fusing protein images and interactions. Nature. 2021.](https://pubmed.ncbi.nlm.nih.gov/34819669/)"
   ]
  },
  {
   "cell_type": "code",
   "id": "47bdf8e5-0bdd-4cac-b57e-2449c75fa178",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.287809Z",
     "start_time": "2025-03-08T18:16:41.178364Z"
    }
   },
   "source": [
    "music_map = {\n",
    "    'cell': {\n",
    "        'cytoplasm': {\n",
    "            'secretory organelles': {\n",
    "                'transmembrane transport systems': {\n",
    "                    'ion transmembrane transport systems': {}}},\n",
    "            'cytoplasmic organelles': {\n",
    "                'subgroup of cytoplasmic organelles': {\n",
    "                    'metabolic organelles': {\n",
    "                        'subgroup of metabolic organelles': {},\n",
    "                        'mitochondrion': {}},\n",
    "                    'ribonucleoprotein complex family': {\n",
    "                        'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
    "                        'ribosome biogenesis community': {\n",
    "                            'ribosome': {\n",
    "                                'ribosomal subunit': {\n",
    "                                    'mitochondirial large ribosomal subunit': {\n",
    "                                        # 'large ribosomal subunit subcomplex 1': {}\n",
    "                                    },\n",
    "                                    'mito-cyto ribosomal cluster': {},\n",
    "                                    'cotranslational protein targeting to membrane system': {},\n",
    "                                    'ribosomal complex 5': {},\n",
    "                                    'ribosomal complex 2': {}}},\n",
    "                            'nucleic acid binding complex': {}}}}},\n",
    "            'transcriptional regulation complex family': {\n",
    "                'transcriptional regulation complexes': {}}},\n",
    "        'nucleus': {\n",
    "            'DNA metabolic assembly': {},\n",
    "            'nuclear lumen': {\n",
    "                'nucleolus': {\n",
    "                    'ribonucleprotein complex family': {},\n",
    "                    'RNA processing complex family': {\n",
    "                        'RNA processing complex 1': {},\n",
    "                        'RNA splicing complex family': {}},\n",
    "                    'splicosomal complex family': {}},\n",
    "                'nuclear splicing speckle': {}},\n",
    "            'nuclear body': {},\n",
    "            'nucleo-plasm 1': {\n",
    "                'chromosome organization complex family': {\n",
    "                    'chromatin organization complex family': {\n",
    "                        'HAT complex family': {\n",
    "                            'NuA4 HAT complex': {}}}}},\n",
    "            'nucleo-plasm 2': {\n",
    "                'chromatin regulation complex': {}},\n",
    "            'nuclear transcriptional speckle': {\n",
    "                'negative regulation of RNA biosynthesis process': {}}}}}\n",
    "music_map1 = replace_regex_recursive(music_map)\n",
    "fig = plot_bigraph(music_map1, **plot_settings, filename='cell_hierarchy')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell_hierarchy\n"
     ]
    }
   ],
   "execution_count": 15
  },
  {
   "cell_type": "code",
   "id": "ab431982-d156-4b7a-9918-384a5ea5132a",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.296943Z",
     "start_time": "2025-03-08T18:16:41.294500Z"
    }
   },
   "source": [
    "music_map"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cell': {'cytoplasm': {'secretory organelles': {'transmembrane transport systems': {'ion transmembrane transport systems': {}}},\n",
       "   'cytoplasmic organelles': {'subgroup of cytoplasmic organelles': {'metabolic organelles': {'subgroup of metabolic organelles': {},\n",
       "      'mitochondrion': {}},\n",
       "     'ribonucleoprotein complex family': {'RNA processing complex 1': {'RNA splicing complex 1': {}},\n",
       "      'ribosome biogenesis community': {'ribosome': {'ribosomal subunit': {'mitochondirial large ribosomal subunit': {},\n",
       "         'mito-cyto ribosomal cluster': {},\n",
       "         'cotranslational protein targeting to membrane system': {},\n",
       "         'ribosomal complex 5': {},\n",
       "         'ribosomal complex 2': {}}},\n",
       "       'nucleic acid binding complex': {}}}}},\n",
       "   'transcriptional regulation complex family': {'transcriptional regulation complexes': {}}},\n",
       "  'nucleus': {'DNA metabolic assembly': {},\n",
       "   'nuclear lumen': {'nucleolus': {'ribonucleprotein complex family': {},\n",
       "     'RNA processing complex family': {'RNA processing complex 1': {},\n",
       "      'RNA splicing complex family': {}},\n",
       "     'splicosomal complex family': {}},\n",
       "    'nuclear splicing speckle': {}},\n",
       "   'nuclear body': {},\n",
       "   'nucleo-plasm 1': {'chromosome organization complex family': {'chromatin organization complex family': {'HAT complex family': {'NuA4 HAT complex': {}}}}},\n",
       "   'nucleo-plasm 2': {'chromatin regulation complex': {}},\n",
       "   'nuclear transcriptional speckle': {'negative regulation of RNA biosynthesis process': {}}}}}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 16
  },
  {
   "cell_type": "code",
   "id": "5a76608a-fff0-4005-9f32-c02afef39d9e",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.462826Z",
     "start_time": "2025-03-08T18:16:41.308381Z"
    }
   },
   "source": [
    "import copy\n",
    "\n",
    "transcription_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'RNA processing': 'any',\n",
    "        'regulation': 'any',\n",
    "        'nucleic acids': 'any',\n",
    "        'negative regulation': 'any'\n",
    "    },\n",
    "    'outputs': {\n",
    "        'RNA processing': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "        'regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "        'nucleic acids': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', \n",
    "            'ribonucleoprotein complex family', 'ribosome biogenesis community', 'nucleic acid binding complex'],\n",
    "        'negative regulation': [\n",
    "            'nucleus', 'nuclear transcriptional speckle', 'negative regulation of RNA biosynthesis process'],\n",
    "    }\n",
    "}\n",
    "\n",
    "translation_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'ribosomes': 'any',\n",
    "        'membrane': 'any'\n",
    "    },\n",
    "    'outputs': {\n",
    "        'ribosomes': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', 'ribosome biogenesis community', 'ribosome'],\n",
    "        # 'membrane': ['],\n",
    "    }\n",
    "}\n",
    "\n",
    "metabolism_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'organelles': 'any',\n",
    "        'mitochondrion': 'any',\n",
    "        'ion transport': 'any',\n",
    "    },\n",
    "    'outputs': {\n",
    "        'organelles': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', ],\n",
    "        'mitochondrion': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'metabolic organelles', 'mitochondrion'],\n",
    "        'ion transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems', 'ion transmembrane transport systems'],\n",
    "    }\n",
    "}\n",
    "\n",
    "cell_cycle_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'DNA': 'any',\n",
    "        'chromosome': 'any',\n",
    "        'chromatin': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'DNA': ['nucleus', 'DNA metabolic assembly'],\n",
    "        'chromosome': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family',],\n",
    "        'chromatin': ['nucleus', 'nucleo-plasm 1', 'chromosome organization complex family', 'chromatin organization complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "signalling_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'transport': 'any',\n",
    "        'trancript regulation': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'trancript regulation': [\n",
    "            'cytoplasm', 'transcriptional regulation complex family', 'transcriptional regulation complexes'],\n",
    "    }\n",
    "}\n",
    "\n",
    "protein_transport_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'transmembrane transport': 'any',\n",
    "        'secretory': 'any',\n",
    "        'cotranslation': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'transmembrane transport': [\n",
    "            'cytoplasm', 'secretory organelles', 'transmembrane transport systems'],\n",
    "        'secretory': [\n",
    "            'cytoplasm', 'secretory organelles'],\n",
    "        'cotranslation': [\n",
    "            'cytoplasm', 'cytoplasmic organelles', 'subgroup of cytoplasmic organelles', 'ribonucleoprotein complex family', \n",
    "            'ribosome biogenesis community', 'ribosome', 'ribosomal subunit', 'cotranslational protein targeting to membrane system']\n",
    "    }\n",
    "}\n",
    "\n",
    "rna_processing_process = {\n",
    "    '_type': 'process',\n",
    "    '_outputs': {\n",
    "        'processing complex': 'any',\n",
    "    },\n",
    "    'outputs': {        \n",
    "        'processing complex': [\n",
    "            'nucleus', 'nuclear lumen', 'nucleolus', 'RNA processing complex family'],\n",
    "    }\n",
    "}\n",
    "\n",
    "# compose schema\n",
    "music_map['cell']['transcription'] = transcription_process\n",
    "music_map['cell']['translation'] = translation_process\n",
    "music_map['cell']['metabolism'] = metabolism_process\n",
    "music_map['cell']['cell cycle'] = cell_cycle_process\n",
    "music_map['cell']['signalling'] = signalling_process\n",
    "music_map['cell']['protein transport'] = protein_transport_process\n",
    "music_map['cell']['RNA processing'] = rna_processing_process\n",
    "\n",
    "music_map2 = replace_regex_recursive(music_map)\n",
    "\n",
    "# plot\n",
    "plot_settingsx = copy.deepcopy(plot_settings)\n",
    "plot_settingsx['label_margin'] = '0.02'\n",
    "plot_settingsx['node_border_colors'] = {\n",
    "    ('cell', 'translation'): 'blue',\n",
    "    ('cell', 'transcription'): 'blue',\n",
    "    ('cell', 'metabolism'): 'blue',\n",
    "    ('cell', 'cell<br/>cycle'): 'blue',\n",
    "    ('cell', 'signalling'): 'blue',\n",
    "    ('cell', 'protein<br/>transport'): 'blue',\n",
    "    ('cell', 'RNA<br/>processing'): 'blue',\n",
    "}\n",
    "# plot_settingsx.update({'out_dir': 'out','dpi': '250'})\n",
    "fig = plot_bigraph(music_map2, **plot_settingsx, filename='cell_hierarchy_functions')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell_hierarchy_functions\n"
     ]
    }
   ],
   "execution_count": 17
  },
  {
   "cell_type": "code",
   "id": "fe3008ad-498e-449f-8521-6ea29fd903f1",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.496867Z",
     "start_time": "2025-03-08T18:16:41.468538Z"
    }
   },
   "source": [
    "new_process = {\n",
    "    'hypothesized mechanism': {\n",
    "        '_type': 'step',\n",
    "        '_inputs': {\n",
    "            '1': 'Any',\n",
    "            '2': 'Any',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "new_process = replace_regex_recursive(new_process)\n",
    "\n",
    "# plot\n",
    "plot_settingsx = copy.deepcopy(plot_settings)\n",
    "plot_settingsx['label_margin'] = '0.03'\n",
    "plot_settingsx['node_border_colors'] = {\n",
    "    ('hypothesized<br/>mechanism',): 'blue',\n",
    "}\n",
    "plot_settingsx['port_labels'] = False\n",
    "fig = plot_bigraph(new_process, **plot_settingsx, filename='hypothesized_mechanism')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/hypothesized_mechanism\n"
     ]
    }
   ],
   "execution_count": 18
  },
  {
   "cell_type": "markdown",
   "id": "ae52f73c-e384-4880-b18f-47b4c25ce203",
   "metadata": {},
   "source": [
    "### Cell structure and function\n",
    "Processes represent their functions, we show a more minimal cell structure with added processes."
   ]
  },
  {
   "cell_type": "code",
   "id": "b39417ce-d9a1-4985-abce-ce22df816897",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.668601Z",
     "start_time": "2025-03-08T18:16:41.503264Z"
    }
   },
   "source": [
    "core = VisualizeTypes()\n",
    "core.register(key='concentrations', schema='float')\n",
    "core.register(key='sequences', schema='float')\n",
    "core.register(key='membrane', schema={\n",
    "    'transporters': 'concentrations',\n",
    "    'lipids': 'concentrations',\n",
    "    'transmembrane<br/>transport': {\n",
    "        '_type': 'process',\n",
    "        '_inputs': {},\n",
    "        '_outputs': {\n",
    "            'transporters': 'concentrations',\n",
    "            'internal': 'concentrations',\n",
    "            'external': 'concentrations'\n",
    "        }\n",
    "    }\n",
    "})\n",
    "\n",
    "core.register(key='cytoplasm', schema={\n",
    "    'metabolites': 'concentrations',\n",
    "    'ribosomal<br/>complexes': 'concentrations',\n",
    "    'transcript<br/>regulation<br/>complex': {\n",
    "        'transcripts': 'concentrations'},\n",
    "    'translation': {\n",
    "        '_type': 'process',\n",
    "        '_outputs': {\n",
    "            'p1': 'concentrations',\n",
    "            'p2': 'concentrations'}\n",
    "    }\n",
    "})\n",
    "\n",
    "core.register(\n",
    "    key='nucleoid',\n",
    "    schema={\n",
    "        'chromosome': {\n",
    "            'genes': 'sequences'}})\n",
    "\n",
    "core.register(\n",
    "    key='cell',\n",
    "    schema={\n",
    "        'membrane': 'membrane',\n",
    "        'cytoplasm': 'cytoplasm',\n",
    "        'nucleoid': 'nucleoid'})\n",
    "\n",
    "# state\n",
    "cell_struct_state = {\n",
    "    'cell<br/>composite': {\n",
    "        'membrane': {\n",
    "            'transmembrane<br/>transport': {\n",
    "                'outputs': {\n",
    "                    'transporters': ['transporters'],\n",
    "                    'internal': ['..', 'cytoplasm', 'metabolites']}}},\n",
    "        'cytoplasm': {\n",
    "            'translation': {\n",
    "                'outputs': {\n",
    "                    'p1': ['ribosomal<br/>complexes'],\n",
    "                    'p2': ['transcript<br/>regulation<br/>complex', 'transcripts']}}}}}\n",
    "\n",
    "plot_bigraph(\n",
    "    cell_struct_state,\n",
    "    dpi='1000',\n",
    "    size='5,4',\n",
    "    aspect_ratio='fill',\n",
    "    label_margin='0.02',\n",
    "    schema={'cell<br/>composite': 'cell'},\n",
    "    core=core,\n",
    "    remove_process_place_edges=True,\n",
    "    out_dir='out',\n",
    "    filename='cell_composite')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell_composite\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 9.0.0 (0)\n -->\n<!-- Title: bigraph Pages: 1 -->\n<svg width=\"4986pt\" height=\"4000pt\"\n viewBox=\"0.00 0.00 359.01 288.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(7.40488 7.40488) rotate(0) translate(4 536.18)\">\n<title>bigraph</title>\n<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-536.18 669.38,-536.18 669.38,4 -4,4\"/>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;,) -->\n<g id=\"node1\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;,)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"408.84\" cy=\"-495\" rx=\"36.91\" ry=\"36.91\"/>\n<text text-anchor=\"start\" x=\"400.18\" y=\"-497.4\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n<text text-anchor=\"start\" x=\"384.18\" y=\"-485.4\" font-family=\"Times,serif\" font-size=\"12.00\">composite</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;) -->\n<g id=\"node2\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"154.84\" cy=\"-392\" rx=\"29.39\" ry=\"29.39\"/>\n<text text-anchor=\"start\" x=\"129.52\" y=\"-388.4\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;,)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;) -->\n<g id=\"edge1\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;,)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M374.05,-480.17C323.1,-459.91 229.19,-422.57 182.58,-404.03\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"node5\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"408.84\" cy=\"-392\" rx=\"28.67\" ry=\"28.67\"/>\n<text text-anchor=\"start\" x=\"384.18\" y=\"-388.4\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;,)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;) -->\n<g id=\"edge5\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;,)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M408.84,-457.25C408.84,-445.52 408.84,-432.72 408.84,-421.65\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;) -->\n<g id=\"node10\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"548.84\" cy=\"-392\" rx=\"24.27\" ry=\"24.27\"/>\n<text text-anchor=\"start\" x=\"528.18\" y=\"-388.4\" font-family=\"Times,serif\" font-size=\"12.00\">nucleoid</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;,)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;) -->\n<g id=\"edge11\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;,)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M438.87,-472.33C466.19,-452.63 505.72,-424.11 529.2,-407.17\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n<g id=\"node3\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"40.84\" cy=\"-289\" rx=\"32.69\" ry=\"32.69\"/>\n<text text-anchor=\"start\" x=\"12.51\" y=\"-285.4\" font-family=\"Times,serif\" font-size=\"12.00\">transporters</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n<g id=\"edge2\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M132.86,-371.53C113.58,-354.44 85.57,-329.63 65.52,-311.87\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n<g id=\"node4\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"126.84\" cy=\"-289\" rx=\"18\" ry=\"18\"/>\n<text text-anchor=\"start\" x=\"113.5\" y=\"-285.4\" font-family=\"Times,serif\" font-size=\"12.00\">lipids</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n<g id=\"edge3\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M147.04,-362.87C142.09,-345 135.84,-322.47 131.61,-307.22\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n<g id=\"node13\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"244.26,-129 167.41,-129 167.41,-93 244.26,-93 244.26,-129\"/>\n<text text-anchor=\"start\" x=\"168.85\" y=\"-113.4\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n<text text-anchor=\"start\" x=\"184.51\" y=\"-101.4\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n<g id=\"edge14\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M59.21,-247.84C73.48,-219.96 95.41,-183.36 122.63,-157.68 135.34,-145.69 151.79,-135.8 166.7,-128.32\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"56.12,-246.19 54.8,-256.7 62.39,-249.31 56.12,-246.19\"/>\n<text text-anchor=\"start\" x=\"122.84\" y=\"-193\" font-family=\"Times,serif\" font-size=\"10.00\">transporters</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n<g id=\"node6\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"258.84\" cy=\"-289\" rx=\"31.96\" ry=\"31.96\"/>\n<text text-anchor=\"start\" x=\"231.18\" y=\"-285.4\" font-family=\"Times,serif\" font-size=\"12.00\">metabolites</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n<g id=\"edge6\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M384.74,-374.77C357.66,-356.54 313.81,-327.02 285.67,-308.06\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;) -->\n<g id=\"node7\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"397.84\" cy=\"-196\" rx=\"38.32\" ry=\"38.32\"/>\n<text text-anchor=\"start\" x=\"373.84\" y=\"-198.4\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n<text text-anchor=\"start\" x=\"372.18\" y=\"-186.4\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;) -->\n<g id=\"edge7\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M407.24,-362.79C405.33,-329.06 402.12,-272.45 399.99,-234.92\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n<g id=\"node8\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"461.84\" cy=\"-289\" rx=\"36.44\" ry=\"36.44\"/>\n<text text-anchor=\"start\" x=\"439.18\" y=\"-297.4\" font-family=\"Times,serif\" font-size=\"12.00\">transcript</text>\n<text text-anchor=\"start\" x=\"437.51\" y=\"-285.4\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n<text text-anchor=\"start\" x=\"441.18\" y=\"-273.4\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n<g id=\"edge8\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M422.08,-365.77C429.01,-352.57 437.56,-336.26 444.97,-322.16\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n<g id=\"node14\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"495.84,-36 441.84,-36 441.84,0 495.84,0 495.84,-36\"/>\n<text text-anchor=\"start\" x=\"443.51\" y=\"-14.4\" font-family=\"Times,serif\" font-size=\"12.00\">translation</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n<g id=\"edge15\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M246.42,-246.76C235.22,-209.58 219.3,-156.7 211.13,-129.56\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"243.06,-247.73 249.29,-256.3 249.76,-245.71 243.06,-247.73\"/>\n<text text-anchor=\"start\" x=\"241.84\" y=\"-193\" font-family=\"Times,serif\" font-size=\"10.00\">internal</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n<g id=\"edge17\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M402.94,-145.76C406.29,-125.37 411.8,-101.88 420.84,-81.97 428.45,-65.2 441.25,-48.73 451.71,-36.84\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"399.5,-145.13 401.48,-155.54 406.42,-146.17 399.5,-145.13\"/>\n<text text-anchor=\"start\" x=\"420.84\" y=\"-108\" font-family=\"Times,serif\" font-size=\"10.00\">p1</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;) -->\n<g id=\"node9\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"468.84\" cy=\"-111\" rx=\"29.03\" ry=\"29.03\"/>\n<text text-anchor=\"start\" x=\"443.85\" y=\"-107.4\" font-family=\"Times,serif\" font-size=\"12.00\">transcripts</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;) -->\n<g id=\"edge9\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M463.27,-251.96C464.58,-219.07 466.49,-171.1 467.7,-140.73\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n<g id=\"edge16\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M468.84,-69.89C468.84,-58.3 468.84,-46.23 468.84,-36.76\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"465.34,-69.69 468.84,-79.69 472.34,-69.69 465.34,-69.69\"/>\n<text text-anchor=\"start\" x=\"468.84\" y=\"-56\" font-family=\"Times,serif\" font-size=\"10.00\">p2</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n<g id=\"node11\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"620.84\" cy=\"-289\" rx=\"35.62\" ry=\"35.62\"/>\n<text text-anchor=\"start\" x=\"589.85\" y=\"-285.4\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n<g id=\"edge12\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M562.72,-371.53C573.35,-356.62 588.17,-335.83 600.18,-318.98\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n<g id=\"node12\" class=\"node\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"620.84\" cy=\"-111\" rx=\"18\" ry=\"18\"/>\n<text text-anchor=\"start\" x=\"607.18\" y=\"-107.4\" font-family=\"Times,serif\" font-size=\"12.00\">genes</text>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n<g id=\"edge13\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M620.84,-252.38C620.84,-215.22 620.84,-158.46 620.84,-129.83\"/>\n</g>\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;, &#39;e&#39;, &#39;x&#39;, &#39;t&#39;, &#39;e&#39;, &#39;r&#39;, &#39;n&#39;, &#39;a&#39;, &#39;l&#39;)_output -->\n<!-- (&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;, &#39;e&#39;, &#39;x&#39;, &#39;t&#39;, &#39;e&#39;, &#39;r&#39;, &#39;n&#39;, &#39;a&#39;, &#39;l&#39;)_output&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n<g id=\"edge18\" class=\"edge\">\n<title>(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;, &#39;e&#39;, &#39;x&#39;, &#39;t&#39;, &#39;e&#39;, &#39;r&#39;, &#39;n&#39;, &#39;a&#39;, &#39;l&#39;)_output&#45;&gt;(&#39;cell&lt;br/&gt;composite&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M328.66,-266.66C321.08,-239.01 304.76,-190.44 276.84,-157.68 267.33,-146.52 254.42,-137.11 242.21,-129.78\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"325.26,-267.5 331.15,-276.3 332.04,-265.75 325.26,-267.5\"/>\n<text text-anchor=\"start\" x=\"317.84\" y=\"-193\" font-family=\"Times,serif\" font-size=\"10.00\">external</text>\n</g>\n</g>\n</svg>\n",
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x11929d0d0>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 19
  },
  {
   "cell_type": "markdown",
   "id": "188eb4a6-c526-4b38-a22d-97f7fb8b7244",
   "metadata": {},
   "source": [
    "### Whole-cell *E. coli* model\n",
    "\n",
    "We here plot the processes and wiring of a whole-cell model of *E. coli* from [Macklin et al. Simultaneous cross-evaluation of heterogeneous E. coli datasets via mechanistic simulation. Science. 2020.](https://pubmed.ncbi.nlm.nih.gov/32703847/)"
   ]
  },
  {
   "cell_type": "code",
   "id": "16960d1b-6e35-421e-bb7d-3efe5052e4fc",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2025-03-08T18:16:41.740473Z",
     "start_time": "2025-03-08T18:16:41.674846Z"
    }
   },
   "source": [
    "ecoli = { \n",
    "    'chromosome-structure': { \n",
    "        'wires': {\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'DnaA_boxes': ('unique', 'DnaA_box')\n",
    "        }\n",
    "    },\n",
    "    'metabolism': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'catalysts': ('bulk',),\n",
    "            'kinetics_enzymes': ('bulk',),\n",
    "            'kinetics_substrates': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'environment': ('environment',),\n",
    "            'amino_acids_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'tf-binding': { \n",
    "        'wires': {\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'active_tfs_total': ('bulk',),\n",
    "            'inactive_tfs_total': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'molecules': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'molecules': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'ntps': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'rna-degradation': { \n",
    "        'wires': {\n",
    "            'charged_trna': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'nmps': ('bulk',),\n",
    "            'fragmentMetabolites': ('bulk',),\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'endoRnases': ('bulk',),\n",
    "            'exoRnases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-initiation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'RNA': ('unique', 'RNA'),\n",
    "            'subunits': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-elongation': { \n",
    "        'wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'molecules': ('bulk',),\n",
    "            'monomers': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'ppgpp_reaction_metabolites': ('bulk',),\n",
    "            'uncharged_trna': ('bulk',),\n",
    "            'charged_trna': ('bulk',),\n",
    "            'charging_molecules': ('bulk',),\n",
    "            'synthetases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules_total': ('bulk',),\n",
    "            'amino_acids_total': ('bulk',),\n",
    "            'charged_trna_total': ('bulk',),\n",
    "            'uncharged_trna_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'complexation': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',), \n",
    "        }\n",
    "    },\n",
    "    'two-component-system': { \n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'equilibrium': {\n",
    "        'wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'protein-degradation': { \n",
    "        'wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'proteins': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'chromosome-replication': { \n",
    "        'wires': {\n",
    "            'replisome_trimers': ('bulk',),\n",
    "            'replisome_monomers': ('bulk',),\n",
    "            'dntps': ('bulk',),\n",
    "            'ppi': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'environment': ('environment',)\n",
    "        }\n",
    "    },\n",
    "    'unique': {\n",
    "        'chromosome_domain': {},\n",
    "        'full_chromosome': {},\n",
    "        'oriC': {},\n",
    "        'active_replisome': {},\n",
    "        'RNA': {},\n",
    "        'active_ribosome': {},\n",
    "        'DnaA_box': {},\n",
    "        'promoter': {},\n",
    "    },\n",
    "    'bulk': {},\n",
    "    'environment': {},\n",
    "}\n",
    "node_groups = [[ #line up the top nodes for visual effect\n",
    "        ('unique',), ('bulk',), ('environment',), \n",
    "    ],\n",
    "]\n",
    "# plot settings\n",
    "plot_settings_ecoli = {\n",
    "    'remove_process_place_edges': True,\n",
    "    'rankdir': 'RL',\n",
    "    'dpi': '100',\n",
    "    'out_dir': 'out',\n",
    "}\n",
    "fig = plot_bigraph(ecoli, **plot_settings_ecoli, node_groups=node_groups, filename='ecoli')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/ecoli\n"
     ]
    }
   ],
   "execution_count": 20
  },
  {
   "cell_type": "markdown",
   "id": "54e1102d-521c-49cc-8ed2-be26c2b84c3d",
   "metadata": {
    "tags": []
   },
   "source": [
    "[See E. coli wiring diagram](https://github.com/vivarium-collective/bigraph-viz/blob/main/doc/_static/ecoli.png?raw=true)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
