{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5338cfab-cf15-4c63-a091-c1cbb51bf8dd",
   "metadata": {},
   "source": [
    "# Bigraph Schema Basics\n",
    "This tutorial demonstrates use of [bigraph-viz](https://github.com/vivarium-collective/bigraph-viz), a visualization tool for bigraph composition schema. The tutorial starts with a single `store` and increases in detail as you scroll towards the bottom.\n",
    "\n",
    "Compositional Bigraph Schemas (CBS) are based on a mathematical formalism [introduced by Robin Milner in 2009](https://www.google.com/search?q=the+space+and+motion+of+communicating+agents+by+robin+milner), and used in Vivarium. Bigraphs consist of networks with embeddable nodes that can be placed within other nodes and dynamically restructured. CBS reimagines the bigraph concept. Variables are contained within Stores (the circles in the figure), which can be embedded in a hierarchy, as shown by the dark edges. Instead of Milner's hyperedges, CBS employs Processes (the rectangles) which connect via wires (dashed edges) to shared variables within the Stores. Processes and Stores form a type of bipartite graph, as illustrated by the dashed edges."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8532c1b2-9af2-46f3-8e85-80a7f5161c26",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Imports and settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2ab3cb32-e007-45c4-807e-b1b442ac1608",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bigraph_viz import plot_bigraph, plot_flow, plot_multitimestep\n",
    "\n",
    "plot_settings = {\n",
    "    'remove_process_place_edges': True,\n",
    "}\n",
    "save_images = False\n",
    "if save_images:\n",
    "    plot_settings.update({\n",
    "        'out_dir': 'out',\n",
    "        'dpi': '250'\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "99be1b87-5e5b-4c06-93f1-7bf772f7b6d5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function plot_bigraph in module bigraph_viz.plot:\n",
      "\n",
      "plot_bigraph(bigraph_schema, size='16,10', node_label_size='12pt', show_values=False, show_types=False, collapse_processes=False, port_labels=True, port_label_size='10pt', engine='dot', rankdir='TB', node_border_colors=None, node_fill_colors=None, node_groups=False, invisible_edges=False, remove_process_place_edges=False, print_source=False, dpi='70', file_format='png', out_dir=None, filename=None)\n",
      "    Plot a bigraph from bigraph schema.\n",
      "    \n",
      "    Args:\n",
      "        bigraph_schema (dict): The bigraph schema dict that will be plotted.\n",
      "        size (str, optional): The size of the output figure (example: '16,10'). Default is '16,10'.\n",
      "        node_label_size (str, optional): The font size for the node labels. Default is None.\n",
      "        show_values (bool, optional): Display on value info in node label. Default is False.\n",
      "        show_types (bool, optional): Display on type info in node label. Default is False.\n",
      "        collapse_processes (bool, optional): Collapse rectangular process nodes to a hyperedge vertex. Default is False.\n",
      "        port_labels (bool, optional): Turn on port labels. Default is True.\n",
      "        port_label_size (str, optional): The font size of the port labels (example: '10pt'). Default is None.\n",
      "        engine (str, optional): Graphviz graphing engine. Try 'dot' or 'neato'. Default is 'dot'.\n",
      "        rankdir (str, optional): Sets direction of graph layout. 'TB'=top-to-bottom, 'LR'=left-to-right.\n",
      "            Default is 'TB'.\n",
      "        node_border_colors (dict, optional): Colors of node borders, with node path tuples mapped to the node color as\n",
      "            the value: {('path', 'to', 'node',): 'color'}. Colors at https://graphviz.org/doc/info/colors.html\n",
      "        node_fill_colors (dict, optional): Colors of node fills, with node path tuples mapped to the node color as\n",
      "            the value: {('path', 'to', 'node',): 'color'}. Colors at https://graphviz.org/doc/info/colors.html\n",
      "        node_groups (list, optional): A list of lists of nodes.\n",
      "            Each sub-list is a grouping of nodes that will be aligned at the same rank.\n",
      "            For example: [[('path to', 'group1 node1',), ('path to', 'group1 node2',)], [another group]]\n",
      "            Default is None.\n",
      "        invisible_edges (list, optional): A list of edge tuples. The edge tuples have the (source, target) node\n",
      "            according to the nodes' paths. For example: [(('top',), ('top', 'inner1')), (another edge)]\n",
      "            Default is None.\n",
      "        remove_process_place_edges (bool, optional): Turn off process place edges from plotting. Default is False.\n",
      "        print_source (bool, optional): Print the graphviz DOT source code as string. Default is False.\n",
      "        file_format (str, optional): File format of the output image. Default is 'png'.\n",
      "        out_dir (bool, optional): The output directory for the bigraph image. Default is None.\n",
      "        filename (bool, optional): The file name for the bigraph image. Default is None.\n",
      "    \n",
      "    Notes:\n",
      "        You can adjust node labels using HTML syntax for fonts, colors, sizes, subscript, superscript. For example:\n",
      "            H<sub><font point-size=\"8\">2</font></sub>O will print H2O with 2 as a subscript with smaller font.\n",
      "        You can also make newlines by adding with <br/>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(plot_bigraph)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e716c23-5fbe-4c08-a868-09611ca40ec4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Stores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a6e9cc2-e0df-4f73-b4c3-eddb1055b088",
   "metadata": {},
   "source": [
    "### Simple stores\n",
    "An store can be declared in a JSON dictionary with the store's name mapped to its value. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d711b952-089a-4938-817d-d1ccad9de56e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"59pt\" height=\"59pt\"\n",
       " viewBox=\"0.00 0.00 60.33 60.33\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 56.33)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-56.33 56.33,-56.33 56.33,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"26.16\" cy=\"-26.16\" rx=\"26.33\" ry=\"26.33\"/>\n",
       "<text text-anchor=\"start\" x=\"11.66\" y=\"-29.56\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "<text text-anchor=\"start\" x=\"18.16\" y=\"-16.56\" font-family=\"Times,serif\" font-size=\"12.00\">1.0</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x106d51790>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simple_store_spec = {\n",
    "    'store1': 1.0,\n",
    "}\n",
    "plot_bigraph(simple_store_spec, **plot_settings, show_values=True, filename='simple_store')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db1cead7-e691-4291-920e-827d69c00727",
   "metadata": {},
   "source": [
    "To include values *and* types in a node label, you can instead have the store name map to a dict with `_value` and `_type` keys."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f3e5fa71-7ffc-4a8f-a1a0-dc547c886f84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"78pt\" height=\"78pt\"\n",
       " viewBox=\"0.00 0.00 80.12 80.12\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 76.12)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-76.12 76.12,-76.12 76.12,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"36.06\" cy=\"-36.06\" rx=\"36.12\" ry=\"36.12\"/>\n",
       "<text text-anchor=\"start\" x=\"21.56\" y=\"-39.46\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "<text text-anchor=\"start\" x=\"14.56\" y=\"-26.46\" font-family=\"Times,serif\" font-size=\"12.00\">1.0::float</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1083557c0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "typed_store_spec = {\n",
    "    'store1': {\n",
    "        '_value': 1.0, \n",
    "        '_type': 'float',\n",
    "    },\n",
    "}\n",
    "plot_bigraph(typed_store_spec, **plot_settings, show_values=True, show_types=True, filename='store')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de1011cb-903b-4c15-baed-a62f4afdd6b9",
   "metadata": {},
   "source": [
    "### Hierarchy\n",
    "A hierarchy is a place graph of nested stores. Stores can be placed within stores using JSON dictionary nesting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0f3f0824-f89d-4c6e-830e-2c5a321f7b33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"246pt\" height=\"362pt\"\n",
       " viewBox=\"0.00 0.00 253.25 372.70\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 368.7)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-368.7 249.25,-368.7 249.25,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"156.86\" cy=\"-337.92\" rx=\"22.59\" ry=\"22.59\"/>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"156.86\" cy=\"-337.92\" rx=\"26.56\" ry=\"26.56\"/>\n",
       "<text text-anchor=\"start\" x=\"142.36\" y=\"-334.82\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"119.86\" cy=\"-246.83\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"100.86\" y=\"-243.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M146.75,-312.58C141.74,-300.52 135.69,-285.96 130.6,-273.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"194.86\" cy=\"-246.83\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"175.86\" y=\"-243.73\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M167.24,-312.58C172.38,-300.52 178.59,-285.96 183.82,-273.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"33.86\" cy=\"-148.65\" rx=\"33.72\" ry=\"33.72\"/>\n",
       "<text text-anchor=\"start\" x=\"10.36\" y=\"-145.55\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M101.2,-224.96C88.06,-210.26 70.39,-190.51 56.45,-174.92\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"119.86\" cy=\"-148.65\" rx=\"33.72\" ry=\"33.72\"/>\n",
       "<text text-anchor=\"start\" x=\"96.36\" y=\"-145.55\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M119.86,-217.75C119.86,-206.88 119.86,-194.36 119.86,-183.03\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"205.86\" cy=\"-148.65\" rx=\"33.72\" ry=\"33.72\"/>\n",
       "<text text-anchor=\"start\" x=\"182.36\" y=\"-145.55\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M138.52,-224.96C151.66,-210.26 169.32,-190.51 183.26,-174.92\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"205.86\" cy=\"-39.4\" rx=\"39.3\" ry=\"39.3\"/>\n",
       "<text text-anchor=\"start\" x=\"177.86\" y=\"-36.3\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1.3.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;, &#39;store1.1.3&#39;, &#39;store1.1.3.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M205.86,-114.01C205.86,-103.08 205.86,-90.85 205.86,-79.54\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108355c40>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hierarchy_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': {\n",
    "            'store1.1.1': {'_value': 1.1, '_type': 'float'},\n",
    "            'store1.1.2': {'_value': 2, '_type': 'int'},\n",
    "            'store1.1.3': {\n",
    "                'store1.1.3.1': {'_value': 0.1, '_type': 'float'},\n",
    "            },\n",
    "        },\n",
    "        'store1.2': {'_value': 1, '_type': 'int'},\n",
    "    },\n",
    "}\n",
    "plot_bigraph(hierarchy_spec, **plot_settings, filename='hierarchy')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c825e07c-8cd5-4df1-8958-aa0f36d61045",
   "metadata": {},
   "source": [
    "## Processes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ea1a83-aeac-4048-9593-7b30588b0ad8",
   "metadata": {},
   "source": [
    "### Single process\n",
    "Processes require ports, which are declared with the `_ports` key mapped to a dict that has port names matched to types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c74237b-0d66-4e8f-a8ca-79174fadad8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"118pt\" height=\"95pt\"\n",
       " viewBox=\"0.00 0.00 121.00 98.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 94)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-94 117,-94 117,4 -4,4\"/>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-63 0,-63 0,-27 54,-27 54,-63\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-41.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.06,-71.7C105.93,-70.05 80.86,-61.99 59.59,-55.15\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"57.22\" cy=\"-54.39\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-69\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.06,-18.3C105.93,-19.95 80.86,-28.01 59.59,-34.85\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"57.22\" cy=\"-35.61\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-33\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108355580>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_spec = {\n",
    "    'process1': {\n",
    "        '_ports': {\n",
    "            'port1': {'_type': 'type'},\n",
    "            'port2': {'_type': 'type'},\n",
    "        },\n",
    "    },\n",
    "}\n",
    "plot_bigraph(process_spec, **plot_settings, rankdir='RL', filename='process')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68a2c480-32fe-43d3-b618-8fc6c647b8ca",
   "metadata": {},
   "source": [
    "### Multiple processes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8c2457a3-fc47-4850-8b9e-9ee98e81567d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"249pt\" height=\"123pt\"\n",
       " viewBox=\"0.00 0.00 256.50 127.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 123)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-123 252.5,-123 252.5,4 -4,4\"/>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"67.5,-119 13.5,-119 13.5,-83 67.5,-83 67.5,-119\"/>\n",
       "<text text-anchor=\"start\" x=\"19.5\" y=\"-97.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"157.5,-119 103.5,-119 103.5,-83 157.5,-83 157.5,-119\"/>\n",
       "<text text-anchor=\"start\" x=\"109.5\" y=\"-97.9\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;process3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"238.5,-119 184.5,-119 184.5,-83 238.5,-83 238.5,-119\"/>\n",
       "<text text-anchor=\"start\" x=\"190.5\" y=\"-97.9\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M1.93,-23.99C4.5,-32.6 10.28,-50.71 17.5,-65 19.71,-69.38 22.36,-73.88 25.05,-78.14\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"26.39\" cy=\"-80.21\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"17.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;process1&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M44.39,-36.98C43.66,-48.83 42.69,-64.45 41.91,-77.17\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"41.76\" cy=\"-79.58\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"43.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process2&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;process2&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M100.97,-27C102.83,-36.32 106.41,-52.1 111.5,-65 113.21,-69.34 115.35,-73.83 117.56,-78.08\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"118.6\" cy=\"-80.03\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"111.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process2&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;process2&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M143,-27.67C140.95,-39.53 137.2,-61.24 134.34,-77.79\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"133.98\" cy=\"-79.84\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"138.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;process3&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;process3&#39;, &#39;port1&#39;)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M192.13,-36.7C191.9,-45.35 192.29,-55.84 194.5,-65 195.52,-69.24 197.12,-73.58 198.92,-77.68\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"199.89\" cy=\"-79.76\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"194.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process3&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;process3&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;process3&#39;, &#39;port2&#39;)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M237.05,-23.35C233.7,-33.41 225.22,-58.83 218.95,-77.66\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"218.2\" cy=\"-79.92\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"226.5\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1083553a0>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_schema = {\n",
    "    '_ports': {\n",
    "        'port1': {'_type': 'type'},\n",
    "        'port2': {'_type': 'type'}\n",
    "    }\n",
    "}\n",
    "\n",
    "processes_spec = {\n",
    "    'process1': process_schema,\n",
    "    'process2': process_schema,\n",
    "    'process3': process_schema,\n",
    "}\n",
    "plot_bigraph(processes_spec, **plot_settings, rankdir='BT', filename='multiple_processes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e5f5a6-8dd4-41c3-ba37-33ed9e3e0830",
   "metadata": {},
   "source": [
    "## Wires\n",
    "To connect a process's ports to stores, add wiring with a `_wires` key that maps port names to relative paths in the store hierarchy. The ports used by the wires need to match the ports in the schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6ecad5d-8948-4636-910c-79284a3e8679",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"180pt\" height=\"134pt\"\n",
       " viewBox=\"0.00 0.00 184.78 137.78\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 133.78)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-133.78 180.78,-133.78 180.78,4 -4,4\"/>\n",
       "<!-- (&#39;node1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;node1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"23.39\" cy=\"-106.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"8.39\" y=\"-103.29\" font-family=\"Times,serif\" font-size=\"12.00\">node1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"115.39,-36 61.39,-36 61.39,0 115.39,0 115.39,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"67.39\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M37.5,-86.65C47.67,-73.13 61.42,-54.85 72.07,-40.7\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"73.48\" cy=\"-38.82\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"62.39\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;node2&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"88.39\" cy=\"-106.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"73.39\" y=\"-103.29\" font-family=\"Times,serif\" font-size=\"12.00\">node2</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M88.39,-82.24C88.39,-69.62 88.39,-54.07 88.39,-41.55\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"88.39\" cy=\"-39.23\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"88.39\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;node3&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"153.39\" cy=\"-106.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"138.39\" y=\"-103.29\" font-family=\"Times,serif\" font-size=\"12.00\">node3</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M139.29,-86.65C129.12,-73.13 115.36,-54.85 104.72,-40.7\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"103.31\" cy=\"-38.82\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"122.39\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port3</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x106d51910>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "connected_process_spec = {\n",
    "    'process1': {\n",
    "        '_ports': {\n",
    "            'port1': {'_type': 'float'},\n",
    "            'port2': {'_type': 'int'},\n",
    "            'port3': {'_type': 'int'},\n",
    "        },\n",
    "        '_wires': {\n",
    "            'port1': 'node1',\n",
    "            'port2': 'node2',\n",
    "            'port3': 'node3',\n",
    "        }\n",
    "    },\n",
    "    'node1': 1.1,\n",
    "    'node2': 2,\n",
    "    'node3': 2,\n",
    "}\n",
    "plot_bigraph(connected_process_spec, **plot_settings, filename='wires')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7f2769-7a83-44bd-9bf6-1e40fae89f1c",
   "metadata": {},
   "source": [
    "### Processes are hyperedges\n",
    "When we collapse processes from the graph, their connections become hyperedges, and the Process Bigraph can be depicted as a standard Milner Bigraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f00bb63a-494c-476e-ac76-0c796baec828",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"180pt\" height=\"100pt\"\n",
       " viewBox=\"0.00 0.00 184.78 102.78\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 98.78)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-98.78 180.78,-98.78 180.78,4 -4,4\"/>\n",
       "<!-- (&#39;node1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;node1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"23.39\" cy=\"-71.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"8.39\" y=\"-68.29\" font-family=\"Times,serif\" font-size=\"12.00\">node1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<!-- (&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;node1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M35.8,-51.15C42.9,-40.96 52.36,-28.6 62.39,-19 71.28,-10.5 84.52,-3.09 87.69,-1.37\"/>\n",
       "<text text-anchor=\"start\" x=\"62.39\" y=\"-22\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;node2&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"88.39\" cy=\"-71.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"73.39\" y=\"-68.29\" font-family=\"Times,serif\" font-size=\"12.00\">node2</text>\n",
       "</g>\n",
       "<!-- (&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;node2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M88.39,-47.29C88.39,-28.99 88.39,-5.84 88.39,-1.66\"/>\n",
       "<text text-anchor=\"start\" x=\"88.39\" y=\"-22\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;node3&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"153.39\" cy=\"-71.39\" rx=\"23.29\" ry=\"23.29\"/>\n",
       "<text text-anchor=\"start\" x=\"138.39\" y=\"-68.29\" font-family=\"Times,serif\" font-size=\"12.00\">node3</text>\n",
       "</g>\n",
       "<!-- (&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;node3&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M140.98,-51.15C133.89,-40.96 124.42,-28.6 114.39,-19 105.51,-10.5 92.27,-3.09 89.09,-1.37\"/>\n",
       "<text text-anchor=\"start\" x=\"123.39\" y=\"-22\" font-family=\"Times,serif\" font-size=\"10.00\">port3</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108355fd0>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot_bigraph(connected_process_spec, collapse_processes=True, **plot_settings, filename='hyperedge')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e037285d-116a-404d-96ec-ba47318a2e8b",
   "metadata": {},
   "source": [
    "### Advanced wiring\n",
    "**TODO**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b17201-2489-45c8-a06b-9fdc29a714a7",
   "metadata": {},
   "source": [
    "## Composites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf5d428-db23-4985-80ff-6b93d22973aa",
   "metadata": {},
   "source": [
    "### Flat composite\n",
    "A flat composite connects stores with multiple connected processes, without nesting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "781a84e6-bbe2-4708-b441-47efc611b46f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"172pt\" height=\"143pt\"\n",
       " viewBox=\"0.00 0.00 176.63 146.63\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 142.63)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-142.63 172.63,-142.63 172.63,4 -4,4\"/>\n",
       "<!-- (&#39;store1.1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1.1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"140.32\" cy=\"-110.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"121.32\" y=\"-107.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;process1&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-123.32 0,-123.32 0,-87.32 54,-87.32 54,-123.32\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-102.22\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.23,-109.06C95.58,-108.35 76.05,-107.48 59.84,-106.75\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"57.43\" cy=\"-106.64\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-110.32\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;process2&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-57.32 0,-57.32 0,-21.32 54,-21.32 54,-57.32\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-36.22\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;store1.1&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M115.71,-95.26C99.1,-84.66 76.7,-70.38 58.7,-58.9\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"56.81\" cy=\"-57.69\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-83.32\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1.2&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"140.32\" cy=\"-28.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"121.32\" y=\"-25.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process1&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M112.26,-35.41C99.33,-39.61 84.09,-45.8 72,-54.32 61.03,-62.05 51.05,-73.1 43.38,-82.93\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"41.97\" cy=\"-84.79\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-57.32\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;store1.2&#39;,)&#45;&gt;(&#39;process2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M112.94,-18.53C100.45,-15.19 85.37,-13.03 72,-16.32 67.7,-17.37 63.35,-18.93 59.14,-20.75\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"56.92\" cy=\"-21.77\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72\" y=\"-19.32\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x10837ec40>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flat_composite_spec = {\n",
    "    'store1.1': {\n",
    "        '_value': 1.1, \n",
    "        '_type': 'float',\n",
    "    },\n",
    "    'store1.2': {\n",
    "        '_value': 2, \n",
    "        '_type': 'int',\n",
    "    },\n",
    "    'process1': {\n",
    "        # '_ports': {\n",
    "        #     'port1': 'type', \n",
    "        #     'port2': 'type',\n",
    "        # },\n",
    "        '_wires': {\n",
    "            'port1': 'store1.1',\n",
    "            'port2': 'store1.2',\n",
    "        }\n",
    "    },\n",
    "    'process2': {\n",
    "        # '_ports': {\n",
    "        #     'port1': 'type', \n",
    "        #     'port2': 'type',\n",
    "        # },\n",
    "        '_wires': {\n",
    "            'port1': 'store1.1',\n",
    "            'port2': 'store1.2',\n",
    "        }\n",
    "    },\n",
    "}\n",
    "plot_bigraph(flat_composite_spec, **plot_settings, rankdir='RL', filename='flat_composite')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398509c7-cf5c-4cce-a179-3a39752b0dde",
   "metadata": {},
   "source": [
    "### Nested composite\n",
    "A nested composite has store hierarchies with multiple connected processes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "701e759d-7f9b-41bb-911c-a8208c892462",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"281pt\" height=\"241pt\"\n",
       " viewBox=\"0.00 0.00 289.00 248.19\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 244.19)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-244.19 285,-244.19 285,4 -4,4\"/>\n",
       "<!-- (&#39;store1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;store1&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143\" cy=\"-213.41\" rx=\"22.59\" ry=\"22.59\"/>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143\" cy=\"-213.41\" rx=\"26.56\" ry=\"26.56\"/>\n",
       "<text text-anchor=\"start\" x=\"128.5\" y=\"-210.31\" font-family=\"Times,serif\" font-size=\"12.00\">store1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"143\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"124\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M143,-186.11C143,-171.98 143,-154.56 143,-140.19\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"68\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"49\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M127.09,-191.18C114.71,-174.66 97.56,-151.77 84.92,-134.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;process1&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-36 0,-36 0,0 54,0 54,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"6\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<!-- (&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;store1&#39;, &#39;process2&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"189,-36 135,-36 135,0 189,0 189,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"141\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<!-- (&#39;process3&#39;,) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;process3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"281,-129.32 227,-129.32 227,-93.32 281,-93.32 281,-129.32\"/>\n",
       "<text text-anchor=\"start\" x=\"233\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">process3</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;,)&#45;&gt;(&#39;process3&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;,)&#45;&gt;(&#39;process3&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M166.94,-200.45C181.26,-192.6 199.26,-181.44 213,-168.63 223.8,-158.57 233.61,-145.31 240.93,-134.18\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"242.16\" cy=\"-132.29\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"222\" y=\"-160.63\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M120.63,-92.71C101.42,-77.59 73.69,-55.75 53.39,-39.78\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"51.66\" cy=\"-38.41\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"85\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M148.7,-82.92C151.44,-69.75 154.69,-54.15 157.29,-41.66\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"157.77\" cy=\"-39.31\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"155\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M46.64,-92.11C39.32,-84.55 31.97,-75.17 28,-65 25.11,-57.61 24.31,-49.02 24.42,-41.28\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"24.5\" cy=\"-39.09\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"28\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;store1&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;store1&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M87.93,-90.95C103.27,-76.06 124.43,-55.5 140.2,-40.17\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"141.89\" cy=\"-38.53\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"124\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x10837e8e0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nested_composite_spec = {\n",
    "    'store1': {\n",
    "        'store1.1': {\n",
    "            '_value': 1.1, \n",
    "            '_type': 'float',\n",
    "        },\n",
    "        'store1.2': {\n",
    "            '_value': 2, \n",
    "            '_type': 'int',\n",
    "        },\n",
    "        'process1': {\n",
    "            # '_ports': {\n",
    "            #     'port1': 'type', \n",
    "            #     'port2': 'type',\n",
    "            # },\n",
    "            '_wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            # '_ports': {\n",
    "            #     'port1': 'type', \n",
    "            #     'port2': 'type',\n",
    "            # },\n",
    "            '_wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "    },\n",
    "    'process3': {\n",
    "        '_wires': {\n",
    "            'port1': 'store1',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(nested_composite_spec, **plot_settings, filename='nested_composite')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c8c1704-2331-44eb-bd05-72e36ccb6858",
   "metadata": {},
   "source": [
    "### Composite process\n",
    "Composite processes are processes with internal stores and internal processes, which can run as their own simulation. The composite process has schema, allowing it to connect to a super-simulation. This allows for improved distributed computation, with composite processes able to run on their own computer and syncronize to the super-simulation via message passing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "638c5dc7-4cfd-46e4-b66a-8c42ca98ff1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"201pt\" height=\"313pt\"\n",
       " viewBox=\"0.00 0.00 206.62 321.63\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 317.63)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-317.63 202.62,-317.63 202.62,4 -4,4\"/>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"124.82\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"105.82\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"192.82,-36 138.82,-36 138.82,0 192.82,0 192.82,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"144.82\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M136.24,-84.88C142.41,-71.13 149.96,-54.31 155.89,-41.11\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"156.81\" cy=\"-39.07\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"148.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"57.82,-36 3.82,-36 3.82,0 57.82,0 57.82,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"9.82\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.1&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M116.64,-83.46C112.53,-73.31 106.72,-62.24 98.82,-54 89,-43.75 75.53,-35.96 63.17,-30.36\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"60.96\" cy=\"-29.4\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"106.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"49.82\" cy=\"-111.32\" rx=\"28.14\" ry=\"28.14\"/>\n",
       "<text text-anchor=\"start\" x=\"30.82\" y=\"-108.22\" font-family=\"Times,serif\" font-size=\"12.00\">store1.2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M55.99,-83.01C59.47,-72.8 64.8,-61.81 72.82,-54 89.38,-37.9 114.05,-29.06 133.82,-24.3\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"135.78\" cy=\"-23.85\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"72.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;store1.2&#39;)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M37.9,-84.92C35.43,-78.55 33.17,-71.64 31.82,-65 30.27,-57.35 29.76,-48.84 29.72,-41.24\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"29.73\" cy=\"-39.14\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"31.82\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;composite_process&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"150.82,-226.63 52.82,-226.63 52.82,-190.63 150.82,-190.63 150.82,-226.63\"/>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"154.82,-230.63 48.82,-230.63 48.82,-186.63 154.82,-186.63 154.82,-230.63\"/>\n",
       "<text text-anchor=\"start\" x=\"56.82\" y=\"-205.53\" font-family=\"Times,serif\" font-size=\"12.00\">composite_process</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M154.92,-185.85C161.25,-181.1 166.87,-175.41 170.82,-168.63 179.85,-153.15 164.71,-137.52 149.44,-126.66\"/>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M108.16,-181.36C111.29,-168.4 115.06,-152.77 118.22,-139.67\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"107.63\" cy=\"-183.57\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"112.82\" y=\"-160.63\" font-family=\"Times,serif\" font-size=\"10.00\">tunnel port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M90.02,-186.01C82.05,-171.39 71.51,-152.08 63.22,-136.88\"/>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;store1.2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M44.17,-190.03C34.89,-184.64 26.49,-177.65 20.82,-168.63 13.39,-156.81 20.47,-142.86 29.49,-131.8\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"46\" cy=\"-191.03\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"20.82\" y=\"-160.63\" font-family=\"Times,serif\" font-size=\"10.00\">tunnel port2</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process1&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;,)&#45;&gt;(&#39;composite_process&#39;, &#39;process2&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port1&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port1&#39;)&#45;&gt;(&#39;composite_process&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;port1&#39;)&#45;&gt;(&#39;composite_process&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M78.33,-278.2C77.98,-269.3 78.29,-258.21 80.82,-248.63 81.94,-244.37 83.62,-240.07 85.53,-235.95\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"86.59\" cy=\"-233.81\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"80.82\" y=\"-251.63\" font-family=\"Times,serif\" font-size=\"10.00\">port1</text>\n",
       "</g>\n",
       "<!-- (&#39;composite_process&#39;, &#39;port2&#39;) -->\n",
       "<!-- (&#39;composite_process&#39;, &#39;port2&#39;)&#45;&gt;(&#39;composite_process&#39;,) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;composite_process&#39;, &#39;port2&#39;)&#45;&gt;(&#39;composite_process&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M123.45,-289.57C120.67,-279.29 114.19,-255.33 109.04,-236.31\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"108.42\" cy=\"-234.01\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"113.82\" y=\"-251.63\" font-family=\"Times,serif\" font-size=\"10.00\">port2</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x108355eb0>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "composite_process_spec = {\n",
    "    'composite_process': {\n",
    "        'store1.1': {\n",
    "            '_value': 1.1, \n",
    "            '_type': 'float',\n",
    "        },\n",
    "        'store1.2': {\n",
    "            '_value': 2, \n",
    "            '_type': 'int',\n",
    "        },\n",
    "        'process1': {\n",
    "            # '_ports': {\n",
    "            #     'port1': 'type', \n",
    "            #     'port2': 'type',\n",
    "            # },\n",
    "            '_wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        'process2': {\n",
    "            # '_ports': {\n",
    "            #     'port1': 'type', \n",
    "            #     'port2': 'type',\n",
    "            # },\n",
    "            '_wires': {\n",
    "                'port1': 'store1.1',\n",
    "                'port2': 'store1.2',\n",
    "            }\n",
    "        },\n",
    "        '_ports': {\n",
    "            'port1': 'type', \n",
    "            'port2': 'type', \n",
    "        },\n",
    "        '_tunnels': {\n",
    "            'port1': 'store1.1',\n",
    "            'port2': 'store1.2',\n",
    "        }\n",
    "    }\n",
    "}\n",
    "plot_bigraph(composite_process_spec, **plot_settings, filename='composite_process')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "847cef26-65a1-4da6-9dd3-eb5839277970",
   "metadata": {},
   "source": [
    "## Orchestration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e76c97a8-3037-464a-929e-ba4414c73c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_settings2={}\n",
    "if save_images:\n",
    "    plot_settings2.update({\n",
    "        'out_dir': 'out',\n",
    "        'dpi': '250'\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48975890-06a4-462a-9b1d-6bddd51779fe",
   "metadata": {},
   "source": [
    "### Multi-timestepping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f0eff7b0-f0e2-4d56-8b16-ef0fc1ce5aba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: flow Pages: 1 -->\n",
       "<svg width=\"231pt\" height=\"165pt\"\n",
       " viewBox=\"0.00 0.00 238.00 170.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 166)\">\n",
       "<title>flow</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-166 234,-166 234,4 -4,4\"/>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"65.01,-38 0.01,-38 0.01,0 65.01,0 65.01,-38\"/>\n",
       "<text text-anchor=\"start\" x=\"8.01\" y=\"-22.8\" font-family=\"Times,serif\" font-size=\"14.00\">temporal</text>\n",
       "<text text-anchor=\"start\" x=\"8.51\" y=\"-7.8\" font-family=\"Times,serif\" font-size=\"14.00\">process2</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"131.43\" cy=\"-19\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"131.43\" y=\"-16.8\" font-family=\"Times,serif\" font-size=\"9.00\">1.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M66,-19C85.73,-19 109.26,-19 122.04,-19\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"213.87\" cy=\"-19\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"213.87\" y=\"-16.8\" font-family=\"Times,serif\" font-size=\"9.00\">2.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M140.89,-19C156.79,-19 188.7,-19 204.52,-19\"/>\n",
       "</g>\n",
       "<!-- (&#39;state&#39;,) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;state&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"161.54\" cy=\"-90.61\" rx=\"21.4\" ry=\"21.4\"/>\n",
       "<text text-anchor=\"start\" x=\"149.04\" y=\"-86.91\" font-family=\"Times,serif\" font-size=\"14.00\">state</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 1.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M137.16,-32.62C140.99,-41.74 146.16,-54.02 150.75,-64.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"138.85,-32.13 135.3,-28.19 135.62,-33.48 138.85,-32.13\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"149.1,-65.54 152.65,-69.47 152.33,-64.19 149.1,-65.54\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) end -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) end</title>\n",
       "<ellipse fill=\"none\" stroke=\"none\" stroke-width=\"2\" cx=\"230.36\" cy=\"-19\" rx=\"0\" ry=\"0\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) end -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process2&#39;,) end</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M223.27,-19C223.43,-19 223.59,-19 223.74,-19\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"221.43,-20.75 226.43,-19 221.43,-17.25 221.43,-20.75\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process2&#39;,) 2.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M205.18,-30.89C197.74,-41.07 186.84,-55.99 177.79,-68.38\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"206.48,-32.07 208.02,-27.01 203.66,-30.01 206.48,-32.07\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"176.53,-67.13 175,-72.2 179.36,-69.19 176.53,-67.13\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"65.01,-161.65 0.01,-161.65 0.01,-123.65 65.01,-123.65 65.01,-161.65\"/>\n",
       "<text text-anchor=\"start\" x=\"8.01\" y=\"-146.45\" font-family=\"Times,serif\" font-size=\"14.00\">temporal</text>\n",
       "<text text-anchor=\"start\" x=\"8.51\" y=\"-131.45\" font-family=\"Times,serif\" font-size=\"14.00\">process1</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"90.22\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"90.22\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">0.5</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,)&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M65.75,-142.65C71.36,-142.65 76.74,-142.65 81.03,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"131.43\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"131.43\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">1.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M99.36,-142.65C106.1,-142.65 115.23,-142.65 122.03,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 0.5&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M102.33,-133.81C112.43,-126.45 127.09,-115.75 139.29,-106.85\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"101.47,-132.28 98.46,-136.64 103.53,-135.1 101.47,-132.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"140.04,-108.46 143.05,-104.1 137.98,-105.64 140.04,-108.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"172.65\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"172.65\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">1.5</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M140.58,-142.65C147.32,-142.65 156.45,-142.65 163.24,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M138.88,-129.78C141.56,-125.14 144.68,-119.75 147.73,-114.48\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"137.51,-128.66 136.52,-133.87 140.54,-130.41 137.51,-128.66\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"149.13,-115.56 150.12,-110.35 146.1,-113.8 149.13,-115.56\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0</title>\n",
       "<ellipse fill=\"white\" stroke=\"none\" stroke-width=\"2\" cx=\"213.87\" cy=\"-142.65\" rx=\"8.47\" ry=\"8.47\"/>\n",
       "<text text-anchor=\"middle\" x=\"213.87\" y=\"-140.45\" font-family=\"Times,serif\" font-size=\"9.00\">2.0</text>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M181.8,-142.65C188.53,-142.65 197.67,-142.65 204.46,-142.65\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 1.5&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M169.55,-128.14C168.86,-124.88 168.1,-121.32 167.33,-117.73\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"167.8,-128.32 170.56,-132.85 171.22,-127.59 167.8,-128.32\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"169.04,-117.37 166.29,-112.85 165.62,-118.1 169.04,-117.37\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) end -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) end</title>\n",
       "<ellipse fill=\"none\" stroke=\"none\" stroke-width=\"2\" cx=\"230.36\" cy=\"-142.65\" rx=\"0\" ry=\"0\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) end -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;temporal&lt;br/&gt;process1&#39;,) end</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M223.27,-142.65C223.43,-142.65 223.59,-142.65 223.74,-142.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"221.43,-144.4 226.43,-142.65 221.43,-140.9 221.43,-144.4\"/>\n",
       "</g>\n",
       "<!-- (&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;state&#39;,) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;temporal&lt;br/&gt;process1&#39;,) 2.0&#45;&gt;(&#39;state&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.5\" stroke-dasharray=\"5,2\" d=\"M203.57,-132.41C197.06,-125.94 188.45,-117.38 180.72,-109.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"202.14,-133.45 206.92,-135.74 204.6,-130.97 202.14,-133.45\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.5\" points=\"182.33,-108.82 177.55,-106.53 179.86,-111.3 182.33,-108.82\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x10837eb80>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multitimestep_spec = {\n",
    "    'temporal<br/>process2': {\n",
    "        '_ports': {'port1': 'type'},\n",
    "        '_wires': {'port1': 'state'},\n",
    "        '_sync_step': 1.0,\n",
    "    },\n",
    "    'temporal<br/>process1': {\n",
    "        '_ports': {'port1': 'type'},\n",
    "        '_wires': {'port1': 'state'},\n",
    "        '_sync_step': 0.5,\n",
    "    },\n",
    "}\n",
    "plot_multitimestep(multitimestep_spec, total_time=2.0, **plot_settings2, filename='multitimestep')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a00cf677-ebac-4309-8ac9-a5b2855724af",
   "metadata": {},
   "source": [
    "### Flows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "db45c99b-b97b-47bc-a73e-c669882f3969",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: flow Pages: 1 -->\n",
       "<svg width=\"130pt\" height=\"183pt\"\n",
       " viewBox=\"0.00 0.00 134.00 188.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 184)\">\n",
       "<title>flow</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-184 130,-184 130,4 -4,4\"/>\n",
       "<!-- (&#39;step1&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;step1&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-180 0,-180 0,-144 54,-144 54,-180\"/>\n",
       "<text text-anchor=\"start\" x=\"12.5\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">step1</text>\n",
       "</g>\n",
       "<!-- (&#39;step2&#39;,) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;step2&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"54,-108 0,-108 0,-72 54,-72 54,-108\"/>\n",
       "<text text-anchor=\"start\" x=\"12.5\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">step2</text>\n",
       "</g>\n",
       "<!-- (&#39;step1&#39;,)&#45;&gt;(&#39;step2&#39;,) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;step1&#39;,)&#45;&gt;(&#39;step2&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M27,-143.34C27,-136.75 27,-129.08 27,-121.67\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"30.5,-121.93 27,-111.93 23.5,-121.93 30.5,-121.93\"/>\n",
       "</g>\n",
       "<!-- (&#39;step4&#39;,) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;step4&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"90,-36 36,-36 36,0 90,0 90,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"48.5\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">step4</text>\n",
       "</g>\n",
       "<!-- (&#39;step2&#39;,)&#45;&gt;(&#39;step4&#39;,) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;step2&#39;,)&#45;&gt;(&#39;step4&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M36.08,-71.34C39.68,-64.34 43.91,-56.12 47.94,-48.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"50.95,-50.09 52.41,-39.59 44.72,-46.89 50.95,-50.09\"/>\n",
       "</g>\n",
       "<!-- (&#39;step3&#39;,) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;step3&#39;,)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"126,-108 72,-108 72,-72 126,-72 126,-108\"/>\n",
       "<text text-anchor=\"start\" x=\"84.5\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">step3</text>\n",
       "</g>\n",
       "<!-- (&#39;step3&#39;,)&#45;&gt;(&#39;step4&#39;,) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;step3&#39;,)&#45;&gt;(&#39;step4&#39;,)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M89.92,-71.34C86.32,-64.34 82.09,-56.12 78.06,-48.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2\" points=\"81.28,-46.89 73.59,-39.59 75.05,-50.09 81.28,-46.89\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1083940d0>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_schema = {\n",
    "    '_type': 'step_process',\n",
    "    '_ports': {\n",
    "        'port1': 'type',\n",
    "        'port2': 'type'\n",
    "    }\n",
    "}\n",
    "\n",
    "flow_spec = {\n",
    "    'step1': {\n",
    "        '_depends_on': [], \n",
    "        **process_schema\n",
    "    },\n",
    "    'step2': {\n",
    "        '_depends_on': 'step1', \n",
    "        **process_schema\n",
    "    },\n",
    "    'step3': {\n",
    "        '_depends_on': [], \n",
    "        **process_schema\n",
    "    },\n",
    "    'step4': {\n",
    "        '_depends_on': ['step2', 'step3'], \n",
    "        **process_schema\n",
    "    },\n",
    "}\n",
    "plot_flow(flow_spec, **plot_settings2, filename='flow')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5882581a-c5ef-40b5-acbd-efddad269807",
   "metadata": {},
   "source": [
    "## Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae52f73c-e384-4880-b18f-47b4c25ce203",
   "metadata": {},
   "source": [
    "### Cell structure and function\n",
    "Bigraphs can represent the hierarchical nesting of cellular structures, and processes represent their functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b39417ce-d9a1-4985-abce-ce22df816897",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/cell\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"564pt\" height=\"540pt\"\n",
       " viewBox=\"0.00 0.00 580.26 554.97\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.972222 0.972222) rotate(0) translate(4 550.97)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-550.97 576.26,-550.97 576.26,4 -4,4\"/>\n",
       "<!-- (&#39;cell&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;cell&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"346.78\" cy=\"-524.97\" rx=\"18\" ry=\"18\"/>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"346.78\" cy=\"-524.97\" rx=\"22\" ry=\"22\"/>\n",
       "<text text-anchor=\"start\" x=\"338.28\" y=\"-521.87\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"141.78\" cy=\"-430.27\" rx=\"35.91\" ry=\"35.91\"/>\n",
       "<text text-anchor=\"start\" x=\"116.78\" y=\"-427.17\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M326.38,-514.75C290.93,-498.72 217.68,-465.59 174.88,-446.24\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"346.78\" cy=\"-430.27\" rx=\"35.18\" ry=\"35.18\"/>\n",
       "<text text-anchor=\"start\" x=\"322.28\" y=\"-427.17\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M346.78,-502.01C346.78,-491.3 346.78,-478.16 346.78,-466.13\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;) -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"468.78\" cy=\"-430.27\" rx=\"30.33\" ry=\"30.33\"/>\n",
       "<text text-anchor=\"start\" x=\"448.28\" y=\"-427.17\" font-family=\"Times,serif\" font-size=\"12.00\">nucleoid</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M364.52,-510.49C385.67,-494.43 421.11,-467.5 444.72,-449.56\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"38.78\" cy=\"-315.09\" rx=\"38.57\" ry=\"38.57\"/>\n",
       "<text text-anchor=\"start\" x=\"11.28\" y=\"-311.99\" font-family=\"Times,serif\" font-size=\"12.00\">transporters</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M117.63,-402.73C101.7,-385.22 80.84,-362.31 64.58,-344.44\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"116.78\" cy=\"-315.09\" rx=\"20.86\" ry=\"20.86\"/>\n",
       "<text text-anchor=\"start\" x=\"103.78\" y=\"-311.99\" font-family=\"Times,serif\" font-size=\"12.00\">lipids</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;lipids&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M134.08,-394.39C129.86,-375.29 124.78,-352.31 121.22,-336.2\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"223.28,-136.09 142.28,-136.09 142.28,-100.09 223.28,-100.09 223.28,-136.09\"/>\n",
       "<text text-anchor=\"start\" x=\"146.28\" y=\"-121.49\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"161.78\" y=\"-108.49\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transporters&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M49.22,-277.17C59.35,-246.27 77.12,-202.36 103.78,-171.18 114.52,-158.62 129.02,-147.88 142.66,-139.48\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"144.72\" cy=\"-138.24\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"103.78\" y=\"-210.4\" font-family=\"Times,serif\" font-size=\"10.00\">transporters</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"231.78\" cy=\"-315.09\" rx=\"38.33\" ry=\"38.33\"/>\n",
       "<text text-anchor=\"start\" x=\"204.78\" y=\"-311.99\" font-family=\"Times,serif\" font-size=\"12.00\">metabolites</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M321.54,-404.43C303.03,-386.21 277.92,-361.5 258.97,-342.85\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"334.78\" cy=\"-212.9\" rx=\"41.94\" ry=\"41.94\"/>\n",
       "<text text-anchor=\"start\" x=\"311.28\" y=\"-216.3\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"309.28\" y=\"-203.3\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M344.83,-394.26C344.18,-382.59 343.45,-369.52 342.78,-357.57 340.87,-323.27 338.7,-284.37 337.11,-255.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"391.78\" cy=\"-315.09\" rx=\"39.7\" ry=\"39.7\"/>\n",
       "<text text-anchor=\"start\" x=\"369.78\" y=\"-324.99\" font-family=\"Times,serif\" font-size=\"12.00\">transcript</text>\n",
       "<text text-anchor=\"start\" x=\"367.78\" y=\"-311.99\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"371.28\" y=\"-298.99\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M359.8,-396.53C365.25,-382.82 371.61,-366.83 377.22,-352.73\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" stroke-width=\"2\" points=\"425.28,-36 368.28,-36 368.28,0 425.28,0 425.28,-36\"/>\n",
       "<text text-anchor=\"start\" x=\"372.28\" y=\"-14.9\" font-family=\"Times,serif\" font-size=\"12.00\">translation</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;metabolites&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M211.58,-281.71C207.04,-273.17 202.71,-263.77 199.78,-254.62 187.55,-216.34 184.07,-169.35 183.1,-141.96\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"183.03\" cy=\"-139.58\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"199.78\" y=\"-210.4\" font-family=\"Times,serif\" font-size=\"10.00\">internal</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;ribosomal&lt;br/&gt;complexes&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M330.38,-170.38C329.21,-144.37 330.67,-110.66 341.78,-83 348.32,-66.74 360.75,-51.7 372.15,-40.29\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"373.86\" cy=\"-38.62\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"341.78\" y=\"-115.59\" font-family=\"Times,serif\" font-size=\"10.00\">p1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;) -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"396.78\" cy=\"-118.09\" rx=\"35.18\" ry=\"35.18\"/>\n",
       "<text text-anchor=\"start\" x=\"372.28\" y=\"-114.99\" font-family=\"Times,serif\" font-size=\"12.00\">transcripts</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M392.79,-274.63C393.7,-239.35 395.01,-188.13 395.88,-154.24\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;) -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcript&lt;br/&gt;regulation&lt;br/&gt;complex&#39;, &#39;transcripts&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;translation&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M396.78,-81.95C396.78,-68.58 396.78,-53.71 396.78,-41.77\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"396.78\" cy=\"-39.35\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"396.78\" y=\"-57\" font-family=\"Times,serif\" font-size=\"10.00\">p2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"529.78\" cy=\"-315.09\" rx=\"42.45\" ry=\"42.45\"/>\n",
       "<text text-anchor=\"start\" x=\"499.28\" y=\"-311.99\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M483.08,-402.73C491.04,-387.97 501.07,-369.36 509.8,-353.17\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"529.78\" cy=\"-118.09\" rx=\"22.32\" ry=\"22.32\"/>\n",
       "<text text-anchor=\"start\" x=\"515.78\" y=\"-114.99\" font-family=\"Times,serif\" font-size=\"12.00\">genes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleoid&#39;, &#39;chromosome&#39;, &#39;genes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M529.78,-271.78C529.78,-231.68 529.78,-173 529.78,-141.34\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;, &#39;external&#39;) -->\n",
       "<!-- (&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;, &#39;external&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;) -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;, &#39;external&#39;)&#45;&gt;(&#39;cell&#39;, &#39;membrane&#39;, &#39;transmembrane&lt;br/&gt;transport&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M287.53,-296.28C286.25,-288.43 283.73,-279.46 278.78,-272.62 269.89,-260.32 258.52,-267.02 249.78,-254.62 227.93,-223.61 249.42,-204.76 231.78,-171.18 225.71,-159.63 216.37,-148.78 207.46,-140.01\"/>\n",
       "<ellipse fill=\"black\" stroke=\"black\" cx=\"206.01\" cy=\"-138.61\" rx=\"2\" ry=\"2\"/>\n",
       "<text text-anchor=\"start\" x=\"249.78\" y=\"-210.4\" font-family=\"Times,serif\" font-size=\"10.00\">external</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1083949d0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_struct_func = {\n",
    "        'cell': {\n",
    "            'membrane': {\n",
    "                'transporters': 'concentrations',\n",
    "                'lipids': 'concentrations',\n",
    "                'transmembrane<br/>transport': {\n",
    "                    '_value': {\n",
    "                        '_process': 'transport URI',\n",
    "                        '_config': {'parameter': 1}\n",
    "                    },\n",
    "                    '_wires': {\n",
    "                        'transporters': 'transporters',\n",
    "                        'internal': ['..', 'cytoplasm', 'metabolites']},\n",
    "                    '_ports': {\n",
    "                        'transporters': 'concentrations',\n",
    "                        'internal': 'concentrations',\n",
    "                        'external': 'concentrations'\n",
    "                    }\n",
    "                }\n",
    "            },\n",
    "            'cytoplasm': {\n",
    "                'metabolites': {\n",
    "                    '_value': 1.1,\n",
    "                    '_type': 'concentrations'\n",
    "                },\n",
    "                'ribosomal<br/>complexes': {\n",
    "                    '_value': 2.2,\n",
    "                    '_type': 'concentrations'\n",
    "                },\n",
    "                'transcript<br/>regulation<br/>complex': {\n",
    "                    '_value': 0.01,\n",
    "                    '_type': 'concentrations',\n",
    "                    'transcripts': {\n",
    "                        '_value': 0.1,\n",
    "                        '_type': 'concentrations'\n",
    "                    }\n",
    "                },\n",
    "                'translation': {\n",
    "                    '_wires': {\n",
    "                        'p1': 'ribosomal<br/>complexes',\n",
    "                        'p2': ['transcript<br/>regulation<br/>complex', 'transcripts']}}},\n",
    "            'nucleoid': {\n",
    "                'chromosome': {\n",
    "                    'genes': 'sequences'\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "plot_bigraph(cell_struct_func, remove_process_place_edges=True, out_dir='out', filename='cell')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff81178f-36e4-4200-982a-ed8155b87934",
   "metadata": {},
   "source": [
    "## Multiscale map of a cell\n",
    "\n",
    "We here plot the multiscale map of a cell from [Qin et al. A multi-scale map of cell structure fusing protein images and interactions. Nature. 2021.](https://pubmed.ncbi.nlm.nih.gov/34819669/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "47bdf8e5-0bdd-4cac-b57e-2449c75fa178",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 7.1.0 (20230121.1956)\n",
       " -->\n",
       "<!-- Title: bigraph Pages: 1 -->\n",
       "<svg width=\"579pt\" height=\"700pt\"\n",
       " viewBox=\"0.00 0.00 595.06 720.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.54369 0.54369) rotate(0) translate(4 1283.5)\">\n",
       "<title>bigraph</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-1283.5 1060.09,-1283.5 1060.09,4 -4,4\"/>\n",
       "<!-- (&#39;cell&#39;,) -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>(&#39;cell&#39;,)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"423.28\" cy=\"-1257.5\" rx=\"18\" ry=\"18\"/>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"423.28\" cy=\"-1257.5\" rx=\"22\" ry=\"22\"/>\n",
       "<text text-anchor=\"start\" x=\"414.78\" y=\"-1254.4\" font-family=\"Times,serif\" font-size=\"12.00\">cell</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"286.28\" cy=\"-1164.41\" rx=\"35.18\" ry=\"35.18\"/>\n",
       "<text text-anchor=\"start\" x=\"261.78\" y=\"-1161.31\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasm</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;) -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M404.8,-1244.21C381.94,-1229.01 342.73,-1202.95 315.71,-1184.98\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"654.28\" cy=\"-1164.41\" rx=\"27.9\" ry=\"27.9\"/>\n",
       "<text text-anchor=\"start\" x=\"635.78\" y=\"-1161.31\" font-family=\"Times,serif\" font-size=\"12.00\">nucleus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;) -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;,)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M444.18,-1248.25C486.39,-1231.61 581.67,-1194.04 628,-1175.77\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"115.28\" cy=\"-1040.29\" rx=\"40.11\" ry=\"40.11\"/>\n",
       "<text text-anchor=\"start\" x=\"93.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">secretory</text>\n",
       "<text text-anchor=\"start\" x=\"90.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M255.25,-1146.46C231.73,-1133.11 199,-1113.45 172.28,-1093.32 162.64,-1086.07 152.74,-1077.48 143.94,-1069.4\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"227.28\" cy=\"-1040.29\" rx=\"45.92\" ry=\"45.92\"/>\n",
       "<text text-anchor=\"start\" x=\"198.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"202.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M271.08,-1131.96C263.78,-1116.84 254.92,-1098.52 247.09,-1082.31\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"344.28\" cy=\"-1040.29\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"310.78\" y=\"-1056.69\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"320.28\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"323.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"328.78\" y=\"-1017.69\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M301.36,-1131.64C307.54,-1118.64 314.82,-1103.31 321.61,-1089.01\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"89.28\" cy=\"-890.44\" rx=\"57.05\" ry=\"57.05\"/>\n",
       "<text text-anchor=\"start\" x=\"52.78\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"68.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"70.78\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M108.38,-1000.06C105.55,-983.98 102.22,-965.06 99.14,-947.56\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"57.28\" cy=\"-728.58\" rx=\"57.05\" ry=\"57.05\"/>\n",
       "<text text-anchor=\"start\" x=\"49.28\" y=\"-744.98\" font-family=\"Times,serif\" font-size=\"12.00\">ion</text>\n",
       "<text text-anchor=\"start\" x=\"20.78\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">transmembrane</text>\n",
       "<text text-anchor=\"start\" x=\"36.28\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">transport</text>\n",
       "<text text-anchor=\"start\" x=\"38.78\" y=\"-705.98\" font-family=\"Times,serif\" font-size=\"12.00\">systems</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;) -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;secretory&lt;br/&gt;organelles&#39;, &#39;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;, &#39;ion&lt;br/&gt;transmembrane&lt;br/&gt;transport&lt;br/&gt;systems&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M78.08,-833.53C74.97,-817.95 71.58,-801.04 68.46,-785.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"227.28\" cy=\"-890.44\" rx=\"45.92\" ry=\"45.92\"/>\n",
       "<text text-anchor=\"start\" x=\"204.78\" y=\"-906.84\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"222.28\" y=\"-893.84\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"198.78\" y=\"-880.84\" font-family=\"Times,serif\" font-size=\"12.00\">cytoplasmic</text>\n",
       "<text text-anchor=\"start\" x=\"202.78\" y=\"-867.84\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M227.28,-993.72C227.28,-975.74 227.28,-955.11 227.28,-937.11\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"173.28\" cy=\"-728.58\" rx=\"40.11\" ry=\"40.11\"/>\n",
       "<text text-anchor=\"start\" x=\"150.28\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"148.78\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M212.66,-846.18C204.35,-821.59 194.05,-791.07 186.06,-767.43\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"296.28\" cy=\"-728.58\" rx=\"65.11\" ry=\"65.11\"/>\n",
       "<text text-anchor=\"start\" x=\"254.28\" y=\"-738.48\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleoprotein</text>\n",
       "<text text-anchor=\"start\" x=\"275.78\" y=\"-725.48\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"280.78\" y=\"-712.48\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M245.4,-847.44C253.07,-829.68 262.14,-808.67 270.48,-789.34\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"69.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"46.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">subgroup</text>\n",
       "<text text-anchor=\"start\" x=\"64.28\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"46.28\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"44.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">organelles</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;) -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;metabolic&lt;br/&gt;organelles&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M150.19,-694.94C133.44,-671.25 110.81,-639.24 93.74,-615.11\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"177.28\" cy=\"-579.51\" rx=\"48.03\" ry=\"48.03\"/>\n",
       "<text text-anchor=\"start\" x=\"142.28\" y=\"-576.41\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondrion</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;) -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;metabolic&lt;br/&gt;organelles&#39;, &#39;mitochondrion&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M174.36,-687.78C174.86,-669.49 175.45,-647.65 175.97,-628.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"292.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"278.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"268.28\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">processng</text>\n",
       "<text text-anchor=\"start\" x=\"271.78\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"288.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M294.51,-662.64C294.14,-649.1 293.76,-635.16 293.42,-622.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"396.28\" cy=\"-579.51\" rx=\"43.68\" ry=\"43.68\"/>\n",
       "<text text-anchor=\"start\" x=\"374.28\" y=\"-589.41\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "<text text-anchor=\"start\" x=\"371.28\" y=\"-576.41\" font-family=\"Times,serif\" font-size=\"12.00\">biogenesis</text>\n",
       "<text text-anchor=\"start\" x=\"369.28\" y=\"-563.41\" font-family=\"Times,serif\" font-size=\"12.00\">community</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;) -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M332.94,-673.65C345.88,-654.63 360.09,-633.73 371.77,-616.55\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"292.28\" cy=\"-453.07\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"278.78\" y=\"-469.47\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"273.78\" y=\"-456.47\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"271.78\" y=\"-443.47\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"288.78\" y=\"-430.47\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processng&lt;br/&gt;complex&lt;br/&gt;1&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M292.28,-536.38C292.28,-523.41 292.28,-509.13 292.28,-496.16\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"390.28\" cy=\"-453.07\" rx=\"32.02\" ry=\"32.02\"/>\n",
       "<text text-anchor=\"start\" x=\"368.28\" y=\"-449.97\" font-family=\"Times,serif\" font-size=\"12.00\">ribosome</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;) -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M394.18,-534.99C393.4,-518.87 392.53,-500.9 391.82,-486\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"482.28\" cy=\"-453.07\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"465.28\" y=\"-469.47\" font-family=\"Times,serif\" font-size=\"12.00\">nucleic</text>\n",
       "<text text-anchor=\"start\" x=\"472.28\" y=\"-456.47\" font-family=\"Times,serif\" font-size=\"12.00\">acid</text>\n",
       "<text text-anchor=\"start\" x=\"463.78\" y=\"-443.47\" font-family=\"Times,serif\" font-size=\"12.00\">binding</text>\n",
       "<text text-anchor=\"start\" x=\"461.78\" y=\"-430.47\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;nucleic&lt;br/&gt;acid&lt;br/&gt;binding&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M421.15,-542.52C432.86,-525.58 446.76,-505.46 458.34,-488.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"390.28\" cy=\"-335.75\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"366.78\" y=\"-339.15\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"372.78\" y=\"-326.15\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M390.28,-420.25C390.28,-406.39 390.28,-390 390.28,-375.44\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"163.28\" cy=\"-200.05\" rx=\"52.15\" ry=\"52.15\"/>\n",
       "<text text-anchor=\"start\" x=\"130.28\" y=\"-216.45\" font-family=\"Times,serif\" font-size=\"12.00\">mitochondiral</text>\n",
       "<text text-anchor=\"start\" x=\"151.28\" y=\"-203.45\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n",
       "<text text-anchor=\"start\" x=\"139.78\" y=\"-190.45\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"145.78\" y=\"-177.45\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;) -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M353.04,-322.52C318.05,-310.05 265.1,-288.62 224.28,-260.86 215.32,-254.77 206.52,-247.29 198.5,-239.71\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"272.28\" cy=\"-200.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"249.28\" y=\"-209.95\" font-family=\"Times,serif\" font-size=\"12.00\">mito&#45;cyto</text>\n",
       "<text text-anchor=\"start\" x=\"248.78\" y=\"-196.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"256.78\" y=\"-183.95\" font-family=\"Times,serif\" font-size=\"12.00\">cluster</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;) -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mito&#45;cyto&lt;br/&gt;ribosomal&lt;br/&gt;cluster&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M363.13,-307.18C349.93,-293.5 333.98,-276.57 320.28,-260.86 312.14,-251.53 303.58,-241.08 295.97,-231.57\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"390.28\" cy=\"-200.05\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"355.78\" y=\"-229.45\" font-family=\"Times,serif\" font-size=\"12.00\">cotranslational</text>\n",
       "<text text-anchor=\"start\" x=\"373.28\" y=\"-216.45\" font-family=\"Times,serif\" font-size=\"12.00\">protein</text>\n",
       "<text text-anchor=\"start\" x=\"369.28\" y=\"-203.45\" font-family=\"Times,serif\" font-size=\"12.00\">targeting</text>\n",
       "<text text-anchor=\"start\" x=\"385.28\" y=\"-190.45\" font-family=\"Times,serif\" font-size=\"12.00\">to</text>\n",
       "<text text-anchor=\"start\" x=\"365.28\" y=\"-177.45\" font-family=\"Times,serif\" font-size=\"12.00\">membrane</text>\n",
       "<text text-anchor=\"start\" x=\"373.78\" y=\"-164.45\" font-family=\"Times,serif\" font-size=\"12.00\">system</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;) -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;cotranslational&lt;br/&gt;protein&lt;br/&gt;targeting&lt;br/&gt;to&lt;br/&gt;membrane&lt;br/&gt;system&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M390.28,-296.08C390.28,-285.28 390.28,-273.28 390.28,-261.55\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"508.28\" cy=\"-200.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"484.78\" y=\"-209.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"487.78\" y=\"-196.95\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"504.78\" y=\"-183.95\" font-family=\"Times,serif\" font-size=\"12.00\">5</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;) -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;5&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M417.42,-307.18C430.63,-293.5 446.57,-276.57 460.28,-260.86 468.42,-251.53 476.98,-241.08 484.58,-231.57\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"604.28\" cy=\"-200.05\" rx=\"38.78\" ry=\"38.78\"/>\n",
       "<text text-anchor=\"start\" x=\"580.78\" y=\"-209.95\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"583.78\" y=\"-196.95\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"600.78\" y=\"-183.95\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;ribosomal&lt;br/&gt;complex&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M427.95,-323.71C463.5,-312.04 517.05,-291.09 556.28,-260.86 566.19,-253.22 575.29,-243.13 582.85,-233.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"163.28\" cy=\"-51.62\" rx=\"51.74\" ry=\"51.74\"/>\n",
       "<text text-anchor=\"start\" x=\"151.28\" y=\"-74.52\" font-family=\"Times,serif\" font-size=\"12.00\">large</text>\n",
       "<text text-anchor=\"start\" x=\"139.78\" y=\"-61.52\" font-family=\"Times,serif\" font-size=\"12.00\">ribosomal</text>\n",
       "<text text-anchor=\"start\" x=\"145.78\" y=\"-48.52\" font-family=\"Times,serif\" font-size=\"12.00\">subunit</text>\n",
       "<text text-anchor=\"start\" x=\"134.28\" y=\"-35.52\" font-family=\"Times,serif\" font-size=\"12.00\">subcomplex</text>\n",
       "<text text-anchor=\"start\" x=\"159.78\" y=\"-22.52\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;subgroup&lt;br/&gt;of&lt;br/&gt;cytoplasmic&lt;br/&gt;organelles&#39;, &#39;ribonucleoprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;ribosome&lt;br/&gt;biogenesis&lt;br/&gt;community&#39;, &#39;ribosome&#39;, &#39;ribosomal&lt;br/&gt;subunit&#39;, &#39;mitochondiral&lt;br/&gt;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&#39;, &#39;large&lt;br/&gt;ribosomal&lt;br/&gt;subunit&lt;br/&gt;subcomplex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M163.28,-147.01C163.28,-133.04 163.28,-117.94 163.28,-104\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcirptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcirptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"344.28\" cy=\"-890.44\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"310.78\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">transcirptional</text>\n",
       "<text text-anchor=\"start\" x=\"320.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"318.78\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">complexes</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcirptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;) -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;cytoplasm&#39;, &#39;transcriptional&lt;br/&gt;regulation&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;transcirptional&lt;br/&gt;regulation&lt;br/&gt;complexes&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M344.28,-986.33C344.28,-972.58 344.28,-957.77 344.28,-944.03\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"488.28\" cy=\"-1040.29\" rx=\"38.37\" ry=\"38.37\"/>\n",
       "<text text-anchor=\"start\" x=\"474.28\" y=\"-1050.19\" font-family=\"Times,serif\" font-size=\"12.00\">DNA</text>\n",
       "<text text-anchor=\"start\" x=\"465.28\" y=\"-1037.19\" font-family=\"Times,serif\" font-size=\"12.00\">metabolic</text>\n",
       "<text text-anchor=\"start\" x=\"466.28\" y=\"-1024.19\" font-family=\"Times,serif\" font-size=\"12.00\">assembly</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;) -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;DNA&lt;br/&gt;metabolic&lt;br/&gt;assembly&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M628.97,-1151.12C604.1,-1138.43 565.71,-1117.17 536.28,-1093.32 528.06,-1086.67 519.95,-1078.61 512.8,-1070.85\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"575.28\" cy=\"-1040.29\" rx=\"30.31\" ry=\"30.31\"/>\n",
       "<text text-anchor=\"start\" x=\"557.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"560.28\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">lumen</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;) -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M639.04,-1139.86C625.41,-1118.79 605.49,-1087.99 591.55,-1066.44\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"654.28\" cy=\"-1040.29\" rx=\"30.31\" ry=\"30.31\"/>\n",
       "<text text-anchor=\"start\" x=\"636.78\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"641.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">body</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;) -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;body&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M654.28,-1135.7C654.28,-1116.56 654.28,-1091.11 654.28,-1071.46\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"753.28\" cy=\"-1040.29\" rx=\"50.82\" ry=\"50.82\"/>\n",
       "<text text-anchor=\"start\" x=\"721.28\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"749.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M671.97,-1141.58C685.66,-1124.7 704.96,-1100.89 721.32,-1080.7\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"873.28\" cy=\"-1040.29\" rx=\"50.82\" ry=\"50.82\"/>\n",
       "<text text-anchor=\"start\" x=\"841.28\" y=\"-1043.69\" font-family=\"Times,serif\" font-size=\"12.00\">nucleo&#45;plasm</text>\n",
       "<text text-anchor=\"start\" x=\"869.78\" y=\"-1030.69\" font-family=\"Times,serif\" font-size=\"12.00\">2</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;) -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M681.1,-1154.39C713.73,-1142.8 769.83,-1120.85 813.28,-1093.32 821.08,-1088.38 828.91,-1082.46 836.21,-1076.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"995.28\" cy=\"-1040.29\" rx=\"53.07\" ry=\"53.07\"/>\n",
       "<text text-anchor=\"start\" x=\"977.78\" y=\"-1050.19\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"961.78\" y=\"-1037.19\" font-family=\"Times,serif\" font-size=\"12.00\">transcriptional</text>\n",
       "<text text-anchor=\"start\" x=\"977.28\" y=\"-1024.19\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M682.67,-1160.54C734.83,-1154.37 848.32,-1136.71 933.28,-1093.32 941.41,-1089.17 949.41,-1083.77 956.82,-1078.02\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.28\" cy=\"-890.44\" rx=\"33.49\" ry=\"33.49\"/>\n",
       "<text text-anchor=\"start\" x=\"538.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">nucleolus</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;) -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M572.44,-1009.37C570.12,-984.82 566.83,-950.08 564.42,-924.67\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"645.28\" cy=\"-890.44\" rx=\"33.47\" ry=\"33.47\"/>\n",
       "<text text-anchor=\"start\" x=\"627.78\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">nuclear</text>\n",
       "<text text-anchor=\"start\" x=\"626.78\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"627.28\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">speckle</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;) -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nuclear&lt;br/&gt;splicing&lt;br/&gt;speckle&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M588.28,-1011.83C600.42,-986.19 618.49,-948.02 631.02,-921.56\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"440.28\" cy=\"-728.58\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"401.28\" y=\"-738.48\" font-family=\"Times,serif\" font-size=\"12.00\">ribonucleprotein</text>\n",
       "<text text-anchor=\"start\" x=\"419.78\" y=\"-725.48\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"424.78\" y=\"-712.48\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;ribonucleprotein&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M541.02,-862.69C523.53,-839.58 497.82,-805.61 476.76,-777.78\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"561.28\" cy=\"-728.58\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"547.78\" y=\"-744.98\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"535.78\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"540.78\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"545.78\" y=\"-705.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M561.28,-856.32C561.28,-831.81 561.28,-798.23 561.28,-771.8\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"667.28\" cy=\"-728.58\" rx=\"45.01\" ry=\"45.01\"/>\n",
       "<text text-anchor=\"start\" x=\"639.28\" y=\"-738.48\" font-family=\"Times,serif\" font-size=\"12.00\">splicosomal</text>\n",
       "<text text-anchor=\"start\" x=\"646.78\" y=\"-725.48\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"651.78\" y=\"-712.48\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;splicosomal&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M579.74,-861.6C597.12,-835.39 623.27,-795.94 642.47,-766.99\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"531.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"517.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"505.78\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">processing</text>\n",
       "<text text-anchor=\"start\" x=\"510.78\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"527.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">1</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;) -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;1&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M552.84,-686.22C548.73,-666.09 543.81,-641.94 539.7,-621.81\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"634.28\" cy=\"-579.51\" rx=\"42.35\" ry=\"42.35\"/>\n",
       "<text text-anchor=\"start\" x=\"620.78\" y=\"-595.91\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"615.78\" y=\"-582.91\" font-family=\"Times,serif\" font-size=\"12.00\">splicing</text>\n",
       "<text text-anchor=\"start\" x=\"613.78\" y=\"-569.91\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"618.78\" y=\"-556.91\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;lumen&#39;, &#39;nucleolus&#39;, &#39;RNA&lt;br/&gt;processing&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;RNA&lt;br/&gt;splicing&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M580.07,-689.71C590.93,-667.83 604.55,-640.41 615.41,-618.51\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"763.28\" cy=\"-890.44\" rx=\"48.58\" ry=\"48.58\"/>\n",
       "<text text-anchor=\"start\" x=\"732.78\" y=\"-906.84\" font-family=\"Times,serif\" font-size=\"12.00\">chromosome</text>\n",
       "<text text-anchor=\"start\" x=\"733.78\" y=\"-893.84\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"742.78\" y=\"-880.84\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"747.78\" y=\"-867.84\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M756.69,-988.82C757.76,-972.93 758.95,-955.42 760.01,-939.69\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"778.28\" cy=\"-728.58\" rx=\"47.25\" ry=\"47.25\"/>\n",
       "<text text-anchor=\"start\" x=\"754.28\" y=\"-744.98\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"748.78\" y=\"-731.98\" font-family=\"Times,serif\" font-size=\"12.00\">organization</text>\n",
       "<text text-anchor=\"start\" x=\"757.78\" y=\"-718.98\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"762.78\" y=\"-705.98\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M767.78,-841.48C769.71,-820.88 771.96,-796.89 773.88,-776.41\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"778.28\" cy=\"-579.51\" rx=\"34.8\" ry=\"34.8\"/>\n",
       "<text text-anchor=\"start\" x=\"765.78\" y=\"-589.41\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"757.78\" y=\"-576.41\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "<text text-anchor=\"start\" x=\"762.78\" y=\"-563.41\" font-family=\"Times,serif\" font-size=\"12.00\">family</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;) -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M778.28,-680.64C778.28,-659.29 778.28,-634.48 778.28,-614.9\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"778.28\" cy=\"-453.07\" rx=\"34.8\" ry=\"34.8\"/>\n",
       "<text text-anchor=\"start\" x=\"762.78\" y=\"-462.97\" font-family=\"Times,serif\" font-size=\"12.00\">NuA4</text>\n",
       "<text text-anchor=\"start\" x=\"765.78\" y=\"-449.97\" font-family=\"Times,serif\" font-size=\"12.00\">HAT</text>\n",
       "<text text-anchor=\"start\" x=\"757.78\" y=\"-436.97\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;1&#39;, &#39;chromosome&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;chromatin&lt;br/&gt;organization&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;HAT&lt;br/&gt;complex&lt;br/&gt;family&#39;, &#39;NuA4&lt;br/&gt;HAT&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M778.28,-543.85C778.28,-526.59 778.28,-505.78 778.28,-488.55\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"873.28\" cy=\"-890.44\" rx=\"39.7\" ry=\"39.7\"/>\n",
       "<text text-anchor=\"start\" x=\"849.28\" y=\"-900.34\" font-family=\"Times,serif\" font-size=\"12.00\">chromatin</text>\n",
       "<text text-anchor=\"start\" x=\"849.28\" y=\"-887.34\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"852.78\" y=\"-874.34\" font-family=\"Times,serif\" font-size=\"12.00\">complex</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;) -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nucleo&#45;plasm&lt;br/&gt;2&#39;, &#39;chromatin&lt;br/&gt;regulation&lt;br/&gt;complex&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M873.28,-988.82C873.28,-969.92 873.28,-948.73 873.28,-930.96\"/>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" stroke-width=\"2\" cx=\"995.28\" cy=\"-890.44\" rx=\"60.62\" ry=\"60.62\"/>\n",
       "<text text-anchor=\"start\" x=\"975.28\" y=\"-919.84\" font-family=\"Times,serif\" font-size=\"12.00\">negative</text>\n",
       "<text text-anchor=\"start\" x=\"971.28\" y=\"-906.84\" font-family=\"Times,serif\" font-size=\"12.00\">regulation</text>\n",
       "<text text-anchor=\"start\" x=\"990.28\" y=\"-893.84\" font-family=\"Times,serif\" font-size=\"12.00\">of</text>\n",
       "<text text-anchor=\"start\" x=\"981.78\" y=\"-880.84\" font-family=\"Times,serif\" font-size=\"12.00\">RNA</text>\n",
       "<text text-anchor=\"start\" x=\"966.28\" y=\"-867.84\" font-family=\"Times,serif\" font-size=\"12.00\">biosynthesis</text>\n",
       "<text text-anchor=\"start\" x=\"977.28\" y=\"-854.84\" font-family=\"Times,serif\" font-size=\"12.00\">process</text>\n",
       "</g>\n",
       "<!-- (&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;) -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;)&#45;&gt;(&#39;cell&#39;, &#39;nucleus&#39;, &#39;nuclear&lt;br/&gt;transcriptional&lt;br/&gt;speckle&#39;, &#39;negative&lt;br/&gt;regulation&lt;br/&gt;of&lt;br/&gt;RNA&lt;br/&gt;biosynthesis&lt;br/&gt;process&#39;)</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2\" d=\"M995.28,-986.33C995.28,-975.15 995.28,-963.26 995.28,-951.83\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x1087694c0>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bigraph_viz.dict_utils import replace_regex_recursive\n",
    "music_map = {\n",
    "    'cell': {\n",
    "        'cytoplasm': {\n",
    "            'secretory organelles': {\n",
    "                'transmembrane transport systems': {\n",
    "                    'ion transmembrane transport systems': {}}},\n",
    "            'cytoplasmic organelles': {\n",
    "                'subgroup of cytoplasmic organelles': {\n",
    "                    'metabolic organelles': {\n",
    "                        'subgroup of metabolic organelles': {},\n",
    "                        'mitochondrion': {}},\n",
    "                    'ribonucleoprotein complex family': {\n",
    "                        'RNA processng complex 1': {'RNA splicing complex 1': {}},\n",
    "                        'ribosome biogenesis community': {\n",
    "                            'ribosome': {\n",
    "                                'ribosomal subunit': {\n",
    "                                    'mitochondiral large ribosomal subunit': {\n",
    "                                        'large ribosomal subunit subcomplex 1': {}},\n",
    "                                    'mito-cyto ribosomal cluster': {},\n",
    "                                    'cotranslational protein targeting to membrane system': {},\n",
    "                                    'ribosomal complex 5': {},\n",
    "                                    'ribosomal complex 2': {}}},\n",
    "                            'nucleic acid binding complex': {}}}}},\n",
    "            'transcriptional regulation complex family': {\n",
    "                'transcirptional regulation complexes': {}}},\n",
    "        'nucleus': {\n",
    "            'DNA metabolic assembly': {},\n",
    "            'nuclear lumen': {\n",
    "                'nucleolus': {\n",
    "                    'ribonucleprotein complex family': {},\n",
    "                    'RNA processing complex family': {\n",
    "                        'RNA processing complex 1': {},\n",
    "                        'RNA splicing complex family': {}},\n",
    "                    'splicosomal complex family': {}},\n",
    "                'nuclear splicing speckle': {}},\n",
    "            'nuclear body': {},\n",
    "            'nucleo-plasm 1': {\n",
    "                'chromosome organization complex family': {\n",
    "                    'chromatin organization complex family': {\n",
    "                        'HAT complex family': {\n",
    "                            'NuA4 HAT complex': {}}}}},\n",
    "            'nucleo-plasm 2': {\n",
    "                'chromatin regulation complex': {}},\n",
    "            'nuclear transcriptional speckle': {\n",
    "                'negative regulation of RNA biosynthesis process': {}}}}}\n",
    "music_map = replace_regex_recursive(music_map)\n",
    "plot_bigraph(music_map, **plot_settings, filename='cell_hierarchy')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "188eb4a6-c526-4b38-a22d-97f7fb8b7244",
   "metadata": {},
   "source": [
    "## Whole-cell E. coli model\n",
    "\n",
    "We here plot the multiscale map of a cell from [Macklin et al. Simultaneous cross-evaluation of heterogeneous E. coli datasets via mechanistic simulation. Science. 2020.](https://pubmed.ncbi.nlm.nih.gov/32703847/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "16960d1b-6e35-421e-bb7d-3efe5052e4fc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing out/ecoli\n"
     ]
    }
   ],
   "source": [
    "ecoli = { \n",
    "    'chromosome-structure': { \n",
    "        '_wires': {\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'DnaA_boxes': ('unique', 'DnaA_box')\n",
    "        }\n",
    "    },\n",
    "    'metabolism': { \n",
    "        '_wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'catalysts': ('bulk',),\n",
    "            'kinetics_enzymes': ('bulk',),\n",
    "            'kinetics_substrates': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'environment': ('environment',),\n",
    "            'amino_acids_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'tf-binding': { \n",
    "        '_wires': {\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'active_tfs': ('bulk',),\n",
    "            'active_tfs_total': ('bulk',),\n",
    "            'inactive_tfs_total': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-initiation': { \n",
    "        '_wires': {\n",
    "            'environment': ('environment',),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'promoters': ('unique', 'promoter'),\n",
    "            'molecules': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'transcript-elongation': { \n",
    "        '_wires': {\n",
    "            'environment': ('environment',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_RNAPs': ('unique', 'active_RNAP'),\n",
    "            'molecules': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'ntps': ('bulk',),\n",
    "        }\n",
    "    },\n",
    "    'rna-degradation': { \n",
    "        '_wires': {\n",
    "            'charged_trna': ('bulk',),\n",
    "            'bulk_RNAs': ('bulk',),\n",
    "            'nmps': ('bulk',),\n",
    "            'fragmentMetabolites': ('bulk',),\n",
    "            'fragmentBases': ('bulk',),\n",
    "            'endoRnases': ('bulk',),\n",
    "            'exoRnases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules': ('bulk',),\n",
    "            'RNAs': ('unique', 'RNA'),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-initiation': { \n",
    "        '_wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'RNA': ('unique', 'RNA'),\n",
    "            'subunits': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'polypeptide-elongation': { \n",
    "        '_wires': {\n",
    "            'environment': ('environment',),\n",
    "            'active_ribosome': ('unique', 'active_ribosome'),\n",
    "            'molecules': ('bulk',),\n",
    "            'monomers': ('bulk',),\n",
    "            'amino_acids': ('bulk',),\n",
    "            'ppgpp_reaction_metabolites': ('bulk',),\n",
    "            'uncharged_trna': ('bulk',),\n",
    "            'charged_trna': ('bulk',),\n",
    "            'charging_molecules': ('bulk',),\n",
    "            'synthetases': ('bulk',),\n",
    "            'subunits': ('bulk',),\n",
    "            'molecules_total': ('bulk',),\n",
    "            'amino_acids_total': ('bulk',),\n",
    "            'charged_trna_total': ('bulk',),\n",
    "            'uncharged_trna_total': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'complexation': {\n",
    "        '_wires': {\n",
    "            'molecules': ('bulk',), \n",
    "        }\n",
    "    },\n",
    "    'two-component-system': { \n",
    "        '_wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'equilibrium': {\n",
    "        '_wires': {\n",
    "            'molecules': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'protein-degradation': { \n",
    "        '_wires': {\n",
    "            'metabolites': ('bulk',),\n",
    "            'proteins': ('bulk',)\n",
    "        }\n",
    "    },\n",
    "    'chromosome-replication': { \n",
    "        '_wires': {\n",
    "            'replisome_trimers': ('bulk',),\n",
    "            'replisome_monomers': ('bulk',),\n",
    "            'dntps': ('bulk',),\n",
    "            'ppi': ('bulk',),\n",
    "            'active_replisomes': ('unique', 'active_replisome'),\n",
    "            'oriCs': ('unique', 'oriC'),\n",
    "            'chromosome_domains': ('unique', 'chromosome_domain'),\n",
    "            'full_chromosomes': ('unique', 'full_chromosome'),\n",
    "            'environment': ('environment',)\n",
    "        }\n",
    "    },\n",
    "    'unique': {\n",
    "        'chromosome_domain': {},\n",
    "        'full_chromosome': {},\n",
    "        'oriC': {},\n",
    "        'active_replisome': {},\n",
    "        'RNA': {},\n",
    "        'active_ribosome': {},\n",
    "        'DnaA_box': {},\n",
    "        'promoter': {},\n",
    "    },\n",
    "    'bulk': {},\n",
    "    'environment': {},\n",
    "}\n",
    "node_groups = [[\n",
    "        ('unique',), ('bulk',), ('environment',), \n",
    "    ],\n",
    "]\n",
    "# plot settings\n",
    "plot_settings_ecoli = {\n",
    "    'remove_process_place_edges': True,\n",
    "    'rankdir': 'RL',\n",
    "    'dpi': '100',\n",
    "    'out_dir': 'out',\n",
    "}\n",
    "fig = plot_bigraph(ecoli, **plot_settings_ecoli, node_groups=node_groups, filename='ecoli')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54e1102d-521c-49cc-8ed2-be26c2b84c3d",
   "metadata": {
    "tags": []
   },
   "source": [
    "[See E. coli wiring diagram](https://github.com/vivarium-collective/bigraph-viz/blob/main/doc/_static/ecoli.png?raw=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e5848a-e92f-45f3-b8a4-30320268640c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
